// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_code.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_mm7Hc1, __expression_SPxO3N, __block_n_tQ9T;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_mm7Hc1 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/code.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_SPxO3N = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/code.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_n_tQ9T = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/code.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_Uu3ZLs = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_wIJdCe = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/code.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_mm7Hc1(0);
    var util = (__expression_SPxO3N(1), require('util')), super_ = (__expression_SPxO3N(2), require('../field'));
}
function code(list, path, options) {
    __block_n_tQ9T(0);
    {
        __statement_mm7Hc1(3);
        this._nativeType = String;
    }
    {
        __statement_mm7Hc1(4);
        this.height = (__expression_SPxO3N(5), options.height || 180);
    }
    {
        __statement_mm7Hc1(6);
        this.lang = options.lang;
    }
    {
        __statement_mm7Hc1(7);
        this.mime = (__expression_SPxO3N(8), getMime(this.lang));
    }
    if (options.initial) {
        __block_n_tQ9T(1);
        {
            __statement_mm7Hc1(9);
            throw new Error((__expression_SPxO3N(10), (__expression_SPxO3N(11), (__expression_SPxO3N(12), (__expression_SPxO3N(13), (__expression_SPxO3N(14), 'Invalid Configuration\n\n' + 'code fields (') + list.key) + '.') + (__expression_SPxO3N(15), path)) + ') do not currently support being used as initial fields.\n'));
        }
    }
    {
        __statement_mm7Hc1(16);
        __extro_wIJdCe(17, __intro_Uu3ZLs(17, code.super_).call(this, list, path, options));
    }
}
{
    __statement_mm7Hc1(18);
    __extro_wIJdCe(19, __intro_Uu3ZLs(19, util).inherits(code, super_));
}
function getMime(lang) {
    __block_n_tQ9T(2);
    {
        __statement_mm7Hc1(20);
        var mime;
    }
    switch (__expression_SPxO3N(21), lang) {
    case 'c': {
            __block_n_tQ9T(3);
            {
                __statement_mm7Hc1(22);
                mime = 'text/x-csrc';
            }
            break;
        }
    case 'c++':
    case 'objetivec': {
            __block_n_tQ9T(4);
            {
                __statement_mm7Hc1(23);
                mime = 'text/x-c++src';
            }
            break;
        }
    case 'css': {
            __block_n_tQ9T(5);
            {
                __statement_mm7Hc1(24);
                mime = 'text/css';
            }
            break;
        }
    case 'asp': {
            __block_n_tQ9T(6);
            {
                __statement_mm7Hc1(25);
                mime = 'application/x-aspx';
            }
            break;
        }
    case 'c#': {
            __block_n_tQ9T(7);
            {
                __statement_mm7Hc1(26);
                mime = 'text/x-csharp';
            }
            break;
        }
    case 'vb': {
            __block_n_tQ9T(8);
            {
                __statement_mm7Hc1(27);
                mime = 'text/x-vb';
            }
            break;
        }
    case 'xml': {
            __block_n_tQ9T(9);
            {
                __statement_mm7Hc1(28);
                mime = 'text/xml';
            }
            break;
        }
    case 'php': {
            __block_n_tQ9T(10);
            {
                __statement_mm7Hc1(29);
                mime = 'application/x-httpd-php';
            }
            break;
        }
    case 'html': {
            __block_n_tQ9T(11);
            {
                __statement_mm7Hc1(30);
                mime = 'text/html';
            }
            break;
        }
    case 'ini': {
            __block_n_tQ9T(12);
            {
                __statement_mm7Hc1(31);
                mime = 'text/x-properties';
            }
            break;
        }
    case 'js': {
            __block_n_tQ9T(13);
            {
                __statement_mm7Hc1(32);
                mime = 'text/javascript';
            }
            break;
        }
    case 'java': {
            __block_n_tQ9T(14);
            {
                __statement_mm7Hc1(33);
                mime = 'text/x-java';
            }
            break;
        }
    case 'coffee': {
            __block_n_tQ9T(15);
            {
                __statement_mm7Hc1(34);
                mime = 'text/x-coffeescript';
            }
            break;
        }
    case 'lisp': {
            __block_n_tQ9T(16);
            {
                __statement_mm7Hc1(35);
                mime = 'text/x-common-lisp';
            }
            break;
        }
    case 'perl': {
            __block_n_tQ9T(17);
            {
                __statement_mm7Hc1(36);
                mime = 'text/x-perl';
            }
            break;
        }
    case 'python': {
            __block_n_tQ9T(18);
            {
                __statement_mm7Hc1(37);
                mime = 'text/x-python';
            }
            break;
        }
    case 'sql': {
            __block_n_tQ9T(19);
            {
                __statement_mm7Hc1(38);
                mime = 'text/x-sql';
            }
            break;
        }
    case 'json': {
            __block_n_tQ9T(20);
            {
                __statement_mm7Hc1(39);
                mime = 'application/json';
            }
            break;
        }
    case 'less': {
            __block_n_tQ9T(21);
            {
                __statement_mm7Hc1(40);
                mime = 'text/x-less';
            }
            break;
        }
    case 'sass': {
            __block_n_tQ9T(22);
            {
                __statement_mm7Hc1(41);
                mime = 'text/x-sass';
            }
            break;
        }
    case 'sh': {
            __block_n_tQ9T(23);
            {
                __statement_mm7Hc1(42);
                mime = 'text/x-sh';
            }
            break;
        }
    case 'ruby': {
            __block_n_tQ9T(24);
            {
                __statement_mm7Hc1(43);
                mime = 'text/x-ruby';
            }
            break;
        }
    case 'jsp': {
            __block_n_tQ9T(25);
            {
                __statement_mm7Hc1(44);
                mime = 'application/x-jsp';
            }
            break;
        }
    case 'tpl': {
            __block_n_tQ9T(26);
            {
                __statement_mm7Hc1(45);
                mime = 'text/x-smarty';
            }
            break;
        }
    case 'jade': {
            __block_n_tQ9T(27);
            {
                __statement_mm7Hc1(46);
                mime = 'text/x-jade';
            }
            break;
        }
    }
    return __expression_SPxO3N(47), mime;
}
{
    __statement_mm7Hc1(48);
    exports = module.exports = code;
}
