// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_cloudinaryimages.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_D0SZ$A, __expression_dldaz0, __block_na8FQt;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_D0SZ$A = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimages.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_dldaz0 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimages.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_na8FQt = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimages.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_qMyCmE = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_QbutNP = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimages.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_D0SZ$A(0);
    var _ = (__expression_dldaz0(1), require('underscore')), keystone = (__expression_dldaz0(2), require('../../')), util = (__expression_dldaz0(3), require('util')), cloudinary = (__expression_dldaz0(4), require('cloudinary')), utils = (__expression_dldaz0(5), require('keystone-utils')), super_ = (__expression_dldaz0(6), require('../field')), async = (__expression_dldaz0(7), require('async'));
}
function cloudinaryimages(list, path, options) {
    __block_na8FQt(0);
    {
        __statement_D0SZ$A(8);
        this._underscoreMethods = [
            'format'
        ];
    }
    {
        __statement_D0SZ$A(9);
        options.nofilter = true;
    }
    if (options.initial) {
        __block_na8FQt(1);
        {
            __statement_D0SZ$A(10);
            throw new Error((__expression_dldaz0(11), (__expression_dldaz0(12), (__expression_dldaz0(13), (__expression_dldaz0(14), (__expression_dldaz0(15), 'Invalid Configuration\n\n' + 'CloudinaryImages fields (') + list.key) + '.') + (__expression_dldaz0(16), path)) + ') do not currently support being used as initial fields.\n'));
        }
    }
    {
        __statement_D0SZ$A(17);
        __extro_QbutNP(18, __intro_qMyCmE(18, cloudinaryimages.super_).call(this, list, path, options));
    }
    if (__expression_dldaz0(19), !__extro_QbutNP(20, __intro_qMyCmE(20, keystone).get('cloudinary config'))) {
        __block_na8FQt(2);
        {
            __statement_D0SZ$A(21);
            throw new Error((__expression_dldaz0(22), (__expression_dldaz0(23), (__expression_dldaz0(24), (__expression_dldaz0(25), (__expression_dldaz0(26), (__expression_dldaz0(27), 'Invalid Configuration\n\n' + 'CloudinaryImages fields (') + list.key) + '.') + this.path) + ') require the "cloudinary config" option to be set.\n\n') + 'See http://keystonejs.com/docs/configuration/#cloudinary for more information.\n'));
        }
    }
}
{
    __statement_D0SZ$A(28);
    __extro_QbutNP(29, __intro_qMyCmE(29, util).inherits(cloudinaryimages, super_));
}
{
    __statement_D0SZ$A(30);
    cloudinaryimages.prototype.addToSchema = function () {
        __block_na8FQt(3);
        {
            __statement_D0SZ$A(31);
            var mongoose = keystone.mongoose;
        }
        {
            __statement_D0SZ$A(32);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_D0SZ$A(33);
            var paths = this.paths = {
                    upload: __extro_QbutNP(34, __intro_qMyCmE(34, this._path).append('_upload')),
                    uploads: __extro_QbutNP(35, __intro_qMyCmE(35, this._path).append('_uploads')),
                    action: __extro_QbutNP(36, __intro_qMyCmE(36, this._path).append('_action')),
                    order: __extro_QbutNP(37, __intro_qMyCmE(37, this._path).append('_order'))
                };
        }
        {
            __statement_D0SZ$A(38);
            var ImageSchema = new mongoose.Schema({
                    public_id: String,
                    version: Number,
                    signature: String,
                    format: String,
                    resource_type: String,
                    url: String,
                    width: Number,
                    height: Number,
                    secure_url: String
                });
        }
        {
            __statement_D0SZ$A(39);
            var src = function (img, options) {
                __block_na8FQt(4);
                if (__extro_QbutNP(40, __intro_qMyCmE(40, keystone).get('cloudinary secure'))) {
                    __block_na8FQt(5);
                    {
                        __statement_D0SZ$A(41);
                        options = (__expression_dldaz0(42), (__expression_dldaz0(43), options) || {});
                    }
                    {
                        __statement_D0SZ$A(44);
                        options.secure = true;
                    }
                }
                return __expression_dldaz0(45), img.public_id ? (__expression_dldaz0(46), __extro_QbutNP(48, __intro_qMyCmE(48, cloudinary).url((__expression_dldaz0(49), (__expression_dldaz0(50), img.public_id + '.') + img.format), options))) : (__expression_dldaz0(47), '');
            };
        }
        {
            __statement_D0SZ$A(51);
            var addSize = function (options, width, height) {
                __block_na8FQt(6);
                if (__expression_dldaz0(52), width) {
                    __block_na8FQt(7);
                    {
                        __statement_D0SZ$A(53);
                        options.width = width;
                    }
                }
                if (__expression_dldaz0(54), height) {
                    __block_na8FQt(8);
                    {
                        __statement_D0SZ$A(55);
                        options.height = height;
                    }
                }
                return __expression_dldaz0(56), options;
            };
        }
        {
            __statement_D0SZ$A(57);
            __extro_QbutNP(58, __intro_qMyCmE(58, ImageSchema).method('src', function (options) {
                __block_na8FQt(9);
                return __expression_dldaz0(59), (__expression_dldaz0(60), src(this, options));
            }));
        }
        {
            __statement_D0SZ$A(61);
            __extro_QbutNP(62, __intro_qMyCmE(62, ImageSchema).method('scale', function (width, height) {
                __block_na8FQt(10);
                return __expression_dldaz0(63), (__expression_dldaz0(64), src(this, (__expression_dldaz0(65), addSize({
                    crop: 'scale'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(66);
            __extro_QbutNP(67, __intro_qMyCmE(67, ImageSchema).method('fill', function (width, height) {
                __block_na8FQt(11);
                return __expression_dldaz0(68), (__expression_dldaz0(69), src(this, (__expression_dldaz0(70), addSize({
                    crop: 'fill',
                    gravity: 'faces'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(71);
            __extro_QbutNP(72, __intro_qMyCmE(72, ImageSchema).method('lfill', function (width, height) {
                __block_na8FQt(12);
                return __expression_dldaz0(73), (__expression_dldaz0(74), src(this, (__expression_dldaz0(75), addSize({
                    crop: 'lfill',
                    gravity: 'faces'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(76);
            __extro_QbutNP(77, __intro_qMyCmE(77, ImageSchema).method('fit', function (width, height) {
                __block_na8FQt(13);
                return __expression_dldaz0(78), (__expression_dldaz0(79), src(this, (__expression_dldaz0(80), addSize({
                    crop: 'fit'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(81);
            __extro_QbutNP(82, __intro_qMyCmE(82, ImageSchema).method('limit', function (width, height) {
                __block_na8FQt(14);
                return __expression_dldaz0(83), (__expression_dldaz0(84), src(this, (__expression_dldaz0(85), addSize({
                    crop: 'limit'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(86);
            __extro_QbutNP(87, __intro_qMyCmE(87, ImageSchema).method('pad', function (width, height) {
                __block_na8FQt(15);
                return __expression_dldaz0(88), (__expression_dldaz0(89), src(this, (__expression_dldaz0(90), addSize({
                    crop: 'pad'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(91);
            __extro_QbutNP(92, __intro_qMyCmE(92, ImageSchema).method('lpad', function (width, height) {
                __block_na8FQt(16);
                return __expression_dldaz0(93), (__expression_dldaz0(94), src(this, (__expression_dldaz0(95), addSize({
                    crop: 'lpad'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(96);
            __extro_QbutNP(97, __intro_qMyCmE(97, ImageSchema).method('crop', function (width, height) {
                __block_na8FQt(17);
                return __expression_dldaz0(98), (__expression_dldaz0(99), src(this, (__expression_dldaz0(100), addSize({
                    crop: 'crop',
                    gravity: 'faces'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(101);
            __extro_QbutNP(102, __intro_qMyCmE(102, ImageSchema).method('thumbnail', function (width, height) {
                __block_na8FQt(18);
                return __expression_dldaz0(103), (__expression_dldaz0(104), src(this, (__expression_dldaz0(105), addSize({
                    crop: 'thumb',
                    gravity: 'faces'
                }, width, height))));
            }));
        }
        {
            __statement_D0SZ$A(106);
            __extro_QbutNP(107, __intro_qMyCmE(107, schema).add(__extro_QbutNP(108, __intro_qMyCmE(108, this._path).addTo({}, [
                ImageSchema
            ]))));
        }
        {
            __statement_D0SZ$A(109);
            this.removeImage = function (item, id, method, callback) {
                __block_na8FQt(19);
                {
                    __statement_D0SZ$A(110);
                    var images = __extro_QbutNP(111, __intro_qMyCmE(111, item).get(field.path));
                }
                if (__expression_dldaz0(112), 'number' != (__expression_dldaz0(113), typeof id)) {
                    __block_na8FQt(20);
                    for (var i = 0; __expression_dldaz0(114), (__expression_dldaz0(115), i) < images.length; __expression_dldaz0(116), i++) {
                        __block_na8FQt(21);
                        if (__expression_dldaz0(117), images[i].public_id == (__expression_dldaz0(118), id)) {
                            __block_na8FQt(22);
                            {
                                __statement_D0SZ$A(119);
                                id = i;
                            }
                            break;
                        }
                    }
                }
                {
                    __statement_D0SZ$A(120);
                    var img = images[id];
                }
                if (__expression_dldaz0(121), !(__expression_dldaz0(122), img)) {
                    __block_na8FQt(23);
                    return __expression_dldaz0(123);
                }
                if (__expression_dldaz0(124), (__expression_dldaz0(125), method) == 'delete') {
                    __block_na8FQt(24);
                    {
                        __statement_D0SZ$A(126);
                        __extro_QbutNP(127, __intro_qMyCmE(127, cloudinary.uploader).destroy(img.public_id, function () {
                            __block_na8FQt(25);
                        }));
                    }
                }
                {
                    __statement_D0SZ$A(128);
                    __extro_QbutNP(129, __intro_qMyCmE(129, images).splice(id, 1));
                }
                if (__expression_dldaz0(130), callback) {
                    __block_na8FQt(26);
                    {
                        __statement_D0SZ$A(131);
                        __extro_QbutNP(132, __intro_qMyCmE(132, item).save((__expression_dldaz0(135), 'function' != (__expression_dldaz0(136), typeof callback)) ? (__expression_dldaz0(133), callback) : (__expression_dldaz0(134), undefined)));
                    }
                }
            };
        }
        {
            __statement_D0SZ$A(137);
            __extro_QbutNP(138, __intro_qMyCmE(138, this).underscoreMethod('remove', function (id, callback) {
                __block_na8FQt(27);
                {
                    __statement_D0SZ$A(139);
                    __extro_QbutNP(140, __intro_qMyCmE(140, field).removeImage(this, id, 'remove', callback));
                }
            }));
        }
        {
            __statement_D0SZ$A(141);
            __extro_QbutNP(142, __intro_qMyCmE(142, this).underscoreMethod('delete', function (id, callback) {
                __block_na8FQt(28);
                {
                    __statement_D0SZ$A(143);
                    __extro_QbutNP(144, __intro_qMyCmE(144, field).removeImage(this, id, 'delete', callback));
                }
            }));
        }
        {
            __statement_D0SZ$A(145);
            __extro_QbutNP(146, __intro_qMyCmE(146, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_D0SZ$A(147);
    cloudinaryimages.prototype.format = function (item) {
        __block_na8FQt(29);
        return __expression_dldaz0(148), __extro_QbutNP(149, __intro_qMyCmE(149, __extro_QbutNP(150, __intro_qMyCmE(150, _).map(__extro_QbutNP(151, __intro_qMyCmE(151, item).get(this.path)), function (img) {
            __block_na8FQt(30);
            return __expression_dldaz0(152), __extro_QbutNP(153, __intro_qMyCmE(153, img).src());
        }))).join(', '));
    };
}
{
    __statement_D0SZ$A(154);
    cloudinaryimages.prototype.isModified = function (item) {
        __block_na8FQt(31);
        return __expression_dldaz0(155), true;
    };
}
{
    __statement_D0SZ$A(156);
    cloudinaryimages.prototype.validateInput = function (data) {
        __block_na8FQt(32);
        return __expression_dldaz0(157), true;
    };
}
{
    __statement_D0SZ$A(158);
    cloudinaryimages.prototype.updateItem = function (item, data) {
        __block_na8FQt(33);
    };
}
{
    __statement_D0SZ$A(159);
    cloudinaryimages.prototype.getRequestHandler = function (item, req, paths, callback) {
        __block_na8FQt(34);
        {
            __statement_D0SZ$A(160);
            var field = this;
        }
        if (__extro_QbutNP(161, __intro_qMyCmE(161, utils).isFunction(paths))) {
            __block_na8FQt(35);
            {
                __statement_D0SZ$A(162);
                callback = paths;
            }
            {
                __statement_D0SZ$A(163);
                paths = field.paths;
            }
        } else if (__expression_dldaz0(164), !(__expression_dldaz0(165), paths)) {
            __block_na8FQt(36);
            {
                __statement_D0SZ$A(166);
                paths = field.paths;
            }
        }
        {
            __statement_D0SZ$A(167);
            callback = (__expression_dldaz0(168), (__expression_dldaz0(169), callback) || function () {
                __block_na8FQt(37);
            });
        }
        return __expression_dldaz0(170), function () {
            __block_na8FQt(38);
            if (req.body[paths.order]) {
                __block_na8FQt(39);
                {
                    __statement_D0SZ$A(171);
                    var images = __extro_QbutNP(172, __intro_qMyCmE(172, item).get(field.path)), newOrder = __extro_QbutNP(173, __intro_qMyCmE(173, req.body[paths.order]).split(','));
                }
                {
                    __statement_D0SZ$A(174);
                    __extro_QbutNP(175, __intro_qMyCmE(175, images).sort(function (a, b) {
                        __block_na8FQt(40);
                        return __expression_dldaz0(176), (__expression_dldaz0(179), __extro_QbutNP(180, __intro_qMyCmE(180, newOrder).indexOf(a.public_id)) > __extro_QbutNP(181, __intro_qMyCmE(181, newOrder).indexOf(b.public_id))) ? (__expression_dldaz0(177), 1) : (__expression_dldaz0(178), (__expression_dldaz0(182), -1));
                    }));
                }
            }
            if (__expression_dldaz0(183), req.body && req.body[paths.action]) {
                __block_na8FQt(41);
                {
                    __statement_D0SZ$A(184);
                    var actions = __extro_QbutNP(185, __intro_qMyCmE(185, req.body[paths.action]).split('|'));
                }
                {
                    __statement_D0SZ$A(186);
                    __extro_QbutNP(187, __intro_qMyCmE(187, actions).forEach(function (action) {
                        __block_na8FQt(42);
                        {
                            __statement_D0SZ$A(188);
                            action = __extro_QbutNP(189, __intro_qMyCmE(189, action).split(':'));
                        }
                        {
                            __statement_D0SZ$A(190);
                            var method = action[0], ids = action[1];
                        }
                        if (__expression_dldaz0(191), (__expression_dldaz0(192), !__extro_QbutNP(193, __intro_qMyCmE(193, method).match(/^(remove|delete)$/))) || (__expression_dldaz0(194), !(__expression_dldaz0(195), ids))) {
                            __block_na8FQt(43);
                            return __expression_dldaz0(196);
                        }
                        {
                            __statement_D0SZ$A(197);
                            __extro_QbutNP(198, __intro_qMyCmE(198, __extro_QbutNP(199, __intro_qMyCmE(199, ids).split(','))).forEach(function (id) {
                                __block_na8FQt(44);
                                {
                                    __statement_D0SZ$A(200);
                                    __extro_QbutNP(201, __intro_qMyCmE(201, field).removeImage(item, id, method));
                                }
                            }));
                        }
                    }));
                }
            }
            if (req.body[paths.uploads]) {
                __block_na8FQt(45);
                {
                    __statement_D0SZ$A(202);
                    var uploads = __extro_QbutNP(203, __intro_qMyCmE(203, JSON).parse(req.body[paths.uploads]));
                }
                {
                    __statement_D0SZ$A(204);
                    __extro_QbutNP(205, __intro_qMyCmE(205, uploads).forEach(function (file) {
                        __block_na8FQt(46);
                        {
                            __statement_D0SZ$A(206);
                            __extro_QbutNP(207, __intro_qMyCmE(207, __extro_QbutNP(208, __intro_qMyCmE(208, item).get(field.path))).push(file));
                        }
                    }));
                }
            }
            if (__expression_dldaz0(209), req.files && req.files[paths.upload]) {
                __block_na8FQt(47);
                {
                    __statement_D0SZ$A(210);
                    var files = __extro_QbutNP(211, __intro_qMyCmE(211, _).flatten(req.files[paths.upload]));
                }
                {
                    __statement_D0SZ$A(212);
                    var tp = (__expression_dldaz0(213), __extro_QbutNP(214, __intro_qMyCmE(214, keystone).get('cloudinary prefix')) || '');
                }
                if (tp.length) {
                    __block_na8FQt(48);
                    {
                        __statement_D0SZ$A(215);
                        tp += '_';
                    }
                }
                {
                    __statement_D0SZ$A(216);
                    var uploadOptions = {
                            tags: [
                                (__expression_dldaz0(217), (__expression_dldaz0(218), (__expression_dldaz0(219), (__expression_dldaz0(220), tp) + field.list.path) + '_') + field.path),
                                (__expression_dldaz0(221), (__expression_dldaz0(222), (__expression_dldaz0(223), (__expression_dldaz0(224), (__expression_dldaz0(225), (__expression_dldaz0(226), tp) + field.list.path) + '_') + field.path) + '_') + item.id)
                            ]
                        };
                }
                if (__extro_QbutNP(227, __intro_qMyCmE(227, keystone).get('cloudinary prefix'))) {
                    __block_na8FQt(49);
                    {
                        __statement_D0SZ$A(228);
                        __extro_QbutNP(229, __intro_qMyCmE(229, uploadOptions.tags).push(__extro_QbutNP(230, __intro_qMyCmE(230, keystone).get('cloudinary prefix'))));
                    }
                }
                if (__expression_dldaz0(231), __extro_QbutNP(232, __intro_qMyCmE(232, keystone).get('env')) != 'production') {
                    __block_na8FQt(50);
                    {
                        __statement_D0SZ$A(233);
                        __extro_QbutNP(234, __intro_qMyCmE(234, uploadOptions.tags).push((__expression_dldaz0(235), (__expression_dldaz0(236), tp) + 'dev')));
                    }
                }
                {
                    __statement_D0SZ$A(237);
                    __extro_QbutNP(238, __intro_qMyCmE(238, async).each(files, function (file, next) {
                        __block_na8FQt(51);
                        if (__expression_dldaz0(239), !file.size) {
                            __block_na8FQt(52);
                            return __expression_dldaz0(240), (__expression_dldaz0(241), next());
                        }
                        {
                            __statement_D0SZ$A(242);
                            __extro_QbutNP(243, __intro_qMyCmE(243, cloudinary.uploader).upload(file.path, function (result) {
                                __block_na8FQt(53);
                                if (result.error) {
                                    __block_na8FQt(54);
                                    return __expression_dldaz0(244), (__expression_dldaz0(245), next(result.error));
                                } else {
                                    __block_na8FQt(55);
                                    {
                                        __statement_D0SZ$A(246);
                                        __extro_QbutNP(247, __intro_qMyCmE(247, __extro_QbutNP(248, __intro_qMyCmE(248, item).get(field.path))).push(result));
                                    }
                                    return __expression_dldaz0(249), (__expression_dldaz0(250), next());
                                }
                            }, uploadOptions));
                        }
                    }, function (err) {
                        __block_na8FQt(56);
                        return __expression_dldaz0(251), (__expression_dldaz0(252), callback(err));
                    }));
                }
            } else {
                __block_na8FQt(57);
                return __expression_dldaz0(253), (__expression_dldaz0(254), callback());
            }
        };
    };
}
{
    __statement_D0SZ$A(255);
    cloudinaryimages.prototype.handleRequest = function (item, req, paths, callback) {
        __block_na8FQt(58);
        {
            __statement_D0SZ$A(256);
            __expression_dldaz0(257), __extro_QbutNP(258, __intro_qMyCmE(258, this).getRequestHandler(item, req, paths, callback))();
        }
    };
}
{
    __statement_D0SZ$A(259);
    exports = module.exports = cloudinaryimages;
}
