// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_localfiles.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_cwld1d, __expression_wNbCNj, __block_BKv82a;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_cwld1d = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfiles.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_wNbCNj = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfiles.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_BKv82a = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfiles.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_LerOvg = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_QLWdfI = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfiles.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_cwld1d(0);
    var fs = (__expression_wNbCNj(1), require('fs')), path = (__expression_wNbCNj(2), require('path')), _ = (__expression_wNbCNj(3), require('underscore')), moment = (__expression_wNbCNj(4), require('moment')), keystone = (__expression_wNbCNj(5), require('../../')), async = (__expression_wNbCNj(6), require('async')), util = (__expression_wNbCNj(7), require('util')), utils = (__expression_wNbCNj(8), require('keystone-utils')), super_ = (__expression_wNbCNj(9), require('../field')), async = (__expression_wNbCNj(10), require('async'));
}
function localfiles(list, path, options) {
    __block_BKv82a(0);
    {
        __statement_cwld1d(11);
        this._underscoreMethods = [
            'format',
            'uploadFiles'
        ];
    }
    {
        __statement_cwld1d(12);
        this._pre = {
            move: []
        };
    }
    {
        __statement_cwld1d(13);
        this._post = {
            move: []
        };
    }
    {
        __statement_cwld1d(14);
        options.nofilter = true;
    }
    if (options.initial) {
        __block_BKv82a(1);
        {
            __statement_cwld1d(15);
            throw new Error((__expression_wNbCNj(16), (__expression_wNbCNj(17), (__expression_wNbCNj(18), (__expression_wNbCNj(19), (__expression_wNbCNj(20), 'Invalid Configuration\n\n' + 'localfiles fields (') + list.key) + '.') + (__expression_wNbCNj(21), path)) + ') do not currently support being used as initial fields.\n'));
        }
    }
    {
        __statement_cwld1d(22);
        __extro_QLWdfI(23, __intro_LerOvg(23, localfiles.super_).call(this, list, path, options));
    }
    if (__expression_wNbCNj(24), !options.dest) {
        __block_BKv82a(2);
        {
            __statement_cwld1d(25);
            throw new Error((__expression_wNbCNj(26), (__expression_wNbCNj(27), (__expression_wNbCNj(28), (__expression_wNbCNj(29), (__expression_wNbCNj(30), 'Invalid Configuration\n\n' + 'localfiles fields (') + list.key) + '.') + (__expression_wNbCNj(31), path)) + ') require the "dest" option to be set.'));
        }
    }
    if (__expression_wNbCNj(32), options.pre && options.pre.move) {
        __block_BKv82a(3);
        {
            __statement_cwld1d(33);
            this._pre.move = __extro_QLWdfI(34, __intro_LerOvg(34, this._pre.move).concat(options.pre.move));
        }
    }
    if (__expression_wNbCNj(35), options.post && options.post.move) {
        __block_BKv82a(4);
        {
            __statement_cwld1d(36);
            this._post.move = __extro_QLWdfI(37, __intro_LerOvg(37, this._post.move).concat(options.post.move));
        }
    }
}
{
    __statement_cwld1d(38);
    __extro_QLWdfI(39, __intro_LerOvg(39, util).inherits(localfiles, super_));
}
{
    __statement_cwld1d(40);
    localfiles.prototype.pre = function (event, fn) {
        __block_BKv82a(5);
        if (__expression_wNbCNj(41), !this._pre[event]) {
            __block_BKv82a(6);
            {
                __statement_cwld1d(42);
                throw new Error((__expression_wNbCNj(43), (__expression_wNbCNj(44), (__expression_wNbCNj(45), (__expression_wNbCNj(46), (__expression_wNbCNj(47), (__expression_wNbCNj(48), (__expression_wNbCNj(49), 'localfiles (' + this.list.key) + '.') + this.path) + ') error: localfiles.pre()\n\n') + 'Event ') + (__expression_wNbCNj(50), event)) + ' is not supported.\n'));
            }
        }
        {
            __statement_cwld1d(51);
            __extro_QLWdfI(52, __intro_LerOvg(52, this._pre[event]).push(fn));
        }
        return __expression_wNbCNj(53), this;
    };
}
{
    __statement_cwld1d(54);
    localfiles.prototype.post = function (event, fn) {
        __block_BKv82a(7);
        if (__expression_wNbCNj(55), !this._post[event]) {
            __block_BKv82a(8);
            {
                __statement_cwld1d(56);
                throw new Error((__expression_wNbCNj(57), (__expression_wNbCNj(58), (__expression_wNbCNj(59), (__expression_wNbCNj(60), (__expression_wNbCNj(61), (__expression_wNbCNj(62), (__expression_wNbCNj(63), 'localfiles (' + this.list.key) + '.') + this.path) + ') error: localfiles.post()\n\n') + 'Event ') + (__expression_wNbCNj(64), event)) + ' is not supported.\n'));
            }
        }
        {
            __statement_cwld1d(65);
            __extro_QLWdfI(66, __intro_LerOvg(66, this._post[event]).push(fn));
        }
        return __expression_wNbCNj(67), this;
    };
}
{
    __statement_cwld1d(68);
    localfiles.prototype.addToSchema = function () {
        __block_BKv82a(9);
        {
            __statement_cwld1d(69);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_cwld1d(70);
            var mongoose = keystone.mongoose;
        }
        {
            __statement_cwld1d(71);
            var paths = this.paths = {
                    filename: __extro_QLWdfI(72, __intro_LerOvg(72, this._path).append('.filename')),
                    path: __extro_QLWdfI(73, __intro_LerOvg(73, this._path).append('.path')),
                    size: __extro_QLWdfI(74, __intro_LerOvg(74, this._path).append('.size')),
                    filetype: __extro_QLWdfI(75, __intro_LerOvg(75, this._path).append('.filetype')),
                    exists: __extro_QLWdfI(76, __intro_LerOvg(76, this._path).append('.exists')),
                    upload: __extro_QLWdfI(77, __intro_LerOvg(77, this._path).append('_upload')),
                    action: __extro_QLWdfI(78, __intro_LerOvg(78, this._path).append('_action')),
                    order: __extro_QLWdfI(79, __intro_LerOvg(79, this._path).append('_order'))
                };
        }
        {
            __statement_cwld1d(80);
            var schemaPaths = new mongoose.Schema({
                    filename: String,
                    path: String,
                    size: Number,
                    filetype: String
                });
        }
        {
            __statement_cwld1d(81);
            __extro_QLWdfI(82, __intro_LerOvg(82, schema).add(__extro_QLWdfI(83, __intro_LerOvg(83, this._path).addTo({}, [
                schemaPaths
            ]))));
        }
        {
            __statement_cwld1d(84);
            var exists = function (item) {
                __block_BKv82a(10);
                {
                    __statement_cwld1d(85);
                    var filepaths = __extro_QLWdfI(86, __intro_LerOvg(86, item).get(paths.path)), filename = __extro_QLWdfI(87, __intro_LerOvg(87, item).get(paths.filename));
                }
                if (__expression_wNbCNj(88), (__expression_wNbCNj(89), !(__expression_wNbCNj(90), filepaths)) || (__expression_wNbCNj(91), !(__expression_wNbCNj(92), filename))) {
                    __block_BKv82a(11);
                    return __expression_wNbCNj(93), false;
                }
                return __expression_wNbCNj(94), __extro_QLWdfI(95, __intro_LerOvg(95, fs).existsSync(__extro_QLWdfI(96, __intro_LerOvg(96, path).join(filepaths, filename))));
            };
        }
        {
            __statement_cwld1d(97);
            __extro_QLWdfI(98, __intro_LerOvg(98, __extro_QLWdfI(99, __intro_LerOvg(99, schema).virtual(paths.exists))).get(function () {
                __block_BKv82a(12);
                return __expression_wNbCNj(100), __extro_QLWdfI(101, __intro_LerOvg(101, schemaMethods.exists).apply(this));
            }));
        }
        {
            __statement_cwld1d(102);
            var reset = function (item) {
                __block_BKv82a(13);
                {
                    __statement_cwld1d(103);
                    __extro_QLWdfI(104, __intro_LerOvg(104, item).set(field.path, {
                        filename: '',
                        path: '',
                        size: 0,
                        filetype: ''
                    }));
                }
            };
        }
        {
            __statement_cwld1d(105);
            var schemaMethods = {
                    exists: function () {
                        __block_BKv82a(14);
                        return __expression_wNbCNj(106), (__expression_wNbCNj(107), exists(this));
                    },
                    reset: function () {
                        __block_BKv82a(15);
                        {
                            __statement_cwld1d(108);
                            __expression_wNbCNj(109), reset(this);
                        }
                    },
                    delete: function () {
                        __block_BKv82a(16);
                        if (__expression_wNbCNj(110), exists(this)) {
                            __block_BKv82a(17);
                            {
                                __statement_cwld1d(111);
                                __extro_QLWdfI(112, __intro_LerOvg(112, fs).unlinkSync(__extro_QLWdfI(113, __intro_LerOvg(113, path).join(__extro_QLWdfI(114, __intro_LerOvg(114, this).get(paths.path)), __extro_QLWdfI(115, __intro_LerOvg(115, this).get(paths.filename))))));
                            }
                        }
                        {
                            __statement_cwld1d(116);
                            __expression_wNbCNj(117), reset(this);
                        }
                    }
                };
        }
        {
            __statement_cwld1d(118);
            __extro_QLWdfI(119, __intro_LerOvg(119, _).each(schemaMethods, function (fn, key) {
                __block_BKv82a(18);
                {
                    __statement_cwld1d(120);
                    __extro_QLWdfI(121, __intro_LerOvg(121, field).underscoreMethod(key, fn));
                }
            }));
        }
        {
            __statement_cwld1d(122);
            this.apply = function (item, method) {
                __block_BKv82a(19);
                return __expression_wNbCNj(123), __extro_QLWdfI(124, __intro_LerOvg(124, schemaMethods[method]).apply(item, __extro_QLWdfI(125, __intro_LerOvg(125, Array.prototype.slice).call(arguments, 2))));
            };
        }
        {
            __statement_cwld1d(126);
            __extro_QLWdfI(127, __intro_LerOvg(127, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_cwld1d(128);
    localfiles.prototype.format = function (item) {
        __block_BKv82a(20);
        return __expression_wNbCNj(129), __extro_QLWdfI(130, __intro_LerOvg(130, path).join(__extro_QLWdfI(131, __intro_LerOvg(131, item).get(this.paths.path)), __extro_QLWdfI(132, __intro_LerOvg(132, item).get(this.paths.filename))));
    };
}
{
    __statement_cwld1d(133);
    localfiles.prototype.isModified = function (item) {
        __block_BKv82a(21);
        return __expression_wNbCNj(134), __extro_QLWdfI(135, __intro_LerOvg(135, item).isModified(this.paths.path));
    };
}
{
    __statement_cwld1d(136);
    localfiles.prototype.validateInput = function (data) {
        __block_BKv82a(22);
        return __expression_wNbCNj(137), true;
    };
}
{
    __statement_cwld1d(138);
    localfiles.prototype.updateItem = function (item, data) {
        __block_BKv82a(23);
    };
}
{
    __statement_cwld1d(139);
    localfiles.prototype.uploadFiles = function (item, files, update, callback) {
        __block_BKv82a(24);
        {
            __statement_cwld1d(140);
            var field = this;
        }
        {
            __statement_cwld1d(141);
            var fileDatas = [];
        }
        {
            __statement_cwld1d(142);
            __extro_QLWdfI(143, __intro_LerOvg(143, _).each(files, function (file) {
                __block_BKv82a(25);
                {
                    __statement_cwld1d(144);
                    var prefix = field.options.datePrefix ? (__expression_wNbCNj(145), (__expression_wNbCNj(147), __extro_QLWdfI(148, __intro_LerOvg(148, (__expression_wNbCNj(149), moment())).format(field.options.datePrefix)) + '-')) : (__expression_wNbCNj(146), ''), name = (__expression_wNbCNj(150), (__expression_wNbCNj(151), prefix) + file.name);
                }
                if (__expression_wNbCNj(152), field.options.allowedTypes && (__expression_wNbCNj(153), !__extro_QLWdfI(154, __intro_LerOvg(154, _).contains(field.options.allowedTypes, file.type)))) {
                    __block_BKv82a(26);
                    return __expression_wNbCNj(155), (__expression_wNbCNj(156), callback(new Error((__expression_wNbCNj(157), 'Unsupported File Type: ' + file.type))));
                }
                if (__expression_wNbCNj(158), 'function' === (__expression_wNbCNj(159), typeof update)) {
                    __block_BKv82a(27);
                    {
                        __statement_cwld1d(160);
                        callback = update;
                    }
                    {
                        __statement_cwld1d(161);
                        update = false;
                    }
                }
                {
                    __statement_cwld1d(162);
                    var doMove = function (callback) {
                        __block_BKv82a(28);
                        if (__expression_wNbCNj(163), 'function' === (__expression_wNbCNj(164), typeof field.options.filename)) {
                            __block_BKv82a(29);
                            {
                                __statement_cwld1d(165);
                                name = __extro_QLWdfI(166, __intro_LerOvg(166, field.options).filename(item, name));
                            }
                        }
                        {
                            __statement_cwld1d(167);
                            __extro_QLWdfI(168, __intro_LerOvg(168, fs).rename(file.path, __extro_QLWdfI(169, __intro_LerOvg(169, path).join(field.options.dest, name)), function (err) {
                                __block_BKv82a(30);
                                if (__expression_wNbCNj(170), err) {
                                    __block_BKv82a(31);
                                    return __expression_wNbCNj(171), (__expression_wNbCNj(172), callback(err));
                                }
                                {
                                    __statement_cwld1d(173);
                                    var fileData = {
                                            filename: name,
                                            path: field.options.dest,
                                            size: file.size,
                                            filetype: file.type
                                        };
                                }
                                if (__expression_wNbCNj(174), update) {
                                    __block_BKv82a(32);
                                    {
                                        __statement_cwld1d(175);
                                        __extro_QLWdfI(176, __intro_LerOvg(176, item).set(field.path, fileData));
                                    }
                                } else {
                                    __block_BKv82a(33);
                                    {
                                        __statement_cwld1d(177);
                                        __extro_QLWdfI(178, __intro_LerOvg(178, __extro_QLWdfI(179, __intro_LerOvg(179, item).get(field.path))).push(fileData));
                                    }
                                }
                                {
                                    __statement_cwld1d(180);
                                    __extro_QLWdfI(181, __intro_LerOvg(181, fileDatas).push(fileData));
                                }
                                {
                                    __statement_cwld1d(182);
                                    __expression_wNbCNj(183), callback(null, fileData);
                                }
                            }));
                        }
                    };
                }
                {
                    __statement_cwld1d(184);
                    __extro_QLWdfI(185, __intro_LerOvg(185, async).eachSeries(this._pre.move, function (fn, next) {
                        __block_BKv82a(34);
                        {
                            __statement_cwld1d(186);
                            __expression_wNbCNj(187), fn(item, file, next);
                        }
                    }, function (err) {
                        __block_BKv82a(35);
                        if (__expression_wNbCNj(188), err) {
                            __block_BKv82a(36);
                            return __expression_wNbCNj(189), (__expression_wNbCNj(190), callback(err));
                        }
                        {
                            __statement_cwld1d(191);
                            __expression_wNbCNj(192), doMove(function (err, fileData) {
                                __block_BKv82a(37);
                                if (__expression_wNbCNj(193), err) {
                                    __block_BKv82a(38);
                                    return __expression_wNbCNj(194), (__expression_wNbCNj(195), callback(err));
                                }
                                {
                                    __statement_cwld1d(196);
                                    __extro_QLWdfI(197, __intro_LerOvg(197, async).eachSeries(field._post.move, function (fn, next) {
                                        __block_BKv82a(39);
                                        {
                                            __statement_cwld1d(198);
                                            __expression_wNbCNj(199), fn(item, file, fileData, next);
                                        }
                                    }, function (err) {
                                        __block_BKv82a(40);
                                        if (__expression_wNbCNj(200), err) {
                                            __block_BKv82a(41);
                                            return __expression_wNbCNj(201), (__expression_wNbCNj(202), callback(err));
                                        }
                                        if (__expression_wNbCNj(203), fileDatas.length === files.length) {
                                            __block_BKv82a(42);
                                            {
                                                __statement_cwld1d(204);
                                                __expression_wNbCNj(205), callback(null, fileDatas);
                                            }
                                        }
                                    }));
                                }
                            });
                        }
                    }));
                }
            }, this));
        }
    };
}
{
    __statement_cwld1d(206);
    localfiles.prototype.getRequestHandler = function (item, req, paths, callback) {
        __block_BKv82a(43);
        {
            __statement_cwld1d(207);
            var field = this;
        }
        if (__extro_QLWdfI(208, __intro_LerOvg(208, utils).isFunction(paths))) {
            __block_BKv82a(44);
            {
                __statement_cwld1d(209);
                callback = paths;
            }
            {
                __statement_cwld1d(210);
                paths = field.paths;
            }
        } else if (__expression_wNbCNj(211), !(__expression_wNbCNj(212), paths)) {
            __block_BKv82a(45);
            {
                __statement_cwld1d(213);
                paths = field.paths;
            }
        }
        {
            __statement_cwld1d(214);
            callback = (__expression_wNbCNj(215), (__expression_wNbCNj(216), callback) || function () {
                __block_BKv82a(46);
            });
        }
        return __expression_wNbCNj(217), function () {
            __block_BKv82a(47);
            if (req.body[paths.order]) {
                __block_BKv82a(48);
                {
                    __statement_cwld1d(218);
                    var files = __extro_QLWdfI(219, __intro_LerOvg(219, item).get(field.path)), newOrder = __extro_QLWdfI(220, __intro_LerOvg(220, req.body[paths.order]).split(','));
                }
                {
                    __statement_cwld1d(221);
                    __extro_QLWdfI(222, __intro_LerOvg(222, files).sort(function (a, b) {
                        __block_BKv82a(49);
                        return __expression_wNbCNj(223), (__expression_wNbCNj(226), __extro_QLWdfI(227, __intro_LerOvg(227, newOrder).indexOf(__extro_QLWdfI(228, __intro_LerOvg(228, a._id).toString()))) > __extro_QLWdfI(229, __intro_LerOvg(229, newOrder).indexOf(__extro_QLWdfI(230, __intro_LerOvg(230, b._id).toString())))) ? (__expression_wNbCNj(224), 1) : (__expression_wNbCNj(225), (__expression_wNbCNj(231), -1));
                    }));
                }
            }
            if (__expression_wNbCNj(232), req.body && req.body[paths.action]) {
                __block_BKv82a(50);
                {
                    __statement_cwld1d(233);
                    var actions = __extro_QLWdfI(234, __intro_LerOvg(234, req.body[paths.action]).split('|'));
                }
                {
                    __statement_cwld1d(235);
                    __extro_QLWdfI(236, __intro_LerOvg(236, actions).forEach(function (action) {
                        __block_BKv82a(51);
                        {
                            __statement_cwld1d(237);
                            action = __extro_QLWdfI(238, __intro_LerOvg(238, action).split(':'));
                        }
                        {
                            __statement_cwld1d(239);
                            var method = action[0], ids = action[1];
                        }
                        if (__expression_wNbCNj(240), (__expression_wNbCNj(241), !__extro_QLWdfI(242, __intro_LerOvg(242, /^(delete|reset)$/).test(method))) || (__expression_wNbCNj(243), !(__expression_wNbCNj(244), ids))) {
                            __block_BKv82a(52);
                            return __expression_wNbCNj(245);
                        }
                        {
                            __statement_cwld1d(246);
                            __extro_QLWdfI(247, __intro_LerOvg(247, __extro_QLWdfI(248, __intro_LerOvg(248, ids).split(','))).forEach(function (id) {
                                __block_BKv82a(53);
                                {
                                    __statement_cwld1d(249);
                                    __extro_QLWdfI(250, __intro_LerOvg(250, field).apply(item, action));
                                }
                            }));
                        }
                    }));
                }
            }
            if (__expression_wNbCNj(251), (__expression_wNbCNj(252), req.files && req.files[paths.upload]) && (__expression_wNbCNj(253), req.files[paths.upload].length > 0)) {
                __block_BKv82a(54);
                return __expression_wNbCNj(254), __extro_QLWdfI(255, __intro_LerOvg(255, field).uploadFiles(item, req.files[paths.upload], false, callback));
            }
            return __expression_wNbCNj(256), (__expression_wNbCNj(257), callback());
        };
    };
}
{
    __statement_cwld1d(258);
    localfiles.prototype.handleRequest = function (item, req, paths, callback) {
        __block_BKv82a(55);
        {
            __statement_cwld1d(259);
            __expression_wNbCNj(260), __extro_QLWdfI(261, __intro_LerOvg(261, this).getRequestHandler(item, req, paths, callback))();
        }
    };
}
{
    __statement_cwld1d(262);
    exports = module.exports = localfiles;
}
