// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_asyncdi.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_e6T_CF, __expression_J0PCzI, __block_veLrTz;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_e6T_CF = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/asyncdi.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_J0PCzI = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/asyncdi.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_veLrTz = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/asyncdi.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_QNp1Mc = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_Crc_WO = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/asyncdi.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_e6T_CF(0);
    var _ = (__expression_J0PCzI(1), require('underscore')), async = (__expression_J0PCzI(2), require('async'));
}
{
    __statement_e6T_CF(3);
    var FN_ARGS = /^function\s*[^\(]*\(\s*([^\)]*)\)/m;
}
{
    __statement_e6T_CF(4);
    exports = module.exports = function (fn) {
        __block_veLrTz(0);
        return __expression_J0PCzI(5), new Wrapper(fn);
    };
}
{
    __statement_e6T_CF(6);
    var Wrapper = exports.Wrapper = function Wrapper(fn, context) {
            __block_veLrTz(1);
            if (__expression_J0PCzI(7), !(__expression_J0PCzI(8), this instanceof (__expression_J0PCzI(9), Wrapper))) {
                __block_veLrTz(2);
                return __expression_J0PCzI(10), new Wrapper(fn, context);
            }
            if (__expression_J0PCzI(11), !__extro_Crc_WO(12, __intro_QNp1Mc(12, _).isFunction(fn))) {
                __block_veLrTz(3);
                {
                    __statement_e6T_CF(13);
                    throw new Error('AsyncDI Wrapper must be initialised with a function');
                }
            }
            {
                __statement_e6T_CF(14);
                this.fn = fn;
            }
            {
                __statement_e6T_CF(15);
                this.deps = __extro_Crc_WO(16, __intro_QNp1Mc(16, __extro_Crc_WO(17, __intro_QNp1Mc(17, __extro_Crc_WO(18, __intro_QNp1Mc(18, __extro_Crc_WO(19, __intro_QNp1Mc(19, fn).toString())).match(FN_ARGS))[1]).split(','))).map(function (i) {
                    __block_veLrTz(4);
                    return __expression_J0PCzI(20), __extro_Crc_WO(21, __intro_QNp1Mc(21, i).trim());
                }));
            }
            {
                __statement_e6T_CF(22);
                this.requires = {};
            }
            {
                __statement_e6T_CF(23);
                __extro_Crc_WO(24, __intro_QNp1Mc(24, this.deps).forEach(function (i) {
                    __block_veLrTz(5);
                    {
                        __statement_e6T_CF(25);
                        this.requires[i] = true;
                    }
                }, this));
            }
            if (__expression_J0PCzI(26), __extro_Crc_WO(27, __intro_QNp1Mc(27, _).last(this.deps)) === 'callback') {
                __block_veLrTz(6);
                {
                    __statement_e6T_CF(28);
                    this.isAsync = true;
                }
                {
                    __statement_e6T_CF(29);
                    __extro_Crc_WO(30, __intro_QNp1Mc(30, this.deps).pop());
                }
            }
            {
                __statement_e6T_CF(31);
                this._context = context;
            }
            {
                __statement_e6T_CF(32);
                this._provides = {};
            }
            {
                __statement_e6T_CF(33);
                this._arguments = [];
            }
        };
}
{
    __statement_e6T_CF(34);
    __extro_Crc_WO(35, __intro_QNp1Mc(35, _).extend(Wrapper.prototype, {
        provides: function (provides) {
            __block_veLrTz(7);
            {
                __statement_e6T_CF(36);
                __extro_Crc_WO(37, __intro_QNp1Mc(37, _).extend(this._provides, provides));
            }
            {
                __statement_e6T_CF(38);
                this._arguments = __extro_Crc_WO(39, __intro_QNp1Mc(39, _).map(this.deps, function (key) {
                    __block_veLrTz(8);
                    return __expression_J0PCzI(40), this._provides[key];
                }, this));
            }
            return __expression_J0PCzI(41), this;
        },
        bind: function (context) {
            __block_veLrTz(9);
            {
                __statement_e6T_CF(42);
                this._context = context;
            }
            return __expression_J0PCzI(43), this;
        },
        call: function (context, callback) {
            __block_veLrTz(10);
            if (__expression_J0PCzI(44), arguments.length === 1) {
                __block_veLrTz(11);
                {
                    __statement_e6T_CF(45);
                    callback = context;
                }
                {
                    __statement_e6T_CF(46);
                    context = this._context;
                }
            }
            if (this.isAsync) {
                __block_veLrTz(12);
                {
                    __statement_e6T_CF(47);
                    var asyncArgs = __extro_Crc_WO(48, __intro_QNp1Mc(48, this._arguments).slice());
                }
                {
                    __statement_e6T_CF(49);
                    __extro_Crc_WO(50, __intro_QNp1Mc(50, asyncArgs).push(callback));
                }
                {
                    __statement_e6T_CF(51);
                    __extro_Crc_WO(52, __intro_QNp1Mc(52, this.fn).apply(context, asyncArgs));
                }
            } else {
                __block_veLrTz(13);
                if (__expression_J0PCzI(53), callback) {
                    __block_veLrTz(14);
                    {
                        __statement_e6T_CF(54);
                        __expression_J0PCzI(55), callback(null, __extro_Crc_WO(56, __intro_QNp1Mc(56, this.fn).apply(context, this._arguments)));
                    }
                } else {
                    __block_veLrTz(15);
                    return __expression_J0PCzI(57), __extro_Crc_WO(58, __intro_QNp1Mc(58, this.fn).apply(context, this._arguments));
                }
            }
        },
        each: function each(arr, callback) {
            __block_veLrTz(16);
            {
                __statement_e6T_CF(59);
                var wrapper = this;
            }
            if (this.isAsync) {
                __block_veLrTz(17);
                return __expression_J0PCzI(60), __extro_Crc_WO(61, __intro_QNp1Mc(61, async).each(arr, function (item, cb) {
                    __block_veLrTz(18);
                    {
                        __statement_e6T_CF(62);
                        __extro_Crc_WO(63, __intro_QNp1Mc(63, wrapper).call(item, cb));
                    }
                }, callback));
            } else {
                __block_veLrTz(19);
                {
                    __statement_e6T_CF(64);
                    __extro_Crc_WO(65, __intro_QNp1Mc(65, arr).each(function (item) {
                        __block_veLrTz(20);
                        {
                            __statement_e6T_CF(66);
                            __extro_Crc_WO(67, __intro_QNp1Mc(67, wrapper).call(item));
                        }
                    }));
                }
                if (__expression_J0PCzI(68), callback) {
                    __block_veLrTz(21);
                    {
                        __statement_e6T_CF(69);
                        __expression_J0PCzI(70), callback();
                    }
                }
            }
        },
        map: function map(arr, callback) {
            __block_veLrTz(22);
            {
                __statement_e6T_CF(71);
                var wrapper = this;
            }
            if (this.isAsync) {
                __block_veLrTz(23);
                {
                    __statement_e6T_CF(72);
                    __extro_Crc_WO(73, __intro_QNp1Mc(73, async).map(arr, function (item, cb) {
                        __block_veLrTz(24);
                        {
                            __statement_e6T_CF(74);
                            __extro_Crc_WO(75, __intro_QNp1Mc(75, wrapper).call(item, cb));
                        }
                    }, callback));
                }
            } else {
                __block_veLrTz(25);
                {
                    __statement_e6T_CF(76);
                    __expression_J0PCzI(77), callback(null, __extro_Crc_WO(78, __intro_QNp1Mc(78, arr).map(function (item) {
                        __block_veLrTz(26);
                        return __expression_J0PCzI(79), __extro_Crc_WO(80, __intro_QNp1Mc(80, wrapper).call(item));
                    })));
                }
            }
        }
    }));
}
