// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_location.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_Vb_Tj$, __expression_m5Db4U, __block_VAv3oa;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_Vb_Tj$ = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/location.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_m5Db4U = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/location.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_VAv3oa = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/location.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_eEByID = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_oK15uW = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/location.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_Vb_Tj$(0);
    var _ = (__expression_m5Db4U(1), require('underscore')), keystone = (__expression_m5Db4U(2), require('../../')), querystring = (__expression_m5Db4U(3), require('querystring')), https = (__expression_m5Db4U(4), require('https')), util = (__expression_m5Db4U(5), require('util')), utils = (__expression_m5Db4U(6), require('keystone-utils')), super_ = (__expression_m5Db4U(7), require('../field'));
}
{
    __statement_Vb_Tj$(8);
    var RADIUS_KM = 6371, RADIUS_MILES = 3959;
}
function location(list, path, options) {
    __block_VAv3oa(0);
    {
        __statement_Vb_Tj$(9);
        this._underscoreMethods = [
            'format',
            'googleLookup',
            'kmFrom',
            'milesFrom'
        ];
    }
    {
        __statement_Vb_Tj$(10);
        this.enableMapsAPI = __extro_oK15uW(13, __intro_eEByID(13, keystone).get('google api key')) ? (__expression_m5Db4U(11), true) : (__expression_m5Db4U(12), false);
    }
    if (__expression_m5Db4U(14), !options.defaults) {
        __block_VAv3oa(1);
        {
            __statement_Vb_Tj$(15);
            options.defaults = {};
        }
    }
    if (options.required) {
        __block_VAv3oa(2);
        if (__extro_oK15uW(16, __intro_eEByID(16, Array).isArray(options.required))) {
            __block_VAv3oa(3);
            {
                __statement_Vb_Tj$(17);
                this.requiredPaths = options.required;
            }
        } else if (__expression_m5Db4U(18), 'string' === (__expression_m5Db4U(19), typeof options.required)) {
            __block_VAv3oa(4);
            {
                __statement_Vb_Tj$(20);
                this.requiredPaths = __extro_oK15uW(21, __intro_eEByID(21, __extro_oK15uW(22, __intro_eEByID(22, options.required).replace(/,/g, ' '))).split(/\s+/));
            }
        }
        {
            __statement_Vb_Tj$(23);
            options.required = true;
        }
    }
    if (__expression_m5Db4U(24), !this.requiredPaths) {
        __block_VAv3oa(5);
        {
            __statement_Vb_Tj$(25);
            this.requiredPaths = [
                'street1',
                'suburb'
            ];
        }
    }
    {
        __statement_Vb_Tj$(26);
        __extro_oK15uW(27, __intro_eEByID(27, location.super_).call(this, list, path, options));
    }
}
{
    __statement_Vb_Tj$(28);
    __extro_oK15uW(29, __intro_eEByID(29, util).inherits(location, super_));
}
{
    __statement_Vb_Tj$(30);
    location.prototype.addToSchema = function () {
        __block_VAv3oa(6);
        {
            __statement_Vb_Tj$(31);
            var field = this, schema = this.list.schema, options = this.options;
        }
        {
            __statement_Vb_Tj$(32);
            var paths = this.paths = {
                    number: __extro_oK15uW(33, __intro_eEByID(33, this._path).append('.number')),
                    name: __extro_oK15uW(34, __intro_eEByID(34, this._path).append('.name')),
                    street1: __extro_oK15uW(35, __intro_eEByID(35, this._path).append('.street1')),
                    street2: __extro_oK15uW(36, __intro_eEByID(36, this._path).append('.street2')),
                    suburb: __extro_oK15uW(37, __intro_eEByID(37, this._path).append('.suburb')),
                    state: __extro_oK15uW(38, __intro_eEByID(38, this._path).append('.state')),
                    postcode: __extro_oK15uW(39, __intro_eEByID(39, this._path).append('.postcode')),
                    country: __extro_oK15uW(40, __intro_eEByID(40, this._path).append('.country')),
                    geo: __extro_oK15uW(41, __intro_eEByID(41, this._path).append('.geo')),
                    geo_lat: __extro_oK15uW(42, __intro_eEByID(42, this._path).append('.geo_lat')),
                    geo_lng: __extro_oK15uW(43, __intro_eEByID(43, this._path).append('.geo_lng')),
                    serialised: __extro_oK15uW(44, __intro_eEByID(44, this._path).append('.serialised')),
                    improve: __extro_oK15uW(45, __intro_eEByID(45, this._path).append('_improve')),
                    overwrite: __extro_oK15uW(46, __intro_eEByID(46, this._path).append('_improve_overwrite'))
                };
        }
        {
            __statement_Vb_Tj$(47);
            var getFieldDef = function (type, key) {
                __block_VAv3oa(7);
                {
                    __statement_Vb_Tj$(48);
                    var def = {
                            type: type
                        };
                }
                if (options.defaults[key]) {
                    __block_VAv3oa(8);
                    {
                        __statement_Vb_Tj$(49);
                        def.default = options.defaults[key];
                    }
                }
                return __expression_m5Db4U(50), def;
            };
        }
        {
            __statement_Vb_Tj$(51);
            schema.nested[this.path] = true;
        }
        {
            __statement_Vb_Tj$(52);
            __extro_oK15uW(53, __intro_eEByID(53, schema).add({
                number: (__expression_m5Db4U(54), getFieldDef(String, 'number')),
                name: (__expression_m5Db4U(55), getFieldDef(String, 'name')),
                street1: (__expression_m5Db4U(56), getFieldDef(String, 'street1')),
                street2: (__expression_m5Db4U(57), getFieldDef(String, 'street2')),
                street3: (__expression_m5Db4U(58), getFieldDef(String, 'street3')),
                suburb: (__expression_m5Db4U(59), getFieldDef(String, 'suburb')),
                state: (__expression_m5Db4U(60), getFieldDef(String, 'state')),
                postcode: (__expression_m5Db4U(61), getFieldDef(String, 'postcode')),
                country: (__expression_m5Db4U(62), getFieldDef(String, 'country')),
                geo: {
                    type: [
                        Number
                    ],
                    index: '2dsphere'
                }
            }, (__expression_m5Db4U(63), this.path + '.')));
        }
        {
            __statement_Vb_Tj$(64);
            __extro_oK15uW(65, __intro_eEByID(65, __extro_oK15uW(66, __intro_eEByID(66, schema).virtual(paths.serialised))).get(function () {
                __block_VAv3oa(9);
                return __expression_m5Db4U(67), __extro_oK15uW(68, __intro_eEByID(68, __extro_oK15uW(69, __intro_eEByID(69, _).compact([
                    __extro_oK15uW(70, __intro_eEByID(70, this).get(paths.number)),
                    __extro_oK15uW(71, __intro_eEByID(71, this).get(paths.name)),
                    __extro_oK15uW(72, __intro_eEByID(72, this).get(paths.street1)),
                    __extro_oK15uW(73, __intro_eEByID(73, this).get(paths.street2)),
                    __extro_oK15uW(74, __intro_eEByID(74, this).get(paths.suburb)),
                    __extro_oK15uW(75, __intro_eEByID(75, this).get(paths.state)),
                    __extro_oK15uW(76, __intro_eEByID(76, this).get(paths.postcode)),
                    __extro_oK15uW(77, __intro_eEByID(77, this).get(paths.country))
                ]))).join(', '));
            }));
        }
        {
            __statement_Vb_Tj$(78);
            __extro_oK15uW(79, __intro_eEByID(79, schema).pre('save', function (next) {
                __block_VAv3oa(10);
                {
                    __statement_Vb_Tj$(80);
                    var obj = __extro_oK15uW(81, __intro_eEByID(81, field._path).get(this));
                }
                if (__expression_m5Db4U(82), __extro_oK15uW(83, __intro_eEByID(83, Array).isArray(obj.geo)) && (__expression_m5Db4U(84), (__expression_m5Db4U(85), obj.geo.length !== 2) || (__expression_m5Db4U(86), (__expression_m5Db4U(87), obj.geo[0] === null) && (__expression_m5Db4U(88), obj.geo[1] === null)))) {
                    __block_VAv3oa(11);
                    {
                        __statement_Vb_Tj$(89);
                        obj.geo = undefined;
                    }
                }
                {
                    __statement_Vb_Tj$(90);
                    __expression_m5Db4U(91), next();
                }
            }));
        }
        {
            __statement_Vb_Tj$(92);
            __extro_oK15uW(93, __intro_eEByID(93, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_Vb_Tj$(94);
    location.prototype.format = function (item, values, delimiter) {
        __block_VAv3oa(12);
        if (__expression_m5Db4U(95), !(__expression_m5Db4U(96), values)) {
            __block_VAv3oa(13);
            return __expression_m5Db4U(97), __extro_oK15uW(98, __intro_eEByID(98, item).get(this.paths.serialised));
        }
        {
            __statement_Vb_Tj$(99);
            var paths = this.paths;
        }
        {
            __statement_Vb_Tj$(100);
            values = __extro_oK15uW(101, __intro_eEByID(101, __extro_oK15uW(102, __intro_eEByID(102, values).split(' '))).map(function (i) {
                __block_VAv3oa(14);
                return __expression_m5Db4U(103), __extro_oK15uW(104, __intro_eEByID(104, item).get(paths[i]));
            }));
        }
        return __expression_m5Db4U(105), __extro_oK15uW(106, __intro_eEByID(106, __extro_oK15uW(107, __intro_eEByID(107, _).compact(values))).join((__expression_m5Db4U(108), (__expression_m5Db4U(109), delimiter) || ', ')));
    };
}
{
    __statement_Vb_Tj$(110);
    location.prototype.isModified = function (item) {
        __block_VAv3oa(15);
        return __expression_m5Db4U(111), (__expression_m5Db4U(112), (__expression_m5Db4U(113), (__expression_m5Db4U(114), (__expression_m5Db4U(115), (__expression_m5Db4U(116), (__expression_m5Db4U(117), (__expression_m5Db4U(118), (__expression_m5Db4U(119), __extro_oK15uW(120, __intro_eEByID(120, item).isModified(this.paths.number)) || __extro_oK15uW(121, __intro_eEByID(121, item).isModified(this.paths.name))) || __extro_oK15uW(122, __intro_eEByID(122, item).isModified(this.paths.street1))) || __extro_oK15uW(123, __intro_eEByID(123, item).isModified(this.paths.street2))) || __extro_oK15uW(124, __intro_eEByID(124, item).isModified(this.paths.suburb))) || __extro_oK15uW(125, __intro_eEByID(125, item).isModified(this.paths.state))) || __extro_oK15uW(126, __intro_eEByID(126, item).isModified(this.paths.postcode))) || __extro_oK15uW(127, __intro_eEByID(127, item).isModified(this.paths.country))) || __extro_oK15uW(128, __intro_eEByID(128, item).isModified(this.paths.geo)));
    };
}
{
    __statement_Vb_Tj$(129);
    location.prototype.validateInput = function (data, required, item) {
        __block_VAv3oa(16);
        if (__expression_m5Db4U(130), !(__expression_m5Db4U(131), required)) {
            __block_VAv3oa(17);
            return __expression_m5Db4U(132), true;
        }
        {
            __statement_Vb_Tj$(133);
            var paths = this.paths, nested = __extro_oK15uW(134, __intro_eEByID(134, this._path).get(data)), values = (__expression_m5Db4U(135), (__expression_m5Db4U(136), nested) || (__expression_m5Db4U(137), data)), valid = true;
        }
        {
            __statement_Vb_Tj$(138);
            __extro_oK15uW(139, __intro_eEByID(139, this.requiredPaths).forEach(function (path) {
                __block_VAv3oa(18);
                if (__expression_m5Db4U(140), nested) {
                    __block_VAv3oa(19);
                    if (__expression_m5Db4U(141), (__expression_m5Db4U(142), (__expression_m5Db4U(143), !(__expression_m5Db4U(144), ((__expression_m5Db4U(145), path) in (__expression_m5Db4U(146), values)))) && (__expression_m5Db4U(147), item)) && __extro_oK15uW(148, __intro_eEByID(148, item).get(paths[path]))) {
                        __block_VAv3oa(20);
                        return __expression_m5Db4U(149);
                    }
                    if (__expression_m5Db4U(150), !values[path]) {
                        __block_VAv3oa(21);
                        {
                            __statement_Vb_Tj$(151);
                            valid = false;
                        }
                    }
                } else {
                    __block_VAv3oa(22);
                    if (__expression_m5Db4U(152), (__expression_m5Db4U(153), (__expression_m5Db4U(154), !(__expression_m5Db4U(155), (paths[path] in (__expression_m5Db4U(156), values)))) && (__expression_m5Db4U(157), item)) && __extro_oK15uW(158, __intro_eEByID(158, item).get(paths[path]))) {
                        __block_VAv3oa(23);
                        return __expression_m5Db4U(159);
                    }
                    if (__expression_m5Db4U(160), !values[paths[path]]) {
                        __block_VAv3oa(24);
                        {
                            __statement_Vb_Tj$(161);
                            valid = false;
                        }
                    }
                }
            }));
        }
        return __expression_m5Db4U(162), valid;
    };
}
{
    __statement_Vb_Tj$(163);
    location.prototype.updateItem = function (item, data) {
        __block_VAv3oa(25);
        {
            __statement_Vb_Tj$(164);
            var paths = this.paths, fieldKeys = [
                    'number',
                    'name',
                    'street1',
                    'street2',
                    'suburb',
                    'state',
                    'postcode',
                    'country'
                ], geoKeys = [
                    'geo',
                    'geo_lat',
                    'geo_lng'
                ], valueKeys = __extro_oK15uW(165, __intro_eEByID(165, fieldKeys).concat(geoKeys)), valuePaths = valueKeys, values = __extro_oK15uW(166, __intro_eEByID(166, this._path).get(data));
        }
        if (__expression_m5Db4U(167), !(__expression_m5Db4U(168), values)) {
            __block_VAv3oa(26);
            {
                __statement_Vb_Tj$(169);
                valuePaths = __extro_oK15uW(170, __intro_eEByID(170, valueKeys).map(function (i) {
                    __block_VAv3oa(27);
                    return __expression_m5Db4U(171), paths[i];
                }));
            }
            {
                __statement_Vb_Tj$(172);
                values = __extro_oK15uW(173, __intro_eEByID(173, _).pick(data, valuePaths));
            }
        }
        {
            __statement_Vb_Tj$(174);
            valuePaths = __extro_oK15uW(175, __intro_eEByID(175, _).object(valueKeys, valuePaths));
        }
        {
            __statement_Vb_Tj$(176);
            var setValue = function (key) {
                __block_VAv3oa(28);
                if (__expression_m5Db4U(177), (__expression_m5Db4U(178), (valuePaths[key] in (__expression_m5Db4U(179), values))) && (__expression_m5Db4U(180), values[valuePaths[key]] !== __extro_oK15uW(181, __intro_eEByID(181, item).get(paths[key])))) {
                    __block_VAv3oa(29);
                    {
                        __statement_Vb_Tj$(182);
                        __extro_oK15uW(183, __intro_eEByID(183, item).set(paths[key], (__expression_m5Db4U(184), values[valuePaths[key]] || null)));
                    }
                }
            };
        }
        {
            __statement_Vb_Tj$(185);
            __extro_oK15uW(186, __intro_eEByID(186, _).each(fieldKeys, setValue));
        }
        if (__expression_m5Db4U(187), (valuePaths.geo in (__expression_m5Db4U(188), values))) {
            __block_VAv3oa(30);
            {
                __statement_Vb_Tj$(189);
                var oldGeo = (__expression_m5Db4U(190), __extro_oK15uW(191, __intro_eEByID(191, item).get(paths.geo)) || []), newGeo = values[valuePaths.geo];
            }
            if (__expression_m5Db4U(192), (__expression_m5Db4U(193), !__extro_oK15uW(194, __intro_eEByID(194, Array).isArray(newGeo))) || (__expression_m5Db4U(195), newGeo.length !== 2)) {
                __block_VAv3oa(31);
                {
                    __statement_Vb_Tj$(196);
                    newGeo = [];
                }
            }
            if (__expression_m5Db4U(197), (__expression_m5Db4U(198), newGeo[0] !== oldGeo[0]) || (__expression_m5Db4U(199), newGeo[1] !== oldGeo[1])) {
                __block_VAv3oa(32);
                {
                    __statement_Vb_Tj$(200);
                    __extro_oK15uW(201, __intro_eEByID(201, item).set(paths.geo, newGeo));
                }
            }
        } else if (__expression_m5Db4U(202), (__expression_m5Db4U(203), (valuePaths.geo_lat in (__expression_m5Db4U(204), values))) && (__expression_m5Db4U(205), (valuePaths.geo_lng in (__expression_m5Db4U(206), values)))) {
            __block_VAv3oa(33);
            {
                __statement_Vb_Tj$(207);
                var lat = __extro_oK15uW(208, __intro_eEByID(208, utils).number(values[valuePaths.geo_lat])), lng = __extro_oK15uW(209, __intro_eEByID(209, utils).number(values[valuePaths.geo_lng]));
            }
            {
                __statement_Vb_Tj$(210);
                __extro_oK15uW(211, __intro_eEByID(211, item).set(paths.geo, (__expression_m5Db4U(214), (__expression_m5Db4U(215), lat) && (__expression_m5Db4U(216), lng)) ? (__expression_m5Db4U(212), [
                    lng,
                    lat
                ]) : (__expression_m5Db4U(213), undefined)));
            }
        }
    };
}
{
    __statement_Vb_Tj$(217);
    location.prototype.getRequestHandler = function (item, req, paths, callback) {
        __block_VAv3oa(34);
        {
            __statement_Vb_Tj$(218);
            var field = this;
        }
        if (__extro_oK15uW(219, __intro_eEByID(219, utils).isFunction(paths))) {
            __block_VAv3oa(35);
            {
                __statement_Vb_Tj$(220);
                callback = paths;
            }
            {
                __statement_Vb_Tj$(221);
                paths = field.paths;
            }
        } else if (__expression_m5Db4U(222), !(__expression_m5Db4U(223), paths)) {
            __block_VAv3oa(36);
            {
                __statement_Vb_Tj$(224);
                paths = field.paths;
            }
        }
        {
            __statement_Vb_Tj$(225);
            callback = (__expression_m5Db4U(226), (__expression_m5Db4U(227), callback) || function () {
                __block_VAv3oa(37);
            });
        }
        return __expression_m5Db4U(228), function () {
            __block_VAv3oa(38);
            {
                __statement_Vb_Tj$(229);
                var update = req.body[paths.overwrite] ? (__expression_m5Db4U(230), 'overwrite') : (__expression_m5Db4U(231), true);
            }
            if (__expression_m5Db4U(232), req.body && req.body[paths.improve]) {
                __block_VAv3oa(39);
                {
                    __statement_Vb_Tj$(233);
                    __extro_oK15uW(234, __intro_eEByID(234, field).googleLookup(item, false, update, function () {
                        __block_VAv3oa(40);
                        {
                            __statement_Vb_Tj$(235);
                            __expression_m5Db4U(236), callback();
                        }
                    }));
                }
            } else {
                __block_VAv3oa(41);
                {
                    __statement_Vb_Tj$(237);
                    __expression_m5Db4U(238), callback();
                }
            }
        };
    };
}
{
    __statement_Vb_Tj$(239);
    location.prototype.handleRequest = function (item, req, paths, callback) {
        __block_VAv3oa(42);
        {
            __statement_Vb_Tj$(240);
            __expression_m5Db4U(241), __extro_oK15uW(242, __intro_eEByID(242, this).getRequestHandler(item, req, paths, callback))();
        }
    };
}
function doGoogleGeocodeRequest(address, region, callback) {
    __block_VAv3oa(43);
    {
        __statement_Vb_Tj$(243);
        var options = {
                sensor: false,
                language: 'en',
                address: address
            };
    }
    if (__expression_m5Db4U(244), (__expression_m5Db4U(245), arguments.length === 2) && __extro_oK15uW(246, __intro_eEByID(246, _).isFunction(region))) {
        __block_VAv3oa(44);
        {
            __statement_Vb_Tj$(247);
            callback = region;
        }
        {
            __statement_Vb_Tj$(248);
            region = null;
        }
    }
    if (__expression_m5Db4U(249), region) {
        __block_VAv3oa(45);
        {
            __statement_Vb_Tj$(250);
            options.region = region;
        }
    }
    {
        __statement_Vb_Tj$(251);
        var endpoint = (__expression_m5Db4U(252), 'https://maps.googleapis.com/maps/api/geocode/json?' + __extro_oK15uW(253, __intro_eEByID(253, querystring).stringify(options)));
    }
    {
        __statement_Vb_Tj$(254);
        __extro_oK15uW(255, __intro_eEByID(255, __extro_oK15uW(256, __intro_eEByID(256, https).get(endpoint, function (res) {
            __block_VAv3oa(46);
            {
                __statement_Vb_Tj$(257);
                var data = [];
            }
            {
                __statement_Vb_Tj$(258);
                __extro_oK15uW(259, __intro_eEByID(259, __extro_oK15uW(260, __intro_eEByID(260, res).on('data', function (chunk) {
                    __block_VAv3oa(47);
                    {
                        __statement_Vb_Tj$(261);
                        __extro_oK15uW(262, __intro_eEByID(262, data).push(chunk));
                    }
                }))).on('end', function () {
                    __block_VAv3oa(48);
                    {
                        __statement_Vb_Tj$(263);
                        var dataBuff = __extro_oK15uW(264, __intro_eEByID(264, __extro_oK15uW(265, __intro_eEByID(265, data).join(''))).trim());
                    }
                    {
                        __statement_Vb_Tj$(266);
                        var result;
                    }
                    try {
                        __block_VAv3oa(49);
                        {
                            __statement_Vb_Tj$(267);
                            result = __extro_oK15uW(268, __intro_eEByID(268, JSON).parse(dataBuff));
                        }
                    } catch (exp) {
                        __block_VAv3oa(50);
                        {
                            __statement_Vb_Tj$(269);
                            result = {
                                'status_code': 500,
                                'status_text': 'JSON Parse Failed',
                                'status': 'UNKNOWN_ERROR'
                            };
                        }
                    }
                    {
                        __statement_Vb_Tj$(270);
                        __expression_m5Db4U(271), callback(null, result);
                    }
                }));
            }
        }))).on('error', function (err) {
            __block_VAv3oa(51);
            {
                __statement_Vb_Tj$(272);
                __expression_m5Db4U(273), callback(err);
            }
        }));
    }
}
{
    __statement_Vb_Tj$(274);
    location.prototype.googleLookup = function (item, region, update, callback) {
        __block_VAv3oa(52);
        if (__extro_oK15uW(275, __intro_eEByID(275, _).isFunction(update))) {
            __block_VAv3oa(53);
            {
                __statement_Vb_Tj$(276);
                callback = update;
            }
            {
                __statement_Vb_Tj$(277);
                update = false;
            }
        }
        {
            __statement_Vb_Tj$(278);
            var field = this, stored = __extro_oK15uW(279, __intro_eEByID(279, item).get(this.path)), address = __extro_oK15uW(280, __intro_eEByID(280, item).get(this.paths.serialised));
        }
        if (__expression_m5Db4U(281), address.length === 0) {
            __block_VAv3oa(54);
            return __expression_m5Db4U(282), (__expression_m5Db4U(283), callback({
                'status_code': 500,
                'status_text': 'No address to geocode',
                'status': 'NO_ADDRESS'
            }));
        }
        {
            __statement_Vb_Tj$(284);
            __expression_m5Db4U(285), doGoogleGeocodeRequest(address, (__expression_m5Db4U(286), (__expression_m5Db4U(287), region) || __extro_oK15uW(288, __intro_eEByID(288, keystone).get('default region'))), function (err, geocode) {
                __block_VAv3oa(55);
                if (__expression_m5Db4U(289), (__expression_m5Db4U(290), err) || (__expression_m5Db4U(291), geocode.status !== 'OK')) {
                    __block_VAv3oa(56);
                    return __expression_m5Db4U(292), (__expression_m5Db4U(293), callback(err));
                }
                {
                    __statement_Vb_Tj$(294);
                    var result = geocode.results[0];
                }
                {
                    __statement_Vb_Tj$(295);
                    var location = {};
                }
                {
                    __statement_Vb_Tj$(296);
                    __extro_oK15uW(297, __intro_eEByID(297, _).each(result.address_components, function (val) {
                        __block_VAv3oa(57);
                        if (__expression_m5Db4U(298), __extro_oK15uW(299, __intro_eEByID(299, _).indexOf(val.types, 'street_number')) >= 0) {
                            __block_VAv3oa(58);
                            {
                                __statement_Vb_Tj$(300);
                                location.street1 = (__expression_m5Db4U(301), location.street1 || []);
                            }
                            {
                                __statement_Vb_Tj$(302);
                                __extro_oK15uW(303, __intro_eEByID(303, location.street1).push(val.long_name));
                            }
                        }
                        if (__expression_m5Db4U(304), __extro_oK15uW(305, __intro_eEByID(305, _).indexOf(val.types, 'route')) >= 0) {
                            __block_VAv3oa(59);
                            {
                                __statement_Vb_Tj$(306);
                                location.street1 = (__expression_m5Db4U(307), location.street1 || []);
                            }
                            {
                                __statement_Vb_Tj$(308);
                                __extro_oK15uW(309, __intro_eEByID(309, location.street1).push(val.short_name));
                            }
                        }
                        if (__expression_m5Db4U(310), (__expression_m5Db4U(311), __extro_oK15uW(312, __intro_eEByID(312, _).indexOf(val.types, 'locality')) >= 0) && (__expression_m5Db4U(313), !location.suburb)) {
                            __block_VAv3oa(60);
                            {
                                __statement_Vb_Tj$(314);
                                location.suburb = val.long_name;
                            }
                        }
                        if (__expression_m5Db4U(315), __extro_oK15uW(316, __intro_eEByID(316, _).indexOf(val.types, 'administrative_area_level_1')) >= 0) {
                            __block_VAv3oa(61);
                            {
                                __statement_Vb_Tj$(317);
                                location.state = val.short_name;
                            }
                        }
                        if (__expression_m5Db4U(318), __extro_oK15uW(319, __intro_eEByID(319, _).indexOf(val.types, 'country')) >= 0) {
                            __block_VAv3oa(62);
                            {
                                __statement_Vb_Tj$(320);
                                location.country = val.long_name;
                            }
                        }
                        if (__expression_m5Db4U(321), __extro_oK15uW(322, __intro_eEByID(322, _).indexOf(val.types, 'postal_code')) >= 0) {
                            __block_VAv3oa(63);
                            {
                                __statement_Vb_Tj$(323);
                                location.postcode = val.short_name;
                            }
                        }
                    }));
                }
                if (__extro_oK15uW(324, __intro_eEByID(324, Array).isArray(location.street1))) {
                    __block_VAv3oa(64);
                    {
                        __statement_Vb_Tj$(325);
                        location.street1 = __extro_oK15uW(326, __intro_eEByID(326, location.street1).join(' '));
                    }
                }
                {
                    __statement_Vb_Tj$(327);
                    location.geo = [
                        result.geometry.location.lng,
                        result.geometry.location.lat
                    ];
                }
                if (__expression_m5Db4U(328), (__expression_m5Db4U(329), update) === 'overwrite') {
                    __block_VAv3oa(65);
                    {
                        __statement_Vb_Tj$(330);
                        __extro_oK15uW(331, __intro_eEByID(331, item).set(field.path, location));
                    }
                } else if (__expression_m5Db4U(332), update) {
                    __block_VAv3oa(66);
                    {
                        __statement_Vb_Tj$(333);
                        __extro_oK15uW(334, __intro_eEByID(334, _).each(location, function (value, key) {
                            __block_VAv3oa(67);
                            if (__expression_m5Db4U(335), (__expression_m5Db4U(336), key) === 'geo') {
                                __block_VAv3oa(68);
                                return __expression_m5Db4U(337);
                            }
                            if (__expression_m5Db4U(338), !stored[key]) {
                                __block_VAv3oa(69);
                                {
                                    __statement_Vb_Tj$(339);
                                    __extro_oK15uW(340, __intro_eEByID(340, item).set(field.paths[key], value));
                                }
                            }
                        }));
                    }
                    if (__expression_m5Db4U(341), (__expression_m5Db4U(342), (__expression_m5Db4U(343), !__extro_oK15uW(344, __intro_eEByID(344, Array).isArray(stored.geo))) || (__expression_m5Db4U(345), !stored.geo[0])) || (__expression_m5Db4U(346), !stored.geo[1])) {
                        __block_VAv3oa(70);
                        {
                            __statement_Vb_Tj$(347);
                            __extro_oK15uW(348, __intro_eEByID(348, item).set(field.paths.geo, location.geo));
                        }
                    }
                }
                {
                    __statement_Vb_Tj$(349);
                    __expression_m5Db4U(350), callback(null, location, result);
                }
            });
        }
    };
}
function calculateDistance(point1, point2) {
    __block_VAv3oa(71);
    {
        __statement_Vb_Tj$(351);
        var dLng = (__expression_m5Db4U(352), (__expression_m5Db4U(353), (__expression_m5Db4U(354), point2[0] - point1[0]) * Math.PI) / 180);
    }
    {
        __statement_Vb_Tj$(355);
        var dLat = (__expression_m5Db4U(356), (__expression_m5Db4U(357), (__expression_m5Db4U(358), point2[1] - point1[1]) * Math.PI) / 180);
    }
    {
        __statement_Vb_Tj$(359);
        var lat1 = (__expression_m5Db4U(360), (__expression_m5Db4U(361), point1[1] * Math.PI) / 180);
    }
    {
        __statement_Vb_Tj$(362);
        var lat2 = (__expression_m5Db4U(363), (__expression_m5Db4U(364), point2[1] * Math.PI) / 180);
    }
    {
        __statement_Vb_Tj$(365);
        var a = (__expression_m5Db4U(366), (__expression_m5Db4U(367), __extro_oK15uW(368, __intro_eEByID(368, Math).sin((__expression_m5Db4U(369), (__expression_m5Db4U(370), dLat) / 2))) * __extro_oK15uW(371, __intro_eEByID(371, Math).sin((__expression_m5Db4U(372), (__expression_m5Db4U(373), dLat) / 2)))) + (__expression_m5Db4U(374), (__expression_m5Db4U(375), (__expression_m5Db4U(376), __extro_oK15uW(377, __intro_eEByID(377, Math).sin((__expression_m5Db4U(378), (__expression_m5Db4U(379), dLng) / 2))) * __extro_oK15uW(380, __intro_eEByID(380, Math).sin((__expression_m5Db4U(381), (__expression_m5Db4U(382), dLng) / 2)))) * __extro_oK15uW(383, __intro_eEByID(383, Math).cos(lat1))) * __extro_oK15uW(384, __intro_eEByID(384, Math).cos(lat2))));
    }
    {
        __statement_Vb_Tj$(385);
        var c = (__expression_m5Db4U(386), 2 * __extro_oK15uW(387, __intro_eEByID(387, Math).atan2(__extro_oK15uW(388, __intro_eEByID(388, Math).sqrt(a)), __extro_oK15uW(389, __intro_eEByID(389, Math).sqrt((__expression_m5Db4U(390), 1 - (__expression_m5Db4U(391), a)))))));
    }
    return __expression_m5Db4U(392), c;
}
{
    __statement_Vb_Tj$(393);
    location.prototype.kmFrom = function (item, point) {
        __block_VAv3oa(72);
        return __expression_m5Db4U(394), (__expression_m5Db4U(395), (__expression_m5Db4U(396), calculateDistance(__extro_oK15uW(397, __intro_eEByID(397, this).get(this.paths.geo)), point)) * (__expression_m5Db4U(398), RADIUS_KM));
    };
}
{
    __statement_Vb_Tj$(399);
    location.prototype.milesFrom = function (item, point) {
        __block_VAv3oa(73);
        return __expression_m5Db4U(400), (__expression_m5Db4U(401), (__expression_m5Db4U(402), calculateDistance(__extro_oK15uW(403, __intro_eEByID(403, this).get(this.paths.geo)), point)) * (__expression_m5Db4U(404), RADIUS_MILES));
    };
}
{
    __statement_Vb_Tj$(405);
    exports = module.exports = location;
}
