// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_markdown.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_mkBE$q, __expression_kvuk$b, __block_SIJsHj;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_mkBE$q = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/markdown.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_kvuk$b = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/markdown.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_SIJsHj = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/markdown.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_oj0ebX = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_wCA8$w = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/markdown.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_mkBE$q(0);
    var util = (__expression_kvuk$b(1), require('util')), marked = (__expression_kvuk$b(2), require('marked')), super_ = (__expression_kvuk$b(3), require('../field'));
}
function markdown(list, path, options) {
    __block_SIJsHj(0);
    {
        __statement_mkBE$q(4);
        options.nofilter = true;
    }
    {
        __statement_mkBE$q(5);
        this.height = (__expression_kvuk$b(6), options.height || 90);
    }
    {
        __statement_mkBE$q(7);
        __extro_wCA8$w(8, __intro_oj0ebX(8, markdown.super_).call(this, list, path, options));
    }
}
{
    __statement_mkBE$q(9);
    __extro_wCA8$w(10, __intro_oj0ebX(10, util).inherits(markdown, super_));
}
{
    __statement_mkBE$q(11);
    markdown.prototype.addToSchema = function () {
        __block_SIJsHj(1);
        {
            __statement_mkBE$q(12);
            var schema = this.list.schema;
        }
        {
            __statement_mkBE$q(13);
            var paths = this.paths = {
                    md: __extro_wCA8$w(14, __intro_oj0ebX(14, this._path).append('.md')),
                    html: __extro_wCA8$w(15, __intro_oj0ebX(15, this._path).append('.html'))
                };
        }
        {
            __statement_mkBE$q(16);
            var setMarkdown = function (value) {
                __block_SIJsHj(2);
                if (__expression_kvuk$b(17), (__expression_kvuk$b(18), value) === __extro_wCA8$w(19, __intro_oj0ebX(19, this).get(paths.md))) {
                    __block_SIJsHj(3);
                    return __expression_kvuk$b(20), value;
                }
                if (__expression_kvuk$b(21), (__expression_kvuk$b(22), typeof value) === 'string') {
                    __block_SIJsHj(4);
                    {
                        __statement_mkBE$q(23);
                        __extro_wCA8$w(24, __intro_oj0ebX(24, this).set(paths.html, (__expression_kvuk$b(25), marked(value))));
                    }
                    return __expression_kvuk$b(26), value;
                } else {
                    __block_SIJsHj(5);
                    {
                        __statement_mkBE$q(27);
                        __extro_wCA8$w(28, __intro_oj0ebX(28, this).set(paths.html, undefined));
                    }
                    return __expression_kvuk$b(29), undefined;
                }
            };
        }
        {
            __statement_mkBE$q(30);
            schema.nested[this.path] = true;
        }
        {
            __statement_mkBE$q(31);
            __extro_wCA8$w(32, __intro_oj0ebX(32, schema).add({
                html: {
                    type: String
                },
                md: {
                    type: String,
                    set: setMarkdown
                }
            }, (__expression_kvuk$b(33), this.path + '.')));
        }
        {
            __statement_mkBE$q(34);
            __extro_wCA8$w(35, __intro_oj0ebX(35, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_mkBE$q(36);
    markdown.prototype.format = function (item) {
        __block_SIJsHj(6);
        return __expression_kvuk$b(37), __extro_wCA8$w(38, __intro_oj0ebX(38, item).get(this.paths.html));
    };
}
{
    __statement_mkBE$q(39);
    markdown.prototype.validateInput = function (data, required, item) {
        __block_SIJsHj(7);
        if (__expression_kvuk$b(40), (__expression_kvuk$b(41), (__expression_kvuk$b(42), !(__expression_kvuk$b(43), (__expression_kvuk$b(44), (this.path in (__expression_kvuk$b(45), data))) || (__expression_kvuk$b(46), (this.paths.md in (__expression_kvuk$b(47), data))))) && (__expression_kvuk$b(48), item)) && __extro_wCA8$w(49, __intro_oj0ebX(49, item).get(this.paths.md))) {
            __block_SIJsHj(8);
            return __expression_kvuk$b(50), true;
        }
        return __expression_kvuk$b(51), (__expression_kvuk$b(54), (__expression_kvuk$b(55), (__expression_kvuk$b(56), !(__expression_kvuk$b(57), required)) || data[this.path]) || data[this.paths.md]) ? (__expression_kvuk$b(52), true) : (__expression_kvuk$b(53), false);
    };
}
{
    __statement_mkBE$q(58);
    markdown.prototype.isModified = function (item) {
        __block_SIJsHj(9);
        return __expression_kvuk$b(59), __extro_wCA8$w(60, __intro_oj0ebX(60, item).isModified(this.paths.md));
    };
}
{
    __statement_mkBE$q(61);
    markdown.prototype.updateItem = function (item, data) {
        __block_SIJsHj(10);
        if (__expression_kvuk$b(62), (this.path in (__expression_kvuk$b(63), data))) {
            __block_SIJsHj(11);
            {
                __statement_mkBE$q(64);
                __extro_wCA8$w(65, __intro_oj0ebX(65, item).set(this.paths.md, data[this.path]));
            }
        } else if (__expression_kvuk$b(66), (this.paths.md in (__expression_kvuk$b(67), data))) {
            __block_SIJsHj(12);
            {
                __statement_mkBE$q(68);
                __extro_wCA8$w(69, __intro_oj0ebX(69, item).set(this.paths.md, data[this.paths.md]));
            }
        }
    };
}
{
    __statement_mkBE$q(70);
    exports = module.exports = markdown;
}
