// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_embedly.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_lDHrQM, __expression_qSott_, __block_Cbs43u;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_lDHrQM = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/embedly.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_qSott_ = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/embedly.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_Cbs43u = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/embedly.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_tSk0v_ = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_ygGuVj = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/embedly.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_lDHrQM(0);
    var _ = (__expression_qSott_(1), require('underscore')), keystone = (__expression_qSott_(2), require('../../')), util = (__expression_qSott_(3), require('util')), EmbedlyAPI = (__expression_qSott_(4), require('embedly')), super_ = (__expression_qSott_(5), require('../field'));
}
function embedly(list, path, options) {
    __block_Cbs43u(0);
    {
        __statement_lDHrQM(6);
        this._underscoreMethods = [
            'reset'
        ];
    }
    {
        __statement_lDHrQM(7);
        this.fromPath = options.from;
    }
    {
        __statement_lDHrQM(8);
        this.embedlyOptions = (__expression_qSott_(9), options.options || {});
    }
    {
        __statement_lDHrQM(10);
        options.nofilter = true;
    }
    if (__expression_qSott_(11), !__extro_ygGuVj(12, __intro_tSk0v_(12, keystone).get('embedly api key'))) {
        __block_Cbs43u(1);
        {
            __statement_lDHrQM(13);
            throw new Error((__expression_qSott_(14), (__expression_qSott_(15), (__expression_qSott_(16), (__expression_qSott_(17), (__expression_qSott_(18), (__expression_qSott_(19), 'Invalid Configuration\n\n' + 'Embedly fields (') + list.key) + '.') + this.path) + ') require the "embedly api key" option to be set.\n\n') + 'See http://keystonejs.com/docs/configuration/#embedly for more information.\n'));
        }
    }
    if (__expression_qSott_(20), !options.from) {
        __block_Cbs43u(2);
        {
            __statement_lDHrQM(21);
            throw new Error((__expression_qSott_(22), (__expression_qSott_(23), (__expression_qSott_(24), (__expression_qSott_(25), (__expression_qSott_(26), (__expression_qSott_(27), 'Invalid Configuration\n\n' + 'Embedly fields (') + list.key) + '.') + (__expression_qSott_(28), path)) + ') require a fromPath option to be set.\n') + 'See http://keystonejs.com/docs/database/#field_embedly for more information.\n'));
        }
    }
    if (options.initial) {
        __block_Cbs43u(3);
        {
            __statement_lDHrQM(29);
            throw new Error((__expression_qSott_(30), (__expression_qSott_(31), (__expression_qSott_(32), (__expression_qSott_(33), (__expression_qSott_(34), 'Invalid Configuration\n\n' + 'Embedly fields (') + list.key) + '.') + (__expression_qSott_(35), path)) + ') cannot be set as initial fields.\n'));
        }
    }
    {
        __statement_lDHrQM(36);
        __extro_ygGuVj(37, __intro_tSk0v_(37, embedly.super_).call(this, list, path, options));
    }
}
{
    __statement_lDHrQM(38);
    __extro_ygGuVj(39, __intro_tSk0v_(39, util).inherits(embedly, super_));
}
{
    __statement_lDHrQM(40);
    embedly.prototype.addToSchema = function () {
        __block_Cbs43u(4);
        {
            __statement_lDHrQM(41);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_lDHrQM(42);
            this.paths = {
                exists: __extro_ygGuVj(43, __intro_tSk0v_(43, this._path).append('.exists')),
                type: __extro_ygGuVj(44, __intro_tSk0v_(44, this._path).append('.type')),
                title: __extro_ygGuVj(45, __intro_tSk0v_(45, this._path).append('.title')),
                url: __extro_ygGuVj(46, __intro_tSk0v_(46, this._path).append('.url')),
                width: __extro_ygGuVj(47, __intro_tSk0v_(47, this._path).append('.width')),
                height: __extro_ygGuVj(48, __intro_tSk0v_(48, this._path).append('.height')),
                version: __extro_ygGuVj(49, __intro_tSk0v_(49, this._path).append('.version')),
                description: __extro_ygGuVj(50, __intro_tSk0v_(50, this._path).append('.description')),
                html: __extro_ygGuVj(51, __intro_tSk0v_(51, this._path).append('.html')),
                authorName: __extro_ygGuVj(52, __intro_tSk0v_(52, this._path).append('.authorName')),
                authorUrl: __extro_ygGuVj(53, __intro_tSk0v_(53, this._path).append('.authorUrl')),
                providerName: __extro_ygGuVj(54, __intro_tSk0v_(54, this._path).append('.providerName')),
                providerUrl: __extro_ygGuVj(55, __intro_tSk0v_(55, this._path).append('.providerUrl')),
                thumbnailUrl: __extro_ygGuVj(56, __intro_tSk0v_(56, this._path).append('.thumbnailUrl')),
                thumbnailWidth: __extro_ygGuVj(57, __intro_tSk0v_(57, this._path).append('.thumbnailWidth')),
                thumbnailHeight: __extro_ygGuVj(58, __intro_tSk0v_(58, this._path).append('.thumbnailHeight'))
            };
        }
        {
            __statement_lDHrQM(59);
            schema.nested[this.path] = true;
        }
        {
            __statement_lDHrQM(60);
            __extro_ygGuVj(61, __intro_tSk0v_(61, schema).add({
                exists: Boolean,
                type: String,
                title: String,
                url: String,
                width: Number,
                height: Number,
                version: String,
                description: String,
                html: String,
                authorName: String,
                authorUrl: String,
                providerName: String,
                providerUrl: String,
                thumbnailUrl: String,
                thumbnailWidth: Number,
                thumbnailHeight: Number
            }, (__expression_qSott_(62), this.path + '.')));
        }
        {
            __statement_lDHrQM(63);
            __extro_ygGuVj(64, __intro_tSk0v_(64, schema).pre('save', function (next) {
                __block_Cbs43u(5);
                if (__expression_qSott_(65), !__extro_ygGuVj(66, __intro_tSk0v_(66, this).isModified(field.fromPath))) {
                    __block_Cbs43u(6);
                    return __expression_qSott_(67), (__expression_qSott_(68), next());
                }
                {
                    __statement_lDHrQM(69);
                    var fromValue = __extro_ygGuVj(70, __intro_tSk0v_(70, this).get(field.fromPath));
                }
                if (__expression_qSott_(71), !(__expression_qSott_(72), fromValue)) {
                    __block_Cbs43u(7);
                    {
                        __statement_lDHrQM(73);
                        __extro_ygGuVj(74, __intro_tSk0v_(74, field).reset(this));
                    }
                    return __expression_qSott_(75), (__expression_qSott_(76), next());
                }
                {
                    __statement_lDHrQM(77);
                    var post = this;
                }
                {
                    __statement_lDHrQM(78);
                    new EmbedlyAPI({
                        key: __extro_ygGuVj(79, __intro_tSk0v_(79, keystone).get('embedly api key'))
                    }, function (err, api) {
                        __block_Cbs43u(8);
                        if (__expression_qSott_(80), err) {
                            __block_Cbs43u(9);
                            {
                                __statement_lDHrQM(81);
                                __extro_ygGuVj(82, __intro_tSk0v_(82, console).error('Error creating Embedly api:'));
                            }
                            {
                                __statement_lDHrQM(83);
                                __extro_ygGuVj(84, __intro_tSk0v_(84, console).error(err, api));
                            }
                            {
                                __statement_lDHrQM(85);
                                __extro_ygGuVj(86, __intro_tSk0v_(86, field).reset(this));
                            }
                            return __expression_qSott_(87), (__expression_qSott_(88), next());
                        }
                        {
                            __statement_lDHrQM(89);
                            var opts = __extro_ygGuVj(90, __intro_tSk0v_(90, _).defaults({
                                    url: fromValue
                                }, field.embedlyOptions));
                        }
                        {
                            __statement_lDHrQM(91);
                            __extro_ygGuVj(92, __intro_tSk0v_(92, api).oembed(opts, function (err, objs) {
                                __block_Cbs43u(10);
                                if (__expression_qSott_(93), err) {
                                    __block_Cbs43u(11);
                                    {
                                        __statement_lDHrQM(94);
                                        __extro_ygGuVj(95, __intro_tSk0v_(95, console).error('Embedly API Error:'));
                                    }
                                    {
                                        __statement_lDHrQM(96);
                                        __extro_ygGuVj(97, __intro_tSk0v_(97, console).error(err, objs));
                                    }
                                    {
                                        __statement_lDHrQM(98);
                                        __extro_ygGuVj(99, __intro_tSk0v_(99, field).reset(post));
                                    }
                                } else {
                                    __block_Cbs43u(12);
                                    {
                                        __statement_lDHrQM(100);
                                        var data = objs[0];
                                    }
                                    if (__expression_qSott_(101), (__expression_qSott_(102), data) && (__expression_qSott_(103), data.type !== 'error')) {
                                        __block_Cbs43u(13);
                                        {
                                            __statement_lDHrQM(104);
                                            __extro_ygGuVj(105, __intro_tSk0v_(105, post).set(field.path, {
                                                exists: true,
                                                type: data.type,
                                                title: data.title,
                                                url: data.url,
                                                width: data.width,
                                                height: data.height,
                                                version: data.version,
                                                description: data.description,
                                                html: data.html,
                                                authorName: data.author_name,
                                                authorUrl: data.author_url,
                                                providerName: data.provider_name,
                                                providerUrl: data.provider_url,
                                                thumbnailUrl: data.thumbnail_url,
                                                thumbnailWidth: data.thumbnail_width,
                                                thumbnailHeight: data.thumbnail_height
                                            }));
                                        }
                                    } else {
                                        __block_Cbs43u(14);
                                        {
                                            __statement_lDHrQM(106);
                                            __extro_ygGuVj(107, __intro_tSk0v_(107, field).reset(post));
                                        }
                                    }
                                }
                                return __expression_qSott_(108), (__expression_qSott_(109), next());
                            }));
                        }
                    });
                }
            }));
        }
        {
            __statement_lDHrQM(110);
            __extro_ygGuVj(111, __intro_tSk0v_(111, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_lDHrQM(112);
    embedly.prototype.reset = function (item) {
        __block_Cbs43u(15);
        return __expression_qSott_(113), __extro_ygGuVj(114, __intro_tSk0v_(114, item).set(__extro_ygGuVj(115, __intro_tSk0v_(115, item).set(this.path, {
            exists: false,
            type: null,
            title: null,
            url: null,
            width: null,
            height: null,
            version: null,
            description: null,
            html: null,
            authorName: null,
            authorUrl: null,
            providerName: null,
            providerUrl: null,
            thumbnailUrl: null,
            thumbnailWidth: null,
            thumbnailHeight: null
        }))));
    };
}
{
    __statement_lDHrQM(116);
    embedly.prototype.format = function (item) {
        __block_Cbs43u(16);
        return __expression_qSott_(117), __extro_ygGuVj(118, __intro_tSk0v_(118, item).get(this.paths.html));
    };
}
{
    __statement_lDHrQM(119);
    embedly.prototype.isModified = function (item) {
        __block_Cbs43u(17);
        return __expression_qSott_(120), __extro_ygGuVj(121, __intro_tSk0v_(121, item).isModified(this.paths.url));
    };
}
{
    __statement_lDHrQM(122);
    embedly.prototype.validateInput = function (data) {
        __block_Cbs43u(18);
        return __expression_qSott_(123), true;
    };
}
{
    __statement_lDHrQM(124);
    embedly.prototype.updateItem = function (item, data) {
        __block_Cbs43u(19);
        return __expression_qSott_(125), __extro_ygGuVj(126, __intro_tSk0v_(126, item).set(__extro_ygGuVj(127, __intro_tSk0v_(127, item).set(this.path, {
            exists: data[this.paths.exists],
            type: data[this.paths.type],
            title: data[this.paths.title],
            url: data[this.paths.url],
            width: data[this.paths.width],
            height: data[this.paths.height],
            version: data[this.paths.version],
            description: data[this.paths.description],
            html: data[this.paths.html],
            authorName: data[this.paths.authorName],
            authorUrl: data[this.paths.authorUrl],
            providerName: data[this.paths.providerName],
            providerUrl: data[this.paths.providerUrl],
            thumbnailUrl: data[this.paths.thumbnailUrl],
            thumbnailWidth: data[this.paths.thumbnailWidth],
            thumbnailHeight: data[this.paths.thumbnailHeight]
        }))));
    };
}
{
    __statement_lDHrQM(128);
    exports = module.exports = embedly;
}
