// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_view.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_XqOK1e, __expression_ya$zc8, __block_RwsLAV;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_XqOK1e = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/view.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_ya$zc8 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/view.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_RwsLAV = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/view.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_yX_t6M = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_H4aDE7 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/view.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_XqOK1e(0);
    var _ = (__expression_ya$zc8(1), require('underscore')), keystone = (__expression_ya$zc8(2), require('../')), async = (__expression_ya$zc8(3), require('async')), utils = (__expression_ya$zc8(4), require('keystone-utils'));
}
function View(req, res) {
    __block_RwsLAV(0);
    if (__expression_ya$zc8(5), (__expression_ya$zc8(6), !(__expression_ya$zc8(7), req)) || (__expression_ya$zc8(8), req.constructor.name !== 'IncomingMessage')) {
        __block_RwsLAV(1);
        {
            __statement_XqOK1e(9);
            throw new Error('Keystone.View Error: Express request object is required.');
        }
    }
    if (__expression_ya$zc8(10), (__expression_ya$zc8(11), !(__expression_ya$zc8(12), res)) || (__expression_ya$zc8(13), res.constructor.name !== 'ServerResponse')) {
        __block_RwsLAV(2);
        {
            __statement_XqOK1e(14);
            throw new Error('Keystone.View Error: Express response object is required.');
        }
    }
    {
        __statement_XqOK1e(15);
        this.req = req;
    }
    {
        __statement_XqOK1e(16);
        this.res = res;
    }
    {
        __statement_XqOK1e(17);
        this.initQueue = [];
    }
    {
        __statement_XqOK1e(18);
        this.actionQueue = [];
    }
    {
        __statement_XqOK1e(19);
        this.queryQueue = [];
    }
    {
        __statement_XqOK1e(20);
        this.renderQueue = [];
    }
}
{
    __statement_XqOK1e(21);
    module.exports = exports = View;
}
{
    __statement_XqOK1e(22);
    View.prototype.on = function (on) {
        __block_RwsLAV(3);
        {
            __statement_XqOK1e(23);
            var req = this.req, callback = arguments[1], values;
        }
        if (__expression_ya$zc8(24), 'function' === (__expression_ya$zc8(25), typeof on)) {
            __block_RwsLAV(4);
            if (__expression_ya$zc8(26), on()) {
                __block_RwsLAV(5);
                {
                    __statement_XqOK1e(27);
                    __extro_H4aDE7(28, __intro_yX_t6M(28, this.actionQueue).push(callback));
                }
            }
        } else if (__extro_H4aDE7(29, __intro_yX_t6M(29, utils).isObject(on))) {
            __block_RwsLAV(6);
            {
                __statement_XqOK1e(30);
                var check = function (value, path) {
                    __block_RwsLAV(7);
                    {
                        __statement_XqOK1e(31);
                        var ctx = req, parts = __extro_H4aDE7(32, __intro_yX_t6M(32, path).split('.'));
                    }
                    for (var i = 0; __expression_ya$zc8(33), (__expression_ya$zc8(34), i) < (__expression_ya$zc8(35), parts.length - 1); __expression_ya$zc8(36), i++) {
                        __block_RwsLAV(8);
                        if (__expression_ya$zc8(37), !ctx[parts[i]]) {
                            __block_RwsLAV(9);
                            return __expression_ya$zc8(38), false;
                        }
                        {
                            __statement_XqOK1e(39);
                            ctx = ctx[parts[i]];
                        }
                    }
                    {
                        __statement_XqOK1e(40);
                        path = __extro_H4aDE7(41, __intro_yX_t6M(41, _).last(parts));
                    }
                    return __expression_ya$zc8(42), (__expression_ya$zc8(45), (__expression_ya$zc8(46), (__expression_ya$zc8(47), value) === true) && (__expression_ya$zc8(48), ((__expression_ya$zc8(49), path) in (__expression_ya$zc8(50), ctx)))) ? (__expression_ya$zc8(43), true) : (__expression_ya$zc8(44), (__expression_ya$zc8(51), ctx[path] === (__expression_ya$zc8(52), value)));
                };
            }
            if (__extro_H4aDE7(53, __intro_yX_t6M(53, _).every(on, check))) {
                __block_RwsLAV(10);
                {
                    __statement_XqOK1e(54);
                    __extro_H4aDE7(55, __intro_yX_t6M(55, this.actionQueue).push(callback));
                }
            }
        } else if (__expression_ya$zc8(56), (__expression_ya$zc8(57), (__expression_ya$zc8(58), (__expression_ya$zc8(59), (__expression_ya$zc8(60), on) === 'get') || (__expression_ya$zc8(61), (__expression_ya$zc8(62), on) === 'post')) || (__expression_ya$zc8(63), (__expression_ya$zc8(64), on) === 'put')) || (__expression_ya$zc8(65), (__expression_ya$zc8(66), on) === 'delete')) {
            __block_RwsLAV(11);
            if (__expression_ya$zc8(67), req.method !== __extro_H4aDE7(68, __intro_yX_t6M(68, on).toUpperCase())) {
                __block_RwsLAV(12);
                return __expression_ya$zc8(69);
            }
            if (__expression_ya$zc8(70), arguments.length === 3) {
                __block_RwsLAV(13);
                if (__extro_H4aDE7(71, __intro_yX_t6M(71, utils).isString(arguments[1]))) {
                    __block_RwsLAV(14);
                    {
                        __statement_XqOK1e(72);
                        values = {};
                    }
                    {
                        __statement_XqOK1e(73);
                        values[arguments[1]] = true;
                    }
                } else {
                    __block_RwsLAV(15);
                    {
                        __statement_XqOK1e(74);
                        values = arguments[1];
                    }
                }
                {
                    __statement_XqOK1e(75);
                    callback = arguments[2];
                }
                {
                    __statement_XqOK1e(76);
                    var ctx = (__expression_ya$zc8(79), (__expression_ya$zc8(80), (__expression_ya$zc8(81), on) === 'post') || (__expression_ya$zc8(82), (__expression_ya$zc8(83), on) === 'put')) ? (__expression_ya$zc8(77), req.body) : (__expression_ya$zc8(78), req.query);
                }
                if (__extro_H4aDE7(84, __intro_yX_t6M(84, _).every((__expression_ya$zc8(85), (__expression_ya$zc8(86), values) || {}), function (value, path) {
                        __block_RwsLAV(16);
                        return __expression_ya$zc8(87), (__expression_ya$zc8(90), (__expression_ya$zc8(91), (__expression_ya$zc8(92), value) === true) && (__expression_ya$zc8(93), ((__expression_ya$zc8(94), path) in (__expression_ya$zc8(95), ctx)))) ? (__expression_ya$zc8(88), true) : (__expression_ya$zc8(89), (__expression_ya$zc8(96), ctx[path] === (__expression_ya$zc8(97), value)));
                    }))) {
                    __block_RwsLAV(17);
                    {
                        __statement_XqOK1e(98);
                        __extro_H4aDE7(99, __intro_yX_t6M(99, this.actionQueue).push(callback));
                    }
                }
            } else {
                __block_RwsLAV(18);
                {
                    __statement_XqOK1e(100);
                    __extro_H4aDE7(101, __intro_yX_t6M(101, this.actionQueue).push(callback));
                }
            }
        } else if (__expression_ya$zc8(102), (__expression_ya$zc8(103), on) === 'init') {
            __block_RwsLAV(19);
            {
                __statement_XqOK1e(104);
                __extro_H4aDE7(105, __intro_yX_t6M(105, this.initQueue).push(callback));
            }
        } else if (__expression_ya$zc8(106), (__expression_ya$zc8(107), on) === 'render') {
            __block_RwsLAV(20);
            {
                __statement_XqOK1e(108);
                __extro_H4aDE7(109, __intro_yX_t6M(109, this.renderQueue).push(callback));
            }
        }
        return __expression_ya$zc8(110), this;
    };
}
{
    __statement_XqOK1e(111);
    var QueryCallbacks = function (options) {
        __block_RwsLAV(21);
        if (__extro_H4aDE7(112, __intro_yX_t6M(112, utils).isString(options))) {
            __block_RwsLAV(22);
            {
                __statement_XqOK1e(113);
                options = {
                    then: options
                };
            }
        } else {
            __block_RwsLAV(23);
            {
                __statement_XqOK1e(114);
                options = (__expression_ya$zc8(115), (__expression_ya$zc8(116), options) || {});
            }
        }
        {
            __statement_XqOK1e(117);
            this.callbacks = {};
        }
        if (options.err) {
            __block_RwsLAV(24);
            {
                __statement_XqOK1e(118);
                this.callbacks.err = options.err;
            }
        }
        if (options.none) {
            __block_RwsLAV(25);
            {
                __statement_XqOK1e(119);
                this.callbacks.none = options.none;
            }
        }
        if (options.then) {
            __block_RwsLAV(26);
            {
                __statement_XqOK1e(120);
                this.callbacks.then = options.then;
            }
        }
        return __expression_ya$zc8(121), this;
    };
}
{
    __statement_XqOK1e(122);
    QueryCallbacks.prototype.has = function (fn) {
        __block_RwsLAV(27);
        return __expression_ya$zc8(123), (__expression_ya$zc8(124), ((__expression_ya$zc8(125), fn) in this.callbacks));
    };
}
{
    __statement_XqOK1e(126);
    QueryCallbacks.prototype.err = function (fn) {
        __block_RwsLAV(28);
        {
            __statement_XqOK1e(127);
            this.callbacks.err = fn;
        }
        return __expression_ya$zc8(128), this;
    };
}
{
    __statement_XqOK1e(129);
    QueryCallbacks.prototype.none = function (fn) {
        __block_RwsLAV(29);
        {
            __statement_XqOK1e(130);
            this.callbacks.none = fn;
        }
        return __expression_ya$zc8(131), this;
    };
}
{
    __statement_XqOK1e(132);
    QueryCallbacks.prototype.then = function (fn) {
        __block_RwsLAV(30);
        {
            __statement_XqOK1e(133);
            this.callbacks.then = fn;
        }
        return __expression_ya$zc8(134), this;
    };
}
{
    __statement_XqOK1e(135);
    View.prototype.query = function (key, query, options) {
        __block_RwsLAV(31);
        {
            __statement_XqOK1e(136);
            var locals = this.res.locals, parts = __extro_H4aDE7(137, __intro_yX_t6M(137, key).split('.')), chain = new QueryCallbacks(options);
        }
        {
            __statement_XqOK1e(138);
            key = __extro_H4aDE7(139, __intro_yX_t6M(139, parts).pop());
        }
        for (var i = 0; __expression_ya$zc8(140), (__expression_ya$zc8(141), i) < parts.length; __expression_ya$zc8(142), i++) {
            __block_RwsLAV(32);
            if (__expression_ya$zc8(143), !locals[parts[i]]) {
                __block_RwsLAV(33);
                {
                    __statement_XqOK1e(144);
                    locals[parts[i]] = {};
                }
            }
            {
                __statement_XqOK1e(145);
                locals = locals[parts[i]];
            }
        }
        {
            __statement_XqOK1e(146);
            __extro_H4aDE7(147, __intro_yX_t6M(147, this.queryQueue).push(function (next) {
                __block_RwsLAV(34);
                {
                    __statement_XqOK1e(148);
                    __extro_H4aDE7(149, __intro_yX_t6M(149, query).exec(function (err, results) {
                        __block_RwsLAV(35);
                        {
                            __statement_XqOK1e(150);
                            locals[key] = results;
                        }
                        {
                            __statement_XqOK1e(151);
                            var callbacks = chain.callbacks;
                        }
                        if (__expression_ya$zc8(152), err) {
                            __block_RwsLAV(36);
                            if (__expression_ya$zc8(153), ('err' in (__expression_ya$zc8(154), callbacks))) {
                                __block_RwsLAV(37);
                                return __expression_ya$zc8(155), __extro_H4aDE7(156, __intro_yX_t6M(156, callbacks).err(err, next));
                            }
                        } else {
                            __block_RwsLAV(38);
                            if (__expression_ya$zc8(157), (__expression_ya$zc8(158), (__expression_ya$zc8(159), !(__expression_ya$zc8(160), results)) || (__expression_ya$zc8(161), __extro_H4aDE7(162, __intro_yX_t6M(162, utils).isArray(results)) && (__expression_ya$zc8(163), !results.length))) && (__expression_ya$zc8(164), ('none' in (__expression_ya$zc8(165), callbacks)))) {
                                __block_RwsLAV(39);
                                return __expression_ya$zc8(166), __extro_H4aDE7(167, __intro_yX_t6M(167, callbacks).none(next));
                            } else if (__expression_ya$zc8(168), ('then' in (__expression_ya$zc8(169), callbacks))) {
                                __block_RwsLAV(40);
                                if (__extro_H4aDE7(170, __intro_yX_t6M(170, utils).isFunction(callbacks.then))) {
                                    __block_RwsLAV(41);
                                    return __expression_ya$zc8(171), __extro_H4aDE7(172, __intro_yX_t6M(172, callbacks).then(err, results, next));
                                } else {
                                    __block_RwsLAV(42);
                                    return __expression_ya$zc8(173), __extro_H4aDE7(174, __intro_yX_t6M(174, keystone).populateRelated(results, callbacks.then, next));
                                }
                            }
                        }
                        return __expression_ya$zc8(175), (__expression_ya$zc8(176), next(err));
                    }));
                }
            }));
        }
        return __expression_ya$zc8(177), chain;
    };
}
{
    __statement_XqOK1e(178);
    View.prototype.render = function (renderFn, locals, callback) {
        __block_RwsLAV(43);
        {
            __statement_XqOK1e(179);
            var req = this.req, res = this.res;
        }
        if (__expression_ya$zc8(180), 'string' === (__expression_ya$zc8(181), typeof renderFn)) {
            __block_RwsLAV(44);
            {
                __statement_XqOK1e(182);
                var viewPath = renderFn;
            }
            {
                __statement_XqOK1e(183);
                renderFn = __extro_H4aDE7(184, __intro_yX_t6M(184, function () {
                    __block_RwsLAV(45);
                    if (__expression_ya$zc8(185), 'function' === (__expression_ya$zc8(186), typeof locals)) {
                        __block_RwsLAV(46);
                        {
                            __statement_XqOK1e(187);
                            locals = (__expression_ya$zc8(188), locals());
                        }
                    }
                    {
                        __statement_XqOK1e(189);
                        __extro_H4aDE7(190, __intro_yX_t6M(190, this.res).render(viewPath, locals, callback));
                    }
                }).bind(this));
            }
        }
        if (__expression_ya$zc8(191), 'function' !== (__expression_ya$zc8(192), typeof renderFn)) {
            __block_RwsLAV(47);
            {
                __statement_XqOK1e(193);
                throw new Error('Keystone.View.render() renderFn must be a templatePath (string) or a function.');
            }
        }
        {
            __statement_XqOK1e(194);
            __extro_H4aDE7(195, __intro_yX_t6M(195, this.initQueue).push(this.actionQueue));
        }
        {
            __statement_XqOK1e(196);
            __extro_H4aDE7(197, __intro_yX_t6M(197, this.initQueue).push(this.queryQueue));
        }
        {
            __statement_XqOK1e(198);
            var preRenderQueue = [];
        }
        {
            __statement_XqOK1e(199);
            __extro_H4aDE7(200, __intro_yX_t6M(200, keystone._pre.render).forEach(function (fn) {
                __block_RwsLAV(48);
                {
                    __statement_XqOK1e(201);
                    __extro_H4aDE7(202, __intro_yX_t6M(202, preRenderQueue).push(function (next) {
                        __block_RwsLAV(49);
                        {
                            __statement_XqOK1e(203);
                            __expression_ya$zc8(204), fn(req, res, next);
                        }
                    }));
                }
            }));
        }
        {
            __statement_XqOK1e(205);
            __extro_H4aDE7(206, __intro_yX_t6M(206, this.initQueue).push(preRenderQueue));
        }
        {
            __statement_XqOK1e(207);
            __extro_H4aDE7(208, __intro_yX_t6M(208, this.initQueue).push(this.renderQueue));
        }
        {
            __statement_XqOK1e(209);
            __extro_H4aDE7(210, __intro_yX_t6M(210, async).eachSeries(this.initQueue, function (i, next) {
                __block_RwsLAV(50);
                if (__extro_H4aDE7(211, __intro_yX_t6M(211, Array).isArray(i))) {
                    __block_RwsLAV(51);
                    {
                        __statement_XqOK1e(212);
                        __extro_H4aDE7(213, __intro_yX_t6M(213, async).parallel(i, next));
                    }
                } else if (__expression_ya$zc8(214), 'function' === (__expression_ya$zc8(215), typeof i)) {
                    __block_RwsLAV(52);
                    {
                        __statement_XqOK1e(216);
                        __expression_ya$zc8(217), i(next);
                    }
                } else {
                    __block_RwsLAV(53);
                    {
                        __statement_XqOK1e(218);
                        throw new Error('Keystone.View.render() events must be functions.');
                    }
                }
            }, function (err) {
                __block_RwsLAV(54);
                {
                    __statement_XqOK1e(219);
                    __expression_ya$zc8(220), renderFn(err, req, res);
                }
            }));
        }
    };
}
