// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_relationship.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_PuVpKh, __expression_QKUl9L, __block_yAFrpZ;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_PuVpKh = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/relationship.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_QKUl9L = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/relationship.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_yAFrpZ = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/relationship.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_Q4UgWO = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_uPXLW0 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/relationship.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_PuVpKh(0);
    var _ = (__expression_QKUl9L(1), require('underscore')), keystone = (__expression_QKUl9L(2), require('../../')), util = (__expression_QKUl9L(3), require('util')), utils = (__expression_QKUl9L(4), require('keystone-utils')), super_ = (__expression_QKUl9L(5), require('../field'));
}
function relationship(list, path, options) {
    __block_yAFrpZ(0);
    {
        __statement_PuVpKh(6);
        this.many = options.many ? (__expression_QKUl9L(7), true) : (__expression_QKUl9L(8), false);
    }
    {
        __statement_PuVpKh(9);
        this.filters = options.filters;
    }
    {
        __statement_PuVpKh(10);
        this._nativeType = keystone.mongoose.Schema.Types.ObjectId;
    }
    {
        __statement_PuVpKh(11);
        this._underscoreMethods = [
            'format'
        ];
    }
    {
        __statement_PuVpKh(12);
        __extro_uPXLW0(13, __intro_Q4UgWO(13, relationship.super_).call(this, list, path, options));
    }
}
{
    __statement_PuVpKh(14);
    __extro_uPXLW0(15, __intro_Q4UgWO(15, util).inherits(relationship, super_));
}
{
    __statement_PuVpKh(16);
    relationship.prototype.addToSchema = function () {
        __block_yAFrpZ(1);
        {
            __statement_PuVpKh(17);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_PuVpKh(18);
            this.paths = {
                refList: (__expression_QKUl9L(19), this.options.refListPath || __extro_uPXLW0(20, __intro_Q4UgWO(20, this._path).append('RefList')))
            };
        }
        {
            __statement_PuVpKh(21);
            var def = {
                    type: this._nativeType,
                    ref: this.options.ref
                };
        }
        {
            __statement_PuVpKh(22);
            __extro_uPXLW0(23, __intro_Q4UgWO(23, schema).path(this.path, this.many ? (__expression_QKUl9L(24), [
                def
            ]) : (__expression_QKUl9L(25), def)));
        }
        {
            __statement_PuVpKh(26);
            __extro_uPXLW0(27, __intro_Q4UgWO(27, __extro_uPXLW0(28, __intro_Q4UgWO(28, schema).virtual(this.paths.refList))).get(function () {
                __block_yAFrpZ(2);
                return __expression_QKUl9L(29), __extro_uPXLW0(30, __intro_Q4UgWO(30, keystone).list(field.options.ref));
            }));
        }
        if (this.many) {
            __block_yAFrpZ(3);
            {
                __statement_PuVpKh(31);
                __extro_uPXLW0(32, __intro_Q4UgWO(32, this).underscoreMethod('contains', function (find) {
                    __block_yAFrpZ(4);
                    {
                        __statement_PuVpKh(33);
                        var value = (__expression_QKUl9L(34), __extro_uPXLW0(35, __intro_Q4UgWO(35, this).populated(field.path)) || __extro_uPXLW0(36, __intro_Q4UgWO(36, this).get(field.path)));
                    }
                    if (__expression_QKUl9L(37), 'object' === (__expression_QKUl9L(38), typeof find)) {
                        __block_yAFrpZ(5);
                        {
                            __statement_PuVpKh(39);
                            find = find.id;
                        }
                    }
                    {
                        __statement_PuVpKh(40);
                        var result = __extro_uPXLW0(41, __intro_Q4UgWO(41, _).some(value, function (value) {
                                __block_yAFrpZ(6);
                                return __expression_QKUl9L(42), (__expression_QKUl9L(43), (__expression_QKUl9L(44), (__expression_QKUl9L(45), value) + '') === (__expression_QKUl9L(46), find));
                            }));
                    }
                    return __expression_QKUl9L(47), result;
                }));
            }
        }
        {
            __statement_PuVpKh(48);
            __extro_uPXLW0(49, __intro_Q4UgWO(49, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_PuVpKh(50);
    relationship.prototype.format = function (item) {
        __block_yAFrpZ(7);
        {
            __statement_PuVpKh(51);
            var value = __extro_uPXLW0(52, __intro_Q4UgWO(52, item).get(this.path));
        }
        return __expression_QKUl9L(53), this.many ? (__expression_QKUl9L(54), __extro_uPXLW0(56, __intro_Q4UgWO(56, value).join(', '))) : (__expression_QKUl9L(55), (__expression_QKUl9L(57), (__expression_QKUl9L(58), (__expression_QKUl9L(59), value) || '') + ''));
    };
}
{
    __statement_PuVpKh(60);
    relationship.prototype.validateInput = function (data, required, item) {
        __block_yAFrpZ(8);
        if (__expression_QKUl9L(61), !(__expression_QKUl9L(62), required)) {
            __block_yAFrpZ(9);
            return __expression_QKUl9L(63), true;
        }
        if (__expression_QKUl9L(64), (__expression_QKUl9L(65), (__expression_QKUl9L(66), !(__expression_QKUl9L(67), (this.path in (__expression_QKUl9L(68), data)))) && (__expression_QKUl9L(69), item)) && (__expression_QKUl9L(70), (__expression_QKUl9L(71), this.many && __extro_uPXLW0(72, __intro_Q4UgWO(72, item).get(this.path)).length) || __extro_uPXLW0(73, __intro_Q4UgWO(73, item).get(this.path)))) {
            __block_yAFrpZ(10);
            return __expression_QKUl9L(74), true;
        }
        if (__expression_QKUl9L(75), 'string' === (__expression_QKUl9L(76), typeof data[this.path])) {
            __block_yAFrpZ(11);
            return __expression_QKUl9L(77), __extro_uPXLW0(80, __intro_Q4UgWO(80, data[this.path]).trim()) ? (__expression_QKUl9L(78), true) : (__expression_QKUl9L(79), false);
        } else {
            __block_yAFrpZ(12);
            return __expression_QKUl9L(81), data[this.path] ? (__expression_QKUl9L(82), true) : (__expression_QKUl9L(83), false);
        }
    };
}
{
    __statement_PuVpKh(84);
    relationship.prototype.updateItem = function (item, data) {
        __block_yAFrpZ(13);
        if (__expression_QKUl9L(85), !(__expression_QKUl9L(86), (this.path in (__expression_QKUl9L(87), data)))) {
            __block_yAFrpZ(14);
            return __expression_QKUl9L(88);
        }
        if (__extro_uPXLW0(89, __intro_Q4UgWO(89, item).populated(this.path))) {
            __block_yAFrpZ(15);
            {
                __statement_PuVpKh(90);
                throw new Error('fieldTypes.relationship.updateItem() Error - You cannot update populated relationships.');
            }
        }
        if (this.many) {
            __block_yAFrpZ(16);
            {
                __statement_PuVpKh(91);
                var arr = __extro_uPXLW0(92, __intro_Q4UgWO(92, item).get(this.path)), _old = __extro_uPXLW0(93, __intro_Q4UgWO(93, arr).map(function (i) {
                        __block_yAFrpZ(17);
                        return __expression_QKUl9L(94), (__expression_QKUl9L(95), String(i));
                    })), _new = data[this.path];
            }
            if (__expression_QKUl9L(96), !__extro_uPXLW0(97, __intro_Q4UgWO(97, utils).isArray(_new))) {
                __block_yAFrpZ(18);
                {
                    __statement_PuVpKh(98);
                    _new = __extro_uPXLW0(99, __intro_Q4UgWO(99, (__expression_QKUl9L(100), String((__expression_QKUl9L(101), (__expression_QKUl9L(102), _new) || '')))).split(','));
                }
            }
            {
                __statement_PuVpKh(103);
                _new = __extro_uPXLW0(104, __intro_Q4UgWO(104, _).compact(_new));
            }
            {
                __statement_PuVpKh(105);
                __extro_uPXLW0(106, __intro_Q4UgWO(106, __extro_uPXLW0(107, __intro_Q4UgWO(107, _).difference(_old, _new))).forEach(function (val) {
                    __block_yAFrpZ(19);
                    {
                        __statement_PuVpKh(108);
                        __extro_uPXLW0(109, __intro_Q4UgWO(109, arr).pull(val));
                    }
                }));
            }
            {
                __statement_PuVpKh(110);
                __extro_uPXLW0(111, __intro_Q4UgWO(111, __extro_uPXLW0(112, __intro_Q4UgWO(112, _).difference(_new, _old))).forEach(function (val) {
                    __block_yAFrpZ(20);
                    {
                        __statement_PuVpKh(113);
                        __extro_uPXLW0(114, __intro_Q4UgWO(114, arr).push(val));
                    }
                }));
            }
        } else {
            __block_yAFrpZ(21);
            if (data[this.path]) {
                __block_yAFrpZ(22);
                if (__expression_QKUl9L(115), data[this.path] !== __extro_uPXLW0(116, __intro_Q4UgWO(116, item).get(this.path))) {
                    __block_yAFrpZ(23);
                    {
                        __statement_PuVpKh(117);
                        __extro_uPXLW0(118, __intro_Q4UgWO(118, item).set(this.path, data[this.path]));
                    }
                }
            } else if (__extro_uPXLW0(119, __intro_Q4UgWO(119, item).get(this.path))) {
                __block_yAFrpZ(24);
                {
                    __statement_PuVpKh(120);
                    __extro_uPXLW0(121, __intro_Q4UgWO(121, item).set(this.path, null));
                }
            }
        }
    };
}
{
    __statement_PuVpKh(122);
    __extro_uPXLW0(123, __intro_Q4UgWO(123, Object).defineProperty(relationship.prototype, 'isValid', {
        get: function () {
            __block_yAFrpZ(25);
            return __expression_QKUl9L(124), __extro_uPXLW0(127, __intro_Q4UgWO(127, keystone).list(this.options.ref)) ? (__expression_QKUl9L(125), true) : (__expression_QKUl9L(126), false);
        }
    }));
}
{
    __statement_PuVpKh(128);
    __extro_uPXLW0(129, __intro_Q4UgWO(129, Object).defineProperty(relationship.prototype, 'refList', {
        get: function () {
            __block_yAFrpZ(26);
            return __expression_QKUl9L(130), __extro_uPXLW0(131, __intro_Q4UgWO(131, keystone).list(this.options.ref));
        }
    }));
}
{
    __statement_PuVpKh(132);
    __extro_uPXLW0(133, __intro_Q4UgWO(133, Object).defineProperty(relationship.prototype, 'hasFilters', {
        get: function () {
            __block_yAFrpZ(27);
            return __expression_QKUl9L(134), (__expression_QKUl9L(135), this.filters && __extro_uPXLW0(136, __intro_Q4UgWO(136, _).keys(this.filters)).length);
        }
    }));
}
{
    __statement_PuVpKh(137);
    relationship.prototype.addFilters = function (query, item) {
        __block_yAFrpZ(28);
        {
            __statement_PuVpKh(138);
            __extro_uPXLW0(139, __intro_Q4UgWO(139, _).each(this.filters, function (filters, path) {
                __block_yAFrpZ(29);
                if (__expression_QKUl9L(140), !__extro_uPXLW0(141, __intro_Q4UgWO(141, utils).isObject(filters))) {
                    __block_yAFrpZ(30);
                    {
                        __statement_PuVpKh(142);
                        filters = {
                            equals: filters
                        };
                    }
                }
                {
                    __statement_PuVpKh(143);
                    __extro_uPXLW0(144, __intro_Q4UgWO(144, query).where(path));
                }
                {
                    __statement_PuVpKh(145);
                    __extro_uPXLW0(146, __intro_Q4UgWO(146, _).each(filters, function (value, method) {
                        __block_yAFrpZ(31);
                        if (__expression_QKUl9L(147), (__expression_QKUl9L(148), 'string' === (__expression_QKUl9L(149), typeof value)) && (__expression_QKUl9L(150), __extro_uPXLW0(151, __intro_Q4UgWO(151, value).substr(0, 1)) === ':')) {
                            __block_yAFrpZ(32);
                            if (__expression_QKUl9L(152), !(__expression_QKUl9L(153), item)) {
                                __block_yAFrpZ(33);
                                return __expression_QKUl9L(154);
                            }
                            {
                                __statement_PuVpKh(155);
                                value = __extro_uPXLW0(156, __intro_Q4UgWO(156, item).get(__extro_uPXLW0(157, __intro_Q4UgWO(157, value).substr(1))));
                            }
                        }
                        {
                            __statement_PuVpKh(158);
                            __extro_uPXLW0(159, __intro_Q4UgWO(159, query)[method](value));
                        }
                    }));
                }
            }));
        }
    };
}
{
    __statement_PuVpKh(160);
    exports = module.exports = relationship;
}
