// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_field.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_W4Rg36, __expression_dxZS4o, __block_nrL548;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_W4Rg36 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/field.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_dxZS4o = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/field.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_nrL548 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/field.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_FT0Eil = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_FSrivI = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/field.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_W4Rg36(0);
    var _ = (__expression_dxZS4o(1), require('underscore')), marked = (__expression_dxZS4o(2), require('marked')), Path = (__expression_dxZS4o(3), require('./path')), fspath = (__expression_dxZS4o(4), require('path')), jade = (__expression_dxZS4o(5), require('jade')), fs = (__expression_dxZS4o(6), require('fs')), keystone = (__expression_dxZS4o(7), require('../')), utils = (__expression_dxZS4o(8), require('keystone-utils')), compiledTemplates = {};
}
function Field(list, path, options) {
    __block_nrL548(0);
    {
        __statement_W4Rg36(9);
        this.list = list;
    }
    {
        __statement_W4Rg36(10);
        this._path = new Path(path);
    }
    {
        __statement_W4Rg36(11);
        this.path = path;
    }
    {
        __statement_W4Rg36(12);
        this.type = this.constructor.name;
    }
    {
        __statement_W4Rg36(13);
        this.options = __extro_FSrivI(14, __intro_FT0Eil(14, utils).options(this.defaults, options));
    }
    {
        __statement_W4Rg36(15);
        this.label = (__expression_dxZS4o(16), options.label || __extro_FSrivI(17, __intro_FT0Eil(17, utils).keyToLabel(this.path)));
    }
    {
        __statement_W4Rg36(18);
        this.typeDescription = (__expression_dxZS4o(19), (__expression_dxZS4o(20), options.typeDescription || this.typeDescription) || this.type);
    }
    {
        __statement_W4Rg36(21);
        __extro_FSrivI(22, __intro_FT0Eil(22, this.list).automap(this));
    }
    {
        __statement_W4Rg36(23);
        __extro_FSrivI(24, __intro_FT0Eil(24, this).addToSchema());
    }
    if (__expression_dxZS4o(25), (__expression_dxZS4o(26), (__expression_dxZS4o(27), (__expression_dxZS4o(28), (__expression_dxZS4o(29), this.options.required && (__expression_dxZS4o(30), this.options.initial === (__expression_dxZS4o(31), undefined))) && (__expression_dxZS4o(32), this.options.default === (__expression_dxZS4o(33), undefined))) && (__expression_dxZS4o(34), !this.options.value)) && (__expression_dxZS4o(35), !__extro_FSrivI(36, __intro_FT0Eil(36, this.list).get('nocreate')))) && (__expression_dxZS4o(37), this.path !== this.list.mappings.name)) {
        __block_nrL548(1);
        {
            __statement_W4Rg36(38);
            __extro_FSrivI(39, __intro_FT0Eil(39, console).error((__expression_dxZS4o(40), (__expression_dxZS4o(41), (__expression_dxZS4o(42), (__expression_dxZS4o(43), (__expression_dxZS4o(44), (__expression_dxZS4o(45), '\nError: Invalid Configuration\n\n' + 'Field (') + list.key) + '.') + (__expression_dxZS4o(46), path)) + ') is required but not initial, and has no default or generated value.\n') + 'Please provide a default, remove the required setting, or set initial: false to override this error.\n')));
        }
        {
            __statement_W4Rg36(47);
            __extro_FSrivI(48, __intro_FT0Eil(48, process).exit(1));
        }
    }
    {
        __statement_W4Rg36(49);
        this.templateDir = __extro_FSrivI(50, __intro_FT0Eil(50, fspath).normalize((__expression_dxZS4o(51), options.templateDir || (__expression_dxZS4o(52), (__expression_dxZS4o(53), (__expression_dxZS4o(54), __dirname) + '../../templates/fields/') + this.type))));
    }
    {
        __statement_W4Rg36(55);
        var defaultTemplates = {
                form: (__expression_dxZS4o(56), (__expression_dxZS4o(57), this.templateDir + '/') + 'form.jade'),
                initial: (__expression_dxZS4o(58), (__expression_dxZS4o(59), this.templateDir + '/') + 'initial.jade')
            };
    }
    {
        __statement_W4Rg36(60);
        this.templates = __extro_FSrivI(61, __intro_FT0Eil(61, utils).options(defaultTemplates, this.options.templates));
    }
    if (this.options.watch) {
        __block_nrL548(2);
        {
            __statement_W4Rg36(62);
            __extro_FSrivI(63, __intro_FT0Eil(63, this.list.schema).pre('save', __extro_FSrivI(64, __intro_FT0Eil(64, this).getPreSaveWatcher())));
        }
    }
    {
        __statement_W4Rg36(65);
        var note = null;
    }
    {
        __statement_W4Rg36(66);
        __extro_FSrivI(67, __intro_FT0Eil(67, Object).defineProperty(this, 'note', {
            get: function () {
                __block_nrL548(3);
                return __expression_dxZS4o(68), (__expression_dxZS4o(71), (__expression_dxZS4o(72), note) === null) ? (__expression_dxZS4o(69), note = this.options.note ? (__expression_dxZS4o(73), (__expression_dxZS4o(75), marked(this.options.note))) : (__expression_dxZS4o(74), '')) : (__expression_dxZS4o(70), note);
            }
        }));
    }
}
{
    __statement_W4Rg36(76);
    Field.prototype.getPreSaveWatcher = function () {
        __block_nrL548(4);
        {
            __statement_W4Rg36(77);
            var field = this, applyValue;
        }
        if (__expression_dxZS4o(78), this.options.watch === true) {
            __block_nrL548(5);
            {
                __statement_W4Rg36(79);
                applyValue = function () {
                    __block_nrL548(6);
                    return __expression_dxZS4o(80), true;
                };
            }
        } else {
            __block_nrL548(7);
            if (__extro_FSrivI(81, __intro_FT0Eil(81, _).isString(this.options.watch))) {
                __block_nrL548(8);
                {
                    __statement_W4Rg36(82);
                    this.options.watch = __extro_FSrivI(83, __intro_FT0Eil(83, this.options.watch).split(' '));
                }
            }
            if (__extro_FSrivI(84, __intro_FT0Eil(84, _).isFunction(this.options.watch))) {
                __block_nrL548(9);
                {
                    __statement_W4Rg36(85);
                    applyValue = this.options.watch;
                }
            } else if (__extro_FSrivI(86, __intro_FT0Eil(86, _).isArray(this.options.watch))) {
                __block_nrL548(10);
                {
                    __statement_W4Rg36(87);
                    applyValue = function (item) {
                        __block_nrL548(11);
                        {
                            __statement_W4Rg36(88);
                            var pass = false;
                        }
                        {
                            __statement_W4Rg36(89);
                            __extro_FSrivI(90, __intro_FT0Eil(90, field.options.watch).forEach(function (path) {
                                __block_nrL548(12);
                                if (__extro_FSrivI(91, __intro_FT0Eil(91, item).isModified(path))) {
                                    __block_nrL548(13);
                                    {
                                        __statement_W4Rg36(92);
                                        pass = true;
                                    }
                                }
                            }));
                        }
                        return __expression_dxZS4o(93), pass;
                    };
                }
            } else if (__extro_FSrivI(94, __intro_FT0Eil(94, _).isObject(this.options.watch))) {
                __block_nrL548(14);
                {
                    __statement_W4Rg36(95);
                    applyValue = function (item) {
                        __block_nrL548(15);
                        {
                            __statement_W4Rg36(96);
                            var pass = false;
                        }
                        {
                            __statement_W4Rg36(97);
                            __extro_FSrivI(98, __intro_FT0Eil(98, _).each(field.options.watch, function (value, path) {
                                __block_nrL548(16);
                                if (__expression_dxZS4o(99), __extro_FSrivI(100, __intro_FT0Eil(100, item).isModified(path)) && (__expression_dxZS4o(101), __extro_FSrivI(102, __intro_FT0Eil(102, item).get(path)) === (__expression_dxZS4o(103), value))) {
                                    __block_nrL548(17);
                                    {
                                        __statement_W4Rg36(104);
                                        pass = true;
                                    }
                                }
                            }));
                        }
                        return __expression_dxZS4o(105), pass;
                    };
                }
            }
        }
        if (__expression_dxZS4o(106), !(__expression_dxZS4o(107), applyValue)) {
            __block_nrL548(18);
            {
                __statement_W4Rg36(108);
                __extro_FSrivI(109, __intro_FT0Eil(109, console).error((__expression_dxZS4o(110), (__expression_dxZS4o(111), (__expression_dxZS4o(112), (__expression_dxZS4o(113), (__expression_dxZS4o(114), (__expression_dxZS4o(115), (__expression_dxZS4o(116), (__expression_dxZS4o(117), (__expression_dxZS4o(118), '\nError: Invalid Configuration\n\n' + 'Invalid watch value (') + this.options.watch) + ') provided for ') + this.list.key) + '.') + this.path) + ' (') + this.type) + ')')));
            }
            {
                __statement_W4Rg36(119);
                __extro_FSrivI(120, __intro_FT0Eil(120, process).exit(1));
            }
        }
        if (__expression_dxZS4o(121), !__extro_FSrivI(122, __intro_FT0Eil(122, _).isFunction(this.options.value))) {
            __block_nrL548(19);
            {
                __statement_W4Rg36(123);
                __extro_FSrivI(124, __intro_FT0Eil(124, console).error((__expression_dxZS4o(125), (__expression_dxZS4o(126), (__expression_dxZS4o(127), (__expression_dxZS4o(128), (__expression_dxZS4o(129), (__expression_dxZS4o(130), (__expression_dxZS4o(131), '\nError: Invalid Configuration\n\n' + 'Watch set with no value method provided for ') + this.list.key) + '.') + this.path) + ' (') + this.type) + ')')));
            }
            {
                __statement_W4Rg36(132);
                __extro_FSrivI(133, __intro_FT0Eil(133, process).exit(1));
            }
        }
        return __expression_dxZS4o(134), function (next) {
            __block_nrL548(20);
            if (__expression_dxZS4o(135), !(__expression_dxZS4o(136), applyValue(this))) {
                __block_nrL548(21);
                return __expression_dxZS4o(137), (__expression_dxZS4o(138), next());
            }
            {
                __statement_W4Rg36(139);
                __extro_FSrivI(140, __intro_FT0Eil(140, this).set(field.path, __extro_FSrivI(141, __intro_FT0Eil(141, field.options.value).call(this))));
            }
            {
                __statement_W4Rg36(142);
                __expression_dxZS4o(143), next();
            }
        };
    };
}
{
    __statement_W4Rg36(144);
    exports = module.exports = Field;
}
{
    __statement_W4Rg36(145);
    __extro_FSrivI(146, __intro_FT0Eil(146, Object).defineProperty(Field.prototype, 'width', {
        get: function () {
            __block_nrL548(22);
            return __expression_dxZS4o(147), (__expression_dxZS4o(148), this.options.width || 'full');
        }
    }));
}
{
    __statement_W4Rg36(149);
    __extro_FSrivI(150, __intro_FT0Eil(150, Object).defineProperty(Field.prototype, 'initial', {
        get: function () {
            __block_nrL548(23);
            return __expression_dxZS4o(151), (__expression_dxZS4o(152), this.options.initial || false);
        }
    }));
}
{
    __statement_W4Rg36(153);
    __extro_FSrivI(154, __intro_FT0Eil(154, Object).defineProperty(Field.prototype, 'required', {
        get: function () {
            __block_nrL548(24);
            return __expression_dxZS4o(155), (__expression_dxZS4o(156), this.options.required || false);
        }
    }));
}
{
    __statement_W4Rg36(157);
    __extro_FSrivI(158, __intro_FT0Eil(158, Object).defineProperty(Field.prototype, 'col', {
        get: function () {
            __block_nrL548(25);
            return __expression_dxZS4o(159), (__expression_dxZS4o(160), this.options.col || false);
        }
    }));
}
{
    __statement_W4Rg36(161);
    __extro_FSrivI(162, __intro_FT0Eil(162, Object).defineProperty(Field.prototype, 'noedit', {
        get: function () {
            __block_nrL548(26);
            return __expression_dxZS4o(163), (__expression_dxZS4o(164), this.options.noedit || false);
        }
    }));
}
{
    __statement_W4Rg36(165);
    __extro_FSrivI(166, __intro_FT0Eil(166, Object).defineProperty(Field.prototype, 'nocol', {
        get: function () {
            __block_nrL548(27);
            return __expression_dxZS4o(167), (__expression_dxZS4o(168), this.options.nocol || false);
        }
    }));
}
{
    __statement_W4Rg36(169);
    __extro_FSrivI(170, __intro_FT0Eil(170, Object).defineProperty(Field.prototype, 'nosort', {
        get: function () {
            __block_nrL548(28);
            return __expression_dxZS4o(171), (__expression_dxZS4o(172), this.options.nosort || false);
        }
    }));
}
{
    __statement_W4Rg36(173);
    __extro_FSrivI(174, __intro_FT0Eil(174, Object).defineProperty(Field.prototype, 'nofilter', {
        get: function () {
            __block_nrL548(29);
            return __expression_dxZS4o(175), (__expression_dxZS4o(176), this.options.nofilter || false);
        }
    }));
}
{
    __statement_W4Rg36(177);
    __extro_FSrivI(178, __intro_FT0Eil(178, Object).defineProperty(Field.prototype, 'collapse', {
        get: function () {
            __block_nrL548(30);
            return __expression_dxZS4o(179), (__expression_dxZS4o(180), this.options.collapse || false);
        }
    }));
}
{
    __statement_W4Rg36(181);
    __extro_FSrivI(182, __intro_FT0Eil(182, Object).defineProperty(Field.prototype, 'hidden', {
        get: function () {
            __block_nrL548(31);
            return __expression_dxZS4o(183), (__expression_dxZS4o(184), this.options.hidden || false);
        }
    }));
}
{
    __statement_W4Rg36(185);
    __extro_FSrivI(186, __intro_FT0Eil(186, Object).defineProperty(Field.prototype, 'dependsOn', {
        get: function () {
            __block_nrL548(32);
            return __expression_dxZS4o(187), (__expression_dxZS4o(188), this.options.dependsOn || false);
        }
    }));
}
{
    __statement_W4Rg36(189);
    Field.prototype.addToSchema = function () {
        __block_nrL548(33);
        {
            __statement_W4Rg36(190);
            var ops = this._nativeType ? (__expression_dxZS4o(191), __extro_FSrivI(193, __intro_FT0Eil(193, _).defaults({
                    type: this._nativeType
                }, this.options))) : (__expression_dxZS4o(192), this.options);
        }
        {
            __statement_W4Rg36(194);
            __extro_FSrivI(195, __intro_FT0Eil(195, this.list.schema).path(this.path, ops));
        }
        {
            __statement_W4Rg36(196);
            __extro_FSrivI(197, __intro_FT0Eil(197, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_W4Rg36(198);
    Field.prototype.bindUnderscoreMethods = function (methods) {
        __block_nrL548(34);
        {
            __statement_W4Rg36(199);
            var field = this;
        }
        {
            __statement_W4Rg36(200);
            __extro_FSrivI(201, __intro_FT0Eil(201, __extro_FSrivI(202, __intro_FT0Eil(202, (__expression_dxZS4o(203), this._underscoreMethods || [])).concat({
                fn: 'updateItem',
                as: 'update'
            }, (__expression_dxZS4o(204), (__expression_dxZS4o(205), methods) || [])))).forEach(function (method) {
                __block_nrL548(35);
                if (__expression_dxZS4o(206), 'string' === (__expression_dxZS4o(207), typeof method)) {
                    __block_nrL548(36);
                    {
                        __statement_W4Rg36(208);
                        method = {
                            fn: method,
                            as: method
                        };
                    }
                }
                if (__expression_dxZS4o(209), 'function' !== (__expression_dxZS4o(210), typeof field[method.fn])) {
                    __block_nrL548(37);
                    {
                        __statement_W4Rg36(211);
                        throw new Error((__expression_dxZS4o(212), (__expression_dxZS4o(213), (__expression_dxZS4o(214), (__expression_dxZS4o(215), (__expression_dxZS4o(216), (__expression_dxZS4o(217), (__expression_dxZS4o(218), (__expression_dxZS4o(219), 'Invalid underscore method (' + method.fn) + ') applied to ') + field.list.key) + '.') + field.path) + ' (') + field.type) + ')'));
                    }
                }
                {
                    __statement_W4Rg36(220);
                    __extro_FSrivI(221, __intro_FT0Eil(221, field).underscoreMethod(method.as, function () {
                        __block_nrL548(38);
                        {
                            __statement_W4Rg36(222);
                            var args = __extro_FSrivI(223, __intro_FT0Eil(223, [
                                    this
                                ]).concat(__extro_FSrivI(224, __intro_FT0Eil(224, Array.prototype.slice).call(arguments))));
                        }
                        return __expression_dxZS4o(225), __extro_FSrivI(226, __intro_FT0Eil(226, field[method.fn]).apply(field, args));
                    }));
                }
            }));
        }
    };
}
{
    __statement_W4Rg36(227);
    Field.prototype.underscoreMethod = function (path, fn) {
        __block_nrL548(39);
        {
            __statement_W4Rg36(228);
            __extro_FSrivI(229, __intro_FT0Eil(229, this.list).underscoreMethod((__expression_dxZS4o(230), (__expression_dxZS4o(231), this.path + '.') + (__expression_dxZS4o(232), path)), function () {
                __block_nrL548(40);
                return __expression_dxZS4o(233), __extro_FSrivI(234, __intro_FT0Eil(234, fn).apply(this, arguments));
            }));
        }
    };
}
{
    __statement_W4Rg36(235);
    Field.prototype.format = function (item) {
        __block_nrL548(41);
        return __expression_dxZS4o(236), __extro_FSrivI(237, __intro_FT0Eil(237, item).get(this.path));
    };
}
{
    __statement_W4Rg36(238);
    Field.prototype.isModified = function (item) {
        __block_nrL548(42);
        return __expression_dxZS4o(239), __extro_FSrivI(240, __intro_FT0Eil(240, item).isModified(this.path));
    };
}
{
    __statement_W4Rg36(241);
    Field.prototype.validateInput = function (data, required, item) {
        __block_nrL548(43);
        if (__expression_dxZS4o(242), !(__expression_dxZS4o(243), required)) {
            __block_nrL548(44);
            return __expression_dxZS4o(244), true;
        }
        if (__expression_dxZS4o(245), (__expression_dxZS4o(246), (__expression_dxZS4o(247), !(__expression_dxZS4o(248), (this.path in (__expression_dxZS4o(249), data)))) && (__expression_dxZS4o(250), item)) && __extro_FSrivI(251, __intro_FT0Eil(251, item).get(this.path))) {
            __block_nrL548(45);
            return __expression_dxZS4o(252), true;
        }
        if (__expression_dxZS4o(253), 'string' === (__expression_dxZS4o(254), typeof data[this.path])) {
            __block_nrL548(46);
            return __expression_dxZS4o(255), __extro_FSrivI(258, __intro_FT0Eil(258, data[this.path]).trim()) ? (__expression_dxZS4o(256), true) : (__expression_dxZS4o(257), false);
        } else {
            __block_nrL548(47);
            return __expression_dxZS4o(259), data[this.path] ? (__expression_dxZS4o(260), true) : (__expression_dxZS4o(261), false);
        }
    };
}
{
    __statement_W4Rg36(262);
    Field.prototype.updateItem = function (item, data) {
        __block_nrL548(48);
        {
            __statement_W4Rg36(263);
            var value = __extro_FSrivI(264, __intro_FT0Eil(264, this).getValueFromData(data));
        }
        if (__expression_dxZS4o(265), (__expression_dxZS4o(266), (__expression_dxZS4o(267), value) !== (__expression_dxZS4o(268), undefined)) && (__expression_dxZS4o(269), (__expression_dxZS4o(270), value) != __extro_FSrivI(271, __intro_FT0Eil(271, item).get(this.path)))) {
            __block_nrL548(49);
            {
                __statement_W4Rg36(272);
                __extro_FSrivI(273, __intro_FT0Eil(273, item).set(this.path, value));
            }
        }
    };
}
{
    __statement_W4Rg36(274);
    Field.prototype.getValueFromData = function (data) {
        __block_nrL548(50);
        return __expression_dxZS4o(275), (__expression_dxZS4o(278), (this.path in (__expression_dxZS4o(279), data))) ? (__expression_dxZS4o(276), data[this.path]) : (__expression_dxZS4o(277), __extro_FSrivI(280, __intro_FT0Eil(280, this._path).get(data)));
    };
}
{
    __statement_W4Rg36(281);
    Field.prototype.compile = function (type, callback) {
        __block_nrL548(51);
        {
            __statement_W4Rg36(282);
            var templatePath = this.templates[type];
        }
        if (__expression_dxZS4o(283), !compiledTemplates[templatePath]) {
            __block_nrL548(52);
            {
                __statement_W4Rg36(284);
                __extro_FSrivI(285, __intro_FT0Eil(285, fs).readFile(templatePath, 'utf8', function (err, file) {
                    __block_nrL548(53);
                    if (__expression_dxZS4o(286), !(__expression_dxZS4o(287), err)) {
                        __block_nrL548(54);
                        {
                            __statement_W4Rg36(288);
                            compiledTemplates[templatePath] = __extro_FSrivI(289, __intro_FT0Eil(289, jade).compile(file, {
                                filename: templatePath,
                                pretty: (__expression_dxZS4o(290), __extro_FSrivI(291, __intro_FT0Eil(291, keystone).get('env')) !== 'production')
                            }));
                        }
                    }
                    if (__expression_dxZS4o(292), callback) {
                        __block_nrL548(55);
                        return __expression_dxZS4o(293), (__expression_dxZS4o(294), callback());
                    }
                }));
            }
        } else if (__expression_dxZS4o(295), callback) {
            __block_nrL548(56);
            return __expression_dxZS4o(296), (__expression_dxZS4o(297), callback());
        }
    };
}
{
    __statement_W4Rg36(298);
    Field.prototype.render = function (type, item, locals) {
        __block_nrL548(57);
        {
            __statement_W4Rg36(299);
            var templatePath = this.templates[type];
        }
        if (__expression_dxZS4o(300), !compiledTemplates[templatePath]) {
            __block_nrL548(58);
            {
                __statement_W4Rg36(301);
                var file = __extro_FSrivI(302, __intro_FT0Eil(302, fs).readFileSync(templatePath, 'utf8'));
            }
            {
                __statement_W4Rg36(303);
                compiledTemplates[templatePath] = __extro_FSrivI(304, __intro_FT0Eil(304, jade).compile(file, {
                    filename: templatePath,
                    pretty: (__expression_dxZS4o(305), __extro_FSrivI(306, __intro_FT0Eil(306, keystone).get('env')) !== 'production')
                }));
            }
        }
        return __expression_dxZS4o(307), __extro_FSrivI(308, __intro_FT0Eil(308, compiledTemplates)[templatePath](__extro_FSrivI(309, __intro_FT0Eil(309, _).extend((__expression_dxZS4o(310), (__expression_dxZS4o(311), locals) || {}), {
            field: this,
            item: item
        }))));
    };
}
