// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_datetime.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_vX0qDi, __expression_JDP20M, __block_qwVMhc;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_vX0qDi = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/datetime.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_JDP20M = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/datetime.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_qwVMhc = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/datetime.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_T_vT3f = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_zWv04B = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/datetime.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_vX0qDi(0);
    var util = (__expression_JDP20M(1), require('util')), moment = (__expression_JDP20M(2), require('moment')), super_ = (__expression_JDP20M(3), require('../field'));
}
{
    __statement_vX0qDi(4);
    var parseFormats = [
            'YYYY-MM-DD',
            'YYYY-MM-DD h:m:s a',
            'YYYY-MM-DD h:m a',
            'YYYY-MM-DD H:m:s',
            'YYYY-MM-DD H:m'
        ];
}
function datetime(list, path, options) {
    __block_qwVMhc(0);
    {
        __statement_vX0qDi(5);
        this._nativeType = Date;
    }
    {
        __statement_vX0qDi(6);
        this._underscoreMethods = [
            'format',
            'moment',
            'parse'
        ];
    }
    {
        __statement_vX0qDi(7);
        this.typeDescription = 'date and time';
    }
    {
        __statement_vX0qDi(8);
        this._formatString = (__expression_JDP20M(11), options.format === false) ? (__expression_JDP20M(9), false) : (__expression_JDP20M(10), (__expression_JDP20M(12), options.format || 'Do MMM YYYY hh:mm:ss a'));
    }
    if (__expression_JDP20M(13), this._formatString && (__expression_JDP20M(14), 'string' !== (__expression_JDP20M(15), typeof this._formatString))) {
        __block_qwVMhc(1);
        {
            __statement_vX0qDi(16);
            throw new Error('FieldType.DateTime: options.format must be a string.');
        }
    }
    {
        __statement_vX0qDi(17);
        __extro_zWv04B(18, __intro_T_vT3f(18, datetime.super_).call(this, list, path, options));
    }
    {
        __statement_vX0qDi(19);
        this.paths = {
            date: __extro_zWv04B(20, __intro_T_vT3f(20, this._path).append('_date')),
            time: __extro_zWv04B(21, __intro_T_vT3f(21, this._path).append('_time'))
        };
    }
}
{
    __statement_vX0qDi(22);
    __extro_zWv04B(23, __intro_T_vT3f(23, util).inherits(datetime, super_));
}
{
    __statement_vX0qDi(24);
    datetime.prototype.format = function (item, format) {
        __block_qwVMhc(2);
        if (__expression_JDP20M(25), (__expression_JDP20M(26), format) || this._formatString) {
            __block_qwVMhc(3);
            return __expression_JDP20M(27), __extro_zWv04B(30, __intro_T_vT3f(30, item).get(this.path)) ? (__expression_JDP20M(28), __extro_zWv04B(31, __intro_T_vT3f(31, (__expression_JDP20M(32), moment(__extro_zWv04B(33, __intro_T_vT3f(33, item).get(this.path))))).format((__expression_JDP20M(34), (__expression_JDP20M(35), format) || this._formatString)))) : (__expression_JDP20M(29), '');
        } else {
            __block_qwVMhc(4);
            return __expression_JDP20M(36), (__expression_JDP20M(37), __extro_zWv04B(38, __intro_T_vT3f(38, item).get(this.path)) || '');
        }
    };
}
{
    __statement_vX0qDi(39);
    datetime.prototype.moment = function (item) {
        __block_qwVMhc(5);
        return __expression_JDP20M(40), (__expression_JDP20M(41), moment(__extro_zWv04B(42, __intro_T_vT3f(42, item).get(this.path))));
    };
}
{
    __statement_vX0qDi(43);
    datetime.prototype.parse = function (item) {
        __block_qwVMhc(6);
        {
            __statement_vX0qDi(44);
            var newValue = __extro_zWv04B(45, __intro_T_vT3f(45, moment).apply(moment, __extro_zWv04B(46, __intro_T_vT3f(46, Array.prototype.slice).call(arguments, 1))));
        }
        {
            __statement_vX0qDi(47);
            __extro_zWv04B(48, __intro_T_vT3f(48, item).set(this.path, (__expression_JDP20M(51), (__expression_JDP20M(52), newValue) && __extro_zWv04B(53, __intro_T_vT3f(53, newValue).isValid())) ? (__expression_JDP20M(49), __extro_zWv04B(54, __intro_T_vT3f(54, newValue).toDate())) : (__expression_JDP20M(50), null)));
        }
        return __expression_JDP20M(55), newValue;
    };
}
{
    __statement_vX0qDi(56);
    datetime.prototype.getInputFromData = function (data) {
        __block_qwVMhc(7);
        if (__expression_JDP20M(57), (__expression_JDP20M(58), (this.paths.date in (__expression_JDP20M(59), data))) && (__expression_JDP20M(60), (this.paths.time in (__expression_JDP20M(61), data)))) {
            __block_qwVMhc(8);
            return __expression_JDP20M(62), __extro_zWv04B(63, __intro_T_vT3f(63, (__expression_JDP20M(64), (__expression_JDP20M(65), data[this.paths.date] + ' ') + data[this.paths.time])).trim());
        } else {
            __block_qwVMhc(9);
            return __expression_JDP20M(66), data[this.path];
        }
    };
}
{
    __statement_vX0qDi(67);
    datetime.prototype.validateInput = function (data, required, item) {
        __block_qwVMhc(10);
        if (__expression_JDP20M(68), (__expression_JDP20M(69), (__expression_JDP20M(70), !(__expression_JDP20M(71), (__expression_JDP20M(72), (this.path in (__expression_JDP20M(73), data))) && (__expression_JDP20M(74), !(__expression_JDP20M(75), (__expression_JDP20M(76), (this.paths.date in (__expression_JDP20M(77), data))) && (__expression_JDP20M(78), (this.paths.time in (__expression_JDP20M(79), data))))))) && (__expression_JDP20M(80), item)) && __extro_zWv04B(81, __intro_T_vT3f(81, item).get(this.path))) {
            __block_qwVMhc(11);
            return __expression_JDP20M(82), true;
        }
        {
            __statement_vX0qDi(83);
            var newValue = (__expression_JDP20M(84), moment(__extro_zWv04B(85, __intro_T_vT3f(85, this).getInputFromData(data)), parseFormats));
        }
        if (__expression_JDP20M(86), (__expression_JDP20M(87), required) && (__expression_JDP20M(88), (__expression_JDP20M(89), !(__expression_JDP20M(90), newValue)) || (__expression_JDP20M(91), !__extro_zWv04B(92, __intro_T_vT3f(92, newValue).isValid())))) {
            __block_qwVMhc(12);
            return __expression_JDP20M(93), false;
        } else if (__expression_JDP20M(94), (__expression_JDP20M(95), __extro_zWv04B(96, __intro_T_vT3f(96, this).getInputFromData(data)) && (__expression_JDP20M(97), newValue)) && (__expression_JDP20M(98), !__extro_zWv04B(99, __intro_T_vT3f(99, newValue).isValid()))) {
            __block_qwVMhc(13);
            return __expression_JDP20M(100), false;
        } else {
            __block_qwVMhc(14);
            return __expression_JDP20M(101), true;
        }
    };
}
{
    __statement_vX0qDi(102);
    datetime.prototype.updateItem = function (item, data) {
        __block_qwVMhc(15);
        if (__expression_JDP20M(103), !(__expression_JDP20M(104), (__expression_JDP20M(105), (this.path in (__expression_JDP20M(106), data))) || (__expression_JDP20M(107), (__expression_JDP20M(108), (this.paths.date in (__expression_JDP20M(109), data))) && (__expression_JDP20M(110), (this.paths.time in (__expression_JDP20M(111), data)))))) {
            __block_qwVMhc(16);
            return __expression_JDP20M(112);
        }
        {
            __statement_vX0qDi(113);
            var newValue = (__expression_JDP20M(114), moment(__extro_zWv04B(115, __intro_T_vT3f(115, this).getInputFromData(data)), parseFormats));
        }
        if (__expression_JDP20M(116), (__expression_JDP20M(117), newValue) && __extro_zWv04B(118, __intro_T_vT3f(118, newValue).isValid())) {
            __block_qwVMhc(17);
            if (__expression_JDP20M(119), (__expression_JDP20M(120), !__extro_zWv04B(121, __intro_T_vT3f(121, item).get(this.path))) || (__expression_JDP20M(122), !__extro_zWv04B(123, __intro_T_vT3f(123, newValue).isSame(__extro_zWv04B(124, __intro_T_vT3f(124, item).get(this.path)))))) {
                __block_qwVMhc(18);
                {
                    __statement_vX0qDi(125);
                    __extro_zWv04B(126, __intro_T_vT3f(126, item).set(this.path, __extro_zWv04B(127, __intro_T_vT3f(127, newValue).toDate())));
                }
            }
        } else if (__extro_zWv04B(128, __intro_T_vT3f(128, item).get(this.path))) {
            __block_qwVMhc(19);
            {
                __statement_vX0qDi(129);
                __extro_zWv04B(130, __intro_T_vT3f(130, item).set(this.path, null));
            }
        }
    };
}
{
    __statement_vX0qDi(131);
    exports = module.exports = datetime;
}
