// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_name.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_sLOEIA, __expression_WY74v5, __block_rD4rjq;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_sLOEIA = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/name.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_WY74v5 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/name.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_rD4rjq = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/name.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_$WrJBL = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_GwxytT = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/name.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_sLOEIA(0);
    var _ = (__expression_WY74v5(1), require('underscore')), util = (__expression_WY74v5(2), require('util')), super_ = (__expression_WY74v5(3), require('../field'));
}
function name(list, path, options) {
    __block_rD4rjq(0);
    {
        __statement_sLOEIA(4);
        options.nofilter = true;
    }
    {
        __statement_sLOEIA(5);
        __extro_GwxytT(6, __intro_$WrJBL(6, name.super_).call(this, list, path, options));
    }
}
{
    __statement_sLOEIA(7);
    __extro_GwxytT(8, __intro_$WrJBL(8, util).inherits(name, super_));
}
{
    __statement_sLOEIA(9);
    name.prototype.addToSchema = function () {
        __block_rD4rjq(1);
        {
            __statement_sLOEIA(10);
            var schema = this.list.schema;
        }
        {
            __statement_sLOEIA(11);
            var paths = this.paths = {
                    first: __extro_GwxytT(12, __intro_$WrJBL(12, this._path).append('.first')),
                    last: __extro_GwxytT(13, __intro_$WrJBL(13, this._path).append('.last')),
                    full: __extro_GwxytT(14, __intro_$WrJBL(14, this._path).append('.full'))
                };
        }
        {
            __statement_sLOEIA(15);
            schema.nested[this.path] = true;
        }
        {
            __statement_sLOEIA(16);
            __extro_GwxytT(17, __intro_$WrJBL(17, schema).add({
                first: String,
                last: String
            }, (__expression_WY74v5(18), this.path + '.')));
        }
        {
            __statement_sLOEIA(19);
            __extro_GwxytT(20, __intro_$WrJBL(20, __extro_GwxytT(21, __intro_$WrJBL(21, schema).virtual(paths.full))).get(function () {
                __block_rD4rjq(2);
                return __expression_WY74v5(22), __extro_GwxytT(23, __intro_$WrJBL(23, __extro_GwxytT(24, __intro_$WrJBL(24, _).compact([
                    __extro_GwxytT(25, __intro_$WrJBL(25, this).get(paths.first)),
                    __extro_GwxytT(26, __intro_$WrJBL(26, this).get(paths.last))
                ]))).join(' '));
            }));
        }
        {
            __statement_sLOEIA(27);
            __extro_GwxytT(28, __intro_$WrJBL(28, __extro_GwxytT(29, __intro_$WrJBL(29, schema).virtual(paths.full))).set(function (value) {
                __block_rD4rjq(3);
                if (__expression_WY74v5(30), (__expression_WY74v5(31), typeof value) !== 'string') {
                    __block_rD4rjq(4);
                    {
                        __statement_sLOEIA(32);
                        __extro_GwxytT(33, __intro_$WrJBL(33, this).set(paths.first, undefined));
                    }
                    {
                        __statement_sLOEIA(34);
                        __extro_GwxytT(35, __intro_$WrJBL(35, this).set(paths.last, undefined));
                    }
                    return __expression_WY74v5(36);
                }
                {
                    __statement_sLOEIA(37);
                    var split = __extro_GwxytT(38, __intro_$WrJBL(38, value).split(' '));
                }
                {
                    __statement_sLOEIA(39);
                    __extro_GwxytT(40, __intro_$WrJBL(40, this).set(paths.first, __extro_GwxytT(41, __intro_$WrJBL(41, split).shift())));
                }
                {
                    __statement_sLOEIA(42);
                    __extro_GwxytT(43, __intro_$WrJBL(43, this).set(paths.last, (__expression_WY74v5(44), __extro_GwxytT(45, __intro_$WrJBL(45, split).join(' ')) || (__expression_WY74v5(46), undefined))));
                }
            }));
        }
        {
            __statement_sLOEIA(47);
            __extro_GwxytT(48, __intro_$WrJBL(48, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_sLOEIA(49);
    name.prototype.format = function (item) {
        __block_rD4rjq(5);
        return __expression_WY74v5(50), __extro_GwxytT(51, __intro_$WrJBL(51, item).get(this.paths.full));
    };
}
{
    __statement_sLOEIA(52);
    name.prototype.validateInput = function (data, required, item) {
        __block_rD4rjq(6);
        if (__expression_WY74v5(53), (__expression_WY74v5(54), (__expression_WY74v5(55), !(__expression_WY74v5(56), (__expression_WY74v5(57), (__expression_WY74v5(58), (__expression_WY74v5(59), (this.path in (__expression_WY74v5(60), data))) || (__expression_WY74v5(61), (this.paths.first in (__expression_WY74v5(62), data)))) || (__expression_WY74v5(63), (this.paths.last in (__expression_WY74v5(64), data)))) || (__expression_WY74v5(65), (this.paths.full in (__expression_WY74v5(66), data))))) && (__expression_WY74v5(67), item)) && __extro_GwxytT(68, __intro_$WrJBL(68, item).get(this.paths.full))) {
            __block_rD4rjq(7);
            return __expression_WY74v5(69), true;
        }
        if (__expression_WY74v5(70), !(__expression_WY74v5(71), required)) {
            __block_rD4rjq(8);
            return __expression_WY74v5(72), true;
        }
        if (__extro_GwxytT(73, __intro_$WrJBL(73, _).isObject(data[this.path]))) {
            __block_rD4rjq(9);
            return __expression_WY74v5(74), (__expression_WY74v5(77), (__expression_WY74v5(78), data[this.path].full || data[this.path].first) || data[this.path].last) ? (__expression_WY74v5(75), true) : (__expression_WY74v5(76), false);
        } else {
            __block_rD4rjq(10);
            return __expression_WY74v5(79), (__expression_WY74v5(82), (__expression_WY74v5(83), data[this.paths.full] || data[this.paths.first]) || data[this.paths.last]) ? (__expression_WY74v5(80), true) : (__expression_WY74v5(81), false);
        }
    };
}
{
    __statement_sLOEIA(84);
    name.prototype.isModified = function (item) {
        __block_rD4rjq(11);
        return __expression_WY74v5(85), (__expression_WY74v5(86), __extro_GwxytT(87, __intro_$WrJBL(87, item).isModified(this.paths.first)) || __extro_GwxytT(88, __intro_$WrJBL(88, item).isModified(this.paths.last)));
    };
}
{
    __statement_sLOEIA(89);
    name.prototype.updateItem = function (item, data) {
        __block_rD4rjq(12);
        if (__expression_WY74v5(90), !__extro_GwxytT(91, __intro_$WrJBL(91, _).isObject(data))) {
            __block_rD4rjq(13);
            return __expression_WY74v5(92);
        }
        {
            __statement_sLOEIA(93);
            var paths = this.paths, setValue;
        }
        if (__expression_WY74v5(94), (__expression_WY74v5(95), (this.path in (__expression_WY74v5(96), data))) && __extro_GwxytT(97, __intro_$WrJBL(97, _).isString(data[this.path]))) {
            __block_rD4rjq(14);
            {
                __statement_sLOEIA(98);
                __extro_GwxytT(99, __intro_$WrJBL(99, item).set(paths.full, data[this.path]));
            }
        } else if (__expression_WY74v5(100), (__expression_WY74v5(101), (this.path in (__expression_WY74v5(102), data))) && __extro_GwxytT(103, __intro_$WrJBL(103, _).isObject(data[this.path]))) {
            __block_rD4rjq(15);
            {
                __statement_sLOEIA(104);
                var valueObj = data[this.path];
            }
            {
                __statement_sLOEIA(105);
                setValue = function (key) {
                    __block_rD4rjq(16);
                    if (__expression_WY74v5(106), (__expression_WY74v5(107), ((__expression_WY74v5(108), key) in (__expression_WY74v5(109), valueObj))) && (__expression_WY74v5(110), valueObj[key] !== __extro_GwxytT(111, __intro_$WrJBL(111, item).get(paths[key])))) {
                        __block_rD4rjq(17);
                        {
                            __statement_sLOEIA(112);
                            __extro_GwxytT(113, __intro_$WrJBL(113, item).set(paths[key], valueObj[key]));
                        }
                    }
                };
            }
        } else {
            __block_rD4rjq(18);
            {
                __statement_sLOEIA(114);
                setValue = function (key) {
                    __block_rD4rjq(19);
                    if (__expression_WY74v5(115), (__expression_WY74v5(116), (paths[key] in (__expression_WY74v5(117), data))) && (__expression_WY74v5(118), data[paths[key]] !== __extro_GwxytT(119, __intro_$WrJBL(119, item).get(paths[key])))) {
                        __block_rD4rjq(20);
                        {
                            __statement_sLOEIA(120);
                            __extro_GwxytT(121, __intro_$WrJBL(121, item).set(paths[key], data[paths[key]]));
                        }
                    }
                };
            }
        }
        if (__expression_WY74v5(122), setValue) {
            __block_rD4rjq(21);
            {
                __statement_sLOEIA(123);
                __extro_GwxytT(124, __intro_$WrJBL(124, _).each([
                    'full',
                    'first',
                    'last'
                ], setValue));
            }
        }
    };
}
{
    __statement_sLOEIA(125);
    exports = module.exports = name;
}
