// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_email.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_cP8ZeA, __expression_PjknUI, __block__SZGrN;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_cP8ZeA = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/email.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_PjknUI = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/email.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block__SZGrN = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/email.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_nZgH1D = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_QWvo6o = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/email.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_cP8ZeA(0);
    var util = (__expression_PjknUI(1), require('util')), utils = (__expression_PjknUI(2), require('keystone-utils')), super_ = (__expression_PjknUI(3), require('../field')), crypto = (__expression_PjknUI(4), require('crypto'));
}
function email(list, path, options) {
    __block__SZGrN(0);
    {
        __statement_cP8ZeA(5);
        this._nativeType = String;
    }
    {
        __statement_cP8ZeA(6);
        this._underscoreMethods = [
            'gravatarUrl'
        ];
    }
    {
        __statement_cP8ZeA(7);
        this.typeDescription = 'email address';
    }
    {
        __statement_cP8ZeA(8);
        __extro_QWvo6o(9, __intro_nZgH1D(9, email.super_).call(this, list, path, options));
    }
}
{
    __statement_cP8ZeA(10);
    __extro_QWvo6o(11, __intro_nZgH1D(11, util).inherits(email, super_));
}
{
    __statement_cP8ZeA(12);
    email.prototype.validateInput = function (data, required, item) {
        __block__SZGrN(1);
        if (data[this.path]) {
            __block__SZGrN(2);
            return __expression_PjknUI(13), __extro_QWvo6o(14, __intro_nZgH1D(14, utils).isEmail(data[this.path]));
        } else {
            __block__SZGrN(3);
            return __expression_PjknUI(15), (__expression_PjknUI(18), (__expression_PjknUI(19), !(__expression_PjknUI(20), required)) || (__expression_PjknUI(21), (__expression_PjknUI(22), (__expression_PjknUI(23), !(__expression_PjknUI(24), (this.path in (__expression_PjknUI(25), data)))) && (__expression_PjknUI(26), item)) && __extro_QWvo6o(27, __intro_nZgH1D(27, item).get(this.path)))) ? (__expression_PjknUI(16), true) : (__expression_PjknUI(17), false);
        }
    };
}
{
    __statement_cP8ZeA(28);
    email.prototype.updateItem = function (item, data) {
        __block__SZGrN(4);
        {
            __statement_cP8ZeA(29);
            var newValue = data[this.path];
        }
        if (__expression_PjknUI(30), 'string' === (__expression_PjknUI(31), typeof newValue)) {
            __block__SZGrN(5);
            {
                __statement_cP8ZeA(32);
                newValue = __extro_QWvo6o(33, __intro_nZgH1D(33, newValue).toLowerCase());
            }
        }
        if (__expression_PjknUI(34), (__expression_PjknUI(35), (this.path in (__expression_PjknUI(36), data))) && (__expression_PjknUI(37), data[this.path] !== __extro_QWvo6o(38, __intro_nZgH1D(38, item).get(this.path)))) {
            __block__SZGrN(6);
            {
                __statement_cP8ZeA(39);
                __extro_QWvo6o(40, __intro_nZgH1D(40, item).set(this.path, data[this.path]));
            }
        }
    };
}
{
    __statement_cP8ZeA(41);
    email.prototype.gravatarUrl = function (item, size, defaultImage, rating) {
        __block__SZGrN(7);
        {
            __statement_cP8ZeA(42);
            var value = __extro_QWvo6o(43, __intro_nZgH1D(43, item).get(this.path));
        }
        if (__expression_PjknUI(44), 'string' !== (__expression_PjknUI(45), typeof value)) {
            __block__SZGrN(8);
            return __expression_PjknUI(46), '';
        }
        return __expression_PjknUI(47), __extro_QWvo6o(48, __intro_nZgH1D(48, [
            '//www.gravatar.com/avatar/',
            __extro_QWvo6o(49, __intro_nZgH1D(49, __extro_QWvo6o(50, __intro_nZgH1D(50, __extro_QWvo6o(51, __intro_nZgH1D(51, crypto).createHash('md5'))).update(__extro_QWvo6o(52, __intro_nZgH1D(52, __extro_QWvo6o(53, __intro_nZgH1D(53, value).toLowerCase())).trim())))).digest('hex')),
            (__expression_PjknUI(54), '?s=' + (__extro_QWvo6o(57, __intro_nZgH1D(57, /^(?:[1-9][0-9]{0,2}|1[0-9]{3}|20[0-3][0-9]|204[0-8])$/).test(size)) ? (__expression_PjknUI(55), size) : (__expression_PjknUI(56), 80))),
            (__expression_PjknUI(58), '&d=' + ((__expression_PjknUI(61), defaultImage) ? (__expression_PjknUI(59), (__expression_PjknUI(62), encodeURIComponent(defaultImage))) : (__expression_PjknUI(60), 'identicon'))),
            (__expression_PjknUI(63), '&r=' + (__extro_QWvo6o(66, __intro_nZgH1D(66, /^(?:g|pg|r|x)$/i).test(rating)) ? (__expression_PjknUI(64), __extro_QWvo6o(67, __intro_nZgH1D(67, rating).toLowerCase())) : (__expression_PjknUI(65), 'g')))
        ]).join(''));
    };
}
{
    __statement_cP8ZeA(68);
    exports = module.exports = email;
}
