// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_content_page.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_W57Jy4, __expression_WWV8jh, __block_nvViJX;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_W57Jy4 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/page.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_WWV8jh = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/page.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_nvViJX = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/page.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_FxjW60 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_dMV_cs = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/page.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_W57Jy4(0);
    var _ = (__expression_WWV8jh(1), require('underscore')), keystone = (__expression_WWV8jh(2), require('../../')), utils = keystone.utils, Type = (__expression_WWV8jh(3), require('./type'));
}
function Page(key, options) {
    __block_nvViJX(0);
    if (__expression_WWV8jh(4), !(__expression_WWV8jh(5), this instanceof (__expression_WWV8jh(6), Page))) {
        __block_nvViJX(1);
        return __expression_WWV8jh(7), new Page(key, options);
    }
    {
        __statement_W57Jy4(8);
        var page = this;
    }
    {
        __statement_W57Jy4(9);
        this.options = __extro_dMV_cs(10, __intro_FxjW60(10, utils).options({}, options));
    }
    {
        __statement_W57Jy4(11);
        this.key = key;
    }
    {
        __statement_W57Jy4(12);
        this.fields = {};
    }
}
{
    __statement_W57Jy4(13);
    __extro_dMV_cs(14, __intro_FxjW60(14, Object).defineProperty(Page.prototype, 'name', {
        get: function () {
            __block_nvViJX(2);
            return __expression_WWV8jh(15), (__expression_WWV8jh(16), __extro_dMV_cs(17, __intro_FxjW60(17, this).get('name')) || __extro_dMV_cs(18, __intro_FxjW60(18, this).set('name', __extro_dMV_cs(19, __intro_FxjW60(19, utils).keyToLabel(this.key)))));
        }
    }));
}
{
    __statement_W57Jy4(20);
    Page.prototype.set = function (key, value) {
        __block_nvViJX(3);
        if (__expression_WWV8jh(21), arguments.length === 1) {
            __block_nvViJX(4);
            return __expression_WWV8jh(22), this.options[key];
        }
        {
            __statement_W57Jy4(23);
            this.options[key] = value;
        }
        return __expression_WWV8jh(24), value;
    };
}
{
    __statement_W57Jy4(25);
    Page.prototype.get = Page.prototype.set;
}
{
    __statement_W57Jy4(26);
    Page.prototype.add = function (fields) {
        __block_nvViJX(5);
        if (__expression_WWV8jh(27), !__extro_dMV_cs(28, __intro_FxjW60(28, utils).isObject(fields))) {
            __block_nvViJX(6);
            {
                __statement_W57Jy4(29);
                throw new Error('keystone.content.Page.add() Error: fields must be an object.');
            }
        }
        {
            __statement_W57Jy4(30);
            __extro_dMV_cs(31, __intro_FxjW60(31, _).each(fields, function (options, path) {
                __block_nvViJX(7);
                if (__expression_WWV8jh(32), 'function' === (__expression_WWV8jh(33), typeof options)) {
                    __block_nvViJX(8);
                    {
                        __statement_W57Jy4(34);
                        options = {
                            type: options
                        };
                    }
                }
                if (__expression_WWV8jh(35), 'function' !== (__expression_WWV8jh(36), typeof options.type)) {
                    __block_nvViJX(9);
                    {
                        __statement_W57Jy4(37);
                        throw new Error('Page fields must be specified with a type function');
                    }
                }
                if (__expression_WWV8jh(38), options.type.prototype.__proto__ !== Type.prototype) {
                    __block_nvViJX(10);
                    if (__expression_WWV8jh(39), options.type === (__expression_WWV8jh(40), String)) {
                        __block_nvViJX(11);
                        {
                            __statement_W57Jy4(41);
                            options.type = keystone.content.Types.Text;
                        }
                    } else {
                        __block_nvViJX(12);
                        {
                            __statement_W57Jy4(42);
                            throw new Error((__expression_WWV8jh(43), 'Unrecognised field constructor: ' + options.type));
                        }
                    }
                }
                {
                    __statement_W57Jy4(44);
                    this.fields[path] = new options.type(path, options);
                }
            }, this));
        }
        return __expression_WWV8jh(45), this;
    };
}
{
    __statement_W57Jy4(46);
    Page.prototype.register = function () {
        __block_nvViJX(13);
        return __expression_WWV8jh(47), __extro_dMV_cs(48, __intro_FxjW60(48, keystone.content).page(this.key, this));
    };
}
{
    __statement_W57Jy4(49);
    Page.prototype.populate = function (data) {
        __block_nvViJX(14);
        if (__expression_WWV8jh(50), !__extro_dMV_cs(51, __intro_FxjW60(51, utils).isObject(data))) {
            __block_nvViJX(15);
            {
                __statement_W57Jy4(52);
                data = {};
            }
        }
        return __expression_WWV8jh(53), data;
    };
}
{
    __statement_W57Jy4(54);
    Page.prototype.validate = function (data) {
        __block_nvViJX(16);
        if (__expression_WWV8jh(55), !__extro_dMV_cs(56, __intro_FxjW60(56, _).isObject(data))) {
            __block_nvViJX(17);
            {
                __statement_W57Jy4(57);
                data = {};
            }
        }
        return __expression_WWV8jh(58), data;
    };
}
{
    __statement_W57Jy4(59);
    Page.prototype.clean = function (data) {
        __block_nvViJX(18);
        if (__expression_WWV8jh(60), !__extro_dMV_cs(61, __intro_FxjW60(61, _).isObject(data))) {
            __block_nvViJX(19);
            {
                __statement_W57Jy4(62);
                data = {};
            }
        }
        return __expression_WWV8jh(63), data;
    };
}
{
    __statement_W57Jy4(64);
    exports = module.exports = Page;
}
