// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_session.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_XC7RB$, __expression_IxbOxx, __block_wRAjD9;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_XC7RB$ = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/session.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_IxbOxx = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/session.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_wRAjD9 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/session.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_fV4iaH = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_x5BXQb = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/session.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_XC7RB$(0);
    var keystone = (__expression_IxbOxx(1), require('../')), crypto = (__expression_IxbOxx(2), require('crypto'));
}
{
    __statement_XC7RB$(3);
    var hash = function (str) {
        __block_wRAjD9(0);
        {
            __statement_XC7RB$(4);
            str = (__expression_IxbOxx(5), '' + (__expression_IxbOxx(6), str));
        }
        {
            __statement_XC7RB$(7);
            str = __extro_x5BXQb(8, __intro_fV4iaH(8, str).substr(0, __extro_x5BXQb(9, __intro_fV4iaH(9, Math).round((__expression_IxbOxx(10), str.length / 2)))));
        }
        return __expression_IxbOxx(11), __extro_x5BXQb(12, __intro_fV4iaH(12, __extro_x5BXQb(13, __intro_fV4iaH(13, __extro_x5BXQb(14, __intro_fV4iaH(14, __extro_x5BXQb(15, __intro_fV4iaH(15, crypto).createHmac('sha256', __extro_x5BXQb(16, __intro_fV4iaH(16, keystone).get('cookie secret'))))).update(str))).digest('base64'))).replace(/\=+$/, ''));
    };
}
{
    __statement_XC7RB$(17);
    exports.signin = function (lookup, req, res, onSuccess, onFail) {
        __block_wRAjD9(1);
        if (__expression_IxbOxx(18), !(__expression_IxbOxx(19), lookup)) {
            __block_wRAjD9(2);
            return __expression_IxbOxx(20), (__expression_IxbOxx(21), onFail(new Error('session.signin requires a User ID or Object as the first argument')));
        }
        {
            __statement_XC7RB$(22);
            var User = __extro_x5BXQb(23, __intro_fV4iaH(23, keystone).list(__extro_x5BXQb(24, __intro_fV4iaH(24, keystone).get('user model'))));
        }
        {
            __statement_XC7RB$(25);
            var doSignin = function (user) {
                __block_wRAjD9(3);
                {
                    __statement_XC7RB$(26);
                    __extro_x5BXQb(27, __intro_fV4iaH(27, req.session).regenerate(function () {
                        __block_wRAjD9(4);
                        {
                            __statement_XC7RB$(28);
                            req.user = user;
                        }
                        {
                            __statement_XC7RB$(29);
                            req.session.userId = user.id;
                        }
                        if (__expression_IxbOxx(30), __extro_x5BXQb(31, __intro_fV4iaH(31, keystone).get('cookie signin')) && user.password) {
                            __block_wRAjD9(5);
                            {
                                __statement_XC7RB$(32);
                                var userToken = (__expression_IxbOxx(33), (__expression_IxbOxx(34), user.id + ':') + (__expression_IxbOxx(35), hash(user.password)));
                            }
                            {
                                __statement_XC7RB$(36);
                                __extro_x5BXQb(37, __intro_fV4iaH(37, res).cookie('keystone.uid', userToken, {
                                    signed: true,
                                    httpOnly: true
                                }));
                            }
                        }
                        {
                            __statement_XC7RB$(38);
                            __expression_IxbOxx(39), onSuccess(user);
                        }
                    }));
                }
            };
        }
        if (__expression_IxbOxx(40), (__expression_IxbOxx(41), 'string' === (__expression_IxbOxx(42), typeof lookup.email)) && (__expression_IxbOxx(43), 'string' === (__expression_IxbOxx(44), typeof lookup.password))) {
            __block_wRAjD9(6);
            {
                __statement_XC7RB$(45);
                __extro_x5BXQb(46, __intro_fV4iaH(46, __extro_x5BXQb(47, __intro_fV4iaH(47, User.model).findOne({
                    email: lookup.email
                }))).exec(function (err, user) {
                    __block_wRAjD9(7);
                    if (__expression_IxbOxx(48), user) {
                        __block_wRAjD9(8);
                        {
                            __statement_XC7RB$(49);
                            __extro_x5BXQb(50, __intro_fV4iaH(50, user._.password).compare(lookup.password, function (err, isMatch) {
                                __block_wRAjD9(9);
                                if (__expression_IxbOxx(51), (__expression_IxbOxx(52), !(__expression_IxbOxx(53), err)) && (__expression_IxbOxx(54), isMatch)) {
                                    __block_wRAjD9(10);
                                    {
                                        __statement_XC7RB$(55);
                                        __expression_IxbOxx(56), doSignin(user);
                                    }
                                } else {
                                    __block_wRAjD9(11);
                                    {
                                        __statement_XC7RB$(57);
                                        __expression_IxbOxx(58), onFail(err);
                                    }
                                }
                            }));
                        }
                    } else {
                        __block_wRAjD9(12);
                        {
                            __statement_XC7RB$(59);
                            __expression_IxbOxx(60), onFail(err);
                        }
                    }
                }));
            }
        } else {
            __block_wRAjD9(13);
            {
                __statement_XC7RB$(61);
                lookup = (__expression_IxbOxx(62), '' + (__expression_IxbOxx(63), lookup));
            }
            {
                __statement_XC7RB$(64);
                var userId = (__expression_IxbOxx(67), __extro_x5BXQb(68, __intro_fV4iaH(68, lookup).indexOf(':')) > 0) ? (__expression_IxbOxx(65), __extro_x5BXQb(69, __intro_fV4iaH(69, lookup).substr(0, __extro_x5BXQb(70, __intro_fV4iaH(70, lookup).indexOf(':'))))) : (__expression_IxbOxx(66), lookup), passwordCheck = (__expression_IxbOxx(73), __extro_x5BXQb(74, __intro_fV4iaH(74, lookup).indexOf(':')) > 0) ? (__expression_IxbOxx(71), __extro_x5BXQb(75, __intro_fV4iaH(75, lookup).substr((__expression_IxbOxx(76), __extro_x5BXQb(77, __intro_fV4iaH(77, lookup).indexOf(':')) + 1)))) : (__expression_IxbOxx(72), false);
            }
            {
                __statement_XC7RB$(78);
                __extro_x5BXQb(79, __intro_fV4iaH(79, __extro_x5BXQb(80, __intro_fV4iaH(80, User.model).findById(userId))).exec(function (err, user) {
                    __block_wRAjD9(14);
                    if (__expression_IxbOxx(81), (__expression_IxbOxx(82), user) && (__expression_IxbOxx(83), (__expression_IxbOxx(84), !(__expression_IxbOxx(85), passwordCheck)) || (__expression_IxbOxx(86), (__expression_IxbOxx(87), passwordCheck) === (__expression_IxbOxx(88), hash(user.password))))) {
                        __block_wRAjD9(15);
                        {
                            __statement_XC7RB$(89);
                            __expression_IxbOxx(90), doSignin(user);
                        }
                    } else {
                        __block_wRAjD9(16);
                        {
                            __statement_XC7RB$(91);
                            __expression_IxbOxx(92), onFail(err);
                        }
                    }
                }));
            }
        }
    };
}
{
    __statement_XC7RB$(93);
    exports.signout = function (req, res, next) {
        __block_wRAjD9(17);
        {
            __statement_XC7RB$(94);
            __extro_x5BXQb(95, __intro_fV4iaH(95, res).clearCookie('keystone.uid'));
        }
        {
            __statement_XC7RB$(96);
            req.user = null;
        }
        {
            __statement_XC7RB$(97);
            __extro_x5BXQb(98, __intro_fV4iaH(98, req.session).regenerate(next));
        }
    };
}
{
    __statement_XC7RB$(99);
    exports.persist = function (req, res, next) {
        __block_wRAjD9(18);
        {
            __statement_XC7RB$(100);
            var User = __extro_x5BXQb(101, __intro_fV4iaH(101, keystone).list(__extro_x5BXQb(102, __intro_fV4iaH(102, keystone).get('user model'))));
        }
        if (__expression_IxbOxx(103), (__expression_IxbOxx(104), (__expression_IxbOxx(105), __extro_x5BXQb(106, __intro_fV4iaH(106, keystone).get('cookie signin')) && (__expression_IxbOxx(107), !req.session.userId)) && req.signedCookies['keystone.uid']) && (__expression_IxbOxx(108), __extro_x5BXQb(109, __intro_fV4iaH(109, req.signedCookies['keystone.uid']).indexOf(':')) > 0)) {
            __block_wRAjD9(19);
            {
                __statement_XC7RB$(110);
                var _next = function () {
                    __block_wRAjD9(20);
                    {
                        __statement_XC7RB$(111);
                        __expression_IxbOxx(112), next();
                    }
                };
            }
            {
                __statement_XC7RB$(113);
                __extro_x5BXQb(114, __intro_fV4iaH(114, exports).signin(req.signedCookies['keystone.uid'], req, res, _next, _next));
            }
        } else if (req.session.userId) {
            __block_wRAjD9(21);
            {
                __statement_XC7RB$(115);
                __extro_x5BXQb(116, __intro_fV4iaH(116, __extro_x5BXQb(117, __intro_fV4iaH(117, User.model).findById(req.session.userId))).exec(function (err, user) {
                    __block_wRAjD9(22);
                    if (__expression_IxbOxx(118), err) {
                        __block_wRAjD9(23);
                        return __expression_IxbOxx(119), (__expression_IxbOxx(120), next(err));
                    }
                    {
                        __statement_XC7RB$(121);
                        req.user = user;
                    }
                    {
                        __statement_XC7RB$(122);
                        __expression_IxbOxx(123), next();
                    }
                }));
            }
        } else {
            __block_wRAjD9(24);
            {
                __statement_XC7RB$(124);
                __expression_IxbOxx(125), next();
            }
        }
    };
}
{
    __statement_XC7RB$(126);
    exports.keystoneAuth = function (req, res, next) {
        __block_wRAjD9(25);
        if (__expression_IxbOxx(127), (__expression_IxbOxx(128), !req.user) || (__expression_IxbOxx(129), !req.user.canAccessKeystone)) {
            __block_wRAjD9(26);
            {
                __statement_XC7RB$(130);
                var from = __extro_x5BXQb(133, __intro_fV4iaH(133, new RegExp('^/keystone/?$', 'i')).test(req.url)) ? (__expression_IxbOxx(131), '') : (__expression_IxbOxx(132), (__expression_IxbOxx(134), '?from=' + req.url));
            }
            return __expression_IxbOxx(135), __extro_x5BXQb(136, __intro_fV4iaH(136, res).redirect((__expression_IxbOxx(137), __extro_x5BXQb(138, __intro_fV4iaH(138, keystone).get('signin url')) + (__expression_IxbOxx(139), from))));
        }
        {
            __statement_XC7RB$(140);
            __expression_IxbOxx(141), next();
        }
    };
}
