// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_security_csrf.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_u7DnI3, __expression_zk4AVK, __block_i$$HKg;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_u7DnI3 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/security/csrf.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_zk4AVK = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/security/csrf.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_i$$HKg = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/security/csrf.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_Wvk9dc = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_qJdyhk = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/security/csrf.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_u7DnI3(0);
    var crypto = (__expression_zk4AVK(1), require('crypto')), utils = (__expression_zk4AVK(2), require('keystone-utils'));
}
{
    __statement_u7DnI3(3);
    exports.TOKEN_KEY = '_csrf';
}
{
    __statement_u7DnI3(4);
    exports.LOCAL_KEY = 'csrf_token_key';
}
{
    __statement_u7DnI3(5);
    exports.LOCAL_VALUE = 'csrf_token_value';
}
{
    __statement_u7DnI3(6);
    exports.SECRET_KEY = (__expression_zk4AVK(7), exports.TOKEN_KEY + '_secret');
}
{
    __statement_u7DnI3(8);
    exports.SECRET_LENGTH = 10;
}
function tokenize(salt, secret) {
    __block_i$$HKg(0);
    return __expression_zk4AVK(9), (__expression_zk4AVK(10), (__expression_zk4AVK(11), salt) + __extro_qJdyhk(12, __intro_Wvk9dc(12, __extro_qJdyhk(13, __intro_Wvk9dc(13, __extro_qJdyhk(14, __intro_Wvk9dc(14, crypto).createHash('sha1'))).update((__expression_zk4AVK(15), (__expression_zk4AVK(16), salt) + (__expression_zk4AVK(17), secret))))).digest('base64')));
}
{
    __statement_u7DnI3(18);
    exports.createSecret = function () {
        __block_i$$HKg(1);
        return __expression_zk4AVK(19), __extro_qJdyhk(20, __intro_Wvk9dc(20, __extro_qJdyhk(21, __intro_Wvk9dc(21, crypto).pseudoRandomBytes(exports.SECRET_LENGTH))).toString('base64'));
    };
}
{
    __statement_u7DnI3(22);
    exports.getSecret = function (req) {
        __block_i$$HKg(2);
        return __expression_zk4AVK(23), (__expression_zk4AVK(24), req.session[exports.SECRET_KEY] || (req.session[exports.SECRET_KEY] = __extro_qJdyhk(25, __intro_Wvk9dc(25, exports).createSecret())));
    };
}
{
    __statement_u7DnI3(26);
    exports.createToken = function (req) {
        __block_i$$HKg(3);
        return __expression_zk4AVK(27), (__expression_zk4AVK(28), tokenize(__extro_qJdyhk(29, __intro_Wvk9dc(29, utils).randomString(exports.SECRET_LENGTH)), __extro_qJdyhk(30, __intro_Wvk9dc(30, exports).getSecret(req))));
    };
}
{
    __statement_u7DnI3(31);
    exports.getToken = function (req, res) {
        __block_i$$HKg(4);
        return __expression_zk4AVK(32), (__expression_zk4AVK(33), res.locals[exports.LOCAL_VALUE] || (res.locals[exports.LOCAL_VALUE] = __extro_qJdyhk(34, __intro_Wvk9dc(34, exports).createToken(req))));
    };
}
{
    __statement_u7DnI3(35);
    exports.requestToken = function (req) {
        __block_i$$HKg(5);
        if (__expression_zk4AVK(36), req.body && req.body[exports.TOKEN_KEY]) {
            __block_i$$HKg(6);
            return __expression_zk4AVK(37), req.body[exports.TOKEN_KEY];
        } else if (__expression_zk4AVK(38), req.query && req.query[exports.TOKEN_KEY]) {
            __block_i$$HKg(7);
            return __expression_zk4AVK(39), req.query[exports.TOKEN_KEY];
        }
        return __expression_zk4AVK(40), '';
    };
}
{
    __statement_u7DnI3(41);
    exports.validate = function (req, token) {
        __block_i$$HKg(8);
        if (__expression_zk4AVK(42), arguments.length === 1) {
            __block_i$$HKg(9);
            {
                __statement_u7DnI3(43);
                token = __extro_qJdyhk(44, __intro_Wvk9dc(44, exports).requestToken(req));
            }
        }
        if (__expression_zk4AVK(45), (__expression_zk4AVK(46), typeof token) !== 'string') {
            __block_i$$HKg(10);
            return __expression_zk4AVK(47), false;
        }
        return __expression_zk4AVK(48), (__expression_zk4AVK(49), (__expression_zk4AVK(50), token) === (__expression_zk4AVK(51), tokenize(__extro_qJdyhk(52, __intro_Wvk9dc(52, token).slice(0, exports.SECRET_LENGTH)), req.session[exports.SECRET_KEY])));
    };
}
{
    __statement_u7DnI3(53);
    exports.middleware = {
        init: function (req, res, next) {
            __block_i$$HKg(11);
            {
                __statement_u7DnI3(54);
                res.locals[exports.LOCAL_KEY] = exports.LOCAL_VALUE;
            }
            {
                __statement_u7DnI3(55);
                __extro_qJdyhk(56, __intro_Wvk9dc(56, exports).getToken(req, res));
            }
            {
                __statement_u7DnI3(57);
                __expression_zk4AVK(58), next();
            }
        },
        validate: function (req, res, next) {
            __block_i$$HKg(12);
            if (__expression_zk4AVK(59), (__expression_zk4AVK(60), (__expression_zk4AVK(61), req.method === 'GET') || (__expression_zk4AVK(62), req.method === 'HEAD')) || (__expression_zk4AVK(63), req.method === 'OPTIONS')) {
                __block_i$$HKg(13);
                return __expression_zk4AVK(64), (__expression_zk4AVK(65), next());
            }
            if (__extro_qJdyhk(66, __intro_Wvk9dc(66, exports).validate(req))) {
                __block_i$$HKg(14);
                {
                    __statement_u7DnI3(67);
                    __expression_zk4AVK(68), next();
                }
            } else {
                __block_i$$HKg(15);
                {
                    __statement_u7DnI3(69);
                    res.statusCode = 403;
                }
                {
                    __statement_u7DnI3(70);
                    __expression_zk4AVK(71), next(new Error('CSRF token mismatch'));
                }
            }
        }
    };
}
