// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_path.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_r0ZuHu, __expression_CwX7Wg, __block_ZNrb1D;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_r0ZuHu = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/path.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_CwX7Wg = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/path.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_ZNrb1D = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/path.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_BwljS7 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_iAjP3g = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/path.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_r0ZuHu(0);
    var utils = (__expression_CwX7Wg(1), require('keystone-utils'));
}
{
    __statement_r0ZuHu(2);
    exports = module.exports = function Path(str) {
        __block_ZNrb1D(0);
        if (__expression_CwX7Wg(3), !(__expression_CwX7Wg(4), this instanceof (__expression_CwX7Wg(5), Path))) {
            __block_ZNrb1D(1);
            return __expression_CwX7Wg(6), new Path(str);
        }
        {
            __statement_r0ZuHu(7);
            this.original = str;
        }
        {
            __statement_r0ZuHu(8);
            var parts = this.parts = __extro_iAjP3g(9, __intro_BwljS7(9, str).split('.'));
        }
        {
            __statement_r0ZuHu(10);
            var last = this.last = this.parts[__expression_CwX7Wg(11), this.parts.length - 1];
        }
        {
            __statement_r0ZuHu(12);
            var exceptLast = [];
        }
        for (var i = 0; __expression_CwX7Wg(13), (__expression_CwX7Wg(14), i) < (__expression_CwX7Wg(15), parts.length - 1); __expression_CwX7Wg(16), i++) {
            __block_ZNrb1D(2);
            {
                __statement_r0ZuHu(17);
                __extro_iAjP3g(18, __intro_BwljS7(18, exceptLast).push(parts[i]));
            }
        }
        {
            __statement_r0ZuHu(19);
            this.exceptLast = __extro_iAjP3g(20, __intro_BwljS7(20, exceptLast).join('.'));
        }
        {
            __statement_r0ZuHu(21);
            this.addTo = function (obj, val) {
                __block_ZNrb1D(3);
                {
                    __statement_r0ZuHu(22);
                    var o = obj;
                }
                for (var i = 0; __expression_CwX7Wg(23), (__expression_CwX7Wg(24), i) < (__expression_CwX7Wg(25), parts.length - 1); __expression_CwX7Wg(26), i++) {
                    __block_ZNrb1D(4);
                    if (__expression_CwX7Wg(27), !__extro_iAjP3g(28, __intro_BwljS7(28, utils).isObject(o[parts[i]]))) {
                        __block_ZNrb1D(5);
                        {
                            __statement_r0ZuHu(29);
                            o[parts[i]] = {};
                        }
                    }
                    {
                        __statement_r0ZuHu(30);
                        o = o[parts[i]];
                    }
                }
                {
                    __statement_r0ZuHu(31);
                    o[last] = val;
                }
                return __expression_CwX7Wg(32), obj;
            };
        }
        {
            __statement_r0ZuHu(33);
            this.get = function (obj) {
                __block_ZNrb1D(6);
                {
                    __statement_r0ZuHu(34);
                    var o = obj;
                }
                for (var i = 0; __expression_CwX7Wg(35), (__expression_CwX7Wg(36), i) < parts.length; __expression_CwX7Wg(37), i++) {
                    __block_ZNrb1D(7);
                    if (__expression_CwX7Wg(38), (__expression_CwX7Wg(39), typeof o) !== 'object') {
                        __block_ZNrb1D(8);
                        return __expression_CwX7Wg(40), undefined;
                    }
                    {
                        __statement_r0ZuHu(41);
                        o = o[parts[i]];
                    }
                }
                return __expression_CwX7Wg(42), o;
            };
        }
        {
            __statement_r0ZuHu(43);
            this.prependToLast = function (prepend, titlecase) {
                __block_ZNrb1D(9);
                {
                    __statement_r0ZuHu(44);
                    var rtn = '';
                }
                for (var i = 0; __expression_CwX7Wg(45), (__expression_CwX7Wg(46), i) < (__expression_CwX7Wg(47), parts.length - 1); __expression_CwX7Wg(48), i++) {
                    __block_ZNrb1D(10);
                    {
                        __statement_r0ZuHu(49);
                        rtn += (__expression_CwX7Wg(50), parts[i] + '.');
                    }
                }
                return __expression_CwX7Wg(51), (__expression_CwX7Wg(52), (__expression_CwX7Wg(53), (__expression_CwX7Wg(54), rtn) + (__expression_CwX7Wg(55), (__expression_CwX7Wg(56), prepend) || '')) + ((__expression_CwX7Wg(59), titlecase) ? (__expression_CwX7Wg(57), __extro_iAjP3g(60, __intro_BwljS7(60, utils).upcase(last))) : (__expression_CwX7Wg(58), last)));
            };
        }
        {
            __statement_r0ZuHu(61);
            this.append = function (append) {
                __block_ZNrb1D(11);
                return __expression_CwX7Wg(62), (__expression_CwX7Wg(63), (__expression_CwX7Wg(64), str) + (__expression_CwX7Wg(65), append));
            };
        }
        {
            __statement_r0ZuHu(66);
            this.flatten = function (titlecase) {
                __block_ZNrb1D(12);
                return __expression_CwX7Wg(67), __extro_iAjP3g(68, __intro_BwljS7(68, utils).camelcase(__extro_iAjP3g(69, __intro_BwljS7(69, parts).join('_')), (__expression_CwX7Wg(72), titlecase) ? (__expression_CwX7Wg(70), true) : (__expression_CwX7Wg(71), false)));
            };
        }
        {
            __statement_r0ZuHu(73);
            this.flattenplural = function (titlecase) {
                __block_ZNrb1D(13);
                return __expression_CwX7Wg(74), __extro_iAjP3g(75, __intro_BwljS7(75, utils).camelcase(__extro_iAjP3g(76, __intro_BwljS7(76, __extro_iAjP3g(77, __intro_BwljS7(77, __extro_iAjP3g(78, __intro_BwljS7(78, []).concat(exceptLast))).concat(__extro_iAjP3g(79, __intro_BwljS7(79, utils).plural(last))))).join('_')), (__expression_CwX7Wg(82), titlecase) ? (__expression_CwX7Wg(80), true) : (__expression_CwX7Wg(81), false)));
            };
        }
        {
            __statement_r0ZuHu(83);
            this.flattensingular = function (titlecase) {
                __block_ZNrb1D(14);
                return __expression_CwX7Wg(84), __extro_iAjP3g(85, __intro_BwljS7(85, utils).camelcase(__extro_iAjP3g(86, __intro_BwljS7(86, __extro_iAjP3g(87, __intro_BwljS7(87, __extro_iAjP3g(88, __intro_BwljS7(88, []).concat(exceptLast))).concat(__extro_iAjP3g(89, __intro_BwljS7(89, utils).singular(last))))).join('_')), (__expression_CwX7Wg(92), titlecase) ? (__expression_CwX7Wg(90), true) : (__expression_CwX7Wg(91), false)));
            };
        }
        return __expression_CwX7Wg(93), this;
    };
}
