// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_select.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_vyOaJv, __expression_IFS52a, __block_w8v8SS;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_vyOaJv = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/select.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_IFS52a = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/select.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_w8v8SS = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/select.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_utQl8d = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_NQ5OA6 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/select.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_vyOaJv(0);
    var _ = (__expression_IFS52a(1), require('underscore')), util = (__expression_IFS52a(2), require('util')), utils = (__expression_IFS52a(3), require('keystone-utils')), super_ = (__expression_IFS52a(4), require('../field'));
}
function select(list, path, options) {
    __block_w8v8SS(0);
    {
        __statement_vyOaJv(5);
        this._nativeType = options.numeric ? (__expression_IFS52a(6), Number) : (__expression_IFS52a(7), String);
    }
    {
        __statement_vyOaJv(8);
        this._underscoreMethods = [
            'format'
        ];
    }
    {
        __statement_vyOaJv(9);
        this.ui = (__expression_IFS52a(10), options.ui || 'select');
    }
    if (__expression_IFS52a(11), (__expression_IFS52a(12), typeof options.options) === 'string') {
        __block_w8v8SS(1);
        {
            __statement_vyOaJv(13);
            options.options = __extro_NQ5OA6(14, __intro_utQl8d(14, options.options).split(','));
        }
    }
    if (__expression_IFS52a(15), !__extro_NQ5OA6(16, __intro_utQl8d(16, Array).isArray(options.options))) {
        __block_w8v8SS(2);
        {
            __statement_vyOaJv(17);
            throw new Error('Select fields require an options array.');
        }
    }
    {
        __statement_vyOaJv(18);
        this.ops = __extro_NQ5OA6(19, __intro_utQl8d(19, options.options).map(function (i) {
            __block_w8v8SS(3);
            {
                __statement_vyOaJv(20);
                var op = __extro_NQ5OA6(23, __intro_utQl8d(23, _).isString(i)) ? (__expression_IFS52a(21), {
                        value: __extro_NQ5OA6(24, __intro_utQl8d(24, i).trim()),
                        label: __extro_NQ5OA6(25, __intro_utQl8d(25, utils).keyToLabel(i))
                    }) : (__expression_IFS52a(22), i);
            }
            if (__expression_IFS52a(26), !__extro_NQ5OA6(27, __intro_utQl8d(27, _).isObject(op))) {
                __block_w8v8SS(4);
                {
                    __statement_vyOaJv(28);
                    op = {
                        label: (__expression_IFS52a(29), '' + (__expression_IFS52a(30), i)),
                        value: (__expression_IFS52a(31), '' + (__expression_IFS52a(32), i))
                    };
                }
            }
            if (__expression_IFS52a(33), options.numeric && (__expression_IFS52a(34), !__extro_NQ5OA6(35, __intro_utQl8d(35, _).isNumber(op.value)))) {
                __block_w8v8SS(5);
                {
                    __statement_vyOaJv(36);
                    op.value = (__expression_IFS52a(37), Number(op.value));
                }
            }
            return __expression_IFS52a(38), op;
        }));
    }
    if (__expression_IFS52a(39), options.emptyOption === (__expression_IFS52a(40), undefined)) {
        __block_w8v8SS(6);
        {
            __statement_vyOaJv(41);
            options.emptyOption = true;
        }
    }
    {
        __statement_vyOaJv(42);
        this.emptyOption = options.emptyOption ? (__expression_IFS52a(43), true) : (__expression_IFS52a(44), false);
    }
    {
        __statement_vyOaJv(45);
        this.map = __extro_NQ5OA6(46, __intro_utQl8d(46, utils).optionsMap(this.ops));
    }
    {
        __statement_vyOaJv(47);
        this.labels = __extro_NQ5OA6(48, __intro_utQl8d(48, utils).optionsMap(this.ops, 'label'));
    }
    {
        __statement_vyOaJv(49);
        this.values = __extro_NQ5OA6(50, __intro_utQl8d(50, _).pluck(this.ops, 'value'));
    }
    {
        __statement_vyOaJv(51);
        __extro_NQ5OA6(52, __intro_utQl8d(52, select.super_).call(this, list, path, options));
    }
}
{
    __statement_vyOaJv(53);
    __extro_NQ5OA6(54, __intro_utQl8d(54, util).inherits(select, super_));
}
{
    __statement_vyOaJv(55);
    select.prototype.addToSchema = function () {
        __block_w8v8SS(7);
        {
            __statement_vyOaJv(56);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_vyOaJv(57);
            this.paths = {
                data: (__expression_IFS52a(58), this.options.dataPath || __extro_NQ5OA6(59, __intro_utQl8d(59, this._path).append('Data'))),
                label: (__expression_IFS52a(60), this.options.labelPath || __extro_NQ5OA6(61, __intro_utQl8d(61, this._path).append('Label'))),
                options: (__expression_IFS52a(62), this.options.optionsPath || __extro_NQ5OA6(63, __intro_utQl8d(63, this._path).append('Options'))),
                map: (__expression_IFS52a(64), this.options.optionsMapPath || __extro_NQ5OA6(65, __intro_utQl8d(65, this._path).append('OptionsMap')))
            };
        }
        {
            __statement_vyOaJv(66);
            __extro_NQ5OA6(67, __intro_utQl8d(67, schema).path(this.path, __extro_NQ5OA6(68, __intro_utQl8d(68, _).defaults({
                type: this._nativeType,
                enum: this.values,
                set: function (val) {
                    __block_w8v8SS(8);
                    return __expression_IFS52a(69), (__expression_IFS52a(72), (__expression_IFS52a(73), val) === '') ? (__expression_IFS52a(70), undefined) : (__expression_IFS52a(71), val);
                }
            }, this.options))));
        }
        {
            __statement_vyOaJv(74);
            __extro_NQ5OA6(75, __intro_utQl8d(75, __extro_NQ5OA6(76, __intro_utQl8d(76, schema).virtual(this.paths.data))).get(function () {
                __block_w8v8SS(9);
                return __expression_IFS52a(77), field.map[__extro_NQ5OA6(78, __intro_utQl8d(78, this).get(field.path))];
            }));
        }
        {
            __statement_vyOaJv(79);
            __extro_NQ5OA6(80, __intro_utQl8d(80, __extro_NQ5OA6(81, __intro_utQl8d(81, schema).virtual(this.paths.label))).get(function () {
                __block_w8v8SS(10);
                return __expression_IFS52a(82), field.labels[__extro_NQ5OA6(83, __intro_utQl8d(83, this).get(field.path))];
            }));
        }
        {
            __statement_vyOaJv(84);
            __extro_NQ5OA6(85, __intro_utQl8d(85, __extro_NQ5OA6(86, __intro_utQl8d(86, schema).virtual(this.paths.options))).get(function () {
                __block_w8v8SS(11);
                return __expression_IFS52a(87), field.ops;
            }));
        }
        {
            __statement_vyOaJv(88);
            __extro_NQ5OA6(89, __intro_utQl8d(89, __extro_NQ5OA6(90, __intro_utQl8d(90, schema).virtual(this.paths.map))).get(function () {
                __block_w8v8SS(12);
                return __expression_IFS52a(91), field.map;
            }));
        }
        {
            __statement_vyOaJv(92);
            __extro_NQ5OA6(93, __intro_utQl8d(93, this).underscoreMethod('pluck', function (property, d) {
                __block_w8v8SS(13);
                {
                    __statement_vyOaJv(94);
                    var option = __extro_NQ5OA6(95, __intro_utQl8d(95, this).get(field.paths.data));
                }
                return __expression_IFS52a(96), (__expression_IFS52a(99), option) ? (__expression_IFS52a(97), option[property]) : (__expression_IFS52a(98), d);
            }));
        }
        {
            __statement_vyOaJv(100);
            __extro_NQ5OA6(101, __intro_utQl8d(101, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_vyOaJv(102);
    select.prototype.cloneOps = function () {
        __block_w8v8SS(14);
        return __expression_IFS52a(103), __extro_NQ5OA6(104, __intro_utQl8d(104, _).map(this.ops, _.clone));
    };
}
{
    __statement_vyOaJv(105);
    select.prototype.cloneMap = function () {
        __block_w8v8SS(15);
        return __expression_IFS52a(106), __extro_NQ5OA6(107, __intro_utQl8d(107, utils).optionsMap(this.ops, true));
    };
}
{
    __statement_vyOaJv(108);
    select.prototype.validateInput = function (data, required, item) {
        __block_w8v8SS(16);
        if (data[this.path]) {
            __block_w8v8SS(17);
            return __expression_IFS52a(109), (__expression_IFS52a(112), (data[this.path] in this.map)) ? (__expression_IFS52a(110), true) : (__expression_IFS52a(111), false);
        } else {
            __block_w8v8SS(18);
            return __expression_IFS52a(113), (__expression_IFS52a(116), (__expression_IFS52a(117), !(__expression_IFS52a(118), required)) || (__expression_IFS52a(119), (__expression_IFS52a(120), (__expression_IFS52a(121), !(__expression_IFS52a(122), (this.path in (__expression_IFS52a(123), data)))) && (__expression_IFS52a(124), item)) && __extro_NQ5OA6(125, __intro_utQl8d(125, item).get(this.path)))) ? (__expression_IFS52a(114), true) : (__expression_IFS52a(115), false);
        }
    };
}
{
    __statement_vyOaJv(126);
    select.prototype.format = function (item) {
        __block_w8v8SS(19);
        return __expression_IFS52a(127), this.labels[__extro_NQ5OA6(128, __intro_utQl8d(128, item).get(this.path))];
    };
}
{
    __statement_vyOaJv(129);
    exports = module.exports = select;
}
