// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_list.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_bxUTm8, __expression_hyFkrD, __block_IT4f3g;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_bxUTm8 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/list.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_hyFkrD = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/list.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_IT4f3g = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/list.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_ZXD72w = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_hNYQv5 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/list.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_bxUTm8(0);
    var _ = (__expression_hyFkrD(1), require('underscore')), async = (__expression_hyFkrD(2), require('async')), moment = (__expression_hyFkrD(3), require('moment')), keystone = (__expression_hyFkrD(4), require('../')), schemaPlugins = (__expression_hyFkrD(5), require('./schemaPlugins')), utils = (__expression_hyFkrD(6), require('keystone-utils')), Field = (__expression_hyFkrD(7), require('./field')), UpdateHandler = (__expression_hyFkrD(8), require('./updateHandler')), queryfilterlib = (__expression_hyFkrD(9), require('queryfilter'));
}
function List(key, options) {
    __block_IT4f3g(0);
    if (__expression_hyFkrD(10), !(__expression_hyFkrD(11), this instanceof (__expression_hyFkrD(12), List))) {
        __block_IT4f3g(1);
        return __expression_hyFkrD(13), new List(key, options);
    }
    {
        __statement_bxUTm8(14);
        this.options = __extro_hNYQv5(15, __intro_ZXD72w(15, utils).options({
            schema: {
                collection: __extro_hNYQv5(16, __intro_ZXD72w(16, keystone).prefixModel(key))
            },
            noedit: false,
            nocreate: false,
            nodelete: false,
            autocreate: false,
            sortable: false,
            hidden: false,
            track: false,
            searchFields: '__name__',
            defaultSort: '__default__',
            defaultColumns: '__name__'
        }, options));
    }
    {
        __statement_bxUTm8(17);
        this.key = key;
    }
    {
        __statement_bxUTm8(18);
        this.path = (__expression_hyFkrD(19), __extro_hNYQv5(20, __intro_ZXD72w(20, this).get('path')) || __extro_hNYQv5(21, __intro_ZXD72w(21, utils).keyToPath(key, true)));
    }
    {
        __statement_bxUTm8(22);
        this.schema = new keystone.mongoose.Schema({}, this.options.schema);
    }
    {
        __statement_bxUTm8(23);
        this.uiElements = [];
    }
    {
        __statement_bxUTm8(24);
        this.underscoreMethods = {};
    }
    {
        __statement_bxUTm8(25);
        this.fields = {};
    }
    {
        __statement_bxUTm8(26);
        this.fieldTypes = {};
    }
    {
        __statement_bxUTm8(27);
        this.relationships = {};
    }
    {
        __statement_bxUTm8(28);
        this.mappings = {
            name: null,
            createdBy: null,
            createdOn: null,
            modifiedBy: null,
            modifiedOn: null
        };
    }
    {
        __statement_bxUTm8(29);
        __extro_hNYQv5(30, __intro_ZXD72w(30, _).each(this.options.map, function (val, key) {
            __block_IT4f3g(2);
            {
                __statement_bxUTm8(31);
                __extro_hNYQv5(32, __intro_ZXD72w(32, this).map(key, val));
            }
        }, this));
    }
    {
        __statement_bxUTm8(33);
        __extro_hNYQv5(34, __intro_ZXD72w(34, Object).defineProperty(this, 'label', {
            get: function () {
                __block_IT4f3g(3);
                return __expression_hyFkrD(35), (__expression_hyFkrD(36), __extro_hNYQv5(37, __intro_ZXD72w(37, this).get('label')) || __extro_hNYQv5(38, __intro_ZXD72w(38, this).set('label', __extro_hNYQv5(39, __intro_ZXD72w(39, utils).plural(__extro_hNYQv5(40, __intro_ZXD72w(40, utils).keyToLabel(key)))))));
            }
        }));
    }
    {
        __statement_bxUTm8(41);
        __extro_hNYQv5(42, __intro_ZXD72w(42, Object).defineProperty(this, 'singular', {
            get: function () {
                __block_IT4f3g(4);
                return __expression_hyFkrD(43), (__expression_hyFkrD(44), __extro_hNYQv5(45, __intro_ZXD72w(45, this).get('singular')) || __extro_hNYQv5(46, __intro_ZXD72w(46, this).set('singular', __extro_hNYQv5(47, __intro_ZXD72w(47, utils).singular(this.label)))));
            }
        }));
    }
    {
        __statement_bxUTm8(48);
        __extro_hNYQv5(49, __intro_ZXD72w(49, Object).defineProperty(this, 'plural', {
            get: function () {
                __block_IT4f3g(5);
                return __expression_hyFkrD(50), (__expression_hyFkrD(51), __extro_hNYQv5(52, __intro_ZXD72w(52, this).get('plural')) || __extro_hNYQv5(53, __intro_ZXD72w(53, this).set('plural', __extro_hNYQv5(54, __intro_ZXD72w(54, utils).plural(this.singular)))));
            }
        }));
    }
    {
        __statement_bxUTm8(55);
        __extro_hNYQv5(56, __intro_ZXD72w(56, Object).defineProperty(this, 'namePath', {
            get: function () {
                __block_IT4f3g(6);
                return __expression_hyFkrD(57), (__expression_hyFkrD(58), this.mappings.name || '_id');
            }
        }));
    }
    {
        __statement_bxUTm8(59);
        __extro_hNYQv5(60, __intro_ZXD72w(60, Object).defineProperty(this, 'nameField', {
            get: function () {
                __block_IT4f3g(7);
                return __expression_hyFkrD(61), this.fields[this.mappings.name];
            }
        }));
    }
    {
        __statement_bxUTm8(62);
        __extro_hNYQv5(63, __intro_ZXD72w(63, Object).defineProperty(this, 'nameIsVirtual', {
            get: function () {
                __block_IT4f3g(8);
                return __expression_hyFkrD(64), this.model.schema.virtuals[this.mappings.name] ? (__expression_hyFkrD(65), true) : (__expression_hyFkrD(66), false);
            }
        }));
    }
    {
        __statement_bxUTm8(67);
        __extro_hNYQv5(68, __intro_ZXD72w(68, Object).defineProperty(this, 'nameIsEditable', {
            get: function () {
                __block_IT4f3g(9);
                return __expression_hyFkrD(69), (__expression_hyFkrD(72), this.fields[this.mappings.name] && (__expression_hyFkrD(73), this.fields[this.mappings.name].type === 'text')) ? (__expression_hyFkrD(70), (__expression_hyFkrD(74), !this.fields[this.mappings.name].noedit)) : (__expression_hyFkrD(71), false);
            }
        }));
    }
    {
        __statement_bxUTm8(75);
        __extro_hNYQv5(76, __intro_ZXD72w(76, Object).defineProperty(this, 'nameIsInitial', {
            get: function () {
                __block_IT4f3g(10);
                return __expression_hyFkrD(77), (__expression_hyFkrD(78), this.fields[this.mappings.name] && (__expression_hyFkrD(79), this.fields[this.mappings.name].options.initial === (__expression_hyFkrD(80), undefined)));
            }
        }));
    }
    {
        __statement_bxUTm8(81);
        var initialFields;
    }
    {
        __statement_bxUTm8(82);
        __extro_hNYQv5(83, __intro_ZXD72w(83, Object).defineProperty(this, 'initialFields', {
            get: function () {
                __block_IT4f3g(11);
                return __expression_hyFkrD(84), (__expression_hyFkrD(85), (__expression_hyFkrD(86), initialFields) || (initialFields = __extro_hNYQv5(87, __intro_ZXD72w(87, _).filter(this.fields, function (i) {
                    __block_IT4f3g(12);
                    return __expression_hyFkrD(88), i.initial;
                }))));
            }
        }));
    }
}
{
    __statement_bxUTm8(89);
    List.prototype.set = function (key, value) {
        __block_IT4f3g(13);
        if (__expression_hyFkrD(90), arguments.length === 1) {
            __block_IT4f3g(14);
            return __expression_hyFkrD(91), this.options[key];
        }
        {
            __statement_bxUTm8(92);
            this.options[key] = value;
        }
        return __expression_hyFkrD(93), value;
    };
}
{
    __statement_bxUTm8(94);
    List.prototype.get = List.prototype.set;
}
{
    __statement_bxUTm8(95);
    List.prototype.map = function (field, path) {
        __block_IT4f3g(15);
        if (__expression_hyFkrD(96), path) {
            __block_IT4f3g(16);
            {
                __statement_bxUTm8(97);
                this.mappings[field] = path;
            }
        }
        return __expression_hyFkrD(98), this.mappings[field];
    };
}
{
    __statement_bxUTm8(99);
    List.prototype.automap = function (field) {
        __block_IT4f3g(17);
        if (__expression_hyFkrD(100), __extro_hNYQv5(101, __intro_ZXD72w(101, _).has(this.mappings, field.path)) && (__expression_hyFkrD(102), !this.mappings[field.path])) {
            __block_IT4f3g(18);
            {
                __statement_bxUTm8(103);
                __extro_hNYQv5(104, __intro_ZXD72w(104, this).map(field.path, field.path));
            }
        }
    };
}
{
    __statement_bxUTm8(105);
    List.prototype.add = function () {
        __block_IT4f3g(19);
        {
            __statement_bxUTm8(106);
            var add = __extro_hNYQv5(107, __intro_ZXD72w(107, function (obj, prefix) {
                    __block_IT4f3g(20);
                    {
                        __statement_bxUTm8(108);
                        prefix = (__expression_hyFkrD(109), (__expression_hyFkrD(110), prefix) || '');
                    }
                    {
                        __statement_bxUTm8(111);
                        var keys = __extro_hNYQv5(112, __intro_ZXD72w(112, Object).keys(obj));
                    }
                    for (var i = 0; __expression_hyFkrD(113), (__expression_hyFkrD(114), i) < keys.length; __expression_hyFkrD(115), ++i) {
                        __block_IT4f3g(21);
                        {
                            __statement_bxUTm8(116);
                            var key = keys[i];
                        }
                        if (__expression_hyFkrD(117), null === obj[key]) {
                            __block_IT4f3g(22);
                            {
                                __statement_bxUTm8(118);
                                throw new Error((__expression_hyFkrD(119), (__expression_hyFkrD(120), (__expression_hyFkrD(121), 'Invalid value for schema path `' + (__expression_hyFkrD(122), prefix)) + (__expression_hyFkrD(123), key)) + '`'));
                            }
                        }
                        if (__expression_hyFkrD(124), (__expression_hyFkrD(125), __extro_hNYQv5(126, __intro_ZXD72w(126, utils).isObject(obj[key])) && (__expression_hyFkrD(127), (__expression_hyFkrD(128), !obj[key].constructor) || (__expression_hyFkrD(129), 'Object' === obj[key].constructor.name))) && (__expression_hyFkrD(130), (__expression_hyFkrD(131), !obj[key].type) || obj[key].type.type)) {
                            __block_IT4f3g(23);
                            if (__extro_hNYQv5(132, __intro_ZXD72w(132, Object).keys(obj[key])).length) {
                                __block_IT4f3g(24);
                                {
                                    __statement_bxUTm8(133);
                                    this.schema.nested[this.path] = true;
                                }
                                {
                                    __statement_bxUTm8(134);
                                    __expression_hyFkrD(135), add(obj[key], (__expression_hyFkrD(136), (__expression_hyFkrD(137), (__expression_hyFkrD(138), prefix) + (__expression_hyFkrD(139), key)) + '.'));
                                }
                            } else {
                                __block_IT4f3g(25);
                                {
                                    __statement_bxUTm8(140);
                                    __expression_hyFkrD(141), addField((__expression_hyFkrD(142), (__expression_hyFkrD(143), prefix) + (__expression_hyFkrD(144), key)), obj[key]);
                                }
                            }
                        } else {
                            __block_IT4f3g(26);
                            {
                                __statement_bxUTm8(145);
                                __expression_hyFkrD(146), addField((__expression_hyFkrD(147), (__expression_hyFkrD(148), prefix) + (__expression_hyFkrD(149), key)), obj[key]);
                            }
                        }
                    }
                }).bind(this));
        }
        {
            __statement_bxUTm8(150);
            var addField = __extro_hNYQv5(151, __intro_ZXD72w(151, function (path, options) {
                    __block_IT4f3g(27);
                    {
                        __statement_bxUTm8(152);
                        __extro_hNYQv5(153, __intro_ZXD72w(153, this.uiElements).push({
                            type: 'field',
                            field: __extro_hNYQv5(154, __intro_ZXD72w(154, this).field(path, options))
                        }));
                    }
                }).bind(this));
        }
        {
            __statement_bxUTm8(155);
            __extro_hNYQv5(156, __intro_ZXD72w(156, _).each(arguments, function (def) {
                __block_IT4f3g(28);
                if (__expression_hyFkrD(157), 'string' === (__expression_hyFkrD(158), typeof def)) {
                    __block_IT4f3g(29);
                    if (__expression_hyFkrD(159), (__expression_hyFkrD(160), def) === '>>>') {
                        __block_IT4f3g(30);
                        {
                            __statement_bxUTm8(161);
                            __extro_hNYQv5(162, __intro_ZXD72w(162, this.uiElements).push({
                                type: 'indent'
                            }));
                        }
                    } else if (__expression_hyFkrD(163), (__expression_hyFkrD(164), def) === '<<<') {
                        __block_IT4f3g(31);
                        {
                            __statement_bxUTm8(165);
                            __extro_hNYQv5(166, __intro_ZXD72w(166, this.uiElements).push({
                                type: 'outdent'
                            }));
                        }
                    } else {
                        __block_IT4f3g(32);
                        {
                            __statement_bxUTm8(167);
                            __extro_hNYQv5(168, __intro_ZXD72w(168, this.uiElements).push({
                                type: 'heading',
                                heading: def,
                                options: {}
                            }));
                        }
                    }
                } else {
                    __block_IT4f3g(33);
                    if (__expression_hyFkrD(169), def.heading && (__expression_hyFkrD(170), 'string' === (__expression_hyFkrD(171), typeof def.heading))) {
                        __block_IT4f3g(34);
                        {
                            __statement_bxUTm8(172);
                            __extro_hNYQv5(173, __intro_ZXD72w(173, this.uiElements).push({
                                type: 'heading',
                                heading: def.heading,
                                options: def
                            }));
                        }
                    } else {
                        __block_IT4f3g(35);
                        {
                            __statement_bxUTm8(174);
                            __expression_hyFkrD(175), add(def);
                        }
                    }
                }
            }, this));
        }
        return __expression_hyFkrD(176), this;
    };
}
{
    __statement_bxUTm8(177);
    List.prototype.addPattern = function (pattern) {
        __block_IT4f3g(36);
        {
            __statement_bxUTm8(178);
            var warningMsg = (__expression_hyFkrD(179), (__expression_hyFkrD(180), 'Use of List.addPattern("standard meta") is now deprecated and will be removed\n' + 'in future versions of KeystoneJS. It has been replaced with the List "track" option.\n\n') + 'See http://keystonejs.com/docs/database/#lists-options for more information.');
        }
        switch (__expression_hyFkrD(181), pattern) {
        case 'standard meta': {
                __block_IT4f3g(37);
                if (__expression_hyFkrD(182), !__extro_hNYQv5(183, __intro_ZXD72w(183, this).get('track'))) {
                    __block_IT4f3g(38);
                    {
                        __statement_bxUTm8(184);
                        __extro_hNYQv5(185, __intro_ZXD72w(185, this).set('track', true));
                    }
                }
                {
                    __statement_bxUTm8(186);
                    __extro_hNYQv5(187, __intro_ZXD72w(187, this).set('track simulate standard meta', true));
                }
                {
                    __statement_bxUTm8(188);
                    __extro_hNYQv5(189, __intro_ZXD72w(189, keystone.console).err('Deprecation Warning', warningMsg));
                }
                break;
            }
        }
        return __expression_hyFkrD(190), this;
    };
}
{
    __statement_bxUTm8(191);
    List.prototype.field = function (path, options) {
        __block_IT4f3g(39);
        if (__expression_hyFkrD(192), arguments.length === 1) {
            __block_IT4f3g(40);
            return __expression_hyFkrD(193), this.fields[path];
        }
        if (__expression_hyFkrD(194), 'function' === (__expression_hyFkrD(195), typeof options)) {
            __block_IT4f3g(41);
            {
                __statement_bxUTm8(196);
                options = {
                    type: options
                };
            }
        }
        if (__extro_hNYQv5(197, __intro_ZXD72w(197, this).get('noedit'))) {
            __block_IT4f3g(42);
            {
                __statement_bxUTm8(198);
                options.noedit = true;
            }
        }
        if (__expression_hyFkrD(199), (__expression_hyFkrD(200), !options.note) && __extro_hNYQv5(201, __intro_ZXD72w(201, this).get('notes'))) {
            __block_IT4f3g(43);
            {
                __statement_bxUTm8(202);
                options.note = __extro_hNYQv5(203, __intro_ZXD72w(203, this).get('notes'))[path];
            }
        }
        if (__expression_hyFkrD(204), 'function' !== (__expression_hyFkrD(205), typeof options.type)) {
            __block_IT4f3g(44);
            {
                __statement_bxUTm8(206);
                throw new Error('Fields must be specified with a type function');
            }
        }
        if (__expression_hyFkrD(207), options.type.prototype.__proto__ !== Field.prototype) {
            __block_IT4f3g(45);
            if (__expression_hyFkrD(208), options.type === (__expression_hyFkrD(209), String)) {
                __block_IT4f3g(46);
                {
                    __statement_bxUTm8(210);
                    options.type = Field.Types.Text;
                }
            } else if (__expression_hyFkrD(211), options.type === (__expression_hyFkrD(212), Number)) {
                __block_IT4f3g(47);
                {
                    __statement_bxUTm8(213);
                    options.type = Field.Types.Number;
                }
            } else if (__expression_hyFkrD(214), options.type === (__expression_hyFkrD(215), Boolean)) {
                __block_IT4f3g(48);
                {
                    __statement_bxUTm8(216);
                    options.type = Field.Types.Boolean;
                }
            } else if (__expression_hyFkrD(217), options.type === (__expression_hyFkrD(218), Date)) {
                __block_IT4f3g(49);
                {
                    __statement_bxUTm8(219);
                    options.type = Field.Types.Datetime;
                }
            } else {
                __block_IT4f3g(50);
                {
                    __statement_bxUTm8(220);
                    throw new Error((__expression_hyFkrD(221), 'Unrecognised field constructor: ' + options.type));
                }
            }
        }
        {
            __statement_bxUTm8(222);
            this.fieldTypes[options.type.name] = true;
        }
        if (__expression_hyFkrD(223), (__expression_hyFkrD(224), options.type.name === 'html') && options.wysiwyg) {
            __block_IT4f3g(51);
            {
                __statement_bxUTm8(225);
                this.fieldTypes.wysiwyg = true;
            }
        }
        {
            __statement_bxUTm8(226);
            var field = new options.type(this, path, options);
        }
        {
            __statement_bxUTm8(227);
            this.fields[path] = field;
        }
        return __expression_hyFkrD(228), field;
    };
}
{
    __statement_bxUTm8(229);
    List.prototype.underscoreMethod = function (path, fn) {
        __block_IT4f3g(52);
        {
            __statement_bxUTm8(230);
            var target = this.underscoreMethods;
        }
        {
            __statement_bxUTm8(231);
            path = __extro_hNYQv5(232, __intro_ZXD72w(232, path).split('.'));
        }
        {
            __statement_bxUTm8(233);
            var last = __extro_hNYQv5(234, __intro_ZXD72w(234, path).pop());
        }
        {
            __statement_bxUTm8(235);
            __extro_hNYQv5(236, __intro_ZXD72w(236, path).forEach(function (part) {
                __block_IT4f3g(53);
                if (__expression_hyFkrD(237), !target[part]) {
                    __block_IT4f3g(54);
                    {
                        __statement_bxUTm8(238);
                        target[part] = {};
                    }
                }
                {
                    __statement_bxUTm8(239);
                    target = target[part];
                }
            }));
        }
        {
            __statement_bxUTm8(240);
            target[last] = fn;
        }
        return __expression_hyFkrD(241), this;
    };
}
{
    __statement_bxUTm8(242);
    __extro_hNYQv5(243, __intro_ZXD72w(243, Object).defineProperty(List.prototype, 'defaultSort', {
        get: function () {
            __block_IT4f3g(55);
            {
                __statement_bxUTm8(244);
                var ds = __extro_hNYQv5(245, __intro_ZXD72w(245, this).get('defaultSort'));
            }
            return __expression_hyFkrD(246), (__expression_hyFkrD(249), (__expression_hyFkrD(250), ds) === '__default__') ? (__expression_hyFkrD(247), __extro_hNYQv5(253, __intro_ZXD72w(253, this).get('sortable')) ? (__expression_hyFkrD(251), 'sortOrder') : (__expression_hyFkrD(252), this.namePath)) : (__expression_hyFkrD(248), ds);
        },
        set: function (value) {
            __block_IT4f3g(56);
            {
                __statement_bxUTm8(254);
                __extro_hNYQv5(255, __intro_ZXD72w(255, this).set('defaultSort', value));
            }
        }
    }));
}
{
    __statement_bxUTm8(256);
    List.prototype.expandColumns = function (cols) {
        __block_IT4f3g(57);
        if (__expression_hyFkrD(257), (__expression_hyFkrD(258), typeof cols) === 'string') {
            __block_IT4f3g(58);
            {
                __statement_bxUTm8(259);
                cols = __extro_hNYQv5(260, __intro_ZXD72w(260, cols).split(','));
            }
        }
        if (__expression_hyFkrD(261), !__extro_hNYQv5(262, __intro_ZXD72w(262, Array).isArray(cols))) {
            __block_IT4f3g(59);
            {
                __statement_bxUTm8(263);
                throw new Error('List.expandColumns: cols must be an array.');
            }
        }
        {
            __statement_bxUTm8(264);
            var list = this, expanded = [], nameCol = false;
        }
        {
            __statement_bxUTm8(265);
            var getCol = function (def) {
                __block_IT4f3g(60);
                if (__expression_hyFkrD(266), def.path === '__name__') {
                    __block_IT4f3g(61);
                    {
                        __statement_bxUTm8(267);
                        def.path = list.namePath;
                    }
                }
                {
                    __statement_bxUTm8(268);
                    var field = list.fields[def.path], col = null;
                }
                if (__expression_hyFkrD(269), field) {
                    __block_IT4f3g(62);
                    {
                        __statement_bxUTm8(270);
                        col = {
                            field: field,
                            path: field.path,
                            type: field.type,
                            label: (__expression_hyFkrD(271), def.label || field.label)
                        };
                    }
                    if (__expression_hyFkrD(272), col.type === 'relationship') {
                        __block_IT4f3g(63);
                        {
                            __statement_bxUTm8(273);
                            col.refList = col.field.refList;
                        }
                        if (col.refList) {
                            __block_IT4f3g(64);
                            {
                                __statement_bxUTm8(274);
                                col.refPath = (__expression_hyFkrD(275), def.subpath || col.refList.namePath);
                            }
                            {
                                __statement_bxUTm8(276);
                                col.subField = col.refList.fields[col.refPath];
                            }
                            {
                                __statement_bxUTm8(277);
                                col.populate = {
                                    path: col.field.path,
                                    subpath: col.refPath
                                };
                            }
                        }
                        if (__expression_hyFkrD(278), (__expression_hyFkrD(279), !def.label) && def.subpath) {
                            __block_IT4f3g(65);
                            {
                                __statement_bxUTm8(280);
                                col.label = (__expression_hyFkrD(281), (__expression_hyFkrD(282), field.label + ': ') + (col.subField ? (__expression_hyFkrD(283), col.subField.label) : (__expression_hyFkrD(284), __extro_hNYQv5(285, __intro_ZXD72w(285, utils).keyToLabel(def.subpath)))));
                            }
                        }
                    }
                } else if (__expression_hyFkrD(286), list.model.schema.paths[def.path] || list.model.schema.virtuals[def.path]) {
                    __block_IT4f3g(66);
                    {
                        __statement_bxUTm8(287);
                        col = {
                            path: def.path,
                            label: (__expression_hyFkrD(288), def.label || __extro_hNYQv5(289, __intro_ZXD72w(289, utils).keyToLabel(def.path)))
                        };
                    }
                }
                if (__expression_hyFkrD(290), col) {
                    __block_IT4f3g(67);
                    {
                        __statement_bxUTm8(291);
                        col.width = def.width;
                    }
                    if (__expression_hyFkrD(292), col.path === list.namePath) {
                        __block_IT4f3g(68);
                        {
                            __statement_bxUTm8(293);
                            col.isName = true;
                        }
                        {
                            __statement_bxUTm8(294);
                            nameCol = col;
                        }
                    }
                    if (__expression_hyFkrD(295), (__expression_hyFkrD(296), field) && field.col) {
                        __block_IT4f3g(69);
                        {
                            __statement_bxUTm8(297);
                            __extro_hNYQv5(298, __intro_ZXD72w(298, _).extend(col, field.col));
                        }
                    }
                }
                return __expression_hyFkrD(299), col;
            };
        }
        for (var i = 0; __expression_hyFkrD(300), (__expression_hyFkrD(301), i) < cols.length; __expression_hyFkrD(302), i++) {
            __block_IT4f3g(70);
            {
                __statement_bxUTm8(303);
                var def = {};
            }
            if (__expression_hyFkrD(304), (__expression_hyFkrD(305), typeof cols[i]) === 'string') {
                __block_IT4f3g(71);
                {
                    __statement_bxUTm8(306);
                    var parts = __extro_hNYQv5(307, __intro_ZXD72w(307, __extro_hNYQv5(308, __intro_ZXD72w(308, cols[i]).trim())).split('|'));
                }
                {
                    __statement_bxUTm8(309);
                    def.width = (__expression_hyFkrD(310), parts[1] || false);
                }
                {
                    __statement_bxUTm8(311);
                    parts = __extro_hNYQv5(312, __intro_ZXD72w(312, parts[0]).split(':'));
                }
                {
                    __statement_bxUTm8(313);
                    def.path = parts[0];
                }
                {
                    __statement_bxUTm8(314);
                    def.subpath = parts[1];
                }
            }
            if (__expression_hyFkrD(315), (__expression_hyFkrD(316), !__extro_hNYQv5(317, __intro_ZXD72w(317, utils).isObject(def))) || (__expression_hyFkrD(318), !def.path)) {
                __block_IT4f3g(72);
                {
                    __statement_bxUTm8(319);
                    throw new Error('List.expandColumns: column definition must contain a path.');
                }
            }
            {
                __statement_bxUTm8(320);
                var col = (__expression_hyFkrD(321), getCol(def));
            }
            if (__expression_hyFkrD(322), col) {
                __block_IT4f3g(73);
                {
                    __statement_bxUTm8(323);
                    __extro_hNYQv5(324, __intro_ZXD72w(324, expanded).push(col));
                }
            }
        }
        if (__expression_hyFkrD(325), !(__expression_hyFkrD(326), nameCol)) {
            __block_IT4f3g(74);
            {
                __statement_bxUTm8(327);
                nameCol = (__expression_hyFkrD(328), getCol({
                    path: list.namePath
                }));
            }
            if (__expression_hyFkrD(329), nameCol) {
                __block_IT4f3g(75);
                {
                    __statement_bxUTm8(330);
                    __extro_hNYQv5(331, __intro_ZXD72w(331, expanded).unshift(nameCol));
                }
            }
        }
        return __expression_hyFkrD(332), expanded;
    };
}
{
    __statement_bxUTm8(333);
    List.prototype.selectColumns = function (q, cols) {
        __block_IT4f3g(76);
        {
            __statement_bxUTm8(334);
            var select = [], populate = {}, path;
        }
        {
            __statement_bxUTm8(335);
            __extro_hNYQv5(336, __intro_ZXD72w(336, cols).forEach(function (col) {
                __block_IT4f3g(77);
                {
                    __statement_bxUTm8(337);
                    __extro_hNYQv5(338, __intro_ZXD72w(338, select).push(col.path));
                }
                if (col.populate) {
                    __block_IT4f3g(78);
                    if (__expression_hyFkrD(339), !populate[col.populate.path]) {
                        __block_IT4f3g(79);
                        {
                            __statement_bxUTm8(340);
                            populate[col.populate.path] = [];
                        }
                    }
                    {
                        __statement_bxUTm8(341);
                        __extro_hNYQv5(342, __intro_ZXD72w(342, populate[col.populate.path]).push(col.populate.subpath));
                    }
                }
            }));
        }
        {
            __statement_bxUTm8(343);
            __extro_hNYQv5(344, __intro_ZXD72w(344, q).select(__extro_hNYQv5(345, __intro_ZXD72w(345, select).join(' '))));
        }
        for (path in populate) {
            __block_IT4f3g(80);
            if (__extro_hNYQv5(346, __intro_ZXD72w(346, populate).hasOwnProperty(path))) {
                __block_IT4f3g(81);
                {
                    __statement_bxUTm8(347);
                    __extro_hNYQv5(348, __intro_ZXD72w(348, q).populate(path, __extro_hNYQv5(349, __intro_ZXD72w(349, populate[path]).join(' '))));
                }
            }
        }
    };
}
{
    __statement_bxUTm8(350);
    __extro_hNYQv5(351, __intro_ZXD72w(351, Object).defineProperty(List.prototype, 'defaultColumns', {
        get: function () {
            __block_IT4f3g(82);
            if (__expression_hyFkrD(352), !this._defaultColumns) {
                __block_IT4f3g(83);
                {
                    __statement_bxUTm8(353);
                    this._defaultColumns = __extro_hNYQv5(354, __intro_ZXD72w(354, this).expandColumns(__extro_hNYQv5(355, __intro_ZXD72w(355, this).get('defaultColumns'))));
                }
            }
            return __expression_hyFkrD(356), this._defaultColumns;
        },
        set: function (value) {
            __block_IT4f3g(84);
            {
                __statement_bxUTm8(357);
                __extro_hNYQv5(358, __intro_ZXD72w(358, this).set('defaultColumns', value));
            }
            {
                __statement_bxUTm8(359);
                __expression_hyFkrD(360), delete this._defaultColumns;
            }
        }
    }));
}
{
    __statement_bxUTm8(361);
    List.prototype.relationship = function (def) {
        __block_IT4f3g(85);
        if (__expression_hyFkrD(362), arguments.length > 1) {
            __block_IT4f3g(86);
            {
                __statement_bxUTm8(363);
                __extro_hNYQv5(364, __intro_ZXD72w(364, _).map(arguments, function (def) {
                    __block_IT4f3g(87);
                    {
                        __statement_bxUTm8(365);
                        __extro_hNYQv5(366, __intro_ZXD72w(366, this).relationship(def));
                    }
                }, this));
            }
            return __expression_hyFkrD(367), this;
        }
        if (__expression_hyFkrD(368), 'string' === (__expression_hyFkrD(369), typeof def)) {
            __block_IT4f3g(88);
            {
                __statement_bxUTm8(370);
                def = {
                    ref: def
                };
            }
        }
        if (__expression_hyFkrD(371), !def.ref) {
            __block_IT4f3g(89);
            {
                __statement_bxUTm8(372);
                throw new Error((__expression_hyFkrD(373), (__expression_hyFkrD(374), 'List Relationships must be specified with an object containing ref (' + this.key) + ')'));
            }
        }
        if (__expression_hyFkrD(375), !def.refPath) {
            __block_IT4f3g(90);
            {
                __statement_bxUTm8(376);
                def.refPath = __extro_hNYQv5(377, __intro_ZXD72w(377, utils).downcase(this.key));
            }
        }
        if (__expression_hyFkrD(378), !def.path) {
            __block_IT4f3g(91);
            {
                __statement_bxUTm8(379);
                def.path = __extro_hNYQv5(380, __intro_ZXD72w(380, utils).keyToProperty(def.ref, true));
            }
        }
        {
            __statement_bxUTm8(381);
            __extro_hNYQv5(382, __intro_ZXD72w(382, Object).defineProperty(def, 'refList', {
                get: function () {
                    __block_IT4f3g(92);
                    return __expression_hyFkrD(383), __extro_hNYQv5(384, __intro_ZXD72w(384, keystone).list(def.ref));
                }
            }));
        }
        {
            __statement_bxUTm8(385);
            __extro_hNYQv5(386, __intro_ZXD72w(386, Object).defineProperty(def, 'isValid', {
                get: function () {
                    __block_IT4f3g(93);
                    return __expression_hyFkrD(387), __extro_hNYQv5(390, __intro_ZXD72w(390, keystone).list(def.ref)) ? (__expression_hyFkrD(388), true) : (__expression_hyFkrD(389), false);
                }
            }));
        }
        {
            __statement_bxUTm8(391);
            this.relationships[def.path] = def;
        }
        return __expression_hyFkrD(392), this;
    };
}
{
    __statement_bxUTm8(393);
    List.prototype.register = function () {
        __block_IT4f3g(94);
        {
            __statement_bxUTm8(394);
            var list = this;
        }
        {
            __statement_bxUTm8(395);
            __extro_hNYQv5(396, __intro_ZXD72w(396, __extro_hNYQv5(397, __intro_ZXD72w(397, this.schema).virtual('list'))).get(function () {
                __block_IT4f3g(95);
                return __expression_hyFkrD(398), list;
            }));
        }
        if (__extro_hNYQv5(399, __intro_ZXD72w(399, this).get('sortable'))) {
            __block_IT4f3g(96);
            {
                __statement_bxUTm8(400);
                __extro_hNYQv5(401, __intro_ZXD72w(401, schemaPlugins.sortable).apply(this));
            }
        }
        if (__extro_hNYQv5(402, __intro_ZXD72w(402, this).get('autokey'))) {
            __block_IT4f3g(97);
            {
                __statement_bxUTm8(403);
                __extro_hNYQv5(404, __intro_ZXD72w(404, schemaPlugins.autokey).apply(this));
            }
        }
        if (__extro_hNYQv5(405, __intro_ZXD72w(405, this).get('track'))) {
            __block_IT4f3g(98);
            {
                __statement_bxUTm8(406);
                __extro_hNYQv5(407, __intro_ZXD72w(407, schemaPlugins.track).apply(this));
            }
        }
        if (__expression_hyFkrD(408), !__extro_hNYQv5(409, __intro_ZXD72w(409, _).isEmpty(this.relationships))) {
            __block_IT4f3g(99);
            {
                __statement_bxUTm8(410);
                this.schema.methods.getRelated = schemaPlugins.methods.getRelated;
            }
            {
                __statement_bxUTm8(411);
                this.schema.methods.populateRelated = schemaPlugins.methods.populateRelated;
            }
            if (__expression_hyFkrD(412), !this.schema.options.toObject) {
                __block_IT4f3g(100);
                {
                    __statement_bxUTm8(413);
                    this.schema.options.toObject = {};
                }
            }
            {
                __statement_bxUTm8(414);
                this.schema.options.toObject.transform = schemaPlugins.options.transform;
            }
        }
        {
            __statement_bxUTm8(415);
            __extro_hNYQv5(416, __intro_ZXD72w(416, __extro_hNYQv5(417, __intro_ZXD72w(417, this.schema).virtual('_'))).get(function () {
                __block_IT4f3g(101);
                if (__expression_hyFkrD(418), !this.__methods) {
                    __block_IT4f3g(102);
                    {
                        __statement_bxUTm8(419);
                        this.__methods = __extro_hNYQv5(420, __intro_ZXD72w(420, utils).bindMethods(list.underscoreMethods, this));
                    }
                }
                return __expression_hyFkrD(421), this.__methods;
            }));
        }
        {
            __statement_bxUTm8(422);
            __extro_hNYQv5(423, __intro_ZXD72w(423, this.schema).method('getUpdateHandler', function (req, res, ops) {
                __block_IT4f3g(103);
                return __expression_hyFkrD(424), new UpdateHandler(list, this, req, res, ops);
            }));
        }
        {
            __statement_bxUTm8(425);
            this.model = __extro_hNYQv5(426, __intro_ZXD72w(426, keystone.mongoose).model(this.key, this.schema));
        }
        {
            __statement_bxUTm8(427);
            __extro_hNYQv5(428, __intro_ZXD72w(428, (__expression_hyFkrD(429), require('../'))).list(this));
        }
        return __expression_hyFkrD(430), this;
    };
}
{
    __statement_bxUTm8(431);
    List.prototype.getDocumentName = function (doc, escape) {
        __block_IT4f3g(104);
        {
            __statement_bxUTm8(432);
            var name = (__expression_hyFkrD(433), String(this.nameField ? (__expression_hyFkrD(434), __extro_hNYQv5(436, __intro_ZXD72w(436, this.nameField).format(doc))) : (__expression_hyFkrD(435), __extro_hNYQv5(437, __intro_ZXD72w(437, doc).get(this.namePath)))));
        }
        return __expression_hyFkrD(438), (__expression_hyFkrD(441), escape) ? (__expression_hyFkrD(439), __extro_hNYQv5(442, __intro_ZXD72w(442, utils).encodeHTMLEntities(name))) : (__expression_hyFkrD(440), name);
    };
}
{
    __statement_bxUTm8(443);
    List.prototype.processFilters = function (q) {
        __block_IT4f3g(105);
        {
            __statement_bxUTm8(444);
            var me = this;
        }
        {
            __statement_bxUTm8(445);
            var filters = {};
        }
        {
            __statement_bxUTm8(446);
            __extro_hNYQv5(447, __intro_ZXD72w(447, __extro_hNYQv5(448, __intro_ZXD72w(448, __extro_hNYQv5(449, __intro_ZXD72w(449, queryfilterlib.QueryFilters).create(q))).getFilters())).forEach(function (filter) {
                __block_IT4f3g(106);
                {
                    __statement_bxUTm8(450);
                    filter.path = filter.key;
                }
                {
                    __statement_bxUTm8(451);
                    filter.field = me.fields[filter.key];
                }
                {
                    __statement_bxUTm8(452);
                    filters[filter.path] = filter;
                }
            }));
        }
        return __expression_hyFkrD(453), filters;
    };
}
{
    __statement_bxUTm8(454);
    List.prototype.getSearchFilters = function (search, add) {
        __block_IT4f3g(107);
        {
            __statement_bxUTm8(455);
            var filters = {}, list = this;
        }
        {
            __statement_bxUTm8(456);
            search = __extro_hNYQv5(457, __intro_ZXD72w(457, (__expression_hyFkrD(458), String((__expression_hyFkrD(459), (__expression_hyFkrD(460), search) || '')))).trim());
        }
        if (search.length) {
            __block_IT4f3g(108);
            {
                __statement_bxUTm8(461);
                var searchFilter, searchParts = __extro_hNYQv5(462, __intro_ZXD72w(462, search).split(' ')), searchRx = new RegExp(__extro_hNYQv5(463, __intro_ZXD72w(463, utils).escapeRegExp(search)), 'i'), splitSearchRx = new RegExp((__expression_hyFkrD(466), searchParts.length > 1) ? (__expression_hyFkrD(464), __extro_hNYQv5(467, __intro_ZXD72w(467, __extro_hNYQv5(468, __intro_ZXD72w(468, _).map(searchParts, utils.escapeRegExp))).join('|'))) : (__expression_hyFkrD(465), search), 'i'), searchFields = __extro_hNYQv5(469, __intro_ZXD72w(469, this).get('searchFields')), searchFilters = [], searchIdField = __extro_hNYQv5(470, __intro_ZXD72w(470, utils).isValidObjectId(search));
            }
            if (__expression_hyFkrD(471), 'string' === (__expression_hyFkrD(472), typeof searchFields)) {
                __block_IT4f3g(109);
                {
                    __statement_bxUTm8(473);
                    searchFields = __extro_hNYQv5(474, __intro_ZXD72w(474, searchFields).split(','));
                }
            }
            {
                __statement_bxUTm8(475);
                __extro_hNYQv5(476, __intro_ZXD72w(476, searchFields).forEach(function (path) {
                    __block_IT4f3g(110);
                    {
                        __statement_bxUTm8(477);
                        path = __extro_hNYQv5(478, __intro_ZXD72w(478, path).trim());
                    }
                    if (__expression_hyFkrD(479), (__expression_hyFkrD(480), path) === '__name__') {
                        __block_IT4f3g(111);
                        {
                            __statement_bxUTm8(481);
                            path = list.mappings.name;
                        }
                    }
                    {
                        __statement_bxUTm8(482);
                        var field = list.fields[path];
                    }
                    if (__expression_hyFkrD(483), (__expression_hyFkrD(484), field) && (__expression_hyFkrD(485), field.type === 'name')) {
                        __block_IT4f3g(112);
                        {
                            __statement_bxUTm8(486);
                            var first = {};
                        }
                        {
                            __statement_bxUTm8(487);
                            first[field.paths.first] = splitSearchRx;
                        }
                        {
                            __statement_bxUTm8(488);
                            var last = {};
                        }
                        {
                            __statement_bxUTm8(489);
                            last[field.paths.last] = splitSearchRx;
                        }
                        {
                            __statement_bxUTm8(490);
                            searchFilter = {};
                        }
                        {
                            __statement_bxUTm8(491);
                            searchFilter.$or = [
                                first,
                                last
                            ];
                        }
                        {
                            __statement_bxUTm8(492);
                            __extro_hNYQv5(493, __intro_ZXD72w(493, searchFilters).push(searchFilter));
                        }
                    } else {
                        __block_IT4f3g(113);
                        {
                            __statement_bxUTm8(494);
                            searchFilter = {};
                        }
                        {
                            __statement_bxUTm8(495);
                            searchFilter[path] = searchRx;
                        }
                        {
                            __statement_bxUTm8(496);
                            __extro_hNYQv5(497, __intro_ZXD72w(497, searchFilters).push(searchFilter));
                        }
                    }
                }));
            }
            if (list.autokey) {
                __block_IT4f3g(114);
                {
                    __statement_bxUTm8(498);
                    searchFilter = {};
                }
                {
                    __statement_bxUTm8(499);
                    searchFilter[list.autokey.path] = searchRx;
                }
                {
                    __statement_bxUTm8(500);
                    __extro_hNYQv5(501, __intro_ZXD72w(501, searchFilters).push(searchFilter));
                }
            }
            if (__expression_hyFkrD(502), searchIdField) {
                __block_IT4f3g(115);
                {
                    __statement_bxUTm8(503);
                    searchFilter = {};
                }
                {
                    __statement_bxUTm8(504);
                    searchFilter._id = search;
                }
                {
                    __statement_bxUTm8(505);
                    __extro_hNYQv5(506, __intro_ZXD72w(506, searchFilters).push(searchFilter));
                }
            }
            if (__expression_hyFkrD(507), searchFilters.length > 1) {
                __block_IT4f3g(116);
                {
                    __statement_bxUTm8(508);
                    filters.$or = searchFilters;
                }
            } else if (searchFilters.length) {
                __block_IT4f3g(117);
                {
                    __statement_bxUTm8(509);
                    filters = searchFilters[0];
                }
            }
        }
        if (__expression_hyFkrD(510), add) {
            __block_IT4f3g(118);
            {
                __statement_bxUTm8(511);
                __extro_hNYQv5(512, __intro_ZXD72w(512, _).each(add, function (filter) {
                    __block_IT4f3g(119);
                    {
                        __statement_bxUTm8(513);
                        var cond, path = filter.key, value = filter.value;
                    }
                    switch (filter.field.type) {
                    case 'boolean': {
                            __block_IT4f3g(120);
                            if (__expression_hyFkrD(514), value) {
                                __block_IT4f3g(121);
                                {
                                    __statement_bxUTm8(515);
                                    filters[path] = true;
                                }
                            } else {
                                __block_IT4f3g(122);
                                {
                                    __statement_bxUTm8(516);
                                    filters[path] = {
                                        $ne: true
                                    };
                                }
                            }
                            break;
                        }
                    case 'localfile':
                    case 'cloudinaryimage':
                    case 'cloudinaryimages':
                    case 's3file':
                    case 'name':
                    case 'password': {
                            __block_IT4f3g(123);
                            break;
                        }
                    case 'location': {
                            __block_IT4f3g(124);
                            {
                                __statement_bxUTm8(517);
                                __extro_hNYQv5(518, __intro_ZXD72w(518, _).each({
                                    address: 'street1',
                                    suburb: 'suburb',
                                    state: 'state',
                                    postcode: 'postcode',
                                    country: 'country'
                                }, function (pathKey, valueKey) {
                                    __block_IT4f3g(125);
                                    {
                                        __statement_bxUTm8(519);
                                        var value = filter[valueKey];
                                    }
                                    if (__expression_hyFkrD(520), value) {
                                        __block_IT4f3g(126);
                                        {
                                            __statement_bxUTm8(521);
                                            filters[filter.field.paths[pathKey]] = new RegExp(__extro_hNYQv5(522, __intro_ZXD72w(522, utils).escapeRegExp(value)), 'i');
                                        }
                                    }
                                }));
                            }
                            break;
                        }
                    case 'relationship': {
                            __block_IT4f3g(127);
                            if (__expression_hyFkrD(523), value) {
                                __block_IT4f3g(128);
                                if (filter.field.many) {
                                    __block_IT4f3g(129);
                                    {
                                        __statement_bxUTm8(524);
                                        filters[path] = filter.inverse ? (__expression_hyFkrD(525), {
                                            $nin: [
                                                value
                                            ]
                                        }) : (__expression_hyFkrD(526), {
                                            $in: [
                                                value
                                            ]
                                        });
                                    }
                                } else {
                                    __block_IT4f3g(130);
                                    {
                                        __statement_bxUTm8(527);
                                        filters[path] = filter.inverse ? (__expression_hyFkrD(528), {
                                            $ne: value
                                        }) : (__expression_hyFkrD(529), value);
                                    }
                                }
                            } else {
                                __block_IT4f3g(131);
                                if (filter.field.many) {
                                    __block_IT4f3g(132);
                                    {
                                        __statement_bxUTm8(530);
                                        filters[path] = filter.inverse ? (__expression_hyFkrD(531), {
                                            $not: {
                                                $size: 0
                                            }
                                        }) : (__expression_hyFkrD(532), {
                                            $size: 0
                                        });
                                    }
                                } else {
                                    __block_IT4f3g(133);
                                    {
                                        __statement_bxUTm8(533);
                                        filters[path] = filter.inverse ? (__expression_hyFkrD(534), {
                                            $ne: null
                                        }) : (__expression_hyFkrD(535), null);
                                    }
                                }
                            }
                            break;
                        }
                    case 'select': {
                            __block_IT4f3g(134);
                            if (filter.value) {
                                __block_IT4f3g(135);
                                {
                                    __statement_bxUTm8(536);
                                    filters[path] = filter.inverse ? (__expression_hyFkrD(537), {
                                        $ne: value
                                    }) : (__expression_hyFkrD(538), value);
                                }
                            } else {
                                __block_IT4f3g(136);
                                {
                                    __statement_bxUTm8(539);
                                    filters[path] = filter.inverse ? (__expression_hyFkrD(540), {
                                        $nin: [
                                            '',
                                            null
                                        ]
                                    }) : (__expression_hyFkrD(541), {
                                        $in: [
                                            '',
                                            null
                                        ]
                                    });
                                }
                            }
                            break;
                        }
                    case 'number':
                    case 'money': {
                            __block_IT4f3g(137);
                            if (__expression_hyFkrD(542), filter.operator === 'bt') {
                                __block_IT4f3g(138);
                                {
                                    __statement_bxUTm8(543);
                                    value = [
                                        __extro_hNYQv5(544, __intro_ZXD72w(544, utils).number(value[0])),
                                        __extro_hNYQv5(545, __intro_ZXD72w(545, utils).number(value[1]))
                                    ];
                                }
                                if (__expression_hyFkrD(546), (__expression_hyFkrD(547), !(__expression_hyFkrD(548), isNaN(value[0]))) && (__expression_hyFkrD(549), !(__expression_hyFkrD(550), isNaN(value[1])))) {
                                    __block_IT4f3g(139);
                                    {
                                        __statement_bxUTm8(551);
                                        filters[path] = {
                                            $gte: value[0],
                                            $lte: value[1]
                                        };
                                    }
                                } else {
                                    __block_IT4f3g(140);
                                    {
                                        __statement_bxUTm8(552);
                                        filters[path] = null;
                                    }
                                }
                            } else {
                                __block_IT4f3g(141);
                                {
                                    __statement_bxUTm8(553);
                                    value = __extro_hNYQv5(554, __intro_ZXD72w(554, utils).number(value));
                                }
                                if (__expression_hyFkrD(555), !(__expression_hyFkrD(556), isNaN(value))) {
                                    __block_IT4f3g(142);
                                    if (__expression_hyFkrD(557), filter.operator === 'gt') {
                                        __block_IT4f3g(143);
                                        {
                                            __statement_bxUTm8(558);
                                            filters[path] = {
                                                $gt: value
                                            };
                                        }
                                    } else if (__expression_hyFkrD(559), filter.operator === 'lt') {
                                        __block_IT4f3g(144);
                                        {
                                            __statement_bxUTm8(560);
                                            filters[path] = {
                                                $lt: value
                                            };
                                        }
                                    } else {
                                        __block_IT4f3g(145);
                                        {
                                            __statement_bxUTm8(561);
                                            filters[path] = value;
                                        }
                                    }
                                } else {
                                    __block_IT4f3g(146);
                                    {
                                        __statement_bxUTm8(562);
                                        filters[path] = null;
                                    }
                                }
                            }
                            break;
                        }
                    case 'date':
                    case 'datetime': {
                            __block_IT4f3g(147);
                            if (__expression_hyFkrD(563), filter.operator === 'bt') {
                                __block_IT4f3g(148);
                                {
                                    __statement_bxUTm8(564);
                                    value = [
                                        (__expression_hyFkrD(565), moment(value[0])),
                                        (__expression_hyFkrD(566), moment(value[1]))
                                    ];
                                }
                                if (__expression_hyFkrD(567), (__expression_hyFkrD(568), value[0] && __extro_hNYQv5(569, __intro_ZXD72w(569, value[0]).isValid())) && (__expression_hyFkrD(570), value[1] && __extro_hNYQv5(571, __intro_ZXD72w(571, value[0]).isValid()))) {
                                    __block_IT4f3g(149);
                                    {
                                        __statement_bxUTm8(572);
                                        filters[path] = {
                                            $gte: __extro_hNYQv5(573, __intro_ZXD72w(573, __extro_hNYQv5(574, __intro_ZXD72w(574, (__expression_hyFkrD(575), moment(value[0]))).startOf('day'))).toDate()),
                                            $lte: __extro_hNYQv5(576, __intro_ZXD72w(576, __extro_hNYQv5(577, __intro_ZXD72w(577, (__expression_hyFkrD(578), moment(value[1]))).endOf('day'))).toDate())
                                        };
                                    }
                                }
                            } else {
                                __block_IT4f3g(150);
                                {
                                    __statement_bxUTm8(579);
                                    value = (__expression_hyFkrD(580), moment(value));
                                }
                                if (__expression_hyFkrD(581), (__expression_hyFkrD(582), value) && __extro_hNYQv5(583, __intro_ZXD72w(583, value).isValid())) {
                                    __block_IT4f3g(151);
                                    {
                                        __statement_bxUTm8(584);
                                        var start = __extro_hNYQv5(585, __intro_ZXD72w(585, __extro_hNYQv5(586, __intro_ZXD72w(586, (__expression_hyFkrD(587), moment(value))).startOf('day'))).toDate());
                                    }
                                    {
                                        __statement_bxUTm8(588);
                                        var end = __extro_hNYQv5(589, __intro_ZXD72w(589, __extro_hNYQv5(590, __intro_ZXD72w(590, (__expression_hyFkrD(591), moment(value))).endOf('day'))).toDate());
                                    }
                                    if (__expression_hyFkrD(592), filter.operator === 'gt') {
                                        __block_IT4f3g(152);
                                        {
                                            __statement_bxUTm8(593);
                                            filters[path] = {
                                                $gt: end
                                            };
                                        }
                                    } else if (__expression_hyFkrD(594), filter.operator === 'lt') {
                                        __block_IT4f3g(153);
                                        {
                                            __statement_bxUTm8(595);
                                            filters[path] = {
                                                $lt: start
                                            };
                                        }
                                    } else {
                                        __block_IT4f3g(154);
                                        {
                                            __statement_bxUTm8(596);
                                            filters[path] = {
                                                $lte: end,
                                                $gte: start
                                            };
                                        }
                                    }
                                }
                            }
                            break;
                        }
                    case 'text':
                    case 'textarea':
                    case 'html':
                    case 'email':
                    case 'url':
                    case 'key': {
                            __block_IT4f3g(155);
                            if (filter.exact) {
                                __block_IT4f3g(156);
                                if (__expression_hyFkrD(597), value) {
                                    __block_IT4f3g(157);
                                    {
                                        __statement_bxUTm8(598);
                                        cond = new RegExp((__expression_hyFkrD(599), (__expression_hyFkrD(600), '^' + __extro_hNYQv5(601, __intro_ZXD72w(601, utils).escapeRegExp(value))) + '$'), 'i');
                                    }
                                    {
                                        __statement_bxUTm8(602);
                                        filters[path] = filter.inverse ? (__expression_hyFkrD(603), {
                                            $not: cond
                                        }) : (__expression_hyFkrD(604), cond);
                                    }
                                } else {
                                    __block_IT4f3g(158);
                                    if (filter.inverse) {
                                        __block_IT4f3g(159);
                                        {
                                            __statement_bxUTm8(605);
                                            filters[path] = {
                                                $nin: [
                                                    '',
                                                    null
                                                ]
                                            };
                                        }
                                    } else {
                                        __block_IT4f3g(160);
                                        {
                                            __statement_bxUTm8(606);
                                            filters[path] = {
                                                $in: [
                                                    '',
                                                    null
                                                ]
                                            };
                                        }
                                    }
                                }
                            } else if (__expression_hyFkrD(607), value) {
                                __block_IT4f3g(161);
                                {
                                    __statement_bxUTm8(608);
                                    cond = new RegExp(__extro_hNYQv5(609, __intro_ZXD72w(609, utils).escapeRegExp(value)), 'i');
                                }
                                {
                                    __statement_bxUTm8(610);
                                    filters[path] = filter.inverse ? (__expression_hyFkrD(611), {
                                        $not: cond
                                    }) : (__expression_hyFkrD(612), cond);
                                }
                            }
                            break;
                        }
                    }
                }));
            }
        }
        return __expression_hyFkrD(613), filters;
    };
}
{
    __statement_bxUTm8(614);
    List.prototype.updateAll = function (data, callback) {
        __block_IT4f3g(162);
        if (__expression_hyFkrD(615), 'function' === (__expression_hyFkrD(616), typeof data)) {
            __block_IT4f3g(163);
            {
                __statement_bxUTm8(617);
                callback = data;
            }
            {
                __statement_bxUTm8(618);
                data = null;
            }
        }
        {
            __statement_bxUTm8(619);
            callback = (__expression_hyFkrD(620), (__expression_hyFkrD(621), callback) || function () {
                __block_IT4f3g(164);
            });
        }
        {
            __statement_bxUTm8(622);
            __extro_hNYQv5(623, __intro_ZXD72w(623, this.model).find(function (err, results) {
                __block_IT4f3g(165);
                if (__expression_hyFkrD(624), err) {
                    __block_IT4f3g(166);
                    return __expression_hyFkrD(625), (__expression_hyFkrD(626), callback(err));
                }
                {
                    __statement_bxUTm8(627);
                    __extro_hNYQv5(628, __intro_ZXD72w(628, async).eachSeries(results, function (doc, next) {
                        __block_IT4f3g(167);
                        if (__expression_hyFkrD(629), data) {
                            __block_IT4f3g(168);
                            {
                                __statement_bxUTm8(630);
                                __extro_hNYQv5(631, __intro_ZXD72w(631, doc).set(data));
                            }
                        }
                        {
                            __statement_bxUTm8(632);
                            __extro_hNYQv5(633, __intro_ZXD72w(633, doc).save(next));
                        }
                    }, function (err) {
                        __block_IT4f3g(169);
                        {
                            __statement_bxUTm8(634);
                            __expression_hyFkrD(635), callback(err);
                        }
                    }));
                }
            }));
        }
    };
}
{
    __statement_bxUTm8(636);
    List.prototype.getUniqueValue = function (path, generator, limit, callback) {
        __block_IT4f3g(170);
        {
            __statement_bxUTm8(637);
            var model = this.model, count = 0, value;
        }
        if (__extro_hNYQv5(638, __intro_ZXD72w(638, utils).isFunction(limit))) {
            __block_IT4f3g(171);
            {
                __statement_bxUTm8(639);
                callback = limit;
            }
            {
                __statement_bxUTm8(640);
                limit = 10;
            }
        }
        if (__extro_hNYQv5(641, __intro_ZXD72w(641, utils).isArray(generator))) {
            __block_IT4f3g(172);
            {
                __statement_bxUTm8(642);
                var fn = generator[0], args = __extro_hNYQv5(643, __intro_ZXD72w(643, generator).slice(1));
            }
            {
                __statement_bxUTm8(644);
                generator = function () {
                    __block_IT4f3g(173);
                    return __expression_hyFkrD(645), __extro_hNYQv5(646, __intro_ZXD72w(646, fn).apply(this, args));
                };
            }
        }
        {
            __statement_bxUTm8(647);
            var check = function () {
                __block_IT4f3g(174);
                if (__expression_hyFkrD(648), (__expression_hyFkrD(649), count++) > 10) {
                    __block_IT4f3g(175);
                    return __expression_hyFkrD(650), (__expression_hyFkrD(651), callback(undefined, undefined));
                }
                {
                    __statement_bxUTm8(652);
                    value = (__expression_hyFkrD(653), generator());
                }
                {
                    __statement_bxUTm8(654);
                    __extro_hNYQv5(655, __intro_ZXD72w(655, __extro_hNYQv5(656, __intro_ZXD72w(656, __extro_hNYQv5(657, __intro_ZXD72w(657, model).count())).where(path, value))).exec(function (err, matches) {
                        __block_IT4f3g(176);
                        if (__expression_hyFkrD(658), err) {
                            __block_IT4f3g(177);
                            return __expression_hyFkrD(659), (__expression_hyFkrD(660), callback(err));
                        }
                        if (__expression_hyFkrD(661), matches) {
                            __block_IT4f3g(178);
                            return __expression_hyFkrD(662), (__expression_hyFkrD(663), check());
                        }
                        {
                            __statement_bxUTm8(664);
                            __expression_hyFkrD(665), callback(undefined, value);
                        }
                    }));
                }
            };
        }
        {
            __statement_bxUTm8(666);
            __expression_hyFkrD(667), check();
        }
    };
}
{
    __statement_bxUTm8(668);
    List.prototype.getPages = function (options, maxPages) {
        __block_IT4f3g(179);
        {
            __statement_bxUTm8(669);
            var surround = __extro_hNYQv5(670, __intro_ZXD72w(670, Math).floor((__expression_hyFkrD(671), (__expression_hyFkrD(672), maxPages) / 2))), firstPage = (__expression_hyFkrD(675), maxPages) ? (__expression_hyFkrD(673), __extro_hNYQv5(676, __intro_ZXD72w(676, Math).max(1, (__expression_hyFkrD(677), options.currentPage - (__expression_hyFkrD(678), surround))))) : (__expression_hyFkrD(674), 1), padRight = __extro_hNYQv5(679, __intro_ZXD72w(679, Math).max((__expression_hyFkrD(680), (__expression_hyFkrD(681), (__expression_hyFkrD(682), options.currentPage - (__expression_hyFkrD(683), surround)) - 1) * (__expression_hyFkrD(684), -1)), 0)), lastPage = (__expression_hyFkrD(687), maxPages) ? (__expression_hyFkrD(685), __extro_hNYQv5(688, __intro_ZXD72w(688, Math).min(options.totalPages, (__expression_hyFkrD(689), (__expression_hyFkrD(690), options.currentPage + (__expression_hyFkrD(691), surround)) + (__expression_hyFkrD(692), padRight))))) : (__expression_hyFkrD(686), options.totalPages), padLeft = __extro_hNYQv5(693, __intro_ZXD72w(693, Math).max((__expression_hyFkrD(694), (__expression_hyFkrD(695), options.currentPage + (__expression_hyFkrD(696), surround)) - (__expression_hyFkrD(697), lastPage)), 0));
        }
        {
            __statement_bxUTm8(698);
            options.pages = [];
        }
        {
            __statement_bxUTm8(699);
            firstPage = __extro_hNYQv5(700, __intro_ZXD72w(700, Math).max(__extro_hNYQv5(701, __intro_ZXD72w(701, Math).min(firstPage, (__expression_hyFkrD(702), (__expression_hyFkrD(703), firstPage) - (__expression_hyFkrD(704), padLeft)))), 1));
        }
        for (var i = firstPage; __expression_hyFkrD(705), (__expression_hyFkrD(706), i) <= (__expression_hyFkrD(707), lastPage); __expression_hyFkrD(708), i++) {
            __block_IT4f3g(180);
            {
                __statement_bxUTm8(709);
                __extro_hNYQv5(710, __intro_ZXD72w(710, options.pages).push(i));
            }
        }
        if (__expression_hyFkrD(711), (__expression_hyFkrD(712), firstPage) !== 1) {
            __block_IT4f3g(181);
            {
                __statement_bxUTm8(713);
                __extro_hNYQv5(714, __intro_ZXD72w(714, options.pages).shift());
            }
            {
                __statement_bxUTm8(715);
                __extro_hNYQv5(716, __intro_ZXD72w(716, options.pages).unshift('...'));
            }
        }
        if (__expression_hyFkrD(717), (__expression_hyFkrD(718), lastPage) !== (__expression_hyFkrD(719), Number(options.totalPages))) {
            __block_IT4f3g(182);
            {
                __statement_bxUTm8(720);
                __extro_hNYQv5(721, __intro_ZXD72w(721, options.pages).pop());
            }
            {
                __statement_bxUTm8(722);
                __extro_hNYQv5(723, __intro_ZXD72w(723, options.pages).push('...'));
            }
        }
    };
}
{
    __statement_bxUTm8(724);
    List.prototype.paginate = function (options, callback) {
        __block_IT4f3g(183);
        {
            __statement_bxUTm8(725);
            var list = this, model = this.model;
        }
        {
            __statement_bxUTm8(726);
            options = (__expression_hyFkrD(727), (__expression_hyFkrD(728), options) || {});
        }
        {
            __statement_bxUTm8(729);
            var query = __extro_hNYQv5(730, __intro_ZXD72w(730, model).find(options.filters));
        }
        {
            __statement_bxUTm8(731);
            query._original_exec = query.exec;
        }
        {
            __statement_bxUTm8(732);
            query._original_sort = query.sort;
        }
        {
            __statement_bxUTm8(733);
            query._original_select = query.select;
        }
        {
            __statement_bxUTm8(734);
            var currentPage = (__expression_hyFkrD(735), (__expression_hyFkrD(736), Number(options.page)) || 1), resultsPerPage = (__expression_hyFkrD(737), (__expression_hyFkrD(738), Number(options.perPage)) || 50), maxPages = (__expression_hyFkrD(739), (__expression_hyFkrD(740), Number(options.maxPages)) || 10), skip = (__expression_hyFkrD(741), (__expression_hyFkrD(742), (__expression_hyFkrD(743), currentPage) - 1) * (__expression_hyFkrD(744), resultsPerPage));
        }
        {
            __statement_bxUTm8(745);
            list.pagination = {
                maxPages: maxPages
            };
        }
        {
            __statement_bxUTm8(746);
            query.select = function () {
                __block_IT4f3g(184);
                {
                    __statement_bxUTm8(747);
                    options.select = arguments[0];
                }
                return __expression_hyFkrD(748), query;
            };
        }
        {
            __statement_bxUTm8(749);
            query.sort = function () {
                __block_IT4f3g(185);
                {
                    __statement_bxUTm8(750);
                    options.sort = arguments[0];
                }
                return __expression_hyFkrD(751), query;
            };
        }
        {
            __statement_bxUTm8(752);
            query.exec = function (callback) {
                __block_IT4f3g(186);
                {
                    __statement_bxUTm8(753);
                    __extro_hNYQv5(754, __intro_ZXD72w(754, query).count(function (err, count) {
                        __block_IT4f3g(187);
                        if (__expression_hyFkrD(755), err) {
                            __block_IT4f3g(188);
                            return __expression_hyFkrD(756), (__expression_hyFkrD(757), callback(err));
                        }
                        {
                            __statement_bxUTm8(758);
                            __extro_hNYQv5(759, __intro_ZXD72w(759, __extro_hNYQv5(760, __intro_ZXD72w(760, __extro_hNYQv5(761, __intro_ZXD72w(761, query).find())).limit(resultsPerPage))).skip(skip));
                        }
                        if (options.select) {
                            __block_IT4f3g(189);
                            {
                                __statement_bxUTm8(762);
                                __extro_hNYQv5(763, __intro_ZXD72w(763, query)._original_select(options.select));
                            }
                        }
                        if (options.sort) {
                            __block_IT4f3g(190);
                            {
                                __statement_bxUTm8(764);
                                __extro_hNYQv5(765, __intro_ZXD72w(765, query)._original_sort(options.sort));
                            }
                        }
                        {
                            __statement_bxUTm8(766);
                            __extro_hNYQv5(767, __intro_ZXD72w(767, query)._original_exec(function (err, results) {
                                __block_IT4f3g(191);
                                if (__expression_hyFkrD(768), err) {
                                    __block_IT4f3g(192);
                                    return __expression_hyFkrD(769), (__expression_hyFkrD(770), callback(err));
                                }
                                {
                                    __statement_bxUTm8(771);
                                    var totalPages = __extro_hNYQv5(772, __intro_ZXD72w(772, Math).ceil((__expression_hyFkrD(773), (__expression_hyFkrD(774), count) / (__expression_hyFkrD(775), resultsPerPage))));
                                }
                                {
                                    __statement_bxUTm8(776);
                                    var rtn = {
                                            total: count,
                                            results: results,
                                            currentPage: currentPage,
                                            totalPages: totalPages,
                                            pages: [],
                                            previous: (__expression_hyFkrD(779), (__expression_hyFkrD(780), currentPage) > 1) ? (__expression_hyFkrD(777), (__expression_hyFkrD(781), (__expression_hyFkrD(782), currentPage) - 1)) : (__expression_hyFkrD(778), false),
                                            next: (__expression_hyFkrD(785), (__expression_hyFkrD(786), currentPage) < (__expression_hyFkrD(787), totalPages)) ? (__expression_hyFkrD(783), (__expression_hyFkrD(788), (__expression_hyFkrD(789), currentPage) + 1)) : (__expression_hyFkrD(784), false),
                                            first: (__expression_hyFkrD(790), (__expression_hyFkrD(791), skip) + 1),
                                            last: (__expression_hyFkrD(792), (__expression_hyFkrD(793), skip) + results.length)
                                        };
                                }
                                {
                                    __statement_bxUTm8(794);
                                    __extro_hNYQv5(795, __intro_ZXD72w(795, list).getPages(rtn, maxPages));
                                }
                                {
                                    __statement_bxUTm8(796);
                                    __expression_hyFkrD(797), callback(err, rtn);
                                }
                            }));
                        }
                    }));
                }
            };
        }
        if (__expression_hyFkrD(798), callback) {
            __block_IT4f3g(193);
            return __expression_hyFkrD(799), (__expression_hyFkrD(800), query(callback));
        } else {
            __block_IT4f3g(194);
            return __expression_hyFkrD(801), query;
        }
    };
}
{
    __statement_bxUTm8(802);
    exports = module.exports = List;
}
