// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_azurefile.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_WIELXX, __expression_hAAjrm, __block_DnoaWA;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_WIELXX = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/azurefile.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_hAAjrm = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/azurefile.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_DnoaWA = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/azurefile.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_HNofZ9 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_o1bl91 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/azurefile.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_WIELXX(0);
    var _ = (__expression_hAAjrm(1), require('underscore')), moment = (__expression_hAAjrm(2), require('moment')), keystone = (__expression_hAAjrm(3), require('../../')), async = (__expression_hAAjrm(4), require('async')), util = (__expression_hAAjrm(5), require('util')), azure = (__expression_hAAjrm(6), require('azure')), utils = (__expression_hAAjrm(7), require('keystone-utils')), super_ = (__expression_hAAjrm(8), require('../field'));
}
function azurefile(list, path, options) {
    __block_DnoaWA(0);
    {
        __statement_WIELXX(9);
        this._underscoreMethods = [
            'format',
            'uploadFile'
        ];
    }
    {
        __statement_WIELXX(10);
        this._pre = {
            upload: []
        };
    }
    {
        __statement_WIELXX(11);
        options.nofilter = true;
    }
    if (options.initial) {
        __block_DnoaWA(1);
        {
            __statement_WIELXX(12);
            throw new Error((__expression_hAAjrm(13), (__expression_hAAjrm(14), (__expression_hAAjrm(15), (__expression_hAAjrm(16), 'Invalid Configuration\n\nAzureFile fields (' + list.key) + '.') + (__expression_hAAjrm(17), path)) + ') do not currently support being used as initial fields.\n'));
        }
    }
    {
        __statement_WIELXX(18);
        __extro_o1bl91(19, __intro_HNofZ9(19, azurefile.super_).call(this, list, path, options));
    }
    if (__expression_hAAjrm(20), !this.azurefileconfig) {
        __block_DnoaWA(2);
        {
            __statement_WIELXX(21);
            throw new Error((__expression_hAAjrm(22), (__expression_hAAjrm(23), (__expression_hAAjrm(24), (__expression_hAAjrm(25), (__expression_hAAjrm(26), (__expression_hAAjrm(27), 'Invalid Configuration\n\n' + 'AzureFile fields (') + list.key) + '.') + (__expression_hAAjrm(28), path)) + ') require the "azurefile config" option to be set.\n\n') + 'See http://keystonejs.com/docs/configuration#services-azure for more information.\n'));
        }
    }
    {
        __statement_WIELXX(29);
        process.env.AZURE_STORAGE_ACCOUNT = this.azurefileconfig.account;
    }
    {
        __statement_WIELXX(30);
        process.env.AZURE_STORAGE_ACCESS_KEY = this.azurefileconfig.key;
    }
    {
        __statement_WIELXX(31);
        this.azurefileconfig.container = (__expression_hAAjrm(32), this.azurefileconfig.container || 'keystone');
    }
    {
        __statement_WIELXX(33);
        var self = this;
    }
    {
        __statement_WIELXX(34);
        options.filenameFormatter = (__expression_hAAjrm(35), options.filenameFormatter || function (item, filename) {
            __block_DnoaWA(3);
            return __expression_hAAjrm(36), filename;
        });
    }
    {
        __statement_WIELXX(37);
        options.containerFormatter = (__expression_hAAjrm(38), options.containerFormatter || function (item, filename) {
            __block_DnoaWA(4);
            return __expression_hAAjrm(39), self.azurefileconfig.container;
        });
    }
    if (__expression_hAAjrm(40), options.pre && options.pre.upload) {
        __block_DnoaWA(5);
        {
            __statement_WIELXX(41);
            this._pre.upload = __extro_o1bl91(42, __intro_HNofZ9(42, this._pre.upload).concat(options.pre.upload));
        }
    }
}
{
    __statement_WIELXX(43);
    __extro_o1bl91(44, __intro_HNofZ9(44, util).inherits(azurefile, super_));
}
{
    __statement_WIELXX(45);
    __extro_o1bl91(46, __intro_HNofZ9(46, Object).defineProperty(azurefile.prototype, 'azurefileconfig', {
        get: function () {
            __block_DnoaWA(6);
            return __expression_hAAjrm(47), (__expression_hAAjrm(48), this.options.azurefileconfig || __extro_o1bl91(49, __intro_HNofZ9(49, keystone).get('azurefile config')));
        }
    }));
}
{
    __statement_WIELXX(50);
    azurefile.prototype.pre = function (event, fn) {
        __block_DnoaWA(7);
        if (__expression_hAAjrm(51), !this._pre[event]) {
            __block_DnoaWA(8);
            {
                __statement_WIELXX(52);
                throw new Error((__expression_hAAjrm(53), (__expression_hAAjrm(54), (__expression_hAAjrm(55), (__expression_hAAjrm(56), (__expression_hAAjrm(57), (__expression_hAAjrm(58), (__expression_hAAjrm(59), 'AzureFile (' + this.list.key) + '.') + this.path) + ') error: azurefile.pre()\n\n') + 'Event ') + (__expression_hAAjrm(60), event)) + ' is not supported.\n'));
            }
        }
        {
            __statement_WIELXX(61);
            __extro_o1bl91(62, __intro_HNofZ9(62, this._pre[event]).push(fn));
        }
        return __expression_hAAjrm(63), this;
    };
}
{
    __statement_WIELXX(64);
    azurefile.prototype.addToSchema = function () {
        __block_DnoaWA(9);
        {
            __statement_WIELXX(65);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_WIELXX(66);
            var paths = this.paths = {
                    filename: __extro_o1bl91(67, __intro_HNofZ9(67, this._path).append('.filename')),
                    path: __extro_o1bl91(68, __intro_HNofZ9(68, this._path).append('.path')),
                    size: __extro_o1bl91(69, __intro_HNofZ9(69, this._path).append('.size')),
                    filetype: __extro_o1bl91(70, __intro_HNofZ9(70, this._path).append('.filetype')),
                    url: __extro_o1bl91(71, __intro_HNofZ9(71, this._path).append('.url')),
                    etag: __extro_o1bl91(72, __intro_HNofZ9(72, this._path).append('.etag')),
                    container: __extro_o1bl91(73, __intro_HNofZ9(73, this._path).append('.container')),
                    exists: __extro_o1bl91(74, __intro_HNofZ9(74, this._path).append('.exists')),
                    upload: __extro_o1bl91(75, __intro_HNofZ9(75, this._path).append('_upload')),
                    action: __extro_o1bl91(76, __intro_HNofZ9(76, this._path).append('_action'))
                };
        }
        {
            __statement_WIELXX(77);
            var schemaPaths = __extro_o1bl91(78, __intro_HNofZ9(78, this._path).addTo({}, {
                    filename: String,
                    path: String,
                    size: Number,
                    filetype: String,
                    url: String,
                    etag: String,
                    container: String
                }));
        }
        {
            __statement_WIELXX(79);
            __extro_o1bl91(80, __intro_HNofZ9(80, schema).add(schemaPaths));
        }
        {
            __statement_WIELXX(81);
            var exists = function (item) {
                __block_DnoaWA(10);
                return __expression_hAAjrm(82), __extro_o1bl91(85, __intro_HNofZ9(85, item).get(paths.url)) ? (__expression_hAAjrm(83), true) : (__expression_hAAjrm(84), false);
            };
        }
        {
            __statement_WIELXX(86);
            __extro_o1bl91(87, __intro_HNofZ9(87, __extro_o1bl91(88, __intro_HNofZ9(88, schema).virtual(paths.exists))).get(function () {
                __block_DnoaWA(11);
                return __expression_hAAjrm(89), __extro_o1bl91(90, __intro_HNofZ9(90, schemaMethods.exists).apply(this));
            }));
        }
        {
            __statement_WIELXX(91);
            var reset = function (item) {
                __block_DnoaWA(12);
                {
                    __statement_WIELXX(92);
                    __extro_o1bl91(93, __intro_HNofZ9(93, item).set(field.path, {
                        filename: '',
                        path: '',
                        size: 0,
                        filetype: '',
                        url: ''
                    }));
                }
            };
        }
        {
            __statement_WIELXX(94);
            var schemaMethods = {
                    exists: function () {
                        __block_DnoaWA(13);
                        return __expression_hAAjrm(95), (__expression_hAAjrm(96), exists(this));
                    },
                    reset: function () {
                        __block_DnoaWA(14);
                        {
                            __statement_WIELXX(97);
                            var self = this;
                        }
                        try {
                            __block_DnoaWA(15);
                            {
                                __statement_WIELXX(98);
                                __extro_o1bl91(99, __intro_HNofZ9(99, __extro_o1bl91(100, __intro_HNofZ9(100, azure).createBlobService())).deleteBlob(__extro_o1bl91(101, __intro_HNofZ9(101, this).get(paths.container)), __extro_o1bl91(102, __intro_HNofZ9(102, this).get(paths.filename)), function (error) {
                                    __block_DnoaWA(16);
                                }));
                            }
                        } catch (e) {
                            __block_DnoaWA(17);
                        }
                        {
                            __statement_WIELXX(103);
                            __expression_hAAjrm(104), reset(self);
                        }
                    },
                    delete: function () {
                        __block_DnoaWA(18);
                        {
                            __statement_WIELXX(105);
                            var self = this;
                        }
                        try {
                            __block_DnoaWA(19);
                            {
                                __statement_WIELXX(106);
                                __extro_o1bl91(107, __intro_HNofZ9(107, __extro_o1bl91(108, __intro_HNofZ9(108, azure).createBlobService()).blobService).deleteBlob(__extro_o1bl91(109, __intro_HNofZ9(109, this).get(paths.container)), __extro_o1bl91(110, __intro_HNofZ9(110, this).get(paths.filename)), function (error) {
                                    __block_DnoaWA(20);
                                    if (__expression_hAAjrm(111), !(__expression_hAAjrm(112), error)) {
                                        __block_DnoaWA(21);
                                    }
                                }));
                            }
                        } catch (e) {
                            __block_DnoaWA(22);
                        }
                        {
                            __statement_WIELXX(113);
                            __expression_hAAjrm(114), reset(self);
                        }
                    }
                };
        }
        {
            __statement_WIELXX(115);
            __extro_o1bl91(116, __intro_HNofZ9(116, _).each(schemaMethods, function (fn, key) {
                __block_DnoaWA(23);
                {
                    __statement_WIELXX(117);
                    __extro_o1bl91(118, __intro_HNofZ9(118, field).underscoreMethod(key, fn));
                }
            }));
        }
        {
            __statement_WIELXX(119);
            this.apply = function (item, method) {
                __block_DnoaWA(24);
                return __expression_hAAjrm(120), __extro_o1bl91(121, __intro_HNofZ9(121, schemaMethods[method]).apply(item, __extro_o1bl91(122, __intro_HNofZ9(122, Array.prototype.slice).call(arguments, 2))));
            };
        }
        {
            __statement_WIELXX(123);
            __extro_o1bl91(124, __intro_HNofZ9(124, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_WIELXX(125);
    azurefile.prototype.format = function (item) {
        __block_DnoaWA(25);
        return __expression_hAAjrm(126), __extro_o1bl91(127, __intro_HNofZ9(127, item).get(this.paths.url));
    };
}
{
    __statement_WIELXX(128);
    azurefile.prototype.isModified = function (item) {
        __block_DnoaWA(26);
        return __expression_hAAjrm(129), __extro_o1bl91(130, __intro_HNofZ9(130, item).isModified(this.paths.url));
    };
}
{
    __statement_WIELXX(131);
    azurefile.prototype.validateInput = function (data) {
        __block_DnoaWA(27);
        return __expression_hAAjrm(132), true;
    };
}
{
    __statement_WIELXX(133);
    azurefile.prototype.updateItem = function (item, data) {
        __block_DnoaWA(28);
    };
}
{
    __statement_WIELXX(134);
    azurefile.prototype.uploadFile = function (item, file, update, callback) {
        __block_DnoaWA(29);
        {
            __statement_WIELXX(135);
            var field = this, prefix = field.options.datePrefix ? (__expression_hAAjrm(136), (__expression_hAAjrm(138), __extro_o1bl91(139, __intro_HNofZ9(139, (__expression_hAAjrm(140), moment())).format(field.options.datePrefix)) + '-')) : (__expression_hAAjrm(137), ''), name = (__expression_hAAjrm(141), (__expression_hAAjrm(142), prefix) + file.name);
        }
        if (__expression_hAAjrm(143), field.options.allowedTypes && (__expression_hAAjrm(144), !__extro_o1bl91(145, __intro_HNofZ9(145, _).contains(field.options.allowedTypes, file.type)))) {
            __block_DnoaWA(30);
            return __expression_hAAjrm(146), (__expression_hAAjrm(147), callback(new Error((__expression_hAAjrm(148), 'Unsupported File Type: ' + file.type))));
        }
        if (__expression_hAAjrm(149), 'function' == (__expression_hAAjrm(150), typeof update)) {
            __block_DnoaWA(31);
            {
                __statement_WIELXX(151);
                callback = update;
            }
            {
                __statement_WIELXX(152);
                update = false;
            }
        }
        {
            __statement_WIELXX(153);
            var doUpload = function () {
                __block_DnoaWA(32);
                {
                    __statement_WIELXX(154);
                    var blobService = __extro_o1bl91(155, __intro_HNofZ9(155, azure).createBlobService());
                }
                {
                    __statement_WIELXX(156);
                    var container = __extro_o1bl91(157, __intro_HNofZ9(157, field.options).containerFormatter(item, file.name));
                }
                {
                    __statement_WIELXX(158);
                    __extro_o1bl91(159, __intro_HNofZ9(159, blobService).createContainerIfNotExists(container, {
                        publicAccessLevel: 'blob'
                    }, function (error) {
                        __block_DnoaWA(33);
                        if (__expression_hAAjrm(160), error) {
                            __block_DnoaWA(34);
                            return __expression_hAAjrm(161), (__expression_hAAjrm(162), callback(error));
                        }
                        {
                            __statement_WIELXX(163);
                            __extro_o1bl91(164, __intro_HNofZ9(164, blobService).createBlockBlobFromFile(container, __extro_o1bl91(165, __intro_HNofZ9(165, field.options).filenameFormatter(item, file.name)), file.path, function (error, blob, res) {
                                __block_DnoaWA(35);
                                if (__expression_hAAjrm(166), error) {
                                    __block_DnoaWA(36);
                                    return __expression_hAAjrm(167), (__expression_hAAjrm(168), callback(error));
                                } else {
                                    __block_DnoaWA(37);
                                    {
                                        __statement_WIELXX(169);
                                        var fileData = {
                                                filename: blob.blob,
                                                size: file.size,
                                                filetype: file.type,
                                                etag: blob.etag,
                                                container: container,
                                                url: (__expression_hAAjrm(170), (__expression_hAAjrm(171), (__expression_hAAjrm(172), (__expression_hAAjrm(173), (__expression_hAAjrm(174), 'http://' + field.azurefileconfig.account) + '.blob.core.windows.net/') + (__expression_hAAjrm(175), container)) + '/') + blob.blob)
                                            };
                                    }
                                    if (__expression_hAAjrm(176), update) {
                                        __block_DnoaWA(38);
                                        {
                                            __statement_WIELXX(177);
                                            __extro_o1bl91(178, __intro_HNofZ9(178, item).set(field.path, fileData));
                                        }
                                    }
                                    {
                                        __statement_WIELXX(179);
                                        __expression_hAAjrm(180), callback(null, fileData);
                                    }
                                }
                            }));
                        }
                    }));
                }
            };
        }
        {
            __statement_WIELXX(181);
            __extro_o1bl91(182, __intro_HNofZ9(182, async).eachSeries(this._pre.upload, function (fn, next) {
                __block_DnoaWA(39);
                {
                    __statement_WIELXX(183);
                    __expression_hAAjrm(184), fn(item, file, next);
                }
            }, function (err) {
                __block_DnoaWA(40);
                if (__expression_hAAjrm(185), err) {
                    __block_DnoaWA(41);
                    return __expression_hAAjrm(186), (__expression_hAAjrm(187), callback(err));
                }
                {
                    __statement_WIELXX(188);
                    __expression_hAAjrm(189), doUpload();
                }
            }));
        }
    };
}
{
    __statement_WIELXX(190);
    azurefile.prototype.getRequestHandler = function (item, req, paths, callback) {
        __block_DnoaWA(42);
        {
            __statement_WIELXX(191);
            var field = this;
        }
        if (__extro_o1bl91(192, __intro_HNofZ9(192, utils).isFunction(paths))) {
            __block_DnoaWA(43);
            {
                __statement_WIELXX(193);
                callback = paths;
            }
            {
                __statement_WIELXX(194);
                paths = field.paths;
            }
        } else if (__expression_hAAjrm(195), !(__expression_hAAjrm(196), paths)) {
            __block_DnoaWA(44);
            {
                __statement_WIELXX(197);
                paths = field.paths;
            }
        }
        {
            __statement_WIELXX(198);
            callback = (__expression_hAAjrm(199), (__expression_hAAjrm(200), callback) || function () {
                __block_DnoaWA(45);
            });
        }
        return __expression_hAAjrm(201), function () {
            __block_DnoaWA(46);
            if (req.body) {
                __block_DnoaWA(47);
                {
                    __statement_WIELXX(202);
                    var action = req.body[paths.action];
                }
                if (__extro_o1bl91(203, __intro_HNofZ9(203, /^(delete|reset)$/).test(action))) {
                    __block_DnoaWA(48);
                    {
                        __statement_WIELXX(204);
                        __extro_o1bl91(205, __intro_HNofZ9(205, field).apply(item, action));
                    }
                }
            }
            if (__expression_hAAjrm(206), (__expression_hAAjrm(207), req.files && req.files[paths.upload]) && req.files[paths.upload].size) {
                __block_DnoaWA(49);
                return __expression_hAAjrm(208), __extro_o1bl91(209, __intro_HNofZ9(209, field).uploadFile(item, req.files[paths.upload], true, callback));
            }
            return __expression_hAAjrm(210), (__expression_hAAjrm(211), callback());
        };
    };
}
{
    __statement_WIELXX(212);
    azurefile.prototype.handleRequest = function (item, req, paths, callback) {
        __block_DnoaWA(50);
        {
            __statement_WIELXX(213);
            __expression_hAAjrm(214), __extro_o1bl91(215, __intro_HNofZ9(215, this).getRequestHandler(item, req, paths, callback))();
        }
    };
}
{
    __statement_WIELXX(216);
    exports = module.exports = azurefile;
}
