// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_core_initNav.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_yJt$FY, __expression_BdkKKh, __block_UBtSOK;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_yJt$FY = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/initNav.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_BdkKKh = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/initNav.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_UBtSOK = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/initNav.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_kK4XO6 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_p$EZBI = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/initNav.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_yJt$FY(0);
    var _ = (__expression_BdkKKh(1), require('underscore')), utils = (__expression_BdkKKh(2), require('keystone-utils'));
}
function initNav(sections) {
    __block_UBtSOK(0);
    {
        __statement_yJt$FY(3);
        var keystone = this;
    }
    {
        __statement_yJt$FY(4);
        var nav = {
                sections: [],
                by: {
                    list: {},
                    section: {}
                }
            };
    }
    if (__expression_BdkKKh(5), !(__expression_BdkKKh(6), sections)) {
        __block_UBtSOK(1);
        {
            __statement_yJt$FY(7);
            sections = {};
        }
        {
            __statement_yJt$FY(8);
            nav.flat = true;
        }
        {
            __statement_yJt$FY(9);
            __extro_p$EZBI(10, __intro_kK4XO6(10, _).each(this.lists, function (list) {
                __block_UBtSOK(2);
                if (__extro_p$EZBI(11, __intro_kK4XO6(11, list).get('hidden'))) {
                    __block_UBtSOK(3);
                    return __expression_BdkKKh(12);
                }
                {
                    __statement_yJt$FY(13);
                    sections[list.path] = [
                        list.path
                    ];
                }
            }));
        }
    }
    {
        __statement_yJt$FY(14);
        __extro_p$EZBI(15, __intro_kK4XO6(15, _).each(sections, function (section, key) {
            __block_UBtSOK(4);
            if (__expression_BdkKKh(16), 'string' === (__expression_BdkKKh(17), typeof section)) {
                __block_UBtSOK(5);
                {
                    __statement_yJt$FY(18);
                    section = [
                        section
                    ];
                }
            }
            {
                __statement_yJt$FY(19);
                section = {
                    lists: section,
                    label: nav.flat ? (__expression_BdkKKh(20), __extro_p$EZBI(22, __intro_kK4XO6(22, keystone).list(section[0])).label) : (__expression_BdkKKh(21), __extro_p$EZBI(23, __intro_kK4XO6(23, utils).keyToLabel(key)))
                };
            }
            {
                __statement_yJt$FY(24);
                section.key = key;
            }
            {
                __statement_yJt$FY(25);
                section.lists = __extro_p$EZBI(26, __intro_kK4XO6(26, _).map(section.lists, function (i) {
                    __block_UBtSOK(6);
                    {
                        __statement_yJt$FY(27);
                        var msg, list = __extro_p$EZBI(28, __intro_kK4XO6(28, keystone).list(i));
                    }
                    if (__expression_BdkKKh(29), !(__expression_BdkKKh(30), list)) {
                        __block_UBtSOK(7);
                        {
                            __statement_yJt$FY(31);
                            msg = (__expression_BdkKKh(32), (__expression_BdkKKh(33), 'Invalid Keystone Option (nav): list ' + (__expression_BdkKKh(34), i)) + ' has not been defined.\n');
                        }
                        {
                            __statement_yJt$FY(35);
                            throw new Error(msg);
                        }
                    }
                    if (__extro_p$EZBI(36, __intro_kK4XO6(36, list).get('hidden'))) {
                        __block_UBtSOK(8);
                        {
                            __statement_yJt$FY(37);
                            msg = (__expression_BdkKKh(38), (__expression_BdkKKh(39), 'Invalid Keystone Option (nav): list ' + (__expression_BdkKKh(40), i)) + ' is hidden.\n');
                        }
                        {
                            __statement_yJt$FY(41);
                            throw new Error(msg);
                        }
                    }
                    {
                        __statement_yJt$FY(42);
                        nav.by.list[list.key] = section;
                    }
                    return __expression_BdkKKh(43), list;
                }));
            }
            if (section.lists.length) {
                __block_UBtSOK(9);
                {
                    __statement_yJt$FY(44);
                    __extro_p$EZBI(45, __intro_kK4XO6(45, nav.sections).push(section));
                }
                {
                    __statement_yJt$FY(46);
                    nav.by.section[section.key] = section;
                }
            }
        }));
    }
    return __expression_BdkKKh(47), nav;
}
{
    __statement_yJt$FY(48);
    module.exports = initNav;
}
