// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_core_createItems.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_CO1mIb, __expression_bjPa6U, __block_k1DDak;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_CO1mIb = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/createItems.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_bjPa6U = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/createItems.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_k1DDak = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/createItems.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_xu8BEL = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_JhVVo9 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/createItems.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_CO1mIb(0);
    var _ = (__expression_bjPa6U(1), require('underscore')), async = (__expression_bjPa6U(2), require('async')), utils = (__expression_bjPa6U(3), require('keystone-utils'));
}
function createItems(data, ops, callback) {
    __block_k1DDak(0);
    {
        __statement_CO1mIb(4);
        var keystone = this;
    }
    {
        __statement_CO1mIb(5);
        var options = {
                verbose: false,
                strict: true
            };
    }
    {
        __statement_CO1mIb(6);
        var dashes = '------------------------------------------------';
    }
    if (__expression_bjPa6U(7), !__extro_JhVVo9(8, __intro_xu8BEL(8, _).isObject(data))) {
        __block_k1DDak(1);
        {
            __statement_CO1mIb(9);
            throw new Error('keystone.createItems() requires a data object as the first argument.');
        }
    }
    if (__extro_JhVVo9(10, __intro_xu8BEL(10, _).isObject(ops))) {
        __block_k1DDak(2);
        {
            __statement_CO1mIb(11);
            __extro_JhVVo9(12, __intro_xu8BEL(12, _).extend(options, ops));
        }
    }
    if (__extro_JhVVo9(13, __intro_xu8BEL(13, _).isFunction(ops))) {
        __block_k1DDak(3);
        {
            __statement_CO1mIb(14);
            callback = ops;
        }
    }
    {
        __statement_CO1mIb(15);
        var lists = __extro_JhVVo9(16, __intro_xu8BEL(16, _).keys(data)), refs = {}, stats = {};
    }
    function writeLog(data) {
        __block_k1DDak(4);
        {
            __statement_CO1mIb(17);
            __extro_JhVVo9(18, __intro_xu8BEL(18, console).log((__expression_bjPa6U(19), (__expression_bjPa6U(20), __extro_JhVVo9(21, __intro_xu8BEL(21, keystone).get('name')) + ': ') + (__expression_bjPa6U(22), data))));
        }
    }
    {
        __statement_CO1mIb(23);
        __extro_JhVVo9(24, __intro_xu8BEL(24, async).waterfall([
            function (next) {
                __block_k1DDak(5);
                {
                    __statement_CO1mIb(25);
                    __extro_JhVVo9(26, __intro_xu8BEL(26, async).eachSeries(lists, function (key, doneList) {
                        __block_k1DDak(6);
                        {
                            __statement_CO1mIb(27);
                            var list = __extro_JhVVo9(28, __intro_xu8BEL(28, keystone).list(key)), relationshipPaths = __extro_JhVVo9(29, __intro_xu8BEL(29, __extro_JhVVo9(30, __intro_xu8BEL(30, _).where(list.fields, {
                                    type: 'relationship'
                                }))).map(function (i) {
                                    __block_k1DDak(7);
                                    return __expression_bjPa6U(31), i.path;
                                }));
                        }
                        if (__expression_bjPa6U(32), !(__expression_bjPa6U(33), list)) {
                            __block_k1DDak(8);
                            if (options.strict) {
                                __block_k1DDak(9);
                                return __expression_bjPa6U(34), (__expression_bjPa6U(35), doneList({
                                    type: 'invalid list',
                                    message: (__expression_bjPa6U(36), (__expression_bjPa6U(37), 'List key ' + (__expression_bjPa6U(38), key)) + ' is invalid.')
                                }));
                            }
                            if (options.verbose) {
                                __block_k1DDak(10);
                                {
                                    __statement_CO1mIb(39);
                                    __expression_bjPa6U(40), writeLog((__expression_bjPa6U(41), 'Skipping invalid list: ' + (__expression_bjPa6U(42), key)));
                                }
                            }
                            return __expression_bjPa6U(43), (__expression_bjPa6U(44), doneList());
                        }
                        {
                            __statement_CO1mIb(45);
                            refs[list.key] = {};
                        }
                        {
                            __statement_CO1mIb(46);
                            stats[list.key] = {
                                singular: list.singular,
                                plural: list.plural,
                                created: 0,
                                warnings: 0
                            };
                        }
                        {
                            __statement_CO1mIb(47);
                            var itemsProcessed = 0, totalItems = data[key].length;
                        }
                        if (options.verbose) {
                            __block_k1DDak(11);
                            {
                                __statement_CO1mIb(48);
                                __expression_bjPa6U(49), writeLog(dashes);
                            }
                            {
                                __statement_CO1mIb(50);
                                __expression_bjPa6U(51), writeLog((__expression_bjPa6U(52), 'Processing list: ' + (__expression_bjPa6U(53), key)));
                            }
                            {
                                __statement_CO1mIb(54);
                                __expression_bjPa6U(55), writeLog((__expression_bjPa6U(56), 'Items to create: ' + (__expression_bjPa6U(57), totalItems)));
                            }
                            {
                                __statement_CO1mIb(58);
                                __expression_bjPa6U(59), writeLog(dashes);
                            }
                        }
                        {
                            __statement_CO1mIb(60);
                            __extro_JhVVo9(61, __intro_xu8BEL(61, async).eachSeries(data[key], function (data, doneItem) {
                                __block_k1DDak(12);
                                {
                                    __statement_CO1mIb(62);
                                    __expression_bjPa6U(63), itemsProcessed++;
                                }
                                {
                                    __statement_CO1mIb(64);
                                    __extro_JhVVo9(65, __intro_xu8BEL(65, __extro_JhVVo9(66, __intro_xu8BEL(66, _).keys(data))).forEach(function (i) {
                                        __block_k1DDak(13);
                                        if (__expression_bjPa6U(67), __extro_JhVVo9(68, __intro_xu8BEL(68, _).isFunction(data[i])) && (__expression_bjPa6U(69), __extro_JhVVo9(70, __intro_xu8BEL(70, relationshipPaths).indexOf(i)) === (__expression_bjPa6U(71), -1))) {
                                            __block_k1DDak(14);
                                            {
                                                __statement_CO1mIb(72);
                                                data[i] = __extro_JhVVo9(73, __intro_xu8BEL(73, data)[i]());
                                            }
                                            if (options.verbose) {
                                                __block_k1DDak(15);
                                                {
                                                    __statement_CO1mIb(74);
                                                    __expression_bjPa6U(75), writeLog((__expression_bjPa6U(76), (__expression_bjPa6U(77), (__expression_bjPa6U(78), 'Generated dynamic value for [' + (__expression_bjPa6U(79), i)) + ']: ') + data[i]));
                                                }
                                            }
                                        }
                                    }));
                                }
                                {
                                    __statement_CO1mIb(80);
                                    var doc = data.__doc = new list.model();
                                }
                                if (data.__ref) {
                                    __block_k1DDak(16);
                                    {
                                        __statement_CO1mIb(81);
                                        refs[list.key][data.__ref] = doc;
                                    }
                                }
                                {
                                    __statement_CO1mIb(82);
                                    __extro_JhVVo9(83, __intro_xu8BEL(83, _).each(list.fields, function (field) {
                                        __block_k1DDak(17);
                                        if (__expression_bjPa6U(84), field.type !== 'relationship') {
                                            __block_k1DDak(18);
                                            {
                                                __statement_CO1mIb(85);
                                                __extro_JhVVo9(86, __intro_xu8BEL(86, field).updateItem(doc, data));
                                            }
                                        }
                                    }));
                                }
                                if (options.verbose) {
                                    __block_k1DDak(19);
                                    {
                                        __statement_CO1mIb(87);
                                        var documentName = __extro_JhVVo9(88, __intro_xu8BEL(88, list).getDocumentName(doc));
                                    }
                                    {
                                        __statement_CO1mIb(89);
                                        __expression_bjPa6U(90), writeLog((__expression_bjPa6U(91), (__expression_bjPa6U(92), (__expression_bjPa6U(93), (__expression_bjPa6U(94), (__expression_bjPa6U(95), 'Creating item [' + (__expression_bjPa6U(96), itemsProcessed)) + ' of ') + (__expression_bjPa6U(97), totalItems)) + '] - ') + (__expression_bjPa6U(98), documentName)));
                                    }
                                }
                                {
                                    __statement_CO1mIb(99);
                                    __extro_JhVVo9(100, __intro_xu8BEL(100, doc).save(doneItem));
                                }
                                {
                                    __statement_CO1mIb(101);
                                    __expression_bjPa6U(102), stats[list.key].created++;
                                }
                            }, doneList));
                        }
                    }, next));
                }
            },
            function (next) {
                __block_k1DDak(20);
                {
                    __statement_CO1mIb(103);
                    __extro_JhVVo9(104, __intro_xu8BEL(104, async).each(lists, function (key, doneList) {
                        __block_k1DDak(21);
                        {
                            __statement_CO1mIb(105);
                            var list = __extro_JhVVo9(106, __intro_xu8BEL(106, keystone).list(key)), relationships = __extro_JhVVo9(107, __intro_xu8BEL(107, _).where(list.fields, {
                                    type: 'relationship'
                                }));
                        }
                        if (__expression_bjPa6U(108), (__expression_bjPa6U(109), !(__expression_bjPa6U(110), list)) || (__expression_bjPa6U(111), !relationships.length)) {
                            __block_k1DDak(22);
                            return __expression_bjPa6U(112), (__expression_bjPa6U(113), doneList());
                        }
                        {
                            __statement_CO1mIb(114);
                            var itemsProcessed = 0, totalItems = data[key].length;
                        }
                        if (options.verbose) {
                            __block_k1DDak(23);
                            {
                                __statement_CO1mIb(115);
                                __expression_bjPa6U(116), writeLog(dashes);
                            }
                            {
                                __statement_CO1mIb(117);
                                __expression_bjPa6U(118), writeLog((__expression_bjPa6U(119), 'Processing relationships for: ' + (__expression_bjPa6U(120), key)));
                            }
                            {
                                __statement_CO1mIb(121);
                                __expression_bjPa6U(122), writeLog((__expression_bjPa6U(123), 'Items to process: ' + (__expression_bjPa6U(124), totalItems)));
                            }
                            {
                                __statement_CO1mIb(125);
                                __expression_bjPa6U(126), writeLog(dashes);
                            }
                        }
                        {
                            __statement_CO1mIb(127);
                            __extro_JhVVo9(128, __intro_xu8BEL(128, async).each(data[key], function (srcData, doneItem) {
                                __block_k1DDak(24);
                                {
                                    __statement_CO1mIb(129);
                                    var doc = srcData.__doc, relationshipsUpdated = 0;
                                }
                                {
                                    __statement_CO1mIb(130);
                                    __expression_bjPa6U(131), itemsProcessed++;
                                }
                                if (options.verbose) {
                                    __block_k1DDak(25);
                                    {
                                        __statement_CO1mIb(132);
                                        var documentName = __extro_JhVVo9(133, __intro_xu8BEL(133, list).getDocumentName(doc));
                                    }
                                    {
                                        __statement_CO1mIb(134);
                                        __expression_bjPa6U(135), writeLog((__expression_bjPa6U(136), (__expression_bjPa6U(137), (__expression_bjPa6U(138), (__expression_bjPa6U(139), (__expression_bjPa6U(140), 'Processing item [' + (__expression_bjPa6U(141), itemsProcessed)) + ' of ') + (__expression_bjPa6U(142), totalItems)) + '] - ') + (__expression_bjPa6U(143), documentName)));
                                    }
                                }
                                {
                                    __statement_CO1mIb(144);
                                    __extro_JhVVo9(145, __intro_xu8BEL(145, async).each(relationships, function (field, doneField) {
                                        __block_k1DDak(26);
                                        {
                                            __statement_CO1mIb(146);
                                            var fieldValue = null, refsLookup = null;
                                        }
                                        if (__expression_bjPa6U(147), !field.path) {
                                            __block_k1DDak(27);
                                            {
                                                __statement_CO1mIb(148);
                                                __expression_bjPa6U(149), writeLog((__expression_bjPa6U(150), (__expression_bjPa6U(151), 'WARNING:  Invalid relationship (undefined list path) [List: ' + (__expression_bjPa6U(152), key)) + ']'));
                                            }
                                            {
                                                __statement_CO1mIb(153);
                                                __expression_bjPa6U(154), stats[list.key].warnings++;
                                            }
                                            return __expression_bjPa6U(155), (__expression_bjPa6U(156), doneField());
                                        } else {
                                            __block_k1DDak(28);
                                            {
                                                __statement_CO1mIb(157);
                                                fieldValue = srcData[field.path];
                                            }
                                        }
                                        if (__expression_bjPa6U(158), !field.refList) {
                                            __block_k1DDak(29);
                                            if (__expression_bjPa6U(159), fieldValue) {
                                                __block_k1DDak(30);
                                                {
                                                    __statement_CO1mIb(160);
                                                    __expression_bjPa6U(161), writeLog((__expression_bjPa6U(162), (__expression_bjPa6U(163), (__expression_bjPa6U(164), (__expression_bjPa6U(165), 'WARNING:  Invalid relationship (undefined reference list) [list: ' + (__expression_bjPa6U(166), key)) + '] [path: ') + (__expression_bjPa6U(167), fieldValue)) + ']'));
                                                }
                                                {
                                                    __statement_CO1mIb(168);
                                                    __expression_bjPa6U(169), stats[list.key].warnings++;
                                                }
                                            }
                                            return __expression_bjPa6U(170), (__expression_bjPa6U(171), doneField());
                                        }
                                        if (__expression_bjPa6U(172), !field.refList.key) {
                                            __block_k1DDak(31);
                                            {
                                                __statement_CO1mIb(173);
                                                __expression_bjPa6U(174), writeLog((__expression_bjPa6U(175), (__expression_bjPa6U(176), (__expression_bjPa6U(177), (__expression_bjPa6U(178), (__expression_bjPa6U(179), (__expression_bjPa6U(180), 'WARNING:  Invalid relationship (undefined ref list key) [list: ' + (__expression_bjPa6U(181), key)) + '] [field.refList: ') + field.refList) + '] [fieldValue: ') + (__expression_bjPa6U(182), fieldValue)) + ']'));
                                            }
                                            {
                                                __statement_CO1mIb(183);
                                                __expression_bjPa6U(184), stats[list.key].warnings++;
                                            }
                                            return __expression_bjPa6U(185), (__expression_bjPa6U(186), doneField());
                                        } else {
                                            __block_k1DDak(32);
                                            {
                                                __statement_CO1mIb(187);
                                                refsLookup = refs[field.refList.key];
                                            }
                                        }
                                        if (__expression_bjPa6U(188), !(__expression_bjPa6U(189), fieldValue)) {
                                            __block_k1DDak(33);
                                            return __expression_bjPa6U(190), (__expression_bjPa6U(191), doneField());
                                        }
                                        if (__extro_JhVVo9(192, __intro_xu8BEL(192, _).isFunction(fieldValue))) {
                                            __block_k1DDak(34);
                                            {
                                                __statement_CO1mIb(193);
                                                __expression_bjPa6U(194), relationshipsUpdated++;
                                            }
                                            {
                                                __statement_CO1mIb(195);
                                                var fn = fieldValue, argsRegExp = /^function\s*[^\(]*\(\s*([^\)]*)\)/m, lists = __extro_JhVVo9(196, __intro_xu8BEL(196, __extro_JhVVo9(197, __intro_xu8BEL(197, __extro_JhVVo9(198, __intro_xu8BEL(198, __extro_JhVVo9(199, __intro_xu8BEL(199, fn).toString())).match(argsRegExp))[1]).split(','))).map(function (i) {
                                                        __block_k1DDak(35);
                                                        return __expression_bjPa6U(200), __extro_JhVVo9(201, __intro_xu8BEL(201, i).trim());
                                                    })), args = __extro_JhVVo9(202, __intro_xu8BEL(202, lists).map(function (i) {
                                                        __block_k1DDak(36);
                                                        return __expression_bjPa6U(203), __extro_JhVVo9(204, __intro_xu8BEL(204, keystone).list(i));
                                                    })), query = __extro_JhVVo9(205, __intro_xu8BEL(205, fn).apply(keystone, args));
                                            }
                                            {
                                                __statement_CO1mIb(206);
                                                __extro_JhVVo9(207, __intro_xu8BEL(207, query).exec(function (err, results) {
                                                    __block_k1DDak(37);
                                                    if (field.many) {
                                                        __block_k1DDak(38);
                                                        {
                                                            __statement_CO1mIb(208);
                                                            __extro_JhVVo9(209, __intro_xu8BEL(209, doc).set(field.path, (__expression_bjPa6U(210), (__expression_bjPa6U(211), results) || [])));
                                                        }
                                                    } else {
                                                        __block_k1DDak(39);
                                                        {
                                                            __statement_CO1mIb(212);
                                                            __extro_JhVVo9(213, __intro_xu8BEL(213, doc).set(field.path, (__expression_bjPa6U(216), (__expression_bjPa6U(217), results) && results.length) ? (__expression_bjPa6U(214), results[0]) : (__expression_bjPa6U(215), undefined)));
                                                        }
                                                    }
                                                    {
                                                        __statement_CO1mIb(218);
                                                        __expression_bjPa6U(219), doneField(err);
                                                    }
                                                }));
                                            }
                                        } else if (__extro_JhVVo9(220, __intro_xu8BEL(220, _).isArray(fieldValue))) {
                                            __block_k1DDak(40);
                                            if (field.many) {
                                                __block_k1DDak(41);
                                                {
                                                    __statement_CO1mIb(221);
                                                    var refsArr = __extro_JhVVo9(222, __intro_xu8BEL(222, _).compact(__extro_JhVVo9(223, __intro_xu8BEL(223, fieldValue).map(function (ref) {
                                                            __block_k1DDak(42);
                                                            return __expression_bjPa6U(224), refsLookup[ref] ? (__expression_bjPa6U(225), refsLookup[ref].id) : (__expression_bjPa6U(226), undefined);
                                                        }))));
                                                }
                                                if (__expression_bjPa6U(227), options.strict && (__expression_bjPa6U(228), refsArr.length !== fieldValue.length)) {
                                                    __block_k1DDak(43);
                                                    return __expression_bjPa6U(229), (__expression_bjPa6U(230), doneField({
                                                        type: 'invalid ref',
                                                        srcData: srcData,
                                                        message: (__expression_bjPa6U(231), (__expression_bjPa6U(232), (__expression_bjPa6U(233), (__expression_bjPa6U(234), 'Relationship ' + list.key) + '.') + field.path) + ' contains an invalid reference.')
                                                    }));
                                                }
                                                {
                                                    __statement_CO1mIb(235);
                                                    __expression_bjPa6U(236), relationshipsUpdated++;
                                                }
                                                {
                                                    __statement_CO1mIb(237);
                                                    __extro_JhVVo9(238, __intro_xu8BEL(238, doc).set(field.path, refsArr));
                                                }
                                                {
                                                    __statement_CO1mIb(239);
                                                    __expression_bjPa6U(240), doneField();
                                                }
                                            } else {
                                                __block_k1DDak(44);
                                                return __expression_bjPa6U(241), (__expression_bjPa6U(242), doneField({
                                                    type: 'invalid data',
                                                    srcData: srcData,
                                                    message: (__expression_bjPa6U(243), (__expression_bjPa6U(244), (__expression_bjPa6U(245), (__expression_bjPa6U(246), 'Single-value relationship ' + list.key) + '.') + field.path) + ' provided as an array.')
                                                }));
                                            }
                                        } else if (__extro_JhVVo9(247, __intro_xu8BEL(247, _).isString(fieldValue))) {
                                            __block_k1DDak(45);
                                            {
                                                __statement_CO1mIb(248);
                                                var refItem = refsLookup[fieldValue];
                                            }
                                            if (__expression_bjPa6U(249), !(__expression_bjPa6U(250), refItem)) {
                                                __block_k1DDak(46);
                                                return __expression_bjPa6U(251), options.strict ? (__expression_bjPa6U(252), (__expression_bjPa6U(254), doneField({
                                                    type: 'invalid ref',
                                                    srcData: srcData,
                                                    message: (__expression_bjPa6U(255), (__expression_bjPa6U(256), (__expression_bjPa6U(257), (__expression_bjPa6U(258), (__expression_bjPa6U(259), (__expression_bjPa6U(260), 'Relationship ' + list.key) + '.') + field.path) + ' contains an invalid reference: "') + (__expression_bjPa6U(261), fieldValue)) + '".')
                                                }))) : (__expression_bjPa6U(253), (__expression_bjPa6U(262), doneField()));
                                            }
                                            {
                                                __statement_CO1mIb(263);
                                                __expression_bjPa6U(264), relationshipsUpdated++;
                                            }
                                            {
                                                __statement_CO1mIb(265);
                                                __extro_JhVVo9(266, __intro_xu8BEL(266, doc).set(field.path, field.many ? (__expression_bjPa6U(267), [
                                                    refItem.id
                                                ]) : (__expression_bjPa6U(268), refItem.id)));
                                            }
                                            {
                                                __statement_CO1mIb(269);
                                                __expression_bjPa6U(270), doneField();
                                            }
                                        } else {
                                            __block_k1DDak(47);
                                            return __expression_bjPa6U(271), (__expression_bjPa6U(272), doneField({
                                                type: 'invalid data',
                                                srcData: srcData,
                                                message: (__expression_bjPa6U(273), (__expression_bjPa6U(274), (__expression_bjPa6U(275), (__expression_bjPa6U(276), 'Relationship ' + list.key) + '.') + field.path) + ' contains an invalid data type.')
                                            }));
                                        }
                                    }, function (err) {
                                        __block_k1DDak(48);
                                        if (__expression_bjPa6U(277), err) {
                                            __block_k1DDak(49);
                                            return __expression_bjPa6U(278), (__expression_bjPa6U(279), doneItem(err));
                                        }
                                        if (options.verbose) {
                                            __block_k1DDak(50);
                                            {
                                                __statement_CO1mIb(280);
                                                __expression_bjPa6U(281), writeLog((__expression_bjPa6U(282), (__expression_bjPa6U(283), 'Populated ' + __extro_JhVVo9(284, __intro_xu8BEL(284, utils).plural(relationshipsUpdated, '* relationship', '* relationships'))) + '.'));
                                            }
                                        }
                                        if (__expression_bjPa6U(285), relationshipsUpdated) {
                                            __block_k1DDak(51);
                                            {
                                                __statement_CO1mIb(286);
                                                __extro_JhVVo9(287, __intro_xu8BEL(287, doc).save(doneItem));
                                            }
                                        } else {
                                            __block_k1DDak(52);
                                            {
                                                __statement_CO1mIb(288);
                                                __expression_bjPa6U(289), doneItem();
                                            }
                                        }
                                    }));
                                }
                            }, doneList));
                        }
                    }, next));
                }
            }
        ], function (err) {
            __block_k1DDak(53);
            if (__expression_bjPa6U(290), err) {
                __block_k1DDak(54);
                return __expression_bjPa6U(291), (__expression_bjPa6U(292), (__expression_bjPa6U(293), callback) && (__expression_bjPa6U(294), callback(err)));
            }
            {
                __statement_CO1mIb(295);
                var msg = '\nSuccessfully created:\n';
            }
            {
                __statement_CO1mIb(296);
                __extro_JhVVo9(297, __intro_xu8BEL(297, _).each(stats, function (list) {
                    __block_k1DDak(55);
                    {
                        __statement_CO1mIb(298);
                        msg += (__expression_bjPa6U(299), '\n*   ' + __extro_JhVVo9(300, __intro_xu8BEL(300, utils).plural(list.created, (__expression_bjPa6U(301), '* ' + list.singular), (__expression_bjPa6U(302), '* ' + list.plural))));
                    }
                    if (list.warnings) {
                        __block_k1DDak(56);
                        {
                            __statement_CO1mIb(303);
                            msg += (__expression_bjPa6U(304), '\n    ' + __extro_JhVVo9(305, __intro_xu8BEL(305, utils).plural(list.warnings, '* warning', '* warnings')));
                        }
                    }
                }));
            }
            {
                __statement_CO1mIb(306);
                stats.message = (__expression_bjPa6U(307), (__expression_bjPa6U(308), msg) + '\n');
            }
            {
                __statement_CO1mIb(309);
                __expression_bjPa6U(310), callback(null, stats);
            }
        }));
    }
}
{
    __statement_CO1mIb(311);
    module.exports = createItems;
}
