// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_date.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_CmkmNK, __expression_rA0Lux, __block_sywKd6;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_CmkmNK = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/date.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_rA0Lux = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/date.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_sywKd6 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/date.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_tj_FI4 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_K1sLe3 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/date.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_CmkmNK(0);
    var util = (__expression_rA0Lux(1), require('util')), moment = (__expression_rA0Lux(2), require('moment')), super_ = (__expression_rA0Lux(3), require('../field'));
}
function date(list, path, options) {
    __block_sywKd6(0);
    {
        __statement_CmkmNK(4);
        this._nativeType = Date;
    }
    {
        __statement_CmkmNK(5);
        this._underscoreMethods = [
            'format',
            'moment',
            'parse'
        ];
    }
    {
        __statement_CmkmNK(6);
        this._formatString = (__expression_rA0Lux(9), options.format === false) ? (__expression_rA0Lux(7), false) : (__expression_rA0Lux(8), (__expression_rA0Lux(10), options.format || 'Do MMM YYYY'));
    }
    if (__expression_rA0Lux(11), this._formatString && (__expression_rA0Lux(12), 'string' !== (__expression_rA0Lux(13), typeof this._formatString))) {
        __block_sywKd6(1);
        {
            __statement_CmkmNK(14);
            throw new Error('FieldType.Date: options.format must be a string.');
        }
    }
    {
        __statement_CmkmNK(15);
        __extro_K1sLe3(16, __intro_tj_FI4(16, date.super_).call(this, list, path, options));
    }
}
{
    __statement_CmkmNK(17);
    __extro_K1sLe3(18, __intro_tj_FI4(18, util).inherits(date, super_));
}
{
    __statement_CmkmNK(19);
    date.prototype.format = function (item, format) {
        __block_sywKd6(2);
        if (__expression_rA0Lux(20), (__expression_rA0Lux(21), format) || this._formatString) {
            __block_sywKd6(3);
            return __expression_rA0Lux(22), __extro_K1sLe3(25, __intro_tj_FI4(25, item).get(this.path)) ? (__expression_rA0Lux(23), __extro_K1sLe3(26, __intro_tj_FI4(26, (__expression_rA0Lux(27), moment(__extro_K1sLe3(28, __intro_tj_FI4(28, item).get(this.path))))).format((__expression_rA0Lux(29), (__expression_rA0Lux(30), format) || this._formatString)))) : (__expression_rA0Lux(24), '');
        } else {
            __block_sywKd6(4);
            return __expression_rA0Lux(31), (__expression_rA0Lux(32), __extro_K1sLe3(33, __intro_tj_FI4(33, item).get(this.path)) || '');
        }
    };
}
{
    __statement_CmkmNK(34);
    date.prototype.moment = function (item) {
        __block_sywKd6(5);
        return __expression_rA0Lux(35), (__expression_rA0Lux(36), moment(__extro_K1sLe3(37, __intro_tj_FI4(37, item).get(this.path))));
    };
}
{
    __statement_CmkmNK(38);
    date.prototype.parse = function (item) {
        __block_sywKd6(6);
        {
            __statement_CmkmNK(39);
            var newValue = __extro_K1sLe3(40, __intro_tj_FI4(40, moment).apply(moment, __extro_K1sLe3(41, __intro_tj_FI4(41, Array.prototype.slice).call(arguments, 1))));
        }
        {
            __statement_CmkmNK(42);
            __extro_K1sLe3(43, __intro_tj_FI4(43, item).set(this.path, (__expression_rA0Lux(46), (__expression_rA0Lux(47), newValue) && __extro_K1sLe3(48, __intro_tj_FI4(48, newValue).isValid())) ? (__expression_rA0Lux(44), __extro_K1sLe3(49, __intro_tj_FI4(49, newValue).toDate())) : (__expression_rA0Lux(45), null)));
        }
        return __expression_rA0Lux(50), newValue;
    };
}
{
    __statement_CmkmNK(51);
    date.prototype.validateInput = function (data, required, item) {
        __block_sywKd6(7);
        if (__expression_rA0Lux(52), (__expression_rA0Lux(53), (__expression_rA0Lux(54), !(__expression_rA0Lux(55), (this.path in (__expression_rA0Lux(56), data)))) && (__expression_rA0Lux(57), item)) && __extro_K1sLe3(58, __intro_tj_FI4(58, item).get(this.path))) {
            __block_sywKd6(8);
            return __expression_rA0Lux(59), true;
        }
        {
            __statement_CmkmNK(60);
            var newValue = (__expression_rA0Lux(61), moment(data[this.path]));
        }
        if (__expression_rA0Lux(62), (__expression_rA0Lux(63), required) && (__expression_rA0Lux(64), (__expression_rA0Lux(65), !(__expression_rA0Lux(66), newValue)) || (__expression_rA0Lux(67), !__extro_K1sLe3(68, __intro_tj_FI4(68, newValue).isValid())))) {
            __block_sywKd6(9);
            return __expression_rA0Lux(69), false;
        } else if (__expression_rA0Lux(70), (__expression_rA0Lux(71), data[this.path] && (__expression_rA0Lux(72), newValue)) && (__expression_rA0Lux(73), !__extro_K1sLe3(74, __intro_tj_FI4(74, newValue).isValid()))) {
            __block_sywKd6(10);
            return __expression_rA0Lux(75), false;
        } else {
            __block_sywKd6(11);
            return __expression_rA0Lux(76), true;
        }
    };
}
{
    __statement_CmkmNK(77);
    date.prototype.updateItem = function (item, data) {
        __block_sywKd6(12);
        if (__expression_rA0Lux(78), !(__expression_rA0Lux(79), (this.path in (__expression_rA0Lux(80), data)))) {
            __block_sywKd6(13);
            return __expression_rA0Lux(81);
        }
        {
            __statement_CmkmNK(82);
            var newValue = (__expression_rA0Lux(83), moment(data[this.path]));
        }
        if (__expression_rA0Lux(84), (__expression_rA0Lux(85), newValue) && __extro_K1sLe3(86, __intro_tj_FI4(86, newValue).isValid())) {
            __block_sywKd6(14);
            if (__expression_rA0Lux(87), (__expression_rA0Lux(88), !__extro_K1sLe3(89, __intro_tj_FI4(89, item).get(this.path))) || (__expression_rA0Lux(90), !__extro_K1sLe3(91, __intro_tj_FI4(91, newValue).isSame(__extro_K1sLe3(92, __intro_tj_FI4(92, item).get(this.path)))))) {
                __block_sywKd6(15);
                {
                    __statement_CmkmNK(93);
                    __extro_K1sLe3(94, __intro_tj_FI4(94, item).set(this.path, __extro_K1sLe3(95, __intro_tj_FI4(95, newValue).toDate())));
                }
            }
        } else if (__extro_K1sLe3(96, __intro_tj_FI4(96, item).get(this.path))) {
            __block_sywKd6(16);
            {
                __statement_CmkmNK(97);
                __extro_K1sLe3(98, __intro_tj_FI4(98, item).set(this.path, null));
            }
        }
    };
}
{
    __statement_CmkmNK(99);
    exports = module.exports = date;
}
