// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_email.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_dzU29k, __expression_BVxb5S, __block_bHVGc$;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_dzU29k = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/email.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_BVxb5S = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/email.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_bHVGc$ = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/email.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_vVnvMg = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_SeD6hf = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/email.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_dzU29k(0);
    var _ = (__expression_BVxb5S(1), require('underscore')), keystone = (__expression_BVxb5S(2), require('../')), fs = (__expression_BVxb5S(3), require('fs')), util = (__expression_BVxb5S(4), require('util')), path = (__expression_BVxb5S(5), require('path')), moment = (__expression_BVxb5S(6), require('moment')), utils = (__expression_BVxb5S(7), require('keystone-utils')), mandrillapi = (__expression_BVxb5S(8), require('mandrill-api'));
}
{
    __statement_dzU29k(9);
    var templateCache = {};
}
{
    __statement_dzU29k(10);
    var defaultConfig = {
            templateExt: 'jade',
            templateEngine: (__expression_BVxb5S(11), require('jade')),
            templateBasePath: __extro_SeD6hf(12, __intro_vVnvMg(12, path).normalize(__extro_SeD6hf(13, __intro_vVnvMg(13, path).join(__dirname, '..', 'templates', 'helpers', 'emails')))),
            mandrill: {
                track_opens: true,
                track_clicks: true,
                preserve_recipients: false,
                inline_css: true
            }
        };
}
{
    __statement_dzU29k(14);
    var ErrorNoEmailTemplateName = function () {
        __block_bHVGc$(0);
        {
            __statement_dzU29k(15);
            __extro_SeD6hf(16, __intro_vVnvMg(16, Error).apply(this, arguments));
        }
        {
            __statement_dzU29k(17);
            __extro_SeD6hf(18, __intro_vVnvMg(18, Error).captureStackTrace(this, arguments.callee));
        }
        {
            __statement_dzU29k(19);
            this.message = 'No email templateName specified.';
        }
        {
            __statement_dzU29k(20);
            this.name = 'ErrorNoEmailTemplateName';
        }
    };
}
{
    __statement_dzU29k(21);
    __extro_SeD6hf(22, __intro_vVnvMg(22, util).inherits(ErrorNoEmailTemplateName, Error));
}
{
    __statement_dzU29k(23);
    var ErrorEmailsPathNotSet = function () {
        __block_bHVGc$(1);
        {
            __statement_dzU29k(24);
            __extro_SeD6hf(25, __intro_vVnvMg(25, Error).apply(this, arguments));
        }
        {
            __statement_dzU29k(26);
            __extro_SeD6hf(27, __intro_vVnvMg(27, Error).captureStackTrace(this, arguments.callee));
        }
        {
            __statement_dzU29k(28);
            this.message = 'Keystone has not been configured for email support. Set the `emails` option in your configuration.';
        }
        {
            __statement_dzU29k(29);
            this.name = 'ErrorEmailsPathNotSet';
        }
    };
}
{
    __statement_dzU29k(30);
    __extro_SeD6hf(31, __intro_vVnvMg(31, util).inherits(ErrorEmailsPathNotSet, Error));
}
{
    __statement_dzU29k(32);
    var ErrorEmailOptionsRequired = function () {
        __block_bHVGc$(2);
        {
            __statement_dzU29k(33);
            __extro_SeD6hf(34, __intro_vVnvMg(34, Error).apply(this, arguments));
        }
        {
            __statement_dzU29k(35);
            __extro_SeD6hf(36, __intro_vVnvMg(36, Error).captureStackTrace(this, arguments.callee));
        }
        {
            __statement_dzU29k(37);
            this.message = 'The keystone.Email class requires a templateName or options argument to be provided.';
        }
        {
            __statement_dzU29k(38);
            this.name = 'ErrorEmailOptionsRequired';
        }
    };
}
{
    __statement_dzU29k(39);
    __extro_SeD6hf(40, __intro_vVnvMg(40, util).inherits(ErrorEmailsPathNotSet, Error));
}
{
    __statement_dzU29k(41);
    var getEmailsPath = function () {
        __block_bHVGc$(3);
        {
            __statement_dzU29k(42);
            var path = __extro_SeD6hf(43, __intro_vVnvMg(43, keystone).getPath('emails'));
        }
        if (__expression_BVxb5S(44), path) {
            __block_bHVGc$(4);
            return __expression_BVxb5S(45), path;
        }
        {
            __statement_dzU29k(46);
            throw new ErrorEmailsPathNotSet();
        }
    };
}
{
    __statement_dzU29k(47);
    var templateCSSMethods = {
            shadeColor: function (color, percent) {
                __block_bHVGc$(5);
                {
                    __statement_dzU29k(48);
                    var num = (__expression_BVxb5S(49), parseInt(__extro_SeD6hf(50, __intro_vVnvMg(50, color).slice(1)), 16)), amt = __extro_SeD6hf(51, __intro_vVnvMg(51, Math).round((__expression_BVxb5S(52), 2.55 * (__expression_BVxb5S(53), percent)))), R = (__expression_BVxb5S(54), (__expression_BVxb5S(55), (__expression_BVxb5S(56), num) >> 16) + (__expression_BVxb5S(57), amt)), G = (__expression_BVxb5S(58), (__expression_BVxb5S(59), (__expression_BVxb5S(60), (__expression_BVxb5S(61), num) >> 8) & 255) + (__expression_BVxb5S(62), amt)), B = (__expression_BVxb5S(63), (__expression_BVxb5S(64), (__expression_BVxb5S(65), num) & 255) + (__expression_BVxb5S(66), amt));
                }
                return __expression_BVxb5S(67), (__expression_BVxb5S(68), '#' + __extro_SeD6hf(69, __intro_vVnvMg(69, __extro_SeD6hf(70, __intro_vVnvMg(70, (__expression_BVxb5S(71), (__expression_BVxb5S(72), (__expression_BVxb5S(73), 16777216 + (__expression_BVxb5S(74), ((__expression_BVxb5S(77), (__expression_BVxb5S(78), R) < 255) ? (__expression_BVxb5S(75), (__expression_BVxb5S(81), (__expression_BVxb5S(82), R) < 1) ? (__expression_BVxb5S(79), 0) : (__expression_BVxb5S(80), R)) : (__expression_BVxb5S(76), 255)) * 65536)) + (__expression_BVxb5S(83), ((__expression_BVxb5S(86), (__expression_BVxb5S(87), G) < 255) ? (__expression_BVxb5S(84), (__expression_BVxb5S(90), (__expression_BVxb5S(91), G) < 1) ? (__expression_BVxb5S(88), 0) : (__expression_BVxb5S(89), G)) : (__expression_BVxb5S(85), 255)) * 256)) + ((__expression_BVxb5S(94), (__expression_BVxb5S(95), B) < 255) ? (__expression_BVxb5S(92), (__expression_BVxb5S(98), (__expression_BVxb5S(99), B) < 1) ? (__expression_BVxb5S(96), 0) : (__expression_BVxb5S(97), B)) : (__expression_BVxb5S(93), 255)))).toString(16))).slice(1)));
            }
        };
}
{
    __statement_dzU29k(100);
    var Email = function (options) {
        __block_bHVGc$(6);
        if (__expression_BVxb5S(101), 'string' === (__expression_BVxb5S(102), typeof options)) {
            __block_bHVGc$(7);
            {
                __statement_dzU29k(103);
                options = {
                    templateName: options
                };
            }
        } else if (__expression_BVxb5S(104), !__extro_SeD6hf(105, __intro_vVnvMg(105, utils).isObject(options))) {
            __block_bHVGc$(8);
            {
                __statement_dzU29k(106);
                throw new ErrorEmailOptionsRequired();
            }
        }
        {
            __statement_dzU29k(107);
            this.templateName = options.templateName;
        }
        {
            __statement_dzU29k(108);
            this.templateExt = (__expression_BVxb5S(109), options.templateExt || Email.defaults.templateExt);
        }
        {
            __statement_dzU29k(110);
            this.templateEngine = (__expression_BVxb5S(111), options.templateEngine || Email.defaults.templateEngine);
        }
        {
            __statement_dzU29k(112);
            this.templateBasePath = (__expression_BVxb5S(113), options.templateBasePath || Email.defaults.templateBasePath);
        }
        if (__expression_BVxb5S(114), !this.templateName) {
            __block_bHVGc$(9);
            {
                __statement_dzU29k(115);
                throw new ErrorNoEmailTemplateName();
            }
        }
        return __expression_BVxb5S(116), this;
    };
}
{
    __statement_dzU29k(117);
    Email.prototype.render = function (locals, callback) {
        __block_bHVGc$(10);
        if (__expression_BVxb5S(118), (__expression_BVxb5S(119), 'function' === (__expression_BVxb5S(120), typeof locals)) && (__expression_BVxb5S(121), !(__expression_BVxb5S(122), callback))) {
            __block_bHVGc$(11);
            {
                __statement_dzU29k(123);
                callback = locals;
            }
            {
                __statement_dzU29k(124);
                locals = {};
            }
        }
        {
            __statement_dzU29k(125);
            locals = (__expression_BVxb5S(128), 'object' === (__expression_BVxb5S(129), typeof locals)) ? (__expression_BVxb5S(126), locals) : (__expression_BVxb5S(127), {});
        }
        {
            __statement_dzU29k(130);
            callback = (__expression_BVxb5S(133), 'function' === (__expression_BVxb5S(134), typeof callback)) ? (__expression_BVxb5S(131), callback) : (__expression_BVxb5S(132), function undefined() {
                __block_bHVGc$(12);
            });
        }
        if (__extro_SeD6hf(135, __intro_vVnvMg(135, keystone).get('email locals'))) {
            __block_bHVGc$(13);
            {
                __statement_dzU29k(136);
                __extro_SeD6hf(137, __intro_vVnvMg(137, _).defaults(locals, __extro_SeD6hf(138, __intro_vVnvMg(138, keystone).get('email locals'))));
            }
        }
        {
            __statement_dzU29k(139);
            __extro_SeD6hf(140, __intro_vVnvMg(140, _).defaults(locals, {
                pretty: true,
                _: _,
                moment: moment,
                utils: utils,
                subject: '(no subject)',
                brand: __extro_SeD6hf(141, __intro_vVnvMg(141, keystone).get('brand')),
                theme: {},
                css: templateCSSMethods
            }));
        }
        if (__expression_BVxb5S(142), !locals.theme.buttons) {
            __block_bHVGc$(14);
            {
                __statement_dzU29k(143);
                locals.theme.buttons = {};
            }
        }
        {
            __statement_dzU29k(144);
            __extro_SeD6hf(145, __intro_vVnvMg(145, this).compileTemplate(__extro_SeD6hf(146, __intro_vVnvMg(146, function (err) {
                __block_bHVGc$(15);
                if (__expression_BVxb5S(147), err) {
                    __block_bHVGc$(16);
                    return __expression_BVxb5S(148), (__expression_BVxb5S(149), callback(err));
                }
                {
                    __statement_dzU29k(150);
                    var html = __extro_SeD6hf(151, __intro_vVnvMg(151, templateCache)[this.templateName](locals));
                }
                {
                    __statement_dzU29k(152);
                    html = __extro_SeD6hf(153, __intro_vVnvMg(153, html).replace(/[\u007f-\uffff]/g, function (c) {
                        __block_bHVGc$(17);
                        return __expression_BVxb5S(154), (__expression_BVxb5S(155), (__expression_BVxb5S(156), '&#x' + __extro_SeD6hf(157, __intro_vVnvMg(157, (__expression_BVxb5S(158), '0000' + __extro_SeD6hf(159, __intro_vVnvMg(159, __extro_SeD6hf(160, __intro_vVnvMg(160, c).charCodeAt(0))).toString(16)))).slice((__expression_BVxb5S(161), -4)))) + ';');
                    }));
                }
                {
                    __statement_dzU29k(162);
                    var rules = __extro_SeD6hf(163, __intro_vVnvMg(163, keystone).get('email rules'));
                }
                if (__expression_BVxb5S(164), rules) {
                    __block_bHVGc$(18);
                    if (__expression_BVxb5S(165), !__extro_SeD6hf(166, __intro_vVnvMg(166, Array).isArray(rules))) {
                        __block_bHVGc$(19);
                        {
                            __statement_dzU29k(167);
                            rules = [
                                rules
                            ];
                        }
                    }
                    {
                        __statement_dzU29k(168);
                        __extro_SeD6hf(169, __intro_vVnvMg(169, _).each(rules, function (rule) {
                            __block_bHVGc$(20);
                            if (__expression_BVxb5S(170), rule.find && rule.replace) {
                                __block_bHVGc$(21);
                                {
                                    __statement_dzU29k(171);
                                    var find = rule.find, replace = rule.replace;
                                }
                                if (__expression_BVxb5S(172), 'string' === (__expression_BVxb5S(173), typeof find)) {
                                    __block_bHVGc$(22);
                                    {
                                        __statement_dzU29k(174);
                                        find = new RegExp(find, 'gi');
                                    }
                                }
                                {
                                    __statement_dzU29k(175);
                                    html = __extro_SeD6hf(176, __intro_vVnvMg(176, html).replace(find, replace));
                                }
                            }
                        }));
                    }
                }
                {
                    __statement_dzU29k(177);
                    __expression_BVxb5S(178), callback(null, {
                        subject: locals.subject,
                        html: html
                    });
                }
            }).bind(this))));
        }
    };
}
{
    __statement_dzU29k(179);
    Email.prototype.loadTemplate = function (callback) {
        __block_bHVGc$(23);
        {
            __statement_dzU29k(180);
            var fsTemplatePath = __extro_SeD6hf(181, __intro_vVnvMg(181, path).join(__extro_SeD6hf(182, __intro_vVnvMg(182, Email).getEmailsPath()), (__expression_BVxb5S(183), (__expression_BVxb5S(184), this.templateName + '.') + this.templateExt)));
        }
        {
            __statement_dzU29k(185);
            __extro_SeD6hf(186, __intro_vVnvMg(186, fs).readFile(fsTemplatePath, __extro_SeD6hf(187, __intro_vVnvMg(187, function (err, contents) {
                __block_bHVGc$(24);
                if (__expression_BVxb5S(188), err) {
                    __block_bHVGc$(25);
                    if (__expression_BVxb5S(189), err.code === 'ENOENT') {
                        __block_bHVGc$(26);
                        {
                            __statement_dzU29k(190);
                            fsTemplatePath = __extro_SeD6hf(191, __intro_vVnvMg(191, path).join(__extro_SeD6hf(192, __intro_vVnvMg(192, Email).getEmailsPath()), this.templateName, (__expression_BVxb5S(193), 'email.' + this.templateExt)));
                        }
                        {
                            __statement_dzU29k(194);
                            __extro_SeD6hf(195, __intro_vVnvMg(195, fs).readFile(fsTemplatePath, function (err, contents) {
                                __block_bHVGc$(27);
                                {
                                    __statement_dzU29k(196);
                                    __expression_BVxb5S(197), callback(err, fsTemplatePath, contents);
                                }
                            }));
                        }
                    } else {
                        __block_bHVGc$(28);
                        return __expression_BVxb5S(198), (__expression_BVxb5S(199), callback(err));
                    }
                } else {
                    __block_bHVGc$(29);
                    return __expression_BVxb5S(200), (__expression_BVxb5S(201), callback(err, fsTemplatePath, contents));
                }
            }).bind(this))));
        }
    };
}
{
    __statement_dzU29k(202);
    Email.prototype.compileTemplate = function (callback) {
        __block_bHVGc$(30);
        if (__expression_BVxb5S(203), (__expression_BVxb5S(204), __extro_SeD6hf(205, __intro_vVnvMg(205, keystone).get('env')) === 'production') && templateCache[this.templateName]) {
            __block_bHVGc$(31);
            return __expression_BVxb5S(206), __extro_SeD6hf(207, __intro_vVnvMg(207, process).nextTick(callback));
        }
        {
            __statement_dzU29k(208);
            __extro_SeD6hf(209, __intro_vVnvMg(209, this).loadTemplate(__extro_SeD6hf(210, __intro_vVnvMg(210, function (err, filename, contents) {
                __block_bHVGc$(32);
                if (__expression_BVxb5S(211), err) {
                    __block_bHVGc$(33);
                    return __expression_BVxb5S(212), (__expression_BVxb5S(213), callback(err));
                }
                {
                    __statement_dzU29k(214);
                    var template = __extro_SeD6hf(215, __intro_vVnvMg(215, this.templateEngine).compile(__extro_SeD6hf(216, __intro_vVnvMg(216, contents).toString()), (__expression_BVxb5S(217), Email.defaults.compilerOptions || {
                            filename: __extro_SeD6hf(218, __intro_vVnvMg(218, fs).realpathSync(filename)),
                            basedir: this.templateBasePath
                        })));
                }
                {
                    __statement_dzU29k(219);
                    templateCache[this.templateName] = template;
                }
                {
                    __statement_dzU29k(220);
                    __expression_BVxb5S(221), callback();
                }
            }).bind(this))));
        }
    };
}
{
    __statement_dzU29k(222);
    Email.prototype.send = function (options, callback) {
        __block_bHVGc$(34);
        {
            __statement_dzU29k(223);
            var locals = options;
        }
        if (__expression_BVxb5S(224), (__expression_BVxb5S(225), arguments.length === 3) || (__expression_BVxb5S(226), !__extro_SeD6hf(227, __intro_vVnvMg(227, utils).isFunction(callback)))) {
            __block_bHVGc$(35);
            {
                __statement_dzU29k(228);
                callback = arguments[2];
            }
            {
                __statement_dzU29k(229);
                options = (__expression_BVxb5S(230), arguments[1] || arguments[0]);
            }
        }
        {
            __statement_dzU29k(231);
            __extro_SeD6hf(232, __intro_vVnvMg(232, this).render(locals, __extro_SeD6hf(233, __intro_vVnvMg(233, function (err, email) {
                __block_bHVGc$(36);
                {
                    __statement_dzU29k(234);
                    callback = (__expression_BVxb5S(237), 'function' === (__expression_BVxb5S(238), typeof callback)) ? (__expression_BVxb5S(235), callback) : (__expression_BVxb5S(236), function undefined() {
                        __block_bHVGc$(37);
                    });
                }
                if (__expression_BVxb5S(239), err) {
                    __block_bHVGc$(38);
                    return __expression_BVxb5S(240), (__expression_BVxb5S(241), callback(err));
                }
                if (__expression_BVxb5S(242), !__extro_SeD6hf(243, __intro_vVnvMg(243, utils).isObject(options))) {
                    __block_bHVGc$(39);
                    return __expression_BVxb5S(244), (__expression_BVxb5S(245), callback({
                        from: 'Email.send',
                        key: 'invalid options',
                        message: 'options object is required'
                    }));
                }
                if (__expression_BVxb5S(246), 'string' === (__expression_BVxb5S(247), typeof options.from)) {
                    __block_bHVGc$(40);
                    {
                        __statement_dzU29k(248);
                        options.fromName = options.from;
                    }
                    {
                        __statement_dzU29k(249);
                        options.fromEmail = options.from;
                    }
                } else if (__extro_SeD6hf(250, __intro_vVnvMg(250, utils).isObject(options.from))) {
                    __block_bHVGc$(41);
                    {
                        __statement_dzU29k(251);
                        options.fromName = __extro_SeD6hf(254, __intro_vVnvMg(254, utils).isObject(options.from.name)) ? (__expression_BVxb5S(252), options.from.name.full) : (__expression_BVxb5S(253), options.from.name);
                    }
                    {
                        __statement_dzU29k(255);
                        options.fromEmail = options.from.email;
                    }
                }
                if (__expression_BVxb5S(256), (__expression_BVxb5S(257), !options.fromName) || (__expression_BVxb5S(258), !options.fromEmail)) {
                    __block_bHVGc$(42);
                    return __expression_BVxb5S(259), (__expression_BVxb5S(260), callback({
                        from: 'Email.send',
                        key: 'invalid options',
                        message: 'options.fromName and options.fromEmail are required'
                    }));
                }
                if (__expression_BVxb5S(261), !options.mandrill) {
                    __block_bHVGc$(43);
                    if (__expression_BVxb5S(262), !__extro_SeD6hf(263, __intro_vVnvMg(263, keystone).get('mandrill api key'))) {
                        __block_bHVGc$(44);
                        return __expression_BVxb5S(264), (__expression_BVxb5S(265), callback({
                            from: 'Email.send',
                            key: 'missing api key',
                            message: 'You must either provide a Mandrill API Instance or set the mandrill api key before sending email.'
                        }));
                    }
                    {
                        __statement_dzU29k(266);
                        options.mandrill = new mandrillapi.Mandrill(__extro_SeD6hf(267, __intro_vVnvMg(267, keystone).get('mandrill api key')));
                    }
                }
                {
                    __statement_dzU29k(268);
                    options.tags = __extro_SeD6hf(271, __intro_vVnvMg(271, utils).isArray(options.tags)) ? (__expression_BVxb5S(269), options.tags) : (__expression_BVxb5S(270), []);
                }
                {
                    __statement_dzU29k(272);
                    __extro_SeD6hf(273, __intro_vVnvMg(273, options.tags).push((__expression_BVxb5S(274), 'sent:' + __extro_SeD6hf(275, __intro_vVnvMg(275, (__expression_BVxb5S(276), moment())).format('YYYY-MM-DD')))));
                }
                {
                    __statement_dzU29k(277);
                    __extro_SeD6hf(278, __intro_vVnvMg(278, options.tags).push(this.templateName));
                }
                if (__expression_BVxb5S(279), __extro_SeD6hf(280, __intro_vVnvMg(280, keystone).get('env')) === 'development') {
                    __block_bHVGc$(45);
                    {
                        __statement_dzU29k(281);
                        __extro_SeD6hf(282, __intro_vVnvMg(282, options.tags).push('development'));
                    }
                }
                {
                    __statement_dzU29k(283);
                    var recipients = [], mergeVars = [];
                }
                {
                    __statement_dzU29k(284);
                    options.to = __extro_SeD6hf(287, __intro_vVnvMg(287, Array).isArray(options.to)) ? (__expression_BVxb5S(285), options.to) : (__expression_BVxb5S(286), [
                        options.to
                    ]);
                }
                for (var i = 0; __expression_BVxb5S(288), (__expression_BVxb5S(289), i) < options.to.length; __expression_BVxb5S(290), i++) {
                    __block_bHVGc$(46);
                    if (__expression_BVxb5S(291), 'string' === (__expression_BVxb5S(292), typeof options.to[i])) {
                        __block_bHVGc$(47);
                        {
                            __statement_dzU29k(293);
                            options.to[i] = {
                                email: options.to[i]
                            };
                        }
                    } else if (__expression_BVxb5S(294), 'object' === (__expression_BVxb5S(295), typeof options.to[i])) {
                        __block_bHVGc$(48);
                        if (__expression_BVxb5S(296), !options.to[i].email) {
                            __block_bHVGc$(49);
                            return __expression_BVxb5S(297), (__expression_BVxb5S(298), callback({
                                from: 'Email.send',
                                key: 'invalid recipient',
                                message: (__expression_BVxb5S(299), (__expression_BVxb5S(300), 'Recipient ' + (__expression_BVxb5S(301), (__expression_BVxb5S(302), i) + 1)) + ' does not have a valid email address.')
                            }));
                        }
                    } else {
                        __block_bHVGc$(50);
                        return __expression_BVxb5S(303), (__expression_BVxb5S(304), callback({
                            from: 'Email.send',
                            key: 'invalid recipient',
                            message: (__expression_BVxb5S(305), (__expression_BVxb5S(306), 'Recipient ' + (__expression_BVxb5S(307), (__expression_BVxb5S(308), i) + 1)) + ' is not a string or an object.')
                        }));
                    }
                    {
                        __statement_dzU29k(309);
                        var recipient = {
                                email: options.to[i].email
                            }, vars = [
                                {
                                    name: 'email',
                                    content: recipient.email
                                }
                            ];
                    }
                    if (__expression_BVxb5S(310), 'string' === (__expression_BVxb5S(311), typeof options.to[i].name)) {
                        __block_bHVGc$(51);
                        {
                            __statement_dzU29k(312);
                            recipient.name = options.to[i].name;
                        }
                        {
                            __statement_dzU29k(313);
                            __extro_SeD6hf(314, __intro_vVnvMg(314, vars).push({
                                name: 'name',
                                content: options.to[i].name
                            }));
                        }
                    } else if (__expression_BVxb5S(315), 'object' === (__expression_BVxb5S(316), typeof options.to[i].name)) {
                        __block_bHVGc$(52);
                        {
                            __statement_dzU29k(317);
                            recipient.name = (__expression_BVxb5S(318), options.to[i].name.full || '');
                        }
                        {
                            __statement_dzU29k(319);
                            __extro_SeD6hf(320, __intro_vVnvMg(320, vars).push({
                                name: 'name',
                                content: (__expression_BVxb5S(321), options.to[i].name.full || '')
                            }));
                        }
                        {
                            __statement_dzU29k(322);
                            __extro_SeD6hf(323, __intro_vVnvMg(323, vars).push({
                                name: 'first_name',
                                content: (__expression_BVxb5S(324), options.to[i].name.first || '')
                            }));
                        }
                        {
                            __statement_dzU29k(325);
                            __extro_SeD6hf(326, __intro_vVnvMg(326, vars).push({
                                name: 'last_name',
                                content: (__expression_BVxb5S(327), options.to[i].name.last || '')
                            }));
                        }
                    }
                    {
                        __statement_dzU29k(328);
                        __extro_SeD6hf(329, __intro_vVnvMg(329, recipients).push(recipient));
                    }
                    {
                        __statement_dzU29k(330);
                        __extro_SeD6hf(331, __intro_vVnvMg(331, mergeVars).push({
                            rcpt: recipient.email,
                            vars: vars
                        }));
                    }
                }
                {
                    __statement_dzU29k(332);
                    var onSuccess = function (info) {
                        __block_bHVGc$(53);
                        {
                            __statement_dzU29k(333);
                            __expression_BVxb5S(334), callback(null, info);
                        }
                    };
                }
                {
                    __statement_dzU29k(335);
                    var onFail = function (info) {
                        __block_bHVGc$(54);
                        {
                            __statement_dzU29k(336);
                            __expression_BVxb5S(337), callback({
                                from: 'Email.send',
                                key: 'send error',
                                message: 'Mandrill encountered an error and did not send the emails.',
                                info: info
                            });
                        }
                    };
                }
                {
                    __statement_dzU29k(338);
                    var message = {
                            html: email.html,
                            subject: email.subject,
                            from_name: options.fromName,
                            from_email: options.fromEmail,
                            tags: options.tags,
                            attachments: options.attachments,
                            to: recipients,
                            merge_vars: mergeVars,
                            async: true
                        };
                }
                {
                    __statement_dzU29k(339);
                    __extro_SeD6hf(340, __intro_vVnvMg(340, _).defaults(message, options.mandrillOptions));
                }
                {
                    __statement_dzU29k(341);
                    __extro_SeD6hf(342, __intro_vVnvMg(342, _).defaults(message, Email.defaults.mandrill));
                }
                return __expression_BVxb5S(343), __extro_SeD6hf(344, __intro_vVnvMg(344, process).nextTick(function () {
                    __block_bHVGc$(55);
                    {
                        __statement_dzU29k(345);
                        __extro_SeD6hf(346, __intro_vVnvMg(346, options.mandrill.messages).send({
                            message: message
                        }, onSuccess, onFail));
                    }
                }));
            }).bind(this))));
        }
    };
}
{
    __statement_dzU29k(347);
    Email.getEmailsPath = getEmailsPath;
}
{
    __statement_dzU29k(348);
    Email.templateCache = templateCache;
}
{
    __statement_dzU29k(349);
    Email.templateCSSMethods = templateCSSMethods;
}
{
    __statement_dzU29k(350);
    Email.defaults = defaultConfig;
}
{
    __statement_dzU29k(351);
    exports = module.exports = Email;
}
