// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_schemaPlugins.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_WMcoz7, __expression_UFQPOZ, __block_P$WU7W;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_WMcoz7 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/schemaPlugins.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_UFQPOZ = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/schemaPlugins.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_P$WU7W = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/schemaPlugins.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_aPs4zS = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_AZ6TvY = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/schemaPlugins.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_WMcoz7(0);
    var keystone = (__expression_UFQPOZ(1), require('../')), Types = (__expression_UFQPOZ(2), require('./fieldTypes')), _ = (__expression_UFQPOZ(3), require('underscore')), async = (__expression_UFQPOZ(4), require('async')), utils = (__expression_UFQPOZ(5), require('keystone-utils'));
}
{
    __statement_WMcoz7(6);
    var methods = module.exports.methods = {};
}
{
    __statement_WMcoz7(7);
    var options = module.exports.options = {};
}
{
    __statement_WMcoz7(8);
    exports.sortable = function () {
        __block_P$WU7W(0);
        {
            __statement_WMcoz7(9);
            var list = this;
        }
        {
            __statement_WMcoz7(10);
            __extro_AZ6TvY(11, __intro_aPs4zS(11, this.schema).add({
                sortOrder: {
                    type: Number,
                    index: true
                }
            }));
        }
        {
            __statement_WMcoz7(12);
            __extro_AZ6TvY(13, __intro_aPs4zS(13, this.schema).pre('save', function (next) {
                __block_P$WU7W(1);
                if (__expression_UFQPOZ(14), (__expression_UFQPOZ(15), typeof this.sortOrder) === 'number') {
                    __block_P$WU7W(2);
                    return __expression_UFQPOZ(16), (__expression_UFQPOZ(17), next());
                }
                {
                    __statement_WMcoz7(18);
                    var item = this;
                }
                {
                    __statement_WMcoz7(19);
                    __extro_AZ6TvY(20, __intro_aPs4zS(20, __extro_AZ6TvY(21, __intro_aPs4zS(21, __extro_AZ6TvY(22, __intro_aPs4zS(22, list.model).findOne())).sort('-sortOrder'))).exec(function (err, max) {
                        __block_P$WU7W(3);
                        {
                            __statement_WMcoz7(23);
                            item.sortOrder = (__expression_UFQPOZ(26), (__expression_UFQPOZ(27), max) && max.sortOrder) ? (__expression_UFQPOZ(24), (__expression_UFQPOZ(28), max.sortOrder + 1)) : (__expression_UFQPOZ(25), 0);
                        }
                        {
                            __statement_WMcoz7(29);
                            __expression_UFQPOZ(30), next();
                        }
                    }));
                }
            }));
        }
    };
}
{
    __statement_WMcoz7(31);
    exports.autokey = function () {
        __block_P$WU7W(4);
        {
            __statement_WMcoz7(32);
            var autokey = this.autokey = __extro_AZ6TvY(33, __intro_aPs4zS(33, this).get('autokey')), list = this, def = {};
        }
        if (__expression_UFQPOZ(34), !autokey.from) {
            __block_P$WU7W(5);
            {
                __statement_WMcoz7(35);
                var fromMsg = (__expression_UFQPOZ(36), (__expression_UFQPOZ(37), 'Invalid List Option (autokey) for ' + list.key) + ' (from is required)\n');
            }
            {
                __statement_WMcoz7(38);
                throw new Error(fromMsg);
            }
        }
        if (__expression_UFQPOZ(39), !autokey.path) {
            __block_P$WU7W(6);
            {
                __statement_WMcoz7(40);
                var pathMsg = (__expression_UFQPOZ(41), (__expression_UFQPOZ(42), 'Invalid List Option (autokey) for ' + list.key) + ' (path is required)\n');
            }
            {
                __statement_WMcoz7(43);
                throw new Error(pathMsg);
            }
        }
        if (__expression_UFQPOZ(44), 'string' === (__expression_UFQPOZ(45), typeof autokey.from)) {
            __block_P$WU7W(7);
            {
                __statement_WMcoz7(46);
                autokey.from = __extro_AZ6TvY(47, __intro_aPs4zS(47, autokey.from).split(' '));
            }
        }
        {
            __statement_WMcoz7(48);
            autokey.from = __extro_AZ6TvY(49, __intro_aPs4zS(49, autokey.from).map(function (i) {
                __block_P$WU7W(8);
                {
                    __statement_WMcoz7(50);
                    i = __extro_AZ6TvY(51, __intro_aPs4zS(51, i).split(':'));
                }
                return __expression_UFQPOZ(52), {
                    path: i[0],
                    format: i[1]
                };
            }));
        }
        {
            __statement_WMcoz7(53);
            def[autokey.path] = {
                type: String,
                index: true
            };
        }
        if (autokey.unique) {
            __block_P$WU7W(9);
            {
                __statement_WMcoz7(54);
                def[autokey.path].index = {
                    unique: true
                };
            }
        }
        {
            __statement_WMcoz7(55);
            __extro_AZ6TvY(56, __intro_aPs4zS(56, this.schema).add(def));
        }
        {
            __statement_WMcoz7(57);
            var getUniqueKey = function (doc, src, callback) {
                __block_P$WU7W(10);
                {
                    __statement_WMcoz7(58);
                    var q = __extro_AZ6TvY(59, __intro_aPs4zS(59, __extro_AZ6TvY(60, __intro_aPs4zS(60, list.model).find())).where(autokey.path, src));
                }
                if (__extro_AZ6TvY(61, __intro_aPs4zS(61, _).isObject(autokey.unique))) {
                    __block_P$WU7W(11);
                    {
                        __statement_WMcoz7(62);
                        __extro_AZ6TvY(63, __intro_aPs4zS(63, _).each(autokey.unique, function (k, v) {
                            __block_P$WU7W(12);
                            if (__expression_UFQPOZ(64), __extro_AZ6TvY(65, __intro_aPs4zS(65, _).isString(v)) && (__expression_UFQPOZ(66), __extro_AZ6TvY(67, __intro_aPs4zS(67, v).charAt(0)) === ':')) {
                                __block_P$WU7W(13);
                                {
                                    __statement_WMcoz7(68);
                                    __extro_AZ6TvY(69, __intro_aPs4zS(69, q).where(k, __extro_AZ6TvY(70, __intro_aPs4zS(70, doc).get(__extro_AZ6TvY(71, __intro_aPs4zS(71, v).substr(1))))));
                                }
                            } else {
                                __block_P$WU7W(14);
                                {
                                    __statement_WMcoz7(72);
                                    __extro_AZ6TvY(73, __intro_aPs4zS(73, q).where(k, v));
                                }
                            }
                        }));
                    }
                }
                {
                    __statement_WMcoz7(74);
                    __extro_AZ6TvY(75, __intro_aPs4zS(75, q).exec(function (err, results) {
                        __block_P$WU7W(15);
                        if (__expression_UFQPOZ(76), err) {
                            __block_P$WU7W(16);
                            {
                                __statement_WMcoz7(77);
                                __expression_UFQPOZ(78), callback(err);
                            }
                        } else if (__expression_UFQPOZ(79), results.length && (__expression_UFQPOZ(80), (__expression_UFQPOZ(81), results.length > 1) || (__expression_UFQPOZ(82), results[0].id != doc.id))) {
                            __block_P$WU7W(17);
                            {
                                __statement_WMcoz7(83);
                                var inc = __extro_AZ6TvY(84, __intro_aPs4zS(84, src).match(/^(.+)\-(\d+)$/));
                            }
                            if (__expression_UFQPOZ(85), (__expression_UFQPOZ(86), inc) && (__expression_UFQPOZ(87), inc.length === 3)) {
                                __block_P$WU7W(18);
                                {
                                    __statement_WMcoz7(88);
                                    src = inc[1];
                                }
                                {
                                    __statement_WMcoz7(89);
                                    inc = (__expression_UFQPOZ(90), '-' + (__expression_UFQPOZ(91), (__expression_UFQPOZ(92), inc[2] * 1) + 1));
                                }
                            } else {
                                __block_P$WU7W(19);
                                {
                                    __statement_WMcoz7(93);
                                    inc = '-1';
                                }
                            }
                            return __expression_UFQPOZ(94), (__expression_UFQPOZ(95), getUniqueKey(doc, (__expression_UFQPOZ(96), (__expression_UFQPOZ(97), src) + (__expression_UFQPOZ(98), inc)), callback));
                        } else {
                            __block_P$WU7W(20);
                            {
                                __statement_WMcoz7(99);
                                __extro_AZ6TvY(100, __intro_aPs4zS(100, doc).set(autokey.path, src));
                            }
                            return __expression_UFQPOZ(101), (__expression_UFQPOZ(102), callback());
                        }
                    }));
                }
            };
        }
        {
            __statement_WMcoz7(103);
            __extro_AZ6TvY(104, __intro_aPs4zS(104, this.schema).pre('save', function (next) {
                __block_P$WU7W(21);
                {
                    __statement_WMcoz7(105);
                    var modified = false, values = [];
                }
                {
                    __statement_WMcoz7(106);
                    __extro_AZ6TvY(107, __intro_aPs4zS(107, autokey.from).forEach(function (ops) {
                        __block_P$WU7W(22);
                        if (list.fields[ops.path]) {
                            __block_P$WU7W(23);
                            {
                                __statement_WMcoz7(108);
                                __extro_AZ6TvY(109, __intro_aPs4zS(109, values).push(__extro_AZ6TvY(110, __intro_aPs4zS(110, list.fields[ops.path]).format(this, ops.format))));
                            }
                            if (__extro_AZ6TvY(111, __intro_aPs4zS(111, list.fields[ops.path]).isModified(this))) {
                                __block_P$WU7W(24);
                                {
                                    __statement_WMcoz7(112);
                                    modified = true;
                                }
                            }
                        } else {
                            __block_P$WU7W(25);
                            {
                                __statement_WMcoz7(113);
                                __extro_AZ6TvY(114, __intro_aPs4zS(114, values).push(__extro_AZ6TvY(115, __intro_aPs4zS(115, this).get(ops.path))));
                            }
                            if (__expression_UFQPOZ(116), (__expression_UFQPOZ(117), (__expression_UFQPOZ(118), ops.path !== 'id') && (__expression_UFQPOZ(119), __extro_AZ6TvY(120, __intro_aPs4zS(120, list.schema).pathType(ops.path)) === 'virtual')) || __extro_AZ6TvY(121, __intro_aPs4zS(121, this).isModified(ops.path))) {
                                __block_P$WU7W(26);
                                {
                                    __statement_WMcoz7(122);
                                    modified = true;
                                }
                            }
                        }
                    }, this));
                }
                if (__expression_UFQPOZ(123), (__expression_UFQPOZ(124), (__expression_UFQPOZ(125), !(__expression_UFQPOZ(126), modified)) || autokey.fixed) && __extro_AZ6TvY(127, __intro_aPs4zS(127, this).get(autokey.path))) {
                    __block_P$WU7W(27);
                    return __expression_UFQPOZ(128), (__expression_UFQPOZ(129), next());
                }
                {
                    __statement_WMcoz7(130);
                    var newKey = __extro_AZ6TvY(131, __intro_aPs4zS(131, utils).slug((__expression_UFQPOZ(132), __extro_AZ6TvY(133, __intro_aPs4zS(133, values).join(' ')) || this.id)));
                }
                if (autokey.unique) {
                    __block_P$WU7W(28);
                    return __expression_UFQPOZ(134), (__expression_UFQPOZ(135), getUniqueKey(this, newKey, next));
                } else {
                    __block_P$WU7W(29);
                    {
                        __statement_WMcoz7(136);
                        __extro_AZ6TvY(137, __intro_aPs4zS(137, this).set(autokey.path, newKey));
                    }
                    return __expression_UFQPOZ(138), (__expression_UFQPOZ(139), next());
                }
            }));
        }
    };
}
{
    __statement_WMcoz7(140);
    exports.track = function () {
        __block_P$WU7W(30);
        {
            __statement_WMcoz7(141);
            var list = this, options = __extro_AZ6TvY(142, __intro_aPs4zS(142, list).get('track')), userModel = __extro_AZ6TvY(143, __intro_aPs4zS(143, keystone).get('user model')), defaultOptions = {
                    createdAt: false,
                    createdBy: false,
                    updatedAt: false,
                    updatedBy: false
                }, fields = {};
        }
        if (__expression_UFQPOZ(144), (__expression_UFQPOZ(145), !__extro_AZ6TvY(146, __intro_aPs4zS(146, _).isBoolean(options))) && (__expression_UFQPOZ(147), !__extro_AZ6TvY(148, __intro_aPs4zS(148, _).isObject(options)))) {
            __block_P$WU7W(31);
            {
                __statement_WMcoz7(149);
                throw new Error((__expression_UFQPOZ(150), (__expression_UFQPOZ(151), (__expression_UFQPOZ(152), (__expression_UFQPOZ(153), 'Invalid List "track" option for ' + list.key) + '\n') + '"track" must be a boolean or an object.\n\n') + 'See http://keystonejs.com/docs/database/#lists-options for more information.'));
            }
        }
        if (__extro_AZ6TvY(154, __intro_aPs4zS(154, _).isBoolean(options))) {
            __block_P$WU7W(32);
            if (__expression_UFQPOZ(155), options) {
                __block_P$WU7W(33);
                {
                    __statement_WMcoz7(156);
                    options = {
                        createdAt: true,
                        createdBy: true,
                        updatedAt: true,
                        updatedBy: true
                    };
                }
            } else {
                __block_P$WU7W(34);
                return __expression_UFQPOZ(157);
            }
        }
        if (__expression_UFQPOZ(158), (__expression_UFQPOZ(159), (__expression_UFQPOZ(160), (__expression_UFQPOZ(161), !options.createdAt) && (__expression_UFQPOZ(162), !options.createdBy)) && (__expression_UFQPOZ(163), !options.updatedAt)) && (__expression_UFQPOZ(164), !options.updatedBy)) {
            __block_P$WU7W(35);
            return __expression_UFQPOZ(165);
        }
        if (__extro_AZ6TvY(166, __intro_aPs4zS(166, list).get('track simulate standard meta'))) {
            __block_P$WU7W(36);
            if (__expression_UFQPOZ(167), !options.createdAt) {
                __block_P$WU7W(37);
                {
                    __statement_WMcoz7(168);
                    options.createdAt = true;
                }
            }
            if (__expression_UFQPOZ(169), !options.updatedAt) {
                __block_P$WU7W(38);
                {
                    __statement_WMcoz7(170);
                    options.updatedAt = true;
                }
            }
        }
        {
            __statement_WMcoz7(171);
            options = __extro_AZ6TvY(172, __intro_aPs4zS(172, _).extend({}, defaultOptions, options));
        }
        {
            __statement_WMcoz7(173);
            __extro_AZ6TvY(174, __intro_aPs4zS(174, _).each(options, function (value, key) {
                __block_P$WU7W(39);
                {
                    __statement_WMcoz7(175);
                    var fieldName;
                }
                if (__extro_AZ6TvY(176, __intro_aPs4zS(176, _).has(defaultOptions, key))) {
                    __block_P$WU7W(40);
                    if (__expression_UFQPOZ(177), (__expression_UFQPOZ(178), !__extro_AZ6TvY(179, __intro_aPs4zS(179, _).isBoolean(value))) && (__expression_UFQPOZ(180), !__extro_AZ6TvY(181, __intro_aPs4zS(181, _).isString(value)))) {
                        __block_P$WU7W(41);
                        {
                            __statement_WMcoz7(182);
                            throw new Error((__expression_UFQPOZ(183), (__expression_UFQPOZ(184), (__expression_UFQPOZ(185), (__expression_UFQPOZ(186), (__expression_UFQPOZ(187), (__expression_UFQPOZ(188), 'Invalid List "track" option for ' + list.key) + '\n') + '"') + (__expression_UFQPOZ(189), key)) + '" must be a boolean or a string.\n\n') + 'See http://keystonejs.com/docs/database/#lists-options for more information.'));
                        }
                    }
                    if (__expression_UFQPOZ(190), value) {
                        __block_P$WU7W(42);
                        {
                            __statement_WMcoz7(191);
                            fieldName = (__expression_UFQPOZ(194), (__expression_UFQPOZ(195), value) === true) ? (__expression_UFQPOZ(192), key) : (__expression_UFQPOZ(193), value);
                        }
                        {
                            __statement_WMcoz7(196);
                            options[key] = fieldName;
                        }
                        switch (__expression_UFQPOZ(197), key) {
                        case 'createdAt':
                        case 'updatedAt': {
                                __block_P$WU7W(43);
                                {
                                    __statement_WMcoz7(198);
                                    fields[fieldName] = {
                                        type: Date,
                                        hidden: true,
                                        index: true
                                    };
                                }
                                break;
                            }
                        case 'createdBy':
                        case 'updatedBy': {
                                __block_P$WU7W(44);
                                {
                                    __statement_WMcoz7(199);
                                    fields[fieldName] = {
                                        type: Types.Relationship,
                                        ref: userModel,
                                        hidden: true,
                                        index: true
                                    };
                                }
                                break;
                            }
                        }
                    }
                } else {
                    __block_P$WU7W(45);
                    {
                        __statement_WMcoz7(200);
                        throw new Error((__expression_UFQPOZ(201), (__expression_UFQPOZ(202), (__expression_UFQPOZ(203), (__expression_UFQPOZ(204), 'Invalid List "track" option for ' + list.key) + '\n') + 'valid field options are "createdAt", "createdBy", "updatedAt", an "updatedBy".\n\n') + 'See http://keystonejs.com/docs/database/#lists-options for more information.'));
                    }
                }
            }));
        }
        if (__extro_AZ6TvY(205, __intro_aPs4zS(205, list).get('track simulate standard meta'))) {
            __block_P$WU7W(46);
            {
                __statement_WMcoz7(206);
                __extro_AZ6TvY(207, __intro_aPs4zS(207, list).map('createdOn', options.createdAt));
            }
            {
                __statement_WMcoz7(208);
                __extro_AZ6TvY(209, __intro_aPs4zS(209, __extro_AZ6TvY(210, __intro_aPs4zS(210, list.schema).virtual('createdOn'))).get(function () {
                    __block_P$WU7W(47);
                    return __expression_UFQPOZ(211), __extro_AZ6TvY(212, __intro_aPs4zS(212, this).get('createdAt'));
                }));
            }
            {
                __statement_WMcoz7(213);
                __extro_AZ6TvY(214, __intro_aPs4zS(214, list).map('updatedOn', options.updatedAt));
            }
            {
                __statement_WMcoz7(215);
                __extro_AZ6TvY(216, __intro_aPs4zS(216, __extro_AZ6TvY(217, __intro_aPs4zS(217, list.schema).virtual('updatedOn'))).get(function () {
                    __block_P$WU7W(48);
                    return __expression_UFQPOZ(218), __extro_AZ6TvY(219, __intro_aPs4zS(219, this).get('updatedAt'));
                }));
            }
        }
        if (__extro_AZ6TvY(220, __intro_aPs4zS(220, _).isEmpty(fields))) {
            __block_P$WU7W(49);
            return __expression_UFQPOZ(221);
        }
        {
            __statement_WMcoz7(222);
            __extro_AZ6TvY(223, __intro_aPs4zS(223, list).add(fields));
        }
        {
            __statement_WMcoz7(224);
            __extro_AZ6TvY(225, __intro_aPs4zS(225, list.schema).pre('save', function (next) {
                __block_P$WU7W(50);
                {
                    __statement_WMcoz7(226);
                    var now = new Date();
                }
                if (this.isNew) {
                    __block_P$WU7W(51);
                    if (options.createdAt) {
                        __block_P$WU7W(52);
                        {
                            __statement_WMcoz7(227);
                            __extro_AZ6TvY(228, __intro_aPs4zS(228, this).set(options.createdAt, now));
                        }
                    }
                    if (__expression_UFQPOZ(229), options.createdBy && this._req_user) {
                        __block_P$WU7W(53);
                        {
                            __statement_WMcoz7(230);
                            __extro_AZ6TvY(231, __intro_aPs4zS(231, this).set(options.createdBy, this._req_user._id));
                        }
                    }
                }
                if (__expression_UFQPOZ(232), this.isNew || __extro_AZ6TvY(233, __intro_aPs4zS(233, this).isModified())) {
                    __block_P$WU7W(54);
                    if (options.updatedAt) {
                        __block_P$WU7W(55);
                        {
                            __statement_WMcoz7(234);
                            __extro_AZ6TvY(235, __intro_aPs4zS(235, this).set(options.updatedAt, now));
                        }
                    }
                    if (__expression_UFQPOZ(236), options.updatedBy && this._req_user) {
                        __block_P$WU7W(56);
                        {
                            __statement_WMcoz7(237);
                            __extro_AZ6TvY(238, __intro_aPs4zS(238, this).set(options.updatedBy, this._req_user._id));
                        }
                    }
                }
                {
                    __statement_WMcoz7(239);
                    __expression_UFQPOZ(240), next();
                }
            }));
        }
    };
}
{
    __statement_WMcoz7(241);
    methods.getRelated = function (paths, callback, nocollapse) {
        __block_P$WU7W(57);
        {
            __statement_WMcoz7(242);
            var item = this, list = this.list, queue = {};
        }
        if (__expression_UFQPOZ(243), 'function' !== (__expression_UFQPOZ(244), typeof callback)) {
            __block_P$WU7W(58);
            {
                __statement_WMcoz7(245);
                throw new Error('List.getRelated(paths, callback, nocollapse) requires a callback function.');
            }
        }
        if (__expression_UFQPOZ(246), 'string' === (__expression_UFQPOZ(247), typeof paths)) {
            __block_P$WU7W(59);
            {
                __statement_WMcoz7(248);
                var pathsArr = __extro_AZ6TvY(249, __intro_aPs4zS(249, paths).split(' '));
            }
            {
                __statement_WMcoz7(250);
                var lastPath = '';
            }
            {
                __statement_WMcoz7(251);
                paths = [];
            }
            for (var i = 0; __expression_UFQPOZ(252), (__expression_UFQPOZ(253), i) < pathsArr.length; __expression_UFQPOZ(254), i++) {
                __block_P$WU7W(60);
                {
                    __statement_WMcoz7(255);
                    lastPath += (__expression_UFQPOZ(256), (lastPath.length ? (__expression_UFQPOZ(257), ' ') : (__expression_UFQPOZ(258), '')) + pathsArr[i]);
                }
                if (__expression_UFQPOZ(259), (__expression_UFQPOZ(260), __extro_AZ6TvY(261, __intro_aPs4zS(261, lastPath).indexOf('[')) < 0) || (__expression_UFQPOZ(262), __extro_AZ6TvY(263, __intro_aPs4zS(263, lastPath).charAt((__expression_UFQPOZ(264), lastPath.length - 1))) === ']')) {
                    __block_P$WU7W(61);
                    {
                        __statement_WMcoz7(265);
                        __extro_AZ6TvY(266, __intro_aPs4zS(266, paths).push(lastPath));
                    }
                    {
                        __statement_WMcoz7(267);
                        lastPath = '';
                    }
                }
            }
        }
        {
            __statement_WMcoz7(268);
            __extro_AZ6TvY(269, __intro_aPs4zS(269, _).each(paths, function (options) {
                __block_P$WU7W(62);
                {
                    __statement_WMcoz7(270);
                    var populateString = '';
                }
                if (__expression_UFQPOZ(271), 'string' === (__expression_UFQPOZ(272), typeof options)) {
                    __block_P$WU7W(63);
                    if (__expression_UFQPOZ(273), __extro_AZ6TvY(274, __intro_aPs4zS(274, options).indexOf('[')) > 0) {
                        __block_P$WU7W(64);
                        {
                            __statement_WMcoz7(275);
                            populateString = __extro_AZ6TvY(276, __intro_aPs4zS(276, options).substring((__expression_UFQPOZ(277), __extro_AZ6TvY(278, __intro_aPs4zS(278, options).indexOf('[')) + 1), __extro_AZ6TvY(279, __intro_aPs4zS(279, options).indexOf(']'))));
                        }
                        {
                            __statement_WMcoz7(280);
                            options = __extro_AZ6TvY(281, __intro_aPs4zS(281, options).substr(0, __extro_AZ6TvY(282, __intro_aPs4zS(282, options).indexOf('['))));
                        }
                    }
                    {
                        __statement_WMcoz7(283);
                        options = {
                            path: options
                        };
                    }
                }
                {
                    __statement_WMcoz7(284);
                    options.populate = (__expression_UFQPOZ(285), options.populate || []);
                }
                {
                    __statement_WMcoz7(286);
                    options.related = (__expression_UFQPOZ(287), options.related || []);
                }
                {
                    __statement_WMcoz7(288);
                    var relationship = list.relationships[options.path];
                }
                if (__expression_UFQPOZ(289), !(__expression_UFQPOZ(290), relationship)) {
                    __block_P$WU7W(65);
                    {
                        __statement_WMcoz7(291);
                        throw new Error((__expression_UFQPOZ(292), (__expression_UFQPOZ(293), (__expression_UFQPOZ(294), (__expression_UFQPOZ(295), 'List.getRelated: list ' + list.key) + ' does not have a relationship ') + options.path) + '.'));
                    }
                }
                {
                    __statement_WMcoz7(296);
                    var refList = __extro_AZ6TvY(297, __intro_aPs4zS(297, keystone).list(relationship.ref));
                }
                if (__expression_UFQPOZ(298), !(__expression_UFQPOZ(299), refList)) {
                    __block_P$WU7W(66);
                    {
                        __statement_WMcoz7(300);
                        throw new Error((__expression_UFQPOZ(301), (__expression_UFQPOZ(302), 'List.getRelated: list ' + relationship.ref) + ' does not exist.'));
                    }
                }
                {
                    __statement_WMcoz7(303);
                    var relField = refList.fields[relationship.refPath];
                }
                if (__expression_UFQPOZ(304), (__expression_UFQPOZ(305), !(__expression_UFQPOZ(306), relField)) || (__expression_UFQPOZ(307), relField.type !== 'relationship')) {
                    __block_P$WU7W(67);
                    {
                        __statement_WMcoz7(308);
                        throw new Error((__expression_UFQPOZ(309), (__expression_UFQPOZ(310), (__expression_UFQPOZ(311), (__expression_UFQPOZ(312), (__expression_UFQPOZ(313), (__expression_UFQPOZ(314), 'List.getRelated: relationship ' + relationship.ref) + ' on list ') + list.key) + ' refers to a path (') + relationship.refPath) + ') which is not a relationship field.'));
                    }
                }
                if (populateString.length) {
                    __block_P$WU7W(68);
                    {
                        __statement_WMcoz7(315);
                        __extro_AZ6TvY(316, __intro_aPs4zS(316, _).each(__extro_AZ6TvY(317, __intro_aPs4zS(317, populateString).split(' ')), function (key) {
                            __block_P$WU7W(69);
                            if (refList.relationships[key]) {
                                __block_P$WU7W(70);
                                {
                                    __statement_WMcoz7(318);
                                    __extro_AZ6TvY(319, __intro_aPs4zS(319, options.related).push(key));
                                }
                            } else {
                                __block_P$WU7W(71);
                                {
                                    __statement_WMcoz7(320);
                                    __extro_AZ6TvY(321, __intro_aPs4zS(321, options.populate).push(key));
                                }
                            }
                        }));
                    }
                }
                {
                    __statement_WMcoz7(322);
                    queue[relationship.path] = function (done) {
                        __block_P$WU7W(72);
                        {
                            __statement_WMcoz7(323);
                            var query = __extro_AZ6TvY(324, __intro_aPs4zS(324, __extro_AZ6TvY(325, __intro_aPs4zS(325, refList.model).find())).where(relField.path));
                        }
                        if (options.populate) {
                            __block_P$WU7W(73);
                            {
                                __statement_WMcoz7(326);
                                __extro_AZ6TvY(327, __intro_aPs4zS(327, query).populate(options.populate));
                            }
                        }
                        if (relField.many) {
                            __block_P$WU7W(74);
                            {
                                __statement_WMcoz7(328);
                                __extro_AZ6TvY(329, __intro_aPs4zS(329, query).in([
                                    item.id
                                ]));
                            }
                        } else {
                            __block_P$WU7W(75);
                            {
                                __statement_WMcoz7(330);
                                __extro_AZ6TvY(331, __intro_aPs4zS(331, query).equals(item.id));
                            }
                        }
                        {
                            __statement_WMcoz7(332);
                            __extro_AZ6TvY(333, __intro_aPs4zS(333, query).sort((__expression_UFQPOZ(334), (__expression_UFQPOZ(335), options.sort || relationship.sort) || refList.defaultSort)));
                        }
                        if (options.related.length) {
                            __block_P$WU7W(76);
                            {
                                __statement_WMcoz7(336);
                                __extro_AZ6TvY(337, __intro_aPs4zS(337, query).exec(function (err, results) {
                                    __block_P$WU7W(77);
                                    if (__expression_UFQPOZ(338), (__expression_UFQPOZ(339), err) || (__expression_UFQPOZ(340), !results.length)) {
                                        __block_P$WU7W(78);
                                        return __expression_UFQPOZ(341), (__expression_UFQPOZ(342), done(err, results));
                                    }
                                    {
                                        __statement_WMcoz7(343);
                                        __extro_AZ6TvY(344, __intro_aPs4zS(344, async).parallel(__extro_AZ6TvY(345, __intro_aPs4zS(345, results).map(function (item) {
                                            __block_P$WU7W(79);
                                            return __expression_UFQPOZ(346), function (done) {
                                                __block_P$WU7W(80);
                                                {
                                                    __statement_WMcoz7(347);
                                                    __extro_AZ6TvY(348, __intro_aPs4zS(348, item).populateRelated(options.related, done));
                                                }
                                            };
                                        })), function (err) {
                                            __block_P$WU7W(81);
                                            {
                                                __statement_WMcoz7(349);
                                                __expression_UFQPOZ(350), done(err, results);
                                            }
                                        }));
                                    }
                                }));
                            }
                        } else {
                            __block_P$WU7W(82);
                            {
                                __statement_WMcoz7(351);
                                __extro_AZ6TvY(352, __intro_aPs4zS(352, query).exec(done));
                            }
                        }
                    };
                }
                if (__expression_UFQPOZ(353), !item._populatedRelationships) {
                    __block_P$WU7W(83);
                    {
                        __statement_WMcoz7(354);
                        item._populatedRelationships = {};
                    }
                }
                {
                    __statement_WMcoz7(355);
                    item._populatedRelationships[relationship.path] = true;
                }
            }));
        }
        {
            __statement_WMcoz7(356);
            __extro_AZ6TvY(357, __intro_aPs4zS(357, async).parallel(queue, function (err, results) {
                __block_P$WU7W(84);
                if (__expression_UFQPOZ(358), (__expression_UFQPOZ(359), (__expression_UFQPOZ(360), !(__expression_UFQPOZ(361), nocollapse)) && (__expression_UFQPOZ(362), results)) && (__expression_UFQPOZ(363), paths.length === 1)) {
                    __block_P$WU7W(85);
                    {
                        __statement_WMcoz7(364);
                        results = results[paths[0]];
                    }
                }
                {
                    __statement_WMcoz7(365);
                    __expression_UFQPOZ(366), callback(err, results);
                }
            }));
        }
    };
}
{
    __statement_WMcoz7(367);
    methods.populateRelated = function (rel, callback) {
        __block_P$WU7W(86);
        {
            __statement_WMcoz7(368);
            var item = this;
        }
        if (__expression_UFQPOZ(369), 'function' !== (__expression_UFQPOZ(370), typeof callback)) {
            __block_P$WU7W(87);
            {
                __statement_WMcoz7(371);
                throw new Error('List.populateRelated(rel, callback) requires a callback function.');
            }
        }
        {
            __statement_WMcoz7(372);
            __extro_AZ6TvY(373, __intro_aPs4zS(373, this).getRelated(rel, function (err, results) {
                __block_P$WU7W(88);
                {
                    __statement_WMcoz7(374);
                    __extro_AZ6TvY(375, __intro_aPs4zS(375, _).each(results, function (data, key) {
                        __block_P$WU7W(89);
                        {
                            __statement_WMcoz7(376);
                            item[key] = data;
                        }
                    }));
                }
                {
                    __statement_WMcoz7(377);
                    __expression_UFQPOZ(378), callback(err, results);
                }
            }, true));
        }
    };
}
{
    __statement_WMcoz7(379);
    options.transform = function (doc, ret) {
        __block_P$WU7W(90);
        if (doc._populatedRelationships) {
            __block_P$WU7W(91);
            {
                __statement_WMcoz7(380);
                __extro_AZ6TvY(381, __intro_aPs4zS(381, _).each(doc._populatedRelationships, function (on, key) {
                    __block_P$WU7W(92);
                    if (__expression_UFQPOZ(382), !(__expression_UFQPOZ(383), on)) {
                        __block_P$WU7W(93);
                        return __expression_UFQPOZ(384);
                    }
                    {
                        __statement_WMcoz7(385);
                        ret[key] = doc[key];
                    }
                }));
            }
        }
    };
}
