// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_cloudinaryimage.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_U0CGMj, __expression_nGbbKl, __block_Yh0ClT;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_U0CGMj = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimage.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_nGbbKl = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimage.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_Yh0ClT = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimage.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_C3ohU_ = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_hkGKpO = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/cloudinaryimage.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_U0CGMj(0);
    var _ = (__expression_nGbbKl(1), require('underscore')), keystone = (__expression_nGbbKl(2), require('../../')), util = (__expression_nGbbKl(3), require('util')), cloudinary = (__expression_nGbbKl(4), require('cloudinary')), utils = (__expression_nGbbKl(5), require('keystone-utils')), super_ = (__expression_nGbbKl(6), require('../field'));
}
function cloudinaryimage(list, path, options) {
    __block_Yh0ClT(0);
    {
        __statement_U0CGMj(7);
        this._underscoreMethods = [
            'format'
        ];
    }
    {
        __statement_U0CGMj(8);
        options.nofilter = true;
    }
    if (options.initial) {
        __block_Yh0ClT(1);
        {
            __statement_U0CGMj(9);
            throw new Error((__expression_nGbbKl(10), (__expression_nGbbKl(11), (__expression_nGbbKl(12), (__expression_nGbbKl(13), (__expression_nGbbKl(14), 'Invalid Configuration\n\n' + 'CloudinaryImage fields (') + list.key) + '.') + (__expression_nGbbKl(15), path)) + ') do not currently support being used as initial fields.\n'));
        }
    }
    {
        __statement_U0CGMj(16);
        __extro_hkGKpO(17, __intro_C3ohU_(17, cloudinaryimage.super_).call(this, list, path, options));
    }
    if (__expression_nGbbKl(18), !__extro_hkGKpO(19, __intro_C3ohU_(19, keystone).get('cloudinary config'))) {
        __block_Yh0ClT(2);
        {
            __statement_U0CGMj(20);
            throw new Error((__expression_nGbbKl(21), (__expression_nGbbKl(22), (__expression_nGbbKl(23), (__expression_nGbbKl(24), (__expression_nGbbKl(25), (__expression_nGbbKl(26), 'Invalid Configuration\n\n' + 'CloudinaryImage fields (') + list.key) + '.') + this.path) + ') require the "cloudinary config" option to be set.\n\n') + 'See http://keystonejs.com/docs/configuration/#cloudinary for more information.\n'));
        }
    }
}
{
    __statement_U0CGMj(27);
    __extro_hkGKpO(28, __intro_C3ohU_(28, util).inherits(cloudinaryimage, super_));
}
{
    __statement_U0CGMj(29);
    cloudinaryimage.prototype.addToSchema = function () {
        __block_Yh0ClT(3);
        {
            __statement_U0CGMj(30);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_U0CGMj(31);
            var paths = this.paths = {
                    public_id: __extro_hkGKpO(32, __intro_C3ohU_(32, this._path).append('.public_id')),
                    version: __extro_hkGKpO(33, __intro_C3ohU_(33, this._path).append('.version')),
                    signature: __extro_hkGKpO(34, __intro_C3ohU_(34, this._path).append('.signature')),
                    format: __extro_hkGKpO(35, __intro_C3ohU_(35, this._path).append('.format')),
                    resource_type: __extro_hkGKpO(36, __intro_C3ohU_(36, this._path).append('.resource_type')),
                    url: __extro_hkGKpO(37, __intro_C3ohU_(37, this._path).append('.url')),
                    width: __extro_hkGKpO(38, __intro_C3ohU_(38, this._path).append('.width')),
                    height: __extro_hkGKpO(39, __intro_C3ohU_(39, this._path).append('.height')),
                    secure_url: __extro_hkGKpO(40, __intro_C3ohU_(40, this._path).append('.secure_url')),
                    exists: __extro_hkGKpO(41, __intro_C3ohU_(41, this._path).append('.exists')),
                    folder: __extro_hkGKpO(42, __intro_C3ohU_(42, this._path).append('.folder')),
                    upload: __extro_hkGKpO(43, __intro_C3ohU_(43, this._path).append('_upload')),
                    action: __extro_hkGKpO(44, __intro_C3ohU_(44, this._path).append('_action')),
                    select: __extro_hkGKpO(45, __intro_C3ohU_(45, this._path).append('_select'))
                };
        }
        {
            __statement_U0CGMj(46);
            var schemaPaths = __extro_hkGKpO(47, __intro_C3ohU_(47, this._path).addTo({}, {
                    public_id: String,
                    version: Number,
                    signature: String,
                    format: String,
                    resource_type: String,
                    url: String,
                    width: Number,
                    height: Number,
                    secure_url: String
                }));
        }
        {
            __statement_U0CGMj(48);
            __extro_hkGKpO(49, __intro_C3ohU_(49, schema).add(schemaPaths));
        }
        {
            __statement_U0CGMj(50);
            var exists = function (item) {
                __block_Yh0ClT(4);
                return __expression_nGbbKl(51), __extro_hkGKpO(54, __intro_C3ohU_(54, item).get(paths.public_id)) ? (__expression_nGbbKl(52), true) : (__expression_nGbbKl(53), false);
            };
        }
        {
            __statement_U0CGMj(55);
            __extro_hkGKpO(56, __intro_C3ohU_(56, __extro_hkGKpO(57, __intro_C3ohU_(57, schema).virtual(paths.exists))).get(function () {
                __block_Yh0ClT(5);
                return __expression_nGbbKl(58), __extro_hkGKpO(59, __intro_C3ohU_(59, schemaMethods.exists).apply(this));
            }));
        }
        {
            __statement_U0CGMj(60);
            var folder = function (item) {
                __block_Yh0ClT(6);
                {
                    __statement_U0CGMj(61);
                    var folderValue = null;
                }
                if (__extro_hkGKpO(62, __intro_C3ohU_(62, keystone).get('cloudinary folders'))) {
                    __block_Yh0ClT(7);
                    if (field.options.folder) {
                        __block_Yh0ClT(8);
                        {
                            __statement_U0CGMj(63);
                            folderValue = field.options.folder;
                        }
                    } else {
                        __block_Yh0ClT(9);
                        {
                            __statement_U0CGMj(64);
                            var folderList = __extro_hkGKpO(67, __intro_C3ohU_(67, keystone).get('cloudinary prefix')) ? (__expression_nGbbKl(65), [
                                    __extro_hkGKpO(68, __intro_C3ohU_(68, keystone).get('cloudinary prefix'))
                                ]) : (__expression_nGbbKl(66), []);
                        }
                        {
                            __statement_U0CGMj(69);
                            __extro_hkGKpO(70, __intro_C3ohU_(70, folderList).push(field.list.path));
                        }
                        {
                            __statement_U0CGMj(71);
                            __extro_hkGKpO(72, __intro_C3ohU_(72, folderList).push(field.path));
                        }
                        {
                            __statement_U0CGMj(73);
                            folderValue = __extro_hkGKpO(74, __intro_C3ohU_(74, folderList).join('/'));
                        }
                    }
                }
                return __expression_nGbbKl(75), folderValue;
            };
        }
        {
            __statement_U0CGMj(76);
            __extro_hkGKpO(77, __intro_C3ohU_(77, __extro_hkGKpO(78, __intro_C3ohU_(78, schema).virtual(paths.folder))).get(function () {
                __block_Yh0ClT(10);
                return __expression_nGbbKl(79), __extro_hkGKpO(80, __intro_C3ohU_(80, schemaMethods.folder).apply(this));
            }));
        }
        {
            __statement_U0CGMj(81);
            var src = function (item, options) {
                __block_Yh0ClT(11);
                if (__expression_nGbbKl(82), !(__expression_nGbbKl(83), exists(item))) {
                    __block_Yh0ClT(12);
                    return __expression_nGbbKl(84), '';
                }
                {
                    __statement_U0CGMj(85);
                    options = (__expression_nGbbKl(88), 'object' === (__expression_nGbbKl(89), typeof options)) ? (__expression_nGbbKl(86), options) : (__expression_nGbbKl(87), {});
                }
                if (__expression_nGbbKl(90), (__expression_nGbbKl(91), !(__expression_nGbbKl(92), ('fetch_format' in (__expression_nGbbKl(93), options)))) && (__expression_nGbbKl(94), __extro_hkGKpO(95, __intro_C3ohU_(95, keystone).get('cloudinary webp')) !== false)) {
                    __block_Yh0ClT(13);
                    {
                        __statement_U0CGMj(96);
                        options.fetch_format = 'auto';
                    }
                }
                if (__expression_nGbbKl(97), (__expression_nGbbKl(98), !(__expression_nGbbKl(99), ('progressive' in (__expression_nGbbKl(100), options)))) && (__expression_nGbbKl(101), __extro_hkGKpO(102, __intro_C3ohU_(102, keystone).get('cloudinary progressive')) !== false)) {
                    __block_Yh0ClT(14);
                    {
                        __statement_U0CGMj(103);
                        options.progressive = true;
                    }
                }
                if (__expression_nGbbKl(104), (__expression_nGbbKl(105), !(__expression_nGbbKl(106), ('secure' in (__expression_nGbbKl(107), options)))) && __extro_hkGKpO(108, __intro_C3ohU_(108, keystone).get('cloudinary secure'))) {
                    __block_Yh0ClT(15);
                    {
                        __statement_U0CGMj(109);
                        options.secure = true;
                    }
                }
                return __expression_nGbbKl(110), __extro_hkGKpO(111, __intro_C3ohU_(111, cloudinary).url((__expression_nGbbKl(112), (__expression_nGbbKl(113), __extro_hkGKpO(114, __intro_C3ohU_(114, item).get(paths.public_id)) + '.') + __extro_hkGKpO(115, __intro_C3ohU_(115, item).get(paths.format))), options));
            };
        }
        {
            __statement_U0CGMj(116);
            var reset = function (item) {
                __block_Yh0ClT(16);
                {
                    __statement_U0CGMj(117);
                    __extro_hkGKpO(118, __intro_C3ohU_(118, item).set(field.path, {
                        public_id: '',
                        version: 0,
                        signature: '',
                        format: '',
                        resource_type: '',
                        url: '',
                        width: 0,
                        height: 0,
                        secure_url: ''
                    }));
                }
            };
        }
        {
            __statement_U0CGMj(119);
            var addSize = function (options, width, height, other) {
                __block_Yh0ClT(17);
                if (__expression_nGbbKl(120), width) {
                    __block_Yh0ClT(18);
                    {
                        __statement_U0CGMj(121);
                        options.width = width;
                    }
                }
                if (__expression_nGbbKl(122), height) {
                    __block_Yh0ClT(19);
                    {
                        __statement_U0CGMj(123);
                        options.height = height;
                    }
                }
                if (__expression_nGbbKl(124), 'object' === (__expression_nGbbKl(125), typeof other)) {
                    __block_Yh0ClT(20);
                    {
                        __statement_U0CGMj(126);
                        __extro_hkGKpO(127, __intro_C3ohU_(127, _).extend(options, other));
                    }
                }
                return __expression_nGbbKl(128), options;
            };
        }
        {
            __statement_U0CGMj(129);
            var schemaMethods = {
                    exists: function () {
                        __block_Yh0ClT(21);
                        return __expression_nGbbKl(130), (__expression_nGbbKl(131), exists(this));
                    },
                    folder: function () {
                        __block_Yh0ClT(22);
                        return __expression_nGbbKl(132), (__expression_nGbbKl(133), folder(this));
                    },
                    src: function (options) {
                        __block_Yh0ClT(23);
                        return __expression_nGbbKl(134), (__expression_nGbbKl(135), src(this, options));
                    },
                    tag: function (options) {
                        __block_Yh0ClT(24);
                        return __expression_nGbbKl(136), (__expression_nGbbKl(139), exists(this)) ? (__expression_nGbbKl(137), __extro_hkGKpO(140, __intro_C3ohU_(140, cloudinary).image(__extro_hkGKpO(141, __intro_C3ohU_(141, this).get(field.path)), options))) : (__expression_nGbbKl(138), '');
                    },
                    scale: function (width, height, options) {
                        __block_Yh0ClT(25);
                        return __expression_nGbbKl(142), (__expression_nGbbKl(143), src(this, (__expression_nGbbKl(144), addSize({
                            crop: 'scale'
                        }, width, height, options))));
                    },
                    fill: function (width, height, options) {
                        __block_Yh0ClT(26);
                        return __expression_nGbbKl(145), (__expression_nGbbKl(146), src(this, (__expression_nGbbKl(147), addSize({
                            crop: 'fill',
                            gravity: 'faces'
                        }, width, height, options))));
                    },
                    lfill: function (width, height, options) {
                        __block_Yh0ClT(27);
                        return __expression_nGbbKl(148), (__expression_nGbbKl(149), src(this, (__expression_nGbbKl(150), addSize({
                            crop: 'lfill',
                            gravity: 'faces'
                        }, width, height, options))));
                    },
                    fit: function (width, height, options) {
                        __block_Yh0ClT(28);
                        return __expression_nGbbKl(151), (__expression_nGbbKl(152), src(this, (__expression_nGbbKl(153), addSize({
                            crop: 'fit'
                        }, width, height, options))));
                    },
                    limit: function (width, height, options) {
                        __block_Yh0ClT(29);
                        return __expression_nGbbKl(154), (__expression_nGbbKl(155), src(this, (__expression_nGbbKl(156), addSize({
                            crop: 'limit'
                        }, width, height, options))));
                    },
                    pad: function (width, height, options) {
                        __block_Yh0ClT(30);
                        return __expression_nGbbKl(157), (__expression_nGbbKl(158), src(this, (__expression_nGbbKl(159), addSize({
                            crop: 'pad'
                        }, width, height, options))));
                    },
                    lpad: function (width, height, options) {
                        __block_Yh0ClT(31);
                        return __expression_nGbbKl(160), (__expression_nGbbKl(161), src(this, (__expression_nGbbKl(162), addSize({
                            crop: 'lpad'
                        }, width, height, options))));
                    },
                    crop: function (width, height, options) {
                        __block_Yh0ClT(32);
                        return __expression_nGbbKl(163), (__expression_nGbbKl(164), src(this, (__expression_nGbbKl(165), addSize({
                            crop: 'crop',
                            gravity: 'faces'
                        }, width, height, options))));
                    },
                    thumbnail: function (width, height, options) {
                        __block_Yh0ClT(33);
                        return __expression_nGbbKl(166), (__expression_nGbbKl(167), src(this, (__expression_nGbbKl(168), addSize({
                            crop: 'thumb',
                            gravity: 'faces'
                        }, width, height, options))));
                    },
                    reset: function () {
                        __block_Yh0ClT(34);
                        {
                            __statement_U0CGMj(169);
                            __expression_nGbbKl(170), reset(this);
                        }
                    },
                    delete: function () {
                        __block_Yh0ClT(35);
                        {
                            __statement_U0CGMj(171);
                            __extro_hkGKpO(172, __intro_C3ohU_(172, cloudinary.uploader).destroy(__extro_hkGKpO(173, __intro_C3ohU_(173, this).get(paths.public_id)), function () {
                                __block_Yh0ClT(36);
                            }));
                        }
                        {
                            __statement_U0CGMj(174);
                            __expression_nGbbKl(175), reset(this);
                        }
                    }
                };
        }
        {
            __statement_U0CGMj(176);
            __extro_hkGKpO(177, __intro_C3ohU_(177, _).each(schemaMethods, function (fn, key) {
                __block_Yh0ClT(37);
                {
                    __statement_U0CGMj(178);
                    __extro_hkGKpO(179, __intro_C3ohU_(179, field).underscoreMethod(key, fn));
                }
            }));
        }
        {
            __statement_U0CGMj(180);
            this.apply = function (item, method) {
                __block_Yh0ClT(38);
                return __expression_nGbbKl(181), __extro_hkGKpO(182, __intro_C3ohU_(182, schemaMethods[method]).apply(item, __extro_hkGKpO(183, __intro_C3ohU_(183, Array.prototype.slice).call(arguments, 2))));
            };
        }
        {
            __statement_U0CGMj(184);
            __extro_hkGKpO(185, __intro_C3ohU_(185, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_U0CGMj(186);
    cloudinaryimage.prototype.format = function (item) {
        __block_Yh0ClT(39);
        return __expression_nGbbKl(187), __extro_hkGKpO(188, __intro_C3ohU_(188, item).get(this.paths.url));
    };
}
{
    __statement_U0CGMj(189);
    cloudinaryimage.prototype.isModified = function (item) {
        __block_Yh0ClT(40);
        return __expression_nGbbKl(190), __extro_hkGKpO(191, __intro_C3ohU_(191, item).isModified(this.paths.url));
    };
}
{
    __statement_U0CGMj(192);
    cloudinaryimage.prototype.validateInput = function (data) {
        __block_Yh0ClT(41);
        return __expression_nGbbKl(193), true;
    };
}
{
    __statement_U0CGMj(194);
    cloudinaryimage.prototype.updateItem = function (item, data) {
        __block_Yh0ClT(42);
        {
            __statement_U0CGMj(195);
            var paths = this.paths;
        }
        {
            __statement_U0CGMj(196);
            var setValue = function (key) {
                __block_Yh0ClT(43);
                if (paths[key]) {
                    __block_Yh0ClT(44);
                    {
                        __statement_U0CGMj(197);
                        var index = __extro_hkGKpO(198, __intro_C3ohU_(198, paths[key]).indexOf('.'));
                    }
                    {
                        __statement_U0CGMj(199);
                        var field = __extro_hkGKpO(200, __intro_C3ohU_(200, paths[key]).substr(0, index));
                    }
                    if (__expression_nGbbKl(201), (__expression_nGbbKl(202), data[field] && data[field][key]) && (__expression_nGbbKl(203), data[field][key] != __extro_hkGKpO(204, __intro_C3ohU_(204, item).get(paths[key])))) {
                        __block_Yh0ClT(45);
                        {
                            __statement_U0CGMj(205);
                            __extro_hkGKpO(206, __intro_C3ohU_(206, item).set(paths[key], (__expression_nGbbKl(207), data[field][key] || null)));
                        }
                    }
                }
            };
        }
        {
            __statement_U0CGMj(208);
            __extro_hkGKpO(209, __intro_C3ohU_(209, _).each([
                'public_id',
                'version',
                'signature',
                'format',
                'resource_type',
                'url',
                'width',
                'height',
                'secure_url'
            ], setValue));
        }
    };
}
{
    __statement_U0CGMj(210);
    cloudinaryimage.prototype.getRequestHandler = function (item, req, paths, callback) {
        __block_Yh0ClT(46);
        {
            __statement_U0CGMj(211);
            var field = this;
        }
        if (__extro_hkGKpO(212, __intro_C3ohU_(212, utils).isFunction(paths))) {
            __block_Yh0ClT(47);
            {
                __statement_U0CGMj(213);
                callback = paths;
            }
            {
                __statement_U0CGMj(214);
                paths = field.paths;
            }
        } else if (__expression_nGbbKl(215), !(__expression_nGbbKl(216), paths)) {
            __block_Yh0ClT(48);
            {
                __statement_U0CGMj(217);
                paths = field.paths;
            }
        }
        {
            __statement_U0CGMj(218);
            callback = (__expression_nGbbKl(219), (__expression_nGbbKl(220), callback) || function () {
                __block_Yh0ClT(49);
            });
        }
        return __expression_nGbbKl(221), function () {
            __block_Yh0ClT(50);
            if (req.body) {
                __block_Yh0ClT(51);
                {
                    __statement_U0CGMj(222);
                    var action = req.body[paths.action];
                }
                if (__extro_hkGKpO(223, __intro_C3ohU_(223, /^(delete|reset)$/).test(action))) {
                    __block_Yh0ClT(52);
                    {
                        __statement_U0CGMj(224);
                        __extro_hkGKpO(225, __intro_C3ohU_(225, field).apply(item, action));
                    }
                }
            }
            if (__expression_nGbbKl(226), req.body && req.body[paths.select]) {
                __block_Yh0ClT(53);
                {
                    __statement_U0CGMj(227);
                    __extro_hkGKpO(228, __intro_C3ohU_(228, cloudinary.api).resource(req.body[paths.select], function (result) {
                        __block_Yh0ClT(54);
                        if (result.error) {
                            __block_Yh0ClT(55);
                            {
                                __statement_U0CGMj(229);
                                __expression_nGbbKl(230), callback(result.error);
                            }
                        } else {
                            __block_Yh0ClT(56);
                            {
                                __statement_U0CGMj(231);
                                __extro_hkGKpO(232, __intro_C3ohU_(232, item).set(field.path, result));
                            }
                            {
                                __statement_U0CGMj(233);
                                __expression_nGbbKl(234), callback();
                            }
                        }
                    }));
                }
            } else if (__expression_nGbbKl(235), (__expression_nGbbKl(236), req.files && req.files[paths.upload]) && req.files[paths.upload].size) {
                __block_Yh0ClT(57);
                {
                    __statement_U0CGMj(237);
                    var tp = (__expression_nGbbKl(238), __extro_hkGKpO(239, __intro_C3ohU_(239, keystone).get('cloudinary prefix')) || '');
                }
                if (tp.length) {
                    __block_Yh0ClT(58);
                    {
                        __statement_U0CGMj(240);
                        tp += '_';
                    }
                }
                {
                    __statement_U0CGMj(241);
                    var uploadOptions = {
                            tags: [
                                (__expression_nGbbKl(242), (__expression_nGbbKl(243), (__expression_nGbbKl(244), (__expression_nGbbKl(245), tp) + field.list.path) + '_') + field.path),
                                (__expression_nGbbKl(246), (__expression_nGbbKl(247), (__expression_nGbbKl(248), (__expression_nGbbKl(249), (__expression_nGbbKl(250), (__expression_nGbbKl(251), tp) + field.list.path) + '_') + field.path) + '_') + item.id)
                            ]
                        };
                }
                if (__extro_hkGKpO(252, __intro_C3ohU_(252, item).get(field.paths.folder))) {
                    __block_Yh0ClT(59);
                    {
                        __statement_U0CGMj(253);
                        uploadOptions.folder = __extro_hkGKpO(254, __intro_C3ohU_(254, item).get(field.paths.folder));
                    }
                }
                if (__extro_hkGKpO(255, __intro_C3ohU_(255, keystone).get('cloudinary prefix'))) {
                    __block_Yh0ClT(60);
                    {
                        __statement_U0CGMj(256);
                        __extro_hkGKpO(257, __intro_C3ohU_(257, uploadOptions.tags).push(__extro_hkGKpO(258, __intro_C3ohU_(258, keystone).get('cloudinary prefix'))));
                    }
                }
                if (__expression_nGbbKl(259), __extro_hkGKpO(260, __intro_C3ohU_(260, keystone).get('env')) !== 'production') {
                    __block_Yh0ClT(61);
                    {
                        __statement_U0CGMj(261);
                        __extro_hkGKpO(262, __intro_C3ohU_(262, uploadOptions.tags).push((__expression_nGbbKl(263), (__expression_nGbbKl(264), tp) + 'dev')));
                    }
                }
                if (field.options.publicID) {
                    __block_Yh0ClT(62);
                    {
                        __statement_U0CGMj(265);
                        var publicIdValue = __extro_hkGKpO(266, __intro_C3ohU_(266, item).get(field.options.publicID));
                    }
                    if (__expression_nGbbKl(267), publicIdValue) {
                        __block_Yh0ClT(63);
                        {
                            __statement_U0CGMj(268);
                            uploadOptions.public_id = publicIdValue;
                        }
                    }
                }
                if (__expression_nGbbKl(269), field.options.autoCleanup && __extro_hkGKpO(270, __intro_C3ohU_(270, item).get(field.paths.exists))) {
                    __block_Yh0ClT(64);
                    {
                        __statement_U0CGMj(271);
                        __extro_hkGKpO(272, __intro_C3ohU_(272, field).apply(item, 'delete'));
                    }
                }
                {
                    __statement_U0CGMj(273);
                    __extro_hkGKpO(274, __intro_C3ohU_(274, cloudinary.uploader).upload(req.files[paths.upload].path, function (result) {
                        __block_Yh0ClT(65);
                        if (result.error) {
                            __block_Yh0ClT(66);
                            {
                                __statement_U0CGMj(275);
                                __expression_nGbbKl(276), callback(result.error);
                            }
                        } else {
                            __block_Yh0ClT(67);
                            {
                                __statement_U0CGMj(277);
                                __extro_hkGKpO(278, __intro_C3ohU_(278, item).set(field.path, result));
                            }
                            {
                                __statement_U0CGMj(279);
                                __expression_nGbbKl(280), callback();
                            }
                        }
                    }, uploadOptions));
                }
            } else {
                __block_Yh0ClT(68);
                {
                    __statement_U0CGMj(281);
                    __expression_nGbbKl(282), callback();
                }
            }
        };
    };
}
{
    __statement_U0CGMj(283);
    cloudinaryimage.prototype.handleRequest = function (item, req, paths, callback) {
        __block_Yh0ClT(69);
        {
            __statement_U0CGMj(284);
            __expression_nGbbKl(285), __extro_hkGKpO(286, __intro_C3ohU_(286, this).getRequestHandler(item, req, paths, callback))();
        }
    };
}
{
    __statement_U0CGMj(287);
    exports = module.exports = cloudinaryimage;
}
