// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_index.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_H0KPpk, __expression_eOHbWY, __block_XBpbDz;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_H0KPpk = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/index.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_eOHbWY = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/index.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_XBpbDz = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/index.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_WPpkoM = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_Dt3QkO = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/index.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_H0KPpk(0);
    exports.AzureFile = (__expression_eOHbWY(1), require('./azurefile'));
}
{
    __statement_H0KPpk(2);
    exports.Boolean = (__expression_eOHbWY(3), require('./boolean'));
}
{
    __statement_H0KPpk(4);
    exports.CloudinaryImage = (__expression_eOHbWY(5), require('./cloudinaryimage'));
}
{
    __statement_H0KPpk(6);
    exports.CloudinaryImages = (__expression_eOHbWY(7), require('./cloudinaryimages'));
}
{
    __statement_H0KPpk(8);
    exports.Code = (__expression_eOHbWY(9), require('./code'));
}
{
    __statement_H0KPpk(10);
    exports.Color = (__expression_eOHbWY(11), require('./color'));
}
{
    __statement_H0KPpk(12);
    exports.Date = (__expression_eOHbWY(13), require('./date'));
}
{
    __statement_H0KPpk(14);
    exports.Datetime = (__expression_eOHbWY(15), require('./datetime'));
}
{
    __statement_H0KPpk(16);
    exports.Email = (__expression_eOHbWY(17), require('./email'));
}
{
    __statement_H0KPpk(18);
    exports.Embedly = (__expression_eOHbWY(19), require('./embedly'));
}
{
    __statement_H0KPpk(20);
    exports.Html = (__expression_eOHbWY(21), require('./html'));
}
{
    __statement_H0KPpk(22);
    exports.Key = (__expression_eOHbWY(23), require('./key'));
}
{
    __statement_H0KPpk(24);
    exports.LocalFile = (__expression_eOHbWY(25), require('./localfile'));
}
{
    __statement_H0KPpk(26);
    exports.Location = (__expression_eOHbWY(27), require('./location'));
}
{
    __statement_H0KPpk(28);
    exports.Markdown = (__expression_eOHbWY(29), require('./markdown'));
}
{
    __statement_H0KPpk(30);
    exports.Money = (__expression_eOHbWY(31), require('./money'));
}
{
    __statement_H0KPpk(32);
    exports.Name = (__expression_eOHbWY(33), require('./name'));
}
{
    __statement_H0KPpk(34);
    exports.Number = (__expression_eOHbWY(35), require('./number'));
}
{
    __statement_H0KPpk(36);
    exports.Password = (__expression_eOHbWY(37), require('./password'));
}
{
    __statement_H0KPpk(38);
    exports.Relationship = (__expression_eOHbWY(39), require('./relationship'));
}
{
    __statement_H0KPpk(40);
    exports.S3File = (__expression_eOHbWY(41), require('./s3file'));
}
{
    __statement_H0KPpk(42);
    exports.Select = (__expression_eOHbWY(43), require('./select'));
}
{
    __statement_H0KPpk(44);
    exports.Text = (__expression_eOHbWY(45), require('./text'));
}
{
    __statement_H0KPpk(46);
    exports.Textarea = (__expression_eOHbWY(47), require('./textarea'));
}
{
    __statement_H0KPpk(48);
    exports.Url = (__expression_eOHbWY(49), require('./url'));
}
{
    __statement_H0KPpk(50);
    exports.LocalFiles = (__expression_eOHbWY(51), require('./localfiles'));
}
