// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_core_mount.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_u7Fa0T, __expression_jVSfEU, __block_bh9T1W;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_u7Fa0T = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/mount.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_jVSfEU = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/mount.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_bh9T1W = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/mount.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_Y9R3RP = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_UAhg37 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/core/mount.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_u7Fa0T(0);
    var _ = (__expression_jVSfEU(1), require('underscore')), express = (__expression_jVSfEU(2), require('express')), path = (__expression_jVSfEU(3), require('path')), utils = (__expression_jVSfEU(4), require('keystone-utils'));
}
{
    __statement_u7Fa0T(5);
    var dashes = '\n------------------------------------------------\n';
}
function mount(mountPath, parentApp, events) {
    __block_bh9T1W(0);
    if (__expression_jVSfEU(6), !this.app) {
        __block_bh9T1W(1);
        {
            __statement_u7Fa0T(7);
            __extro_UAhg37(8, __intro_Y9R3RP(8, console).error('\nKeystoneJS Initialisaton Error:\n\napp must be initialised. Call keystone.init() or keystone.connect(new Express()) first.\n'));
        }
        {
            __statement_u7Fa0T(9);
            __extro_UAhg37(10, __intro_Y9R3RP(10, process).exit(1));
        }
    }
    {
        __statement_u7Fa0T(11);
        var keystone = this, app = this.app;
    }
    {
        __statement_u7Fa0T(12);
        this.nativeApp = true;
    }
    if (__expression_jVSfEU(13), !__extro_UAhg37(14, __intro_Y9R3RP(14, this).get('mongo'))) {
        __block_bh9T1W(2);
        {
            __statement_u7Fa0T(15);
            var dbName = (__expression_jVSfEU(16), __extro_UAhg37(17, __intro_Y9R3RP(17, this).get('db name')) || __extro_UAhg37(18, __intro_Y9R3RP(18, utils).slug(__extro_UAhg37(19, __intro_Y9R3RP(19, this).get('name')))));
        }
        {
            __statement_u7Fa0T(20);
            var dbUrl = (__expression_jVSfEU(21), (__expression_jVSfEU(22), (__expression_jVSfEU(23), (__expression_jVSfEU(24), process.env.MONGO_URI || process.env.MONGO_URL) || process.env.MONGOLAB_URI) || process.env.MONGOLAB_URL) || (__expression_jVSfEU(25), (__expression_jVSfEU(26), process.env.OPENSHIFT_MONGODB_DB_URL || 'mongodb://localhost/') + (__expression_jVSfEU(27), dbName)));
        }
        {
            __statement_u7Fa0T(28);
            __extro_UAhg37(29, __intro_Y9R3RP(29, this).set('mongo', dbUrl));
        }
    }
    if (__expression_jVSfEU(30), !__extro_UAhg37(31, __intro_Y9R3RP(31, this).get('cookie secret'))) {
        __block_bh9T1W(3);
        {
            __statement_u7Fa0T(32);
            __extro_UAhg37(33, __intro_Y9R3RP(33, console).error('\nKeystoneJS Configuration Error:\n\nPlease provide a `cookie secret` value for session encryption.\n'));
        }
        {
            __statement_u7Fa0T(34);
            __extro_UAhg37(35, __intro_Y9R3RP(35, process).exit(1));
        }
    }
    {
        __statement_u7Fa0T(36);
        var sessionOptions = __extro_UAhg37(37, __intro_Y9R3RP(37, this).get('session options'));
    }
    if (__expression_jVSfEU(38), !__extro_UAhg37(39, __intro_Y9R3RP(39, _).isObject(sessionOptions))) {
        __block_bh9T1W(4);
        {
            __statement_u7Fa0T(40);
            sessionOptions = {};
        }
    }
    if (__expression_jVSfEU(41), !sessionOptions.key) {
        __block_bh9T1W(5);
        {
            __statement_u7Fa0T(42);
            sessionOptions.key = 'keystone.sid';
        }
    }
    {
        __statement_u7Fa0T(43);
        sessionOptions.cookieParser = __extro_UAhg37(44, __intro_Y9R3RP(44, express).cookieParser(__extro_UAhg37(45, __intro_Y9R3RP(45, this).get('cookie secret'))));
    }
    {
        __statement_u7Fa0T(46);
        var sessionStore = __extro_UAhg37(47, __intro_Y9R3RP(47, this).get('session store'));
    }
    if (__expression_jVSfEU(48), sessionStore) {
        __block_bh9T1W(6);
        {
            __statement_u7Fa0T(49);
            var sessionStoreOptions = (__expression_jVSfEU(50), __extro_UAhg37(51, __intro_Y9R3RP(51, this).get('session store options')) || {});
        }
        switch (__expression_jVSfEU(52), sessionStore) {
        case 'mongo': {
                __block_bh9T1W(7);
                {
                    __statement_u7Fa0T(53);
                    sessionStore = 'connect-mongo';
                }
            }
        case 'connect-mongo': {
                __block_bh9T1W(8);
                {
                    __statement_u7Fa0T(54);
                    __extro_UAhg37(55, __intro_Y9R3RP(55, _).defaults(sessionStoreOptions, {
                        collection: 'app_sessions',
                        url: __extro_UAhg37(56, __intro_Y9R3RP(56, this).get('mongo'))
                    }));
                }
                break;
            }
        case 'connect-mongostore': {
                __block_bh9T1W(9);
                {
                    __statement_u7Fa0T(57);
                    __extro_UAhg37(58, __intro_Y9R3RP(58, _).defaults(sessionStoreOptions, {
                        collection: 'app_sessions'
                    }));
                }
                if (__expression_jVSfEU(59), !sessionStoreOptions.db) {
                    __block_bh9T1W(10);
                    {
                        __statement_u7Fa0T(60);
                        __extro_UAhg37(61, __intro_Y9R3RP(61, console).error((__expression_jVSfEU(62), (__expression_jVSfEU(63), (__expression_jVSfEU(64), (__expression_jVSfEU(65), (__expression_jVSfEU(66), '\nERROR: ' + (__expression_jVSfEU(67), sessionStore)) + ' requires `session store options` to be set.') + '\n') + '\nSee http://localhost:8080/docs/configuration#options-database for details.') + '\n')));
                    }
                    {
                        __statement_u7Fa0T(68);
                        __extro_UAhg37(69, __intro_Y9R3RP(69, process).exit(1));
                    }
                }
                break;
            }
        case 'redis': {
                __block_bh9T1W(11);
                {
                    __statement_u7Fa0T(70);
                    sessionStore = 'connect-redis';
                }
            }
        case 'connect-redis': {
                __block_bh9T1W(12);
                break;
            }
        default: {
                __block_bh9T1W(13);
                {
                    __statement_u7Fa0T(71);
                    __extro_UAhg37(72, __intro_Y9R3RP(72, console).error((__expression_jVSfEU(73), (__expression_jVSfEU(74), (__expression_jVSfEU(75), (__expression_jVSfEU(76), (__expression_jVSfEU(77), '\nERROR: unsupported session store ' + (__expression_jVSfEU(78), sessionStore)) + '.') + '\n') + '\nSee http://localhost:8080/docs/configuration#options-database for details.') + '\n')));
                }
                {
                    __statement_u7Fa0T(79);
                    __extro_UAhg37(80, __intro_Y9R3RP(80, process).exit(1));
                }
                break;
            }
        }
        try {
            __block_bh9T1W(14);
            {
                __statement_u7Fa0T(81);
                var _SessionStore = (__expression_jVSfEU(82), (__expression_jVSfEU(83), require(sessionStore))(express));
            }
            {
                __statement_u7Fa0T(84);
                sessionOptions.store = new _SessionStore(sessionStoreOptions);
            }
        } catch (e) {
            __block_bh9T1W(15);
            if (__expression_jVSfEU(85), e.code === 'MODULE_NOT_FOUND') {
                __block_bh9T1W(16);
                {
                    __statement_u7Fa0T(86);
                    var installName = (__expression_jVSfEU(89), (__expression_jVSfEU(90), sessionStore) === 'connect-redis') ? (__expression_jVSfEU(87), (__expression_jVSfEU(91), (__expression_jVSfEU(92), sessionStore) + '@1.4.7')) : (__expression_jVSfEU(88), sessionStore);
                }
                {
                    __statement_u7Fa0T(93);
                    __extro_UAhg37(94, __intro_Y9R3RP(94, console).error((__expression_jVSfEU(95), (__expression_jVSfEU(96), (__expression_jVSfEU(97), (__expression_jVSfEU(98), (__expression_jVSfEU(99), (__expression_jVSfEU(100), (__expression_jVSfEU(101), (__expression_jVSfEU(102), (__expression_jVSfEU(103), '\nERROR: ' + (__expression_jVSfEU(104), sessionStore)) + ' not found.\n') + '\nPlease install ') + (__expression_jVSfEU(105), sessionStore)) + ' from npm to use it as a `session store` option.') + '\nYou can do this by running "npm install ') + (__expression_jVSfEU(106), installName)) + ' --save".') + '\n')));
                }
                {
                    __statement_u7Fa0T(107);
                    __extro_UAhg37(108, __intro_Y9R3RP(108, process).exit(1));
                }
            } else {
                __block_bh9T1W(17);
                {
                    __statement_u7Fa0T(109);
                    throw e;
                }
            }
        }
    }
    {
        __statement_u7Fa0T(110);
        __extro_UAhg37(111, __intro_Y9R3RP(111, this).set('session options', sessionOptions));
    }
    if (__expression_jVSfEU(112), arguments.length === 1) {
        __block_bh9T1W(18);
        {
            __statement_u7Fa0T(113);
            events = arguments[0];
        }
        {
            __statement_u7Fa0T(114);
            mountPath = null;
        }
    }
    if (__expression_jVSfEU(115), 'function' === (__expression_jVSfEU(116), typeof events)) {
        __block_bh9T1W(19);
        {
            __statement_u7Fa0T(117);
            events = {
                onMount: events
            };
        }
    }
    if (__expression_jVSfEU(118), !(__expression_jVSfEU(119), events)) {
        __block_bh9T1W(20);
        {
            __statement_u7Fa0T(120);
            events = {};
        }
    }
    if (__expression_jVSfEU(121), mountPath) {
        __block_bh9T1W(21);
        {
            __statement_u7Fa0T(122);
            __extro_UAhg37(123, __intro_Y9R3RP(123, parentApp).all(/^\/keystone($|\/*)/, function (req, res, next) {
                __block_bh9T1W(22);
                {
                    __statement_u7Fa0T(124);
                    req.url = (__expression_jVSfEU(125), (__expression_jVSfEU(126), mountPath) + req.url);
                }
                {
                    __statement_u7Fa0T(127);
                    __expression_jVSfEU(128), next();
                }
            }));
        }
        {
            __statement_u7Fa0T(129);
            __extro_UAhg37(130, __intro_Y9R3RP(130, parentApp).use(mountPath, app));
        }
    }
    if (__extro_UAhg37(131, __intro_Y9R3RP(131, this).get('custom engine'))) {
        __block_bh9T1W(23);
        {
            __statement_u7Fa0T(132);
            __extro_UAhg37(133, __intro_Y9R3RP(133, app).engine(__extro_UAhg37(134, __intro_Y9R3RP(134, this).get('view engine')), __extro_UAhg37(135, __intro_Y9R3RP(135, this).get('custom engine'))));
        }
    }
    {
        __statement_u7Fa0T(136);
        __extro_UAhg37(137, __intro_Y9R3RP(137, app).set('views', (__expression_jVSfEU(138), __extro_UAhg37(139, __intro_Y9R3RP(139, this).getPath('views')) || (__expression_jVSfEU(140), path.sep + 'views'))));
    }
    {
        __statement_u7Fa0T(141);
        __extro_UAhg37(142, __intro_Y9R3RP(142, app).set('view engine', __extro_UAhg37(143, __intro_Y9R3RP(143, this).get('view engine'))));
    }
    if (__extro_UAhg37(144, __intro_Y9R3RP(144, utils).isObject(__extro_UAhg37(145, __intro_Y9R3RP(145, this).get('locals'))))) {
        __block_bh9T1W(24);
        {
            __statement_u7Fa0T(146);
            __extro_UAhg37(147, __intro_Y9R3RP(147, _).extend(app.locals, __extro_UAhg37(148, __intro_Y9R3RP(148, this).get('locals'))));
        }
    }
    if (__expression_jVSfEU(149), __extro_UAhg37(150, __intro_Y9R3RP(150, this).get('env')) !== 'production') {
        __block_bh9T1W(25);
        {
            __statement_u7Fa0T(151);
            app.locals.pretty = true;
        }
    }
    {
        __statement_u7Fa0T(152);
        __extro_UAhg37(153, __intro_Y9R3RP(153, app).set('view cache', (__expression_jVSfEU(156), __extro_UAhg37(157, __intro_Y9R3RP(157, this).get('env')) === 'production') ? (__expression_jVSfEU(154), true) : (__expression_jVSfEU(155), false)));
    }
    if (__expression_jVSfEU(158), __extro_UAhg37(159, __intro_Y9R3RP(159, this).get('view cache')) !== (__expression_jVSfEU(160), undefined)) {
        __block_bh9T1W(26);
        {
            __statement_u7Fa0T(161);
            __extro_UAhg37(162, __intro_Y9R3RP(162, app).set('view cache', __extro_UAhg37(163, __intro_Y9R3RP(163, this).get('view cache'))));
        }
    }
    if (__extro_UAhg37(164, __intro_Y9R3RP(164, this).get('compress'))) {
        __block_bh9T1W(27);
        {
            __statement_u7Fa0T(165);
            __extro_UAhg37(166, __intro_Y9R3RP(166, app).use(__extro_UAhg37(167, __intro_Y9R3RP(167, express).compress())));
        }
    }
    if (__extro_UAhg37(168, __intro_Y9R3RP(168, this).get('favico'))) {
        __block_bh9T1W(28);
        {
            __statement_u7Fa0T(169);
            __extro_UAhg37(170, __intro_Y9R3RP(170, app).use(__extro_UAhg37(171, __intro_Y9R3RP(171, express).favicon(__extro_UAhg37(172, __intro_Y9R3RP(172, this).getPath('favico'))))));
        }
    }
    if (__extro_UAhg37(173, __intro_Y9R3RP(173, this).get('less'))) {
        __block_bh9T1W(29);
        {
            __statement_u7Fa0T(174);
            __extro_UAhg37(175, __intro_Y9R3RP(175, app).use((__expression_jVSfEU(176), (__expression_jVSfEU(177), require('less-middleware'))(__extro_UAhg37(178, __intro_Y9R3RP(178, this).getPath('less'))))));
        }
    }
    if (__extro_UAhg37(179, __intro_Y9R3RP(179, this).get('sass'))) {
        __block_bh9T1W(30);
        {
            __statement_u7Fa0T(180);
            var sass;
        }
        try {
            __block_bh9T1W(31);
            {
                __statement_u7Fa0T(181);
                sass = (__expression_jVSfEU(182), require('node-sass'));
            }
        } catch (e) {
            __block_bh9T1W(32);
            if (__expression_jVSfEU(183), e.code === 'MODULE_NOT_FOUND') {
                __block_bh9T1W(33);
                {
                    __statement_u7Fa0T(184);
                    __extro_UAhg37(185, __intro_Y9R3RP(185, console).error((__expression_jVSfEU(186), (__expression_jVSfEU(187), '\nERROR: node-sass not found.\n' + '\nPlease install the node-sass from npm to use the `sass` option.') + '\nYou can do this by running "npm install node-sass --save".\n')));
                }
                {
                    __statement_u7Fa0T(188);
                    __extro_UAhg37(189, __intro_Y9R3RP(189, process).exit(1));
                }
            } else {
                __block_bh9T1W(34);
                {
                    __statement_u7Fa0T(190);
                    throw e;
                }
            }
        }
        {
            __statement_u7Fa0T(191);
            __extro_UAhg37(192, __intro_Y9R3RP(192, app).use(__extro_UAhg37(193, __intro_Y9R3RP(193, sass).middleware({
                src: __extro_UAhg37(194, __intro_Y9R3RP(194, this).getPath('sass')),
                dest: __extro_UAhg37(195, __intro_Y9R3RP(195, this).getPath('sass')),
                outputStyle: (__expression_jVSfEU(198), __extro_UAhg37(199, __intro_Y9R3RP(199, this).get('env')) === 'production') ? (__expression_jVSfEU(196), 'compressed') : (__expression_jVSfEU(197), 'nested')
            }))));
        }
    }
    {
        __statement_u7Fa0T(200);
        var staticPaths = __extro_UAhg37(201, __intro_Y9R3RP(201, this).get('static'));
    }
    if (__extro_UAhg37(202, __intro_Y9R3RP(202, _).isString(staticPaths))) {
        __block_bh9T1W(35);
        {
            __statement_u7Fa0T(203);
            staticPaths = [
                staticPaths
            ];
        }
    }
    if (__extro_UAhg37(204, __intro_Y9R3RP(204, _).isArray(staticPaths))) {
        __block_bh9T1W(36);
        {
            __statement_u7Fa0T(205);
            __extro_UAhg37(206, __intro_Y9R3RP(206, _).each(staticPaths, function (value) {
                __block_bh9T1W(37);
                {
                    __statement_u7Fa0T(207);
                    __extro_UAhg37(208, __intro_Y9R3RP(208, app).use(__extro_UAhg37(209, __intro_Y9R3RP(209, express).static(__extro_UAhg37(210, __intro_Y9R3RP(210, this).expandPath(value))))));
                }
            }, this));
        }
    }
    if (__expression_jVSfEU(211), !__extro_UAhg37(212, __intro_Y9R3RP(212, this).get('headless'))) {
        __block_bh9T1W(38);
        {
            __statement_u7Fa0T(213);
            __extro_UAhg37(214, __intro_Y9R3RP(214, this).static(app));
        }
    }
    if (__extro_UAhg37(215, __intro_Y9R3RP(215, this).get('logger'))) {
        __block_bh9T1W(39);
        {
            __statement_u7Fa0T(216);
            __extro_UAhg37(217, __intro_Y9R3RP(217, app).use(__extro_UAhg37(218, __intro_Y9R3RP(218, express).logger(__extro_UAhg37(219, __intro_Y9R3RP(219, this).get('logger'))))));
        }
    }
    if (__extro_UAhg37(220, __intro_Y9R3RP(220, this).get('file limit'))) {
        __block_bh9T1W(40);
        {
            __statement_u7Fa0T(221);
            __extro_UAhg37(222, __intro_Y9R3RP(222, app).use(__extro_UAhg37(223, __intro_Y9R3RP(223, express).limit(__extro_UAhg37(224, __intro_Y9R3RP(224, this).get('file limit'))))));
        }
    }
    {
        __statement_u7Fa0T(225);
        __extro_UAhg37(226, __intro_Y9R3RP(226, app).use(__extro_UAhg37(227, __intro_Y9R3RP(227, express).bodyParser())));
    }
    {
        __statement_u7Fa0T(228);
        __extro_UAhg37(229, __intro_Y9R3RP(229, app).use(__extro_UAhg37(230, __intro_Y9R3RP(230, express).methodOverride())));
    }
    {
        __statement_u7Fa0T(231);
        __extro_UAhg37(232, __intro_Y9R3RP(232, app).use(sessionOptions.cookieParser));
    }
    {
        __statement_u7Fa0T(233);
        __extro_UAhg37(234, __intro_Y9R3RP(234, app).use(__extro_UAhg37(235, __intro_Y9R3RP(235, express).session(sessionOptions))));
    }
    {
        __statement_u7Fa0T(236);
        __extro_UAhg37(237, __intro_Y9R3RP(237, app).use((__expression_jVSfEU(238), (__expression_jVSfEU(239), require('connect-flash'))())));
    }
    if (__expression_jVSfEU(240), __extro_UAhg37(241, __intro_Y9R3RP(241, this).get('session')) === true) {
        __block_bh9T1W(41);
        {
            __statement_u7Fa0T(242);
            __extro_UAhg37(243, __intro_Y9R3RP(243, app).use(this.session.persist));
        }
    } else if (__expression_jVSfEU(244), 'function' === (__expression_jVSfEU(245), typeof __extro_UAhg37(246, __intro_Y9R3RP(246, this).get('session')))) {
        __block_bh9T1W(42);
        {
            __statement_u7Fa0T(247);
            __extro_UAhg37(248, __intro_Y9R3RP(248, app).use(__extro_UAhg37(249, __intro_Y9R3RP(249, this).get('session'))));
        }
    }
    if (__expression_jVSfEU(250), __extro_UAhg37(251, __intro_Y9R3RP(251, this).get('trust proxy')) === true) {
        __block_bh9T1W(43);
        {
            __statement_u7Fa0T(252);
            __extro_UAhg37(253, __intro_Y9R3RP(253, app).enable('trust proxy'));
        }
    } else {
        __block_bh9T1W(44);
        {
            __statement_u7Fa0T(254);
            __extro_UAhg37(255, __intro_Y9R3RP(255, app).disable('trust proxy'));
        }
    }
    if (__extro_UAhg37(256, __intro_Y9R3RP(256, this).get('allowed ip ranges'))) {
        __block_bh9T1W(45);
        if (__expression_jVSfEU(257), !__extro_UAhg37(258, __intro_Y9R3RP(258, app).get('trust proxy'))) {
            __block_bh9T1W(46);
            {
                __statement_u7Fa0T(259);
                __extro_UAhg37(260, __intro_Y9R3RP(260, console).log((__expression_jVSfEU(261), 'KeystoneJS Initialisaton Error:\n\n' + 'to set IP range restrictions the "trust proxy" setting must be enabled.\n\n')));
            }
            {
                __statement_u7Fa0T(262);
                __extro_UAhg37(263, __intro_Y9R3RP(263, process).exit(1));
            }
        }
        {
            __statement_u7Fa0T(264);
            var ipRangeMiddleware = (__expression_jVSfEU(265), (__expression_jVSfEU(266), require('./lib/security/ipRangeRestrict'))(__extro_UAhg37(267, __intro_Y9R3RP(267, this).get('allowed ip ranges')), this.wrapHTMLError));
        }
        {
            __statement_u7Fa0T(268);
            __extro_UAhg37(269, __intro_Y9R3RP(269, this).pre('routes', ipRangeMiddleware));
        }
    }
    {
        __statement_u7Fa0T(270);
        __extro_UAhg37(271, __intro_Y9R3RP(271, this._pre.routes).forEach(function (fn) {
            __block_bh9T1W(47);
            try {
                __block_bh9T1W(48);
                {
                    __statement_u7Fa0T(272);
                    __extro_UAhg37(273, __intro_Y9R3RP(273, app).use(fn));
                }
            } catch (e) {
                __block_bh9T1W(49);
                if (__extro_UAhg37(274, __intro_Y9R3RP(274, keystone).get('logger'))) {
                    __block_bh9T1W(50);
                    {
                        __statement_u7Fa0T(275);
                        __extro_UAhg37(276, __intro_Y9R3RP(276, console).log('Invalid pre-route middleware provided'));
                    }
                }
                {
                    __statement_u7Fa0T(277);
                    throw e;
                }
            }
        }));
    }
    {
        __statement_u7Fa0T(278);
        __extro_UAhg37(279, __intro_Y9R3RP(279, app).use(app.router));
    }
    if (__expression_jVSfEU(280), !__extro_UAhg37(281, __intro_Y9R3RP(281, this).get('headless'))) {
        __block_bh9T1W(51);
        {
            __statement_u7Fa0T(282);
            __extro_UAhg37(283, __intro_Y9R3RP(283, this).routes(app));
        }
    }
    if (__expression_jVSfEU(284), 'function' === (__expression_jVSfEU(285), typeof __extro_UAhg37(286, __intro_Y9R3RP(286, this).get('routes')))) {
        __block_bh9T1W(52);
        {
            __statement_u7Fa0T(287);
            __expression_jVSfEU(288), __extro_UAhg37(289, __intro_Y9R3RP(289, this).get('routes'))(app);
        }
    }
    {
        __statement_u7Fa0T(290);
        var setHandlers = function () {
            __block_bh9T1W(53);
            if (__extro_UAhg37(291, __intro_Y9R3RP(291, Object).keys(keystone._redirects)).length) {
                __block_bh9T1W(54);
                {
                    __statement_u7Fa0T(292);
                    __extro_UAhg37(293, __intro_Y9R3RP(293, app).use(function (req, res, next) {
                        __block_bh9T1W(55);
                        if (keystone._redirects[req.path]) {
                            __block_bh9T1W(56);
                            {
                                __statement_u7Fa0T(294);
                                __extro_UAhg37(295, __intro_Y9R3RP(295, res).redirect(keystone._redirects[req.path]));
                            }
                        } else {
                            __block_bh9T1W(57);
                            {
                                __statement_u7Fa0T(296);
                                __expression_jVSfEU(297), next();
                            }
                        }
                    }));
                }
            }
            {
                __statement_u7Fa0T(298);
                var default404Handler = function (req, res, next) {
                    __block_bh9T1W(58);
                    {
                        __statement_u7Fa0T(299);
                        __extro_UAhg37(300, __intro_Y9R3RP(300, __extro_UAhg37(301, __intro_Y9R3RP(301, res).status(404))).send(__extro_UAhg37(302, __intro_Y9R3RP(302, keystone).wrapHTMLError('Sorry, no page could be found at this address (404)'))));
                    }
                };
            }
            {
                __statement_u7Fa0T(303);
                __extro_UAhg37(304, __intro_Y9R3RP(304, app).use(function (req, res, next) {
                    __block_bh9T1W(59);
                    {
                        __statement_u7Fa0T(305);
                        var err404 = __extro_UAhg37(306, __intro_Y9R3RP(306, keystone).get('404'));
                    }
                    if (__expression_jVSfEU(307), err404) {
                        __block_bh9T1W(60);
                        try {
                            __block_bh9T1W(61);
                            if (__expression_jVSfEU(308), 'function' === (__expression_jVSfEU(309), typeof err404)) {
                                __block_bh9T1W(62);
                                {
                                    __statement_u7Fa0T(310);
                                    __expression_jVSfEU(311), err404(req, res, next);
                                }
                            } else if (__expression_jVSfEU(312), 'string' === (__expression_jVSfEU(313), typeof err404)) {
                                __block_bh9T1W(63);
                                {
                                    __statement_u7Fa0T(314);
                                    __extro_UAhg37(315, __intro_Y9R3RP(315, __extro_UAhg37(316, __intro_Y9R3RP(316, res).status(404))).render(err404));
                                }
                            } else {
                                __block_bh9T1W(64);
                                if (__extro_UAhg37(317, __intro_Y9R3RP(317, keystone).get('logger'))) {
                                    __block_bh9T1W(65);
                                    {
                                        __statement_u7Fa0T(318);
                                        __extro_UAhg37(319, __intro_Y9R3RP(319, console).log((__expression_jVSfEU(320), (__expression_jVSfEU(321), (__expression_jVSfEU(322), (__expression_jVSfEU(323), (__expression_jVSfEU(324), dashes) + 'Error handling 404 (not found): Invalid type (') + (__expression_jVSfEU(325), typeof err404)) + ') for 404 setting.') + (__expression_jVSfEU(326), dashes))));
                                    }
                                }
                                {
                                    __statement_u7Fa0T(327);
                                    __expression_jVSfEU(328), default404Handler(req, res, next);
                                }
                            }
                        } catch (e) {
                            __block_bh9T1W(66);
                            if (__extro_UAhg37(329, __intro_Y9R3RP(329, keystone).get('logger'))) {
                                __block_bh9T1W(67);
                                {
                                    __statement_u7Fa0T(330);
                                    __extro_UAhg37(331, __intro_Y9R3RP(331, console).log((__expression_jVSfEU(332), (__expression_jVSfEU(333), dashes) + 'Error handling 404 (not found):')));
                                }
                                {
                                    __statement_u7Fa0T(334);
                                    __extro_UAhg37(335, __intro_Y9R3RP(335, console).log(e));
                                }
                                {
                                    __statement_u7Fa0T(336);
                                    __extro_UAhg37(337, __intro_Y9R3RP(337, console).log(dashes));
                                }
                            }
                            {
                                __statement_u7Fa0T(338);
                                __expression_jVSfEU(339), default404Handler(req, res, next);
                            }
                        }
                    } else {
                        __block_bh9T1W(68);
                        {
                            __statement_u7Fa0T(340);
                            __expression_jVSfEU(341), default404Handler(req, res, next);
                        }
                    }
                }));
            }
            {
                __statement_u7Fa0T(342);
                var default500Handler = function (err, req, res, next) {
                    __block_bh9T1W(69);
                    if (__extro_UAhg37(343, __intro_Y9R3RP(343, keystone).get('logger'))) {
                        __block_bh9T1W(70);
                        if (__expression_jVSfEU(344), (__expression_jVSfEU(345), err) instanceof (__expression_jVSfEU(346), Error)) {
                            __block_bh9T1W(71);
                            {
                                __statement_u7Fa0T(347);
                                __extro_UAhg37(348, __intro_Y9R3RP(348, console).log((__expression_jVSfEU(349), (__expression_jVSfEU(350), (err.type ? (__expression_jVSfEU(351), (__expression_jVSfEU(353), err.type + ' ')) : (__expression_jVSfEU(352), '')) + 'Error thrown for request: ') + req.url)));
                            }
                        } else {
                            __block_bh9T1W(72);
                            {
                                __statement_u7Fa0T(354);
                                __extro_UAhg37(355, __intro_Y9R3RP(355, console).log((__expression_jVSfEU(356), 'Error thrown for request: ' + req.url)));
                            }
                        }
                        {
                            __statement_u7Fa0T(357);
                            __extro_UAhg37(358, __intro_Y9R3RP(358, console).log((__expression_jVSfEU(359), err.stack || (__expression_jVSfEU(360), err))));
                        }
                    }
                    {
                        __statement_u7Fa0T(361);
                        var msg = '';
                    }
                    if (__expression_jVSfEU(362), __extro_UAhg37(363, __intro_Y9R3RP(363, keystone).get('env')) === 'development') {
                        __block_bh9T1W(73);
                        if (__expression_jVSfEU(364), (__expression_jVSfEU(365), err) instanceof (__expression_jVSfEU(366), Error)) {
                            __block_bh9T1W(74);
                            if (err.type) {
                                __block_bh9T1W(75);
                                {
                                    __statement_u7Fa0T(367);
                                    msg += (__expression_jVSfEU(368), (__expression_jVSfEU(369), '<h2>' + err.type) + '</h2>');
                                }
                            }
                            {
                                __statement_u7Fa0T(370);
                                msg += __extro_UAhg37(371, __intro_Y9R3RP(371, utils).textToHTML(err.message));
                            }
                        } else if (__expression_jVSfEU(372), 'object' === (__expression_jVSfEU(373), typeof err)) {
                            __block_bh9T1W(76);
                            {
                                __statement_u7Fa0T(374);
                                msg += (__expression_jVSfEU(375), (__expression_jVSfEU(376), '<code>' + __extro_UAhg37(377, __intro_Y9R3RP(377, JSON).stringify(err))) + '</code>');
                            }
                        } else if (__expression_jVSfEU(378), err) {
                            __block_bh9T1W(77);
                            {
                                __statement_u7Fa0T(379);
                                msg += err;
                            }
                        }
                    }
                    {
                        __statement_u7Fa0T(380);
                        __extro_UAhg37(381, __intro_Y9R3RP(381, __extro_UAhg37(382, __intro_Y9R3RP(382, res).status(500))).send(__extro_UAhg37(383, __intro_Y9R3RP(383, keystone).wrapHTMLError('Sorry, an error occurred loading the page (500)', msg))));
                    }
                };
            }
            {
                __statement_u7Fa0T(384);
                __extro_UAhg37(385, __intro_Y9R3RP(385, app).use(function (err, req, res, next) {
                    __block_bh9T1W(78);
                    {
                        __statement_u7Fa0T(386);
                        var err500 = __extro_UAhg37(387, __intro_Y9R3RP(387, keystone).get('500'));
                    }
                    if (__expression_jVSfEU(388), err500) {
                        __block_bh9T1W(79);
                        try {
                            __block_bh9T1W(80);
                            if (__expression_jVSfEU(389), 'function' === (__expression_jVSfEU(390), typeof err500)) {
                                __block_bh9T1W(81);
                                {
                                    __statement_u7Fa0T(391);
                                    __expression_jVSfEU(392), err500(err, req, res, next);
                                }
                            } else if (__expression_jVSfEU(393), 'string' === (__expression_jVSfEU(394), typeof err500)) {
                                __block_bh9T1W(82);
                                {
                                    __statement_u7Fa0T(395);
                                    res.locals.err = err;
                                }
                                {
                                    __statement_u7Fa0T(396);
                                    __extro_UAhg37(397, __intro_Y9R3RP(397, __extro_UAhg37(398, __intro_Y9R3RP(398, res).status(500))).render(err500));
                                }
                            } else {
                                __block_bh9T1W(83);
                                if (__extro_UAhg37(399, __intro_Y9R3RP(399, keystone).get('logger'))) {
                                    __block_bh9T1W(84);
                                    {
                                        __statement_u7Fa0T(400);
                                        __extro_UAhg37(401, __intro_Y9R3RP(401, console).log((__expression_jVSfEU(402), (__expression_jVSfEU(403), (__expression_jVSfEU(404), (__expression_jVSfEU(405), (__expression_jVSfEU(406), dashes) + 'Error handling 500 (error): Invalid type (') + (__expression_jVSfEU(407), typeof err500)) + ') for 500 setting.') + (__expression_jVSfEU(408), dashes))));
                                    }
                                }
                                {
                                    __statement_u7Fa0T(409);
                                    __expression_jVSfEU(410), default500Handler(err, req, res, next);
                                }
                            }
                        } catch (e) {
                            __block_bh9T1W(85);
                            if (__extro_UAhg37(411, __intro_Y9R3RP(411, keystone).get('logger'))) {
                                __block_bh9T1W(86);
                                {
                                    __statement_u7Fa0T(412);
                                    __extro_UAhg37(413, __intro_Y9R3RP(413, console).log((__expression_jVSfEU(414), (__expression_jVSfEU(415), dashes) + 'Error handling 500 (error):')));
                                }
                                {
                                    __statement_u7Fa0T(416);
                                    __extro_UAhg37(417, __intro_Y9R3RP(417, console).log(e));
                                }
                                {
                                    __statement_u7Fa0T(418);
                                    __extro_UAhg37(419, __intro_Y9R3RP(419, console).log(dashes));
                                }
                            }
                            {
                                __statement_u7Fa0T(420);
                                __expression_jVSfEU(421), default500Handler(err, req, res, next);
                            }
                        }
                    } else {
                        __block_bh9T1W(87);
                        {
                            __statement_u7Fa0T(422);
                            __expression_jVSfEU(423), default500Handler(err, req, res, next);
                        }
                    }
                }));
            }
        };
    }
    {
        __statement_u7Fa0T(424);
        var mongoConnectionOpen = false;
    }
    if (__extro_UAhg37(425, __intro_Y9R3RP(425, this).get('mongo replica set'))) {
        __block_bh9T1W(88);
        {
            __statement_u7Fa0T(426);
            var replicaData = __extro_UAhg37(427, __intro_Y9R3RP(427, this).get('mongo replica set'));
        }
        {
            __statement_u7Fa0T(428);
            var replica = '';
        }
        {
            __statement_u7Fa0T(429);
            var credentials = (__expression_jVSfEU(432), replicaData.username && replicaData.password) ? (__expression_jVSfEU(430), (__expression_jVSfEU(433), (__expression_jVSfEU(434), (__expression_jVSfEU(435), replicaData.username + ':') + replicaData.password) + '@')) : (__expression_jVSfEU(431), '');
        }
        {
            __statement_u7Fa0T(436);
            __extro_UAhg37(437, __intro_Y9R3RP(437, replicaData.db.servers).forEach(function (server) {
                __block_bh9T1W(89);
                {
                    __statement_u7Fa0T(438);
                    replica += (__expression_jVSfEU(439), (__expression_jVSfEU(440), (__expression_jVSfEU(441), (__expression_jVSfEU(442), (__expression_jVSfEU(443), (__expression_jVSfEU(444), (__expression_jVSfEU(445), 'mongodb://' + (__expression_jVSfEU(446), credentials)) + server['host']) + ':') + server['port']) + '/') + replicaData.db.name) + ',');
                }
            }));
        }
        {
            __statement_u7Fa0T(447);
            var options = {
                    auth: {
                        authSource: replicaData.authSource
                    },
                    replset: {
                        rs_name: replicaData.db.replicaSetOptions.rs_name,
                        readPreference: replicaData.db.replicaSetOptions.readPreference
                    }
                };
        }
        {
            __statement_u7Fa0T(448);
            __extro_UAhg37(449, __intro_Y9R3RP(449, this.mongoose).connect(replica, options));
        }
    } else {
        __block_bh9T1W(90);
        {
            __statement_u7Fa0T(450);
            __extro_UAhg37(451, __intro_Y9R3RP(451, this.mongoose).connect(__extro_UAhg37(452, __intro_Y9R3RP(452, this).get('mongo'))));
        }
    }
    {
        __statement_u7Fa0T(453);
        __extro_UAhg37(454, __intro_Y9R3RP(454, __extro_UAhg37(455, __intro_Y9R3RP(455, this.mongoose.connection).on('error', function (err) {
            __block_bh9T1W(91);
            if (__extro_UAhg37(456, __intro_Y9R3RP(456, keystone).get('logger'))) {
                __block_bh9T1W(92);
                {
                    __statement_u7Fa0T(457);
                    __extro_UAhg37(458, __intro_Y9R3RP(458, console).log('------------------------------------------------'));
                }
                {
                    __statement_u7Fa0T(459);
                    __extro_UAhg37(460, __intro_Y9R3RP(460, console).log('Mongo Error:\n'));
                }
                {
                    __statement_u7Fa0T(461);
                    __extro_UAhg37(462, __intro_Y9R3RP(462, console).log(err));
                }
            }
            if (__expression_jVSfEU(463), mongoConnectionOpen) {
                __block_bh9T1W(93);
                {
                    __statement_u7Fa0T(464);
                    throw new Error('Mongo Error');
                }
            } else {
                __block_bh9T1W(94);
                {
                    __statement_u7Fa0T(465);
                    throw new Error((__expression_jVSfEU(466), (__expression_jVSfEU(467), 'KeystoneJS (' + __extro_UAhg37(468, __intro_Y9R3RP(468, keystone).get('name'))) + ') failed to start'));
                }
            }
        }))).on('open', function () {
            __block_bh9T1W(95);
            {
                __statement_u7Fa0T(469);
                mongoConnectionOpen = true;
            }
            if (__extro_UAhg37(470, __intro_Y9R3RP(470, keystone).get('auto update'))) {
                __block_bh9T1W(96);
                {
                    __statement_u7Fa0T(471);
                    var mounted = function () {
                        __block_bh9T1W(97);
                        {
                            __statement_u7Fa0T(472);
                            __extro_UAhg37(473, __intro_Y9R3RP(473, events).onMount());
                        }
                        {
                            __statement_u7Fa0T(474);
                            __expression_jVSfEU(475), setHandlers();
                        }
                    };
                }
                {
                    __statement_u7Fa0T(476);
                    __extro_UAhg37(477, __intro_Y9R3RP(477, keystone).applyUpdates(mounted));
                }
            } else {
                __block_bh9T1W(98);
                {
                    __statement_u7Fa0T(478);
                    __expression_jVSfEU(479), events.onMount && __extro_UAhg37(480, __intro_Y9R3RP(480, events).onMount());
                }
                {
                    __statement_u7Fa0T(481);
                    __expression_jVSfEU(482), setHandlers();
                }
            }
        }));
    }
}
{
    __statement_u7Fa0T(483);
    module.exports = mount;
}
