// keystone/.coverdebug/_Users_Jed_Development_Keystone_index.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_zRynHy, __expression_SYuMxd, __block_Ix52Ab;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_zRynHy = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/index.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_SYuMxd = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/index.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_Ix52Ab = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/index.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_RMcQoX = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_JLTOgl = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/index.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_zRynHy(0);
    var fs = (__expression_SYuMxd(1), require('fs')), path = (__expression_SYuMxd(2), require('path')), _ = (__expression_SYuMxd(3), require('underscore')), express = (__expression_SYuMxd(4), require('express')), async = (__expression_SYuMxd(5), require('async')), jade = (__expression_SYuMxd(6), require('jade')), moment = (__expression_SYuMxd(7), require('moment')), numeral = (__expression_SYuMxd(8), require('numeral')), cloudinary = (__expression_SYuMxd(9), require('cloudinary')), utils = (__expression_SYuMxd(10), require('keystone-utils'));
}
{
    __statement_zRynHy(11);
    var templateCache = {};
}
{
    __statement_zRynHy(12);
    var moduleRoot = (__expression_SYuMxd(13), function (_rootPath) {
            __block_Ix52Ab(0);
            {
                __statement_zRynHy(14);
                var parts = __extro_JLTOgl(15, __intro_RMcQoX(15, _rootPath).split(path.sep));
            }
            {
                __statement_zRynHy(16);
                __extro_JLTOgl(17, __intro_RMcQoX(17, parts).pop());
            }
            return __expression_SYuMxd(18), __extro_JLTOgl(19, __intro_RMcQoX(19, parts).join(path.sep));
        }(module.parent ? (__expression_SYuMxd(20), module.parent.paths[0]) : (__expression_SYuMxd(21), module.paths[0])));
}
{
    __statement_zRynHy(22);
    var Keystone = function () {
        __block_Ix52Ab(1);
        {
            __statement_zRynHy(23);
            this.lists = {};
        }
        {
            __statement_zRynHy(24);
            this.paths = {};
        }
        {
            __statement_zRynHy(25);
            this._options = {
                'name': 'Keystone',
                'brand': 'Keystone',
                'compress': true,
                'headless': false,
                'logger': 'dev',
                'auto update': false,
                'model prefix': null
            };
        }
        {
            __statement_zRynHy(26);
            this._pre = {
                routes: [],
                render: []
            };
        }
        {
            __statement_zRynHy(27);
            this._redirects = {};
        }
        {
            __statement_zRynHy(28);
            this.express = express;
        }
        {
            __statement_zRynHy(29);
            __extro_JLTOgl(30, __intro_RMcQoX(30, this).set('env', (__expression_SYuMxd(31), process.env.NODE_ENV || 'development')));
        }
        {
            __statement_zRynHy(32);
            __extro_JLTOgl(33, __intro_RMcQoX(33, this).set('port', (__expression_SYuMxd(34), process.env.PORT || process.env.OPENSHIFT_NODEJS_PORT)));
        }
        {
            __statement_zRynHy(35);
            __extro_JLTOgl(36, __intro_RMcQoX(36, this).set('host', (__expression_SYuMxd(37), (__expression_SYuMxd(38), process.env.HOST || process.env.IP) || process.env.OPENSHIFT_NODEJS_IP)));
        }
        {
            __statement_zRynHy(39);
            __extro_JLTOgl(40, __intro_RMcQoX(40, this).set('listen', process.env.LISTEN));
        }
        {
            __statement_zRynHy(41);
            __extro_JLTOgl(42, __intro_RMcQoX(42, this).set('ssl', process.env.SSL));
        }
        {
            __statement_zRynHy(43);
            __extro_JLTOgl(44, __intro_RMcQoX(44, this).set('ssl port', process.env.SSL_PORT));
        }
        {
            __statement_zRynHy(45);
            __extro_JLTOgl(46, __intro_RMcQoX(46, this).set('ssl host', (__expression_SYuMxd(47), process.env.SSL_HOST || process.env.SSL_IP)));
        }
        {
            __statement_zRynHy(48);
            __extro_JLTOgl(49, __intro_RMcQoX(49, this).set('ssl key', process.env.SSL_KEY));
        }
        {
            __statement_zRynHy(50);
            __extro_JLTOgl(51, __intro_RMcQoX(51, this).set('ssl cert', process.env.SSL_CERT));
        }
        {
            __statement_zRynHy(52);
            __extro_JLTOgl(53, __intro_RMcQoX(53, this).set('cookie secret', process.env.COOKIE_SECRET));
        }
        {
            __statement_zRynHy(54);
            __extro_JLTOgl(55, __intro_RMcQoX(55, this).set('cookie signin', (__expression_SYuMxd(58), __extro_JLTOgl(59, __intro_RMcQoX(59, this).get('env')) === 'development') ? (__expression_SYuMxd(56), true) : (__expression_SYuMxd(57), false)));
        }
        {
            __statement_zRynHy(60);
            __extro_JLTOgl(61, __intro_RMcQoX(61, this).set('embedly api key', (__expression_SYuMxd(62), process.env.EMBEDLY_API_KEY || process.env.EMBEDLY_APIKEY)));
        }
        {
            __statement_zRynHy(63);
            __extro_JLTOgl(64, __intro_RMcQoX(64, this).set('mandrill api key', (__expression_SYuMxd(65), process.env.MANDRILL_API_KEY || process.env.MANDRILL_APIKEY)));
        }
        {
            __statement_zRynHy(66);
            __extro_JLTOgl(67, __intro_RMcQoX(67, this).set('mandrill username', process.env.MANDRILL_USERNAME));
        }
        {
            __statement_zRynHy(68);
            __extro_JLTOgl(69, __intro_RMcQoX(69, this).set('google api key', process.env.GOOGLE_BROWSER_KEY));
        }
        {
            __statement_zRynHy(70);
            __extro_JLTOgl(71, __intro_RMcQoX(71, this).set('google server api key', process.env.GOOGLE_SERVER_KEY));
        }
        {
            __statement_zRynHy(72);
            __extro_JLTOgl(73, __intro_RMcQoX(73, this).set('ga property', process.env.GA_PROPERTY));
        }
        {
            __statement_zRynHy(74);
            __extro_JLTOgl(75, __intro_RMcQoX(75, this).set('ga domain', process.env.GA_DOMAIN));
        }
        {
            __statement_zRynHy(76);
            __extro_JLTOgl(77, __intro_RMcQoX(77, this).set('chartbeat property', process.env.CHARTBEAT_PROPERTY));
        }
        {
            __statement_zRynHy(78);
            __extro_JLTOgl(79, __intro_RMcQoX(79, this).set('chartbeat domain', process.env.CHARTBEAT_DOMAIN));
        }
        {
            __statement_zRynHy(80);
            __extro_JLTOgl(81, __intro_RMcQoX(81, this).set('allowed ip ranges', process.env.ALLOWED_IP_RANGES));
        }
        if (__expression_SYuMxd(82), (__expression_SYuMxd(83), process.env.S3_BUCKET && process.env.S3_KEY) && process.env.S3_SECRET) {
            __block_Ix52Ab(2);
            {
                __statement_zRynHy(84);
                __extro_JLTOgl(85, __intro_RMcQoX(85, this).set('s3 config', {
                    bucket: process.env.S3_BUCKET,
                    key: process.env.S3_KEY,
                    secret: process.env.S3_SECRET,
                    region: process.env.S3_REGION
                }));
            }
        }
        if (__expression_SYuMxd(86), process.env.AZURE_STORAGE_ACCOUNT && process.env.AZURE_STORAGE_ACCESS_KEY) {
            __block_Ix52Ab(3);
            {
                __statement_zRynHy(87);
                __extro_JLTOgl(88, __intro_RMcQoX(88, this).set('azurefile config', {
                    account: process.env.AZURE_STORAGE_ACCOUNT,
                    key: process.env.AZURE_STORAGE_ACCESS_KEY
                }));
            }
        }
        if (process.env.CLOUDINARY_URL) {
            __block_Ix52Ab(4);
            {
                __statement_zRynHy(89);
                __extro_JLTOgl(90, __intro_RMcQoX(90, this).set('cloudinary config', true));
            }
        }
        {
            __statement_zRynHy(91);
            this.initAPI = (__expression_SYuMxd(92), (__expression_SYuMxd(93), require('./lib/middleware/initAPI'))(this));
        }
    };
}
{
    __statement_zRynHy(94);
    __extro_JLTOgl(95, __intro_RMcQoX(95, _).extend(Keystone.prototype, (__expression_SYuMxd(96), (__expression_SYuMxd(97), require('./lib/core/options'))(moduleRoot))));
}
{
    __statement_zRynHy(98);
    Keystone.prototype.pre = function (event, fn) {
        __block_Ix52Ab(5);
        if (__expression_SYuMxd(99), !this._pre[event]) {
            __block_Ix52Ab(6);
            {
                __statement_zRynHy(100);
                throw new Error((__expression_SYuMxd(101), (__expression_SYuMxd(102), 'keystone.pre() Error: event ' + (__expression_SYuMxd(103), event)) + ' does not exist.'));
            }
        }
        {
            __statement_zRynHy(104);
            __extro_JLTOgl(105, __intro_RMcQoX(105, this._pre[event]).push(fn));
        }
        return __expression_SYuMxd(106), this;
    };
}
{
    __statement_zRynHy(107);
    Keystone.prototype.prefixModel = function (key) {
        __block_Ix52Ab(7);
        {
            __statement_zRynHy(108);
            var modelPrefix = __extro_JLTOgl(109, __intro_RMcQoX(109, this).get('model prefix'));
        }
        if (__expression_SYuMxd(110), modelPrefix) {
            __block_Ix52Ab(8);
            {
                __statement_zRynHy(111);
                key = (__expression_SYuMxd(112), (__expression_SYuMxd(113), (__expression_SYuMxd(114), modelPrefix) + '_') + (__expression_SYuMxd(115), key));
            }
        }
        return __expression_SYuMxd(116), __extro_JLTOgl(117, __intro_RMcQoX(117, (__expression_SYuMxd(118), require('mongoose/lib/utils'))).toCollectionName(key));
    };
}
{
    __statement_zRynHy(119);
    Keystone.prototype.init = (__expression_SYuMxd(120), require('./lib/core/init'));
}
{
    __statement_zRynHy(121);
    Keystone.prototype.initNav = (__expression_SYuMxd(122), require('./lib/core/initNav'));
}
{
    __statement_zRynHy(123);
    Keystone.prototype.connect = (__expression_SYuMxd(124), require('./lib/core/connect'));
}
{
    __statement_zRynHy(125);
    Keystone.prototype.start = (__expression_SYuMxd(126), require('./lib/core/start'));
}
{
    __statement_zRynHy(127);
    Keystone.prototype.mount = (__expression_SYuMxd(128), require('./lib/core/mount'));
}
{
    __statement_zRynHy(129);
    Keystone.prototype.routes = (__expression_SYuMxd(130), require('./lib/core/routes'));
}
{
    __statement_zRynHy(131);
    Keystone.prototype.static = (__expression_SYuMxd(132), require('./lib/core/static'));
}
{
    __statement_zRynHy(133);
    Keystone.prototype.importer = (__expression_SYuMxd(134), require('./lib/core/importer'));
}
{
    __statement_zRynHy(135);
    Keystone.prototype.createItems = (__expression_SYuMxd(136), require('./lib/core/createItems'));
}
{
    __statement_zRynHy(137);
    Keystone.prototype.redirect = (__expression_SYuMxd(138), require('./lib/core/redirect'));
}
{
    __statement_zRynHy(139);
    Keystone.prototype.list = (__expression_SYuMxd(140), require('./lib/core/list'));
}
{
    __statement_zRynHy(141);
    Keystone.prototype.getOrphanedLists = (__expression_SYuMxd(142), require('./lib/core/getOrphanedLists'));
}
{
    __statement_zRynHy(143);
    Keystone.prototype.bindEmailTestRoutes = (__expression_SYuMxd(144), require('./lib/core/bindEmailTestRoutes'));
}
{
    __statement_zRynHy(145);
    Keystone.prototype.wrapHTMLError = (__expression_SYuMxd(146), require('./lib/core/wrapHTMLError'));
}
{
    __statement_zRynHy(147);
    var keystone = module.exports = exports = new Keystone();
}
{
    __statement_zRynHy(148);
    keystone.utils = utils;
}
{
    __statement_zRynHy(149);
    keystone.content = (__expression_SYuMxd(150), require('./lib/content'));
}
{
    __statement_zRynHy(151);
    keystone.List = (__expression_SYuMxd(152), require('./lib/list'));
}
{
    __statement_zRynHy(153);
    keystone.Field = (__expression_SYuMxd(154), require('./lib/field'));
}
{
    __statement_zRynHy(155);
    keystone.Field.Types = (__expression_SYuMxd(156), require('./lib/fieldTypes'));
}
{
    __statement_zRynHy(157);
    keystone.View = (__expression_SYuMxd(158), require('./lib/view'));
}
{
    __statement_zRynHy(159);
    keystone.Email = (__expression_SYuMxd(160), require('./lib/email'));
}
{
    __statement_zRynHy(161);
    keystone.security = {
        csrf: (__expression_SYuMxd(162), require('./lib/security/csrf'))
    };
}
{
    __statement_zRynHy(163);
    Keystone.prototype.import = function (dirname) {
        __block_Ix52Ab(9);
        {
            __statement_zRynHy(164);
            var initialPath = __extro_JLTOgl(165, __intro_RMcQoX(165, path).join(moduleRoot, dirname));
        }
        {
            __statement_zRynHy(166);
            var doImport = function (fromPath) {
                __block_Ix52Ab(10);
                {
                    __statement_zRynHy(167);
                    var imported = {};
                }
                {
                    __statement_zRynHy(168);
                    __extro_JLTOgl(169, __intro_RMcQoX(169, __extro_JLTOgl(170, __intro_RMcQoX(170, fs).readdirSync(fromPath))).forEach(function (name) {
                        __block_Ix52Ab(11);
                        {
                            __statement_zRynHy(171);
                            var fsPath = __extro_JLTOgl(172, __intro_RMcQoX(172, path).join(fromPath, name)), info = __extro_JLTOgl(173, __intro_RMcQoX(173, fs).statSync(fsPath));
                        }
                        if (__extro_JLTOgl(174, __intro_RMcQoX(174, info).isDirectory())) {
                            __block_Ix52Ab(12);
                            {
                                __statement_zRynHy(175);
                                imported[name] = (__expression_SYuMxd(176), doImport(fsPath));
                            }
                        } else {
                            __block_Ix52Ab(13);
                            {
                                __statement_zRynHy(177);
                                var parts = __extro_JLTOgl(178, __intro_RMcQoX(178, name).split('.'));
                            }
                            {
                                __statement_zRynHy(179);
                                var ext = __extro_JLTOgl(180, __intro_RMcQoX(180, parts).pop());
                            }
                            if (__expression_SYuMxd(181), (__expression_SYuMxd(182), (__expression_SYuMxd(183), ext) === 'js') || (__expression_SYuMxd(184), (__expression_SYuMxd(185), ext) === 'coffee')) {
                                __block_Ix52Ab(14);
                                {
                                    __statement_zRynHy(186);
                                    imported[__extro_JLTOgl(187, __intro_RMcQoX(187, parts).join('-'))] = (__expression_SYuMxd(188), require(fsPath));
                                }
                            }
                        }
                    }));
                }
                return __expression_SYuMxd(189), imported;
            };
        }
        return __expression_SYuMxd(190), (__expression_SYuMxd(191), doImport(initialPath));
    };
}
{
    __statement_zRynHy(192);
    Keystone.prototype.applyUpdates = function (callback) {
        __block_Ix52Ab(15);
        {
            __statement_zRynHy(193);
            __extro_JLTOgl(194, __intro_RMcQoX(194, (__expression_SYuMxd(195), require('./lib/updates'))).apply(callback));
        }
    };
}
{
    __statement_zRynHy(196);
    Keystone.prototype.render = function (req, res, view, ext) {
        __block_Ix52Ab(16);
        {
            __statement_zRynHy(197);
            var keystone = this;
        }
        {
            __statement_zRynHy(198);
            var templatePath = (__expression_SYuMxd(199), (__expression_SYuMxd(200), (__expression_SYuMxd(201), (__expression_SYuMxd(202), __dirname) + '/templates/views/') + (__expression_SYuMxd(203), view)) + '.jade');
        }
        {
            __statement_zRynHy(204);
            var jadeOptions = {
                    filename: templatePath,
                    pretty: (__expression_SYuMxd(205), __extro_JLTOgl(206, __intro_RMcQoX(206, keystone).get('env')) !== 'production')
                };
        }
        {
            __statement_zRynHy(207);
            var compileTemplate = function () {
                __block_Ix52Ab(17);
                return __expression_SYuMxd(208), __extro_JLTOgl(209, __intro_RMcQoX(209, jade).compile(__extro_JLTOgl(210, __intro_RMcQoX(210, fs).readFileSync(templatePath, 'utf8')), jadeOptions));
            };
        }
        {
            __statement_zRynHy(211);
            var template = __extro_JLTOgl(214, __intro_RMcQoX(214, this).get('viewCache')) ? (__expression_SYuMxd(212), (__expression_SYuMxd(215), templateCache[view] || (templateCache[view] = (__expression_SYuMxd(216), compileTemplate())))) : (__expression_SYuMxd(213), (__expression_SYuMxd(217), compileTemplate()));
        }
        {
            __statement_zRynHy(218);
            var flashMessages = {
                    info: __extro_JLTOgl(219, __intro_RMcQoX(219, res.req).flash('info')),
                    success: __extro_JLTOgl(220, __intro_RMcQoX(220, res.req).flash('success')),
                    warning: __extro_JLTOgl(221, __intro_RMcQoX(221, res.req).flash('warning')),
                    error: __extro_JLTOgl(222, __intro_RMcQoX(222, res.req).flash('error')),
                    hilight: __extro_JLTOgl(223, __intro_RMcQoX(223, res.req).flash('hilight'))
                };
        }
        {
            __statement_zRynHy(224);
            var locals = {
                    _: _,
                    moment: moment,
                    numeral: numeral,
                    env: __extro_JLTOgl(225, __intro_RMcQoX(225, this).get('env')),
                    brand: __extro_JLTOgl(226, __intro_RMcQoX(226, keystone).get('brand')),
                    nav: keystone.nav,
                    messages: __extro_JLTOgl(229, __intro_RMcQoX(229, _).any(flashMessages, function (msgs) {
                        __block_Ix52Ab(18);
                        return __expression_SYuMxd(230), msgs.length;
                    })) ? (__expression_SYuMxd(227), flashMessages) : (__expression_SYuMxd(228), false),
                    lists: keystone.lists,
                    js: 'javascript:;',
                    utils: utils,
                    user: req.user,
                    title: 'Keystone',
                    signout: __extro_JLTOgl(231, __intro_RMcQoX(231, this).get('signout url')),
                    backUrl: (__expression_SYuMxd(232), __extro_JLTOgl(233, __intro_RMcQoX(233, this).get('back url')) || '/'),
                    section: {},
                    version: this.version,
                    csrf_token_key: keystone.security.csrf.TOKEN_KEY,
                    csrf_token_value: __extro_JLTOgl(234, __intro_RMcQoX(234, keystone.security.csrf).getToken(req, res)),
                    csrf_query: (__expression_SYuMxd(235), (__expression_SYuMxd(236), (__expression_SYuMxd(237), '&' + keystone.security.csrf.TOKEN_KEY) + '=') + __extro_JLTOgl(238, __intro_RMcQoX(238, keystone.security.csrf).getToken(req, res))),
                    ga: {
                        property: __extro_JLTOgl(239, __intro_RMcQoX(239, this).get('ga property')),
                        domain: __extro_JLTOgl(240, __intro_RMcQoX(240, this).get('ga domain'))
                    },
                    wysiwygOptions: {
                        enableImages: __extro_JLTOgl(243, __intro_RMcQoX(243, keystone).get('wysiwyg images')) ? (__expression_SYuMxd(241), true) : (__expression_SYuMxd(242), false),
                        enableCloudinaryUploads: __extro_JLTOgl(246, __intro_RMcQoX(246, keystone).get('wysiwyg cloudinary images')) ? (__expression_SYuMxd(244), true) : (__expression_SYuMxd(245), false),
                        additionalButtons: (__expression_SYuMxd(247), __extro_JLTOgl(248, __intro_RMcQoX(248, keystone).get('wysiwyg additional buttons')) || ''),
                        additionalPlugins: (__expression_SYuMxd(249), __extro_JLTOgl(250, __intro_RMcQoX(250, keystone).get('wysiwyg additional plugins')) || ''),
                        additionalOptions: (__expression_SYuMxd(251), __extro_JLTOgl(252, __intro_RMcQoX(252, keystone).get('wysiwyg additional options')) || {}),
                        overrideToolbar: __extro_JLTOgl(253, __intro_RMcQoX(253, keystone).get('wysiwyg override toolbar')),
                        skin: (__expression_SYuMxd(254), __extro_JLTOgl(255, __intro_RMcQoX(255, keystone).get('wysiwyg skin')) || 'keystone'),
                        menubar: __extro_JLTOgl(256, __intro_RMcQoX(256, keystone).get('wysiwyg menubar'))
                    }
                };
        }
        {
            __statement_zRynHy(257);
            __extro_JLTOgl(258, __intro_RMcQoX(258, _).extend(locals, ext));
        }
        if (__extro_JLTOgl(259, __intro_RMcQoX(259, keystone).get('cloudinary config'))) {
            __block_Ix52Ab(19);
            try {
                __block_Ix52Ab(20);
                {
                    __statement_zRynHy(260);
                    var cloudinaryUpload = __extro_JLTOgl(261, __intro_RMcQoX(261, cloudinary.uploader).direct_upload());
                }
                {
                    __statement_zRynHy(262);
                    locals.cloudinary = {
                        cloud_name: __extro_JLTOgl(263, __intro_RMcQoX(263, keystone).get('cloudinary config')).cloud_name,
                        api_key: __extro_JLTOgl(264, __intro_RMcQoX(264, keystone).get('cloudinary config')).api_key,
                        timestamp: cloudinaryUpload.hidden_fields.timestamp,
                        signature: cloudinaryUpload.hidden_fields.signature,
                        prefix: (__expression_SYuMxd(265), __extro_JLTOgl(266, __intro_RMcQoX(266, keystone).get('cloudinary prefix')) || ''),
                        uploader: cloudinary.uploader
                    };
                }
                {
                    __statement_zRynHy(267);
                    locals.cloudinary_js_config = __extro_JLTOgl(268, __intro_RMcQoX(268, cloudinary).cloudinary_js_config());
                }
            } catch (e) {
                __block_Ix52Ab(21);
                if (__expression_SYuMxd(269), (__expression_SYuMxd(270), e) === 'Must supply api_key') {
                    __block_Ix52Ab(22);
                    {
                        __statement_zRynHy(271);
                        throw new Error((__expression_SYuMxd(272), 'Invalid Cloudinary Config Provided\n\n' + 'See http://keystonejs.com/docs/configuration/#cloudinary for more information.'));
                    }
                } else {
                    __block_Ix52Ab(23);
                    {
                        __statement_zRynHy(273);
                        throw e;
                    }
                }
            }
        }
        {
            __statement_zRynHy(274);
            locals.fieldLocals = __extro_JLTOgl(275, __intro_RMcQoX(275, _).pick(locals, '_', 'moment', 'numeral', 'env', 'js', 'utils', 'user', 'cloudinary'));
        }
        {
            __statement_zRynHy(276);
            var html = (__expression_SYuMxd(277), template(__extro_JLTOgl(278, __intro_RMcQoX(278, _).extend(locals, ext))));
        }
        {
            __statement_zRynHy(279);
            __extro_JLTOgl(280, __intro_RMcQoX(280, res).send(html));
        }
    };
}
{
    __statement_zRynHy(281);
    Keystone.prototype.populateRelated = function (docs, relationships, callback) {
        __block_Ix52Ab(24);
        if (__extro_JLTOgl(282, __intro_RMcQoX(282, Array).isArray(docs))) {
            __block_Ix52Ab(25);
            {
                __statement_zRynHy(283);
                __extro_JLTOgl(284, __intro_RMcQoX(284, async).each(docs, function (doc, done) {
                    __block_Ix52Ab(26);
                    {
                        __statement_zRynHy(285);
                        __extro_JLTOgl(286, __intro_RMcQoX(286, doc).populateRelated(relationships, done));
                    }
                }, callback));
            }
        } else if (__expression_SYuMxd(287), (__expression_SYuMxd(288), docs) && docs.populateRelated) {
            __block_Ix52Ab(27);
            {
                __statement_zRynHy(289);
                __extro_JLTOgl(290, __intro_RMcQoX(290, docs).populateRelated(relationships, callback));
            }
        } else {
            __block_Ix52Ab(28);
            {
                __statement_zRynHy(291);
                __expression_SYuMxd(292), callback();
            }
        }
    };
}
{
    __statement_zRynHy(293);
    Keystone.prototype.console = {};
}
{
    __statement_zRynHy(294);
    Keystone.prototype.console.err = function (type, msg) {
        __block_Ix52Ab(29);
        if (__extro_JLTOgl(295, __intro_RMcQoX(295, keystone).get('logger'))) {
            __block_Ix52Ab(30);
            {
                __statement_zRynHy(296);
                var dashes = '\n------------------------------------------------\n';
            }
            {
                __statement_zRynHy(297);
                __extro_JLTOgl(298, __intro_RMcQoX(298, console).log((__expression_SYuMxd(299), (__expression_SYuMxd(300), (__expression_SYuMxd(301), (__expression_SYuMxd(302), (__expression_SYuMxd(303), (__expression_SYuMxd(304), dashes) + 'KeystoneJS: ') + (__expression_SYuMxd(305), type)) + ':\n\n') + (__expression_SYuMxd(306), msg)) + (__expression_SYuMxd(307), dashes))));
            }
        }
    };
}
{
    __statement_zRynHy(308);
    keystone.version = (__expression_SYuMxd(309), require('./package.json')).version;
}
{
    __statement_zRynHy(310);
    keystone.session = (__expression_SYuMxd(311), require('./lib/session'));
}
