// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_updateHandler.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_nKv42A, __expression_Y50xNb, __block_E$QQbv;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_nKv42A = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/updateHandler.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_Y50xNb = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/updateHandler.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_E$QQbv = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/updateHandler.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_UChpcW = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro__z5NQ$ = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/updateHandler.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_nKv42A(0);
    var _ = (__expression_Y50xNb(1), require('underscore')), keystone = (__expression_Y50xNb(2), require('../'));
}
function UpdateHandler(list, item, req, res, options) {
    __block_E$QQbv(0);
    if (__expression_Y50xNb(3), !(__expression_Y50xNb(4), this instanceof (__expression_Y50xNb(5), UpdateHandler))) {
        __block_E$QQbv(1);
        return __expression_Y50xNb(6), new UpdateHandler(list, item);
    }
    {
        __statement_nKv42A(7);
        this.list = list;
    }
    {
        __statement_nKv42A(8);
        this.item = item;
    }
    {
        __statement_nKv42A(9);
        this.req = req;
    }
    {
        __statement_nKv42A(10);
        this.res = res;
    }
    {
        __statement_nKv42A(11);
        this.user = req.user;
    }
    {
        __statement_nKv42A(12);
        this.options = (__expression_Y50xNb(13), (__expression_Y50xNb(14), options) || {});
    }
    if (__expression_Y50xNb(15), !this.options.errorMessage) {
        __block_E$QQbv(2);
        {
            __statement_nKv42A(16);
            this.options.errorMessage = 'There was a problem saving your changes:';
        }
    }
    if (this.options.user) {
        __block_E$QQbv(3);
        {
            __statement_nKv42A(17);
            this.user = this.options.user;
        }
    }
    {
        __statement_nKv42A(18);
        this.validationMethods = {};
    }
    {
        __statement_nKv42A(19);
        this.validationErrors = {};
    }
}
{
    __statement_nKv42A(20);
    UpdateHandler.prototype.validate = function (path, fn) {
        __block_E$QQbv(4);
        {
            __statement_nKv42A(21);
            this.validationMethods[path] = fn;
        }
        return __expression_Y50xNb(22), this;
    };
}
{
    __statement_nKv42A(23);
    UpdateHandler.prototype.addValidationError = function (path, msg, type) {
        __block_E$QQbv(5);
        {
            __statement_nKv42A(24);
            this.validationErrors[path] = {
                name: 'ValidatorError',
                path: path,
                message: msg,
                type: (__expression_Y50xNb(25), (__expression_Y50xNb(26), type) || 'required')
            };
        }
        return __expression_Y50xNb(27), this;
    };
}
{
    __statement_nKv42A(28);
    UpdateHandler.prototype.process = function (data, options, callback) {
        __block_E$QQbv(6);
        {
            __statement_nKv42A(29);
            var usingDefaultFields = false;
        }
        if (__expression_Y50xNb(30), 'function' === (__expression_Y50xNb(31), typeof options)) {
            __block_E$QQbv(7);
            {
                __statement_nKv42A(32);
                callback = options;
            }
            {
                __statement_nKv42A(33);
                options = null;
            }
        }
        if (__expression_Y50xNb(34), !(__expression_Y50xNb(35), options)) {
            __block_E$QQbv(8);
            {
                __statement_nKv42A(36);
                options = {};
            }
        } else if (__expression_Y50xNb(37), 'string' === (__expression_Y50xNb(38), typeof options)) {
            __block_E$QQbv(9);
            {
                __statement_nKv42A(39);
                options = {
                    fields: options
                };
            }
        }
        if (__expression_Y50xNb(40), !options.fields) {
            __block_E$QQbv(10);
            {
                __statement_nKv42A(41);
                options.fields = __extro__z5NQ$(42, __intro_UChpcW(42, _).keys(this.list.fields));
            }
            {
                __statement_nKv42A(43);
                usingDefaultFields = true;
            }
        } else if (__expression_Y50xNb(44), 'string' === (__expression_Y50xNb(45), typeof options.fields)) {
            __block_E$QQbv(11);
            {
                __statement_nKv42A(46);
                options.fields = __extro__z5NQ$(47, __intro_UChpcW(47, __extro__z5NQ$(48, __intro_UChpcW(48, options.fields).split(','))).map(function (i) {
                    __block_E$QQbv(12);
                    return __expression_Y50xNb(49), __extro__z5NQ$(50, __intro_UChpcW(50, i).trim());
                }));
            }
        }
        {
            __statement_nKv42A(51);
            options.required = (__expression_Y50xNb(52), options.required || {});
        }
        {
            __statement_nKv42A(53);
            options.errorMessage = (__expression_Y50xNb(54), options.errorMessage || this.options.errorMessage);
        }
        {
            __statement_nKv42A(55);
            options.invalidMessages = (__expression_Y50xNb(56), options.invalidMessages || {});
        }
        {
            __statement_nKv42A(57);
            options.requiredMessages = (__expression_Y50xNb(58), options.requiredMessages || {});
        }
        if (__expression_Y50xNb(59), 'string' === (__expression_Y50xNb(60), typeof options.required)) {
            __block_E$QQbv(13);
            {
                __statement_nKv42A(61);
                var requiredFields = __extro__z5NQ$(62, __intro_UChpcW(62, __extro__z5NQ$(63, __intro_UChpcW(63, options.required).split(','))).map(function (i) {
                        __block_E$QQbv(14);
                        return __expression_Y50xNb(64), __extro__z5NQ$(65, __intro_UChpcW(65, i).trim());
                    }));
            }
            {
                __statement_nKv42A(66);
                options.required = {};
            }
            {
                __statement_nKv42A(67);
                __extro__z5NQ$(68, __intro_UChpcW(68, requiredFields).forEach(function (path) {
                    __block_E$QQbv(15);
                    {
                        __statement_nKv42A(69);
                        options.required[path] = true;
                    }
                }));
            }
        }
        {
            __statement_nKv42A(70);
            __extro__z5NQ$(71, __intro_UChpcW(71, options.fields).forEach(function (path) {
                __block_E$QQbv(16);
                {
                    __statement_nKv42A(72);
                    var field = (__expression_Y50xNb(75), (__expression_Y50xNb(76), path) instanceof keystone.Field) ? (__expression_Y50xNb(73), path) : (__expression_Y50xNb(74), __extro__z5NQ$(77, __intro_UChpcW(77, this.list).field(path)));
                }
                if (__expression_Y50xNb(78), (__expression_Y50xNb(79), field) && field.required) {
                    __block_E$QQbv(17);
                    {
                        __statement_nKv42A(80);
                        options.required[path] = true;
                    }
                }
            }, this));
        }
        {
            __statement_nKv42A(81);
            var actionQueue = [], addValidationError = __extro__z5NQ$(82, __intro_UChpcW(82, this.addValidationError).bind(this)), validationErrors = this.validationErrors;
        }
        {
            __statement_nKv42A(83);
            var progress = __extro__z5NQ$(84, __intro_UChpcW(84, function (err) {
                    __block_E$QQbv(18);
                    if (__expression_Y50xNb(85), err) {
                        __block_E$QQbv(19);
                        if (options.logErrors) {
                            __block_E$QQbv(20);
                            {
                                __statement_nKv42A(86);
                                __extro__z5NQ$(87, __intro_UChpcW(87, console).log((__expression_Y50xNb(88), (__expression_Y50xNb(89), (__expression_Y50xNb(90), (__expression_Y50xNb(91), 'Error saving changes to ' + this.item.list.singular) + ' ') + this.item.id) + ':')));
                            }
                            {
                                __statement_nKv42A(92);
                                __extro__z5NQ$(93, __intro_UChpcW(93, console).log(err));
                            }
                        }
                        {
                            __statement_nKv42A(94);
                            __expression_Y50xNb(95), callback(err, this);
                        }
                    } else if (__extro__z5NQ$(96, __intro_UChpcW(96, _).size(validationErrors))) {
                        __block_E$QQbv(21);
                        if (options.flashErrors) {
                            __block_E$QQbv(22);
                            {
                                __statement_nKv42A(97);
                                __extro__z5NQ$(98, __intro_UChpcW(98, this.req).flash('error', {
                                    type: 'ValidationError',
                                    title: options.errorMessage,
                                    list: __extro__z5NQ$(99, __intro_UChpcW(99, _).pluck(validationErrors, 'message'))
                                }));
                            }
                        }
                        {
                            __statement_nKv42A(100);
                            __expression_Y50xNb(101), callback({
                                message: 'Validation failed',
                                name: 'ValidationError',
                                errors: validationErrors
                            }, this);
                        }
                    } else if (actionQueue.length) {
                        __block_E$QQbv(23);
                        {
                            __statement_nKv42A(102);
                            __expression_Y50xNb(103), __extro__z5NQ$(104, __intro_UChpcW(104, actionQueue).pop())();
                        }
                    } else {
                        __block_E$QQbv(24);
                        {
                            __statement_nKv42A(105);
                            __expression_Y50xNb(106), saveItem();
                        }
                    }
                }).bind(this));
        }
        {
            __statement_nKv42A(107);
            var saveItem = __extro__z5NQ$(108, __intro_UChpcW(108, function () {
                    __block_E$QQbv(25);
                    {
                        __statement_nKv42A(109);
                        this.item._req_user = this.user;
                    }
                    {
                        __statement_nKv42A(110);
                        __extro__z5NQ$(111, __intro_UChpcW(111, this.item).save(__extro__z5NQ$(112, __intro_UChpcW(112, function (err) {
                            __block_E$QQbv(26);
                            if (__expression_Y50xNb(113), err) {
                                __block_E$QQbv(27);
                                if (__expression_Y50xNb(114), err.name === 'ValidationError') {
                                    __block_E$QQbv(28);
                                    if (options.flashErrors) {
                                        __block_E$QQbv(29);
                                        {
                                            __statement_nKv42A(115);
                                            __extro__z5NQ$(116, __intro_UChpcW(116, this.req).flash('error', {
                                                type: 'ValidationError',
                                                title: options.errorMessage,
                                                list: __extro__z5NQ$(117, __intro_UChpcW(117, _).pluck(err.errors, 'message'))
                                            }));
                                        }
                                    }
                                } else {
                                    __block_E$QQbv(30);
                                    if (options.logErrors) {
                                        __block_E$QQbv(31);
                                        {
                                            __statement_nKv42A(118);
                                            __extro__z5NQ$(119, __intro_UChpcW(119, console).log((__expression_Y50xNb(120), (__expression_Y50xNb(121), (__expression_Y50xNb(122), (__expression_Y50xNb(123), 'Error saving changes to ' + this.item.list.singular) + ' ') + this.item.id) + ':')));
                                        }
                                        {
                                            __statement_nKv42A(124);
                                            __extro__z5NQ$(125, __intro_UChpcW(125, console).log(err));
                                        }
                                    }
                                    if (options.flashErrors) {
                                        __block_E$QQbv(32);
                                        {
                                            __statement_nKv42A(126);
                                            __extro__z5NQ$(127, __intro_UChpcW(127, this.req).flash('error', (__expression_Y50xNb(128), (__expression_Y50xNb(129), (__expression_Y50xNb(130), (__expression_Y50xNb(131), (__expression_Y50xNb(132), 'There was an error saving your changes: ' + err.message) + ' (') + err.name) + (err.type ? (__expression_Y50xNb(133), (__expression_Y50xNb(135), ': ' + err.type)) : (__expression_Y50xNb(134), ''))) + ')')));
                                        }
                                    }
                                }
                            }
                            return __expression_Y50xNb(136), (__expression_Y50xNb(137), callback(err, this));
                        }).bind(this))));
                    }
                }).bind(this));
        }
        {
            __statement_nKv42A(138);
            __extro__z5NQ$(139, __intro_UChpcW(139, options.fields).forEach(function (path) {
                __block_E$QQbv(33);
                {
                    __statement_nKv42A(140);
                    var message;
                }
                {
                    __statement_nKv42A(141);
                    var field = (__expression_Y50xNb(144), (__expression_Y50xNb(145), path) instanceof keystone.Field) ? (__expression_Y50xNb(142), path) : (__expression_Y50xNb(143), __extro__z5NQ$(146, __intro_UChpcW(146, this.list).field(path))), invalidated = false;
                }
                if (__expression_Y50xNb(147), !(__expression_Y50xNb(148), field)) {
                    __block_E$QQbv(34);
                    {
                        __statement_nKv42A(149);
                        throw new Error((__expression_Y50xNb(150), 'UpdateHandler.process called with invalid field path: ' + (__expression_Y50xNb(151), path)));
                    }
                }
                if (__expression_Y50xNb(152), (__expression_Y50xNb(153), (__expression_Y50xNb(154), usingDefaultFields) && field.noedit) && (__expression_Y50xNb(155), !options.ignoreNoedit)) {
                    __block_E$QQbv(35);
                    return __expression_Y50xNb(156);
                }
                switch (field.type) {
                case 'localfile':
                case 'localfiles':
                case 'cloudinaryimage':
                case 'cloudinaryimages':
                case 'azurefile':
                case 's3file': {
                        __block_E$QQbv(36);
                        {
                            __statement_nKv42A(157);
                            __extro__z5NQ$(158, __intro_UChpcW(158, actionQueue).push(__extro__z5NQ$(159, __intro_UChpcW(159, field).getRequestHandler(this.item, this.req, options.paths, __extro__z5NQ$(160, __intro_UChpcW(160, function (err) {
                                __block_E$QQbv(37);
                                if (__expression_Y50xNb(161), (__expression_Y50xNb(162), err) && options.flashErrors) {
                                    __block_E$QQbv(38);
                                    {
                                        __statement_nKv42A(163);
                                        __extro__z5NQ$(164, __intro_UChpcW(164, this.req).flash('error', (__expression_Y50xNb(165), (__expression_Y50xNb(166), field.label + ' upload failed - ') + err.message)));
                                    }
                                }
                                {
                                    __statement_nKv42A(167);
                                    __expression_Y50xNb(168), progress(err);
                                }
                            }).bind(this))))));
                        }
                        break;
                    }
                case 'location': {
                        __block_E$QQbv(39);
                        {
                            __statement_nKv42A(169);
                            __extro__z5NQ$(170, __intro_UChpcW(170, actionQueue).push(__extro__z5NQ$(171, __intro_UChpcW(171, field).getRequestHandler(this.item, this.req, options.paths, __extro__z5NQ$(172, __intro_UChpcW(172, function (err) {
                                __block_E$QQbv(40);
                                if (__expression_Y50xNb(173), (__expression_Y50xNb(174), err) && options.flashErrors) {
                                    __block_E$QQbv(41);
                                    {
                                        __statement_nKv42A(175);
                                        __extro__z5NQ$(176, __intro_UChpcW(176, this.req).flash('error', (__expression_Y50xNb(177), (__expression_Y50xNb(178), field.label + ' improve failed - ') + (__expression_Y50xNb(179), err.status_text || err.status))));
                                    }
                                }
                                {
                                    __statement_nKv42A(180);
                                    __expression_Y50xNb(181), progress(err);
                                }
                            }).bind(this))))));
                        }
                        break;
                    }
                case 'password': {
                        __block_E$QQbv(42);
                        if (__expression_Y50xNb(182), (__expression_Y50xNb(183), !data[field.path]) && (__expression_Y50xNb(184), (__expression_Y50xNb(185), !options.required[field.path]) || __extro__z5NQ$(186, __intro_UChpcW(186, this.item).get(field.path)))) {
                            __block_E$QQbv(43);
                            return __expression_Y50xNb(187);
                        }
                        if (__expression_Y50xNb(188), data[field.path] !== data[field.paths.confirm]) {
                            __block_E$QQbv(44);
                            {
                                __statement_nKv42A(189);
                                message = (__expression_Y50xNb(190), options.invalidMessages[__expression_Y50xNb(191), field.path + '_match'] || 'Passwords must match');
                            }
                            {
                                __statement_nKv42A(192);
                                __expression_Y50xNb(193), addValidationError(field.path, message);
                            }
                            {
                                __statement_nKv42A(194);
                                invalidated = true;
                            }
                        }
                        break;
                    }
                }
                if (__expression_Y50xNb(195), (__expression_Y50xNb(196), !(__expression_Y50xNb(197), invalidated)) && (__expression_Y50xNb(198), !__extro__z5NQ$(199, __intro_UChpcW(199, field).validateInput(data)))) {
                    __block_E$QQbv(45);
                    {
                        __statement_nKv42A(200);
                        message = (__expression_Y50xNb(201), (__expression_Y50xNb(202), options.invalidMessages[field.path] || field.options.invalidMessage) || (__expression_Y50xNb(203), (__expression_Y50xNb(204), (__expression_Y50xNb(205), (__expression_Y50xNb(206), 'Please enter a valid ' + field.typeDescription) + ' in the ') + field.label) + ' field'));
                    }
                    {
                        __statement_nKv42A(207);
                        __expression_Y50xNb(208), addValidationError(field.path, message);
                    }
                    {
                        __statement_nKv42A(209);
                        invalidated = true;
                    }
                }
                if (__expression_Y50xNb(210), (__expression_Y50xNb(211), (__expression_Y50xNb(212), !(__expression_Y50xNb(213), invalidated)) && options.required[field.path]) && (__expression_Y50xNb(214), !__extro__z5NQ$(215, __intro_UChpcW(215, field).validateInput(data, true, this.item)))) {
                    __block_E$QQbv(46);
                    {
                        __statement_nKv42A(216);
                        message = (__expression_Y50xNb(217), (__expression_Y50xNb(218), options.requiredMessages[field.path] || field.options.requiredMessage) || (__expression_Y50xNb(219), field.label + ' is required'));
                    }
                    {
                        __statement_nKv42A(220);
                        __expression_Y50xNb(221), addValidationError(field.path, message);
                    }
                    {
                        __statement_nKv42A(222);
                        invalidated = true;
                    }
                }
                if (__expression_Y50xNb(223), (__expression_Y50xNb(224), !(__expression_Y50xNb(225), invalidated)) && this.validationMethods[field.path]) {
                    __block_E$QQbv(47);
                    {
                        __statement_nKv42A(226);
                        message = __extro__z5NQ$(227, __intro_UChpcW(227, this.validationMethods)[field.path](data));
                    }
                    if (__expression_Y50xNb(228), message) {
                        __block_E$QQbv(48);
                        {
                            __statement_nKv42A(229);
                            __expression_Y50xNb(230), addValidationError(field.path, message);
                        }
                    }
                }
                {
                    __statement_nKv42A(231);
                    __extro__z5NQ$(232, __intro_UChpcW(232, field).updateItem(this.item, data));
                }
            }, this));
        }
        {
            __statement_nKv42A(233);
            __expression_Y50xNb(234), progress();
        }
    };
}
{
    __statement_nKv42A(235);
    exports = module.exports = UpdateHandler;
}
