// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_content_index.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_WPYcoF, __expression_h0CxSg, __block_P2zHyo;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_WPYcoF = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/index.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_h0CxSg = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/index.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_P2zHyo = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/index.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_AKuR92 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_ysr_ge = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/content/index.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_WPYcoF(0);
    var _ = (__expression_h0CxSg(1), require('underscore')), keystone = (__expression_h0CxSg(2), require('../../')), utils = keystone.utils;
}
{
    __statement_WPYcoF(3);
    var Content = function () {
        __block_P2zHyo(0);
    };
}
{
    __statement_WPYcoF(4);
    Content.prototype.fetch = function (page, callback) {
        __block_P2zHyo(1);
        if (__extro_ysr_ge(5, __intro_AKuR92(5, utils).isFunction(page))) {
            __block_P2zHyo(2);
            {
                __statement_WPYcoF(6);
                callback = page;
            }
            {
                __statement_WPYcoF(7);
                page = null;
            }
        }
        {
            __statement_WPYcoF(8);
            var content = this;
        }
        if (__expression_h0CxSg(9), !this.AppContent) {
            __block_P2zHyo(3);
            return __expression_h0CxSg(10), (__expression_h0CxSg(11), callback({
                error: 'invalid page',
                message: 'No pages have been registered.'
            }));
        }
        if (__expression_h0CxSg(12), page) {
            __block_P2zHyo(4);
            if (__expression_h0CxSg(13), !this.pages[page]) {
                __block_P2zHyo(5);
                return __expression_h0CxSg(14), (__expression_h0CxSg(15), callback({
                    error: 'invalid page',
                    message: (__expression_h0CxSg(16), (__expression_h0CxSg(17), 'The page ' + (__expression_h0CxSg(18), page)) + ' does not exist.')
                }));
            }
            {
                __statement_WPYcoF(19);
                __extro_ysr_ge(20, __intro_AKuR92(20, this.AppContent).findOne({
                    key: page
                }, function (err, result) {
                    __block_P2zHyo(6);
                    if (__expression_h0CxSg(21), err) {
                        __block_P2zHyo(7);
                        return __expression_h0CxSg(22), (__expression_h0CxSg(23), callback(err));
                    }
                    return __expression_h0CxSg(24), (__expression_h0CxSg(25), callback(null, __extro_ysr_ge(26, __intro_AKuR92(26, content.pages[page]).populate((__expression_h0CxSg(29), result) ? (__expression_h0CxSg(27), result.content.data) : (__expression_h0CxSg(28), {})))));
                }));
            }
        } else {
            __block_P2zHyo(8);
            {
                __statement_WPYcoF(30);
                __extro_ysr_ge(31, __intro_AKuR92(31, this.AppContent).find(function (err, results) {
                    __block_P2zHyo(9);
                    if (__expression_h0CxSg(32), err) {
                        __block_P2zHyo(10);
                        return __expression_h0CxSg(33), (__expression_h0CxSg(34), callback(err));
                    }
                    {
                        __statement_WPYcoF(35);
                        var data = {};
                    }
                    {
                        __statement_WPYcoF(36);
                        __extro_ysr_ge(37, __intro_AKuR92(37, results).forEach(function (i) {
                            __block_P2zHyo(11);
                            if (content.pages[i.key]) {
                                __block_P2zHyo(12);
                                {
                                    __statement_WPYcoF(38);
                                    data[i.key] = __extro_ysr_ge(39, __intro_AKuR92(39, content.pages[i.key]).populate(i.content.data));
                                }
                            }
                        }));
                    }
                    {
                        __statement_WPYcoF(40);
                        __extro_ysr_ge(41, __intro_AKuR92(41, _).each(content.pages, function (i) {
                            __block_P2zHyo(13);
                            if (__expression_h0CxSg(42), !data[i.key]) {
                                __block_P2zHyo(14);
                                {
                                    __statement_WPYcoF(43);
                                    data[i.key] = __extro_ysr_ge(44, __intro_AKuR92(44, i).populate());
                                }
                            }
                        }));
                    }
                    return __expression_h0CxSg(45), data;
                }));
            }
        }
    };
}
{
    __statement_WPYcoF(46);
    Content.prototype.store = function (page, content, callback) {
        __block_P2zHyo(15);
        if (__expression_h0CxSg(47), !this.pages[page]) {
            __block_P2zHyo(16);
            return __expression_h0CxSg(48), (__expression_h0CxSg(49), callback({
                error: 'invalid page',
                message: (__expression_h0CxSg(50), (__expression_h0CxSg(51), 'The page ' + (__expression_h0CxSg(52), page)) + ' does not exist.')
            }));
        }
        {
            __statement_WPYcoF(53);
            content = __extro_ysr_ge(54, __intro_AKuR92(54, this.pages[page]).validate(content));
        }
        {
            __statement_WPYcoF(55);
            __extro_ysr_ge(56, __intro_AKuR92(56, this.AppContent).findOne({
                key: page
            }, function (err, doc) {
                __block_P2zHyo(17);
                if (__expression_h0CxSg(57), err) {
                    __block_P2zHyo(18);
                    return __expression_h0CxSg(58), (__expression_h0CxSg(59), callback(err));
                }
                if (__expression_h0CxSg(60), doc) {
                    __block_P2zHyo(19);
                    if (doc.content) {
                        __block_P2zHyo(20);
                        {
                            __statement_WPYcoF(61);
                            __extro_ysr_ge(62, __intro_AKuR92(62, doc.history).push(doc.content));
                        }
                    }
                    {
                        __statement_WPYcoF(63);
                        __extro_ysr_ge(64, __intro_AKuR92(64, _).defaults(content, doc.content));
                    }
                } else {
                    __block_P2zHyo(21);
                    {
                        __statement_WPYcoF(65);
                        doc = new content.AppContent({
                            key: page
                        });
                    }
                }
                {
                    __statement_WPYcoF(66);
                    doc.content = {
                        data: __extro_ysr_ge(67, __intro_AKuR92(67, this.pages[page]).clean(content))
                    };
                }
                {
                    __statement_WPYcoF(68);
                    doc.lastChangeDate = __extro_ysr_ge(69, __intro_AKuR92(69, Date).now());
                }
                {
                    __statement_WPYcoF(70);
                    __extro_ysr_ge(71, __intro_AKuR92(71, doc).save(callback));
                }
            }));
        }
    };
}
{
    __statement_WPYcoF(72);
    Content.prototype.page = function (key, page) {
        __block_P2zHyo(22);
        if (__expression_h0CxSg(73), !this.pages) {
            __block_P2zHyo(23);
            {
                __statement_WPYcoF(74);
                this.pages = {};
            }
        }
        if (__expression_h0CxSg(75), arguments.length === 1) {
            __block_P2zHyo(24);
            if (__expression_h0CxSg(76), !this.pages[key]) {
                __block_P2zHyo(25);
                {
                    __statement_WPYcoF(77);
                    throw new Error((__expression_h0CxSg(78), (__expression_h0CxSg(79), 'keystone.content.page() Error: page ' + (__expression_h0CxSg(80), key)) + ' cannot be registered more than once.'));
                }
            }
            return __expression_h0CxSg(81), this.pages[key];
        }
        {
            __statement_WPYcoF(82);
            __extro_ysr_ge(83, __intro_AKuR92(83, this).initModel());
        }
        if (this.pages[key]) {
            __block_P2zHyo(26);
            {
                __statement_WPYcoF(84);
                throw new Error((__expression_h0CxSg(85), (__expression_h0CxSg(86), 'keystone.content.page() Error: page ' + (__expression_h0CxSg(87), key)) + ' cannot be registered more than once.'));
            }
        }
        {
            __statement_WPYcoF(88);
            this.pages[key] = page;
        }
        return __expression_h0CxSg(89), page;
    };
}
{
    __statement_WPYcoF(90);
    Content.prototype.initModel = function () {
        __block_P2zHyo(27);
        if (this.AppContent) {
            __block_P2zHyo(28);
            return __expression_h0CxSg(91);
        }
        {
            __statement_WPYcoF(92);
            var contentSchemaDef = {
                    createdAt: {
                        type: Date,
                        default: Date.now
                    },
                    data: {
                        type: keystone.mongoose.Schema.Types.Mixed
                    }
                };
        }
        {
            __statement_WPYcoF(93);
            var ContentSchema = new keystone.mongoose.Schema(contentSchemaDef);
        }
        {
            __statement_WPYcoF(94);
            var PageSchema = new keystone.mongoose.Schema({
                    page: {
                        type: String,
                        index: true
                    },
                    lastChangeDate: {
                        type: Date,
                        index: true
                    },
                    content: contentSchemaDef,
                    history: [
                        ContentSchema
                    ]
                }, {
                    collection: 'app_content'
                });
        }
        {
            __statement_WPYcoF(95);
            this.AppContent = __extro_ysr_ge(96, __intro_AKuR92(96, keystone.mongoose).model('App_Content', PageSchema));
        }
    };
}
{
    __statement_WPYcoF(97);
    Content.prototype.editable = function (user, options) {
        __block_P2zHyo(29);
        if (__expression_h0CxSg(98), (__expression_h0CxSg(99), !(__expression_h0CxSg(100), user)) || (__expression_h0CxSg(101), !user.canAccessKeystone)) {
            __block_P2zHyo(30);
            return __expression_h0CxSg(102), undefined;
        }
        if (options.list) {
            __block_P2zHyo(31);
            {
                __statement_WPYcoF(103);
                var list = __extro_ysr_ge(104, __intro_AKuR92(104, keystone).list(options.list));
            }
            if (__expression_h0CxSg(105), !(__expression_h0CxSg(106), list)) {
                __block_P2zHyo(32);
                return __expression_h0CxSg(107), __extro_ysr_ge(108, __intro_AKuR92(108, JSON).stringify({
                    type: 'error',
                    err: 'list not found'
                }));
            }
            {
                __statement_WPYcoF(109);
                var data = {
                        type: 'list',
                        path: list.path,
                        singular: list.singular,
                        plural: list.plural
                    };
            }
            if (options.id) {
                __block_P2zHyo(33);
                {
                    __statement_WPYcoF(110);
                    data.id = options.id;
                }
            }
            return __expression_h0CxSg(111), __extro_ysr_ge(112, __intro_AKuR92(112, JSON).stringify(data));
        }
    };
}
{
    __statement_WPYcoF(113);
    module.exports = exports = new Content();
}
{
    __statement_WPYcoF(114);
    exports.Page = (__expression_h0CxSg(115), require('./page'));
}
{
    __statement_WPYcoF(116);
    exports.Types = (__expression_h0CxSg(117), require('./types'));
}
