// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_password.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_V_i1iR, __expression_weOqkq, __block_u_br_1;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_V_i1iR = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/password.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_weOqkq = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/password.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_u_br_1 = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/password.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_ru_qd3 = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_nBr8G0 = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/password.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_V_i1iR(0);
    var _ = (__expression_weOqkq(1), require('underscore')), util = (__expression_weOqkq(2), require('util')), bcrypt = (__expression_weOqkq(3), require('bcrypt-nodejs')), super_ = (__expression_weOqkq(4), require('../field'));
}
function password(list, path, options) {
    __block_u_br_1(0);
    {
        __statement_V_i1iR(5);
        this.workFactor = (__expression_weOqkq(6), options.workFactor || 10);
    }
    {
        __statement_V_i1iR(7);
        this._nativeType = String;
    }
    {
        __statement_V_i1iR(8);
        this._underscoreMethods = [
            'format',
            'compare'
        ];
    }
    {
        __statement_V_i1iR(9);
        options.nosort = true;
    }
    {
        __statement_V_i1iR(10);
        options.nofilter = true;
    }
    {
        __statement_V_i1iR(11);
        __extro_nBr8G0(12, __intro_ru_qd3(12, password.super_).call(this, list, path, options));
    }
}
{
    __statement_V_i1iR(13);
    __extro_nBr8G0(14, __intro_ru_qd3(14, util).inherits(password, super_));
}
{
    __statement_V_i1iR(15);
    password.prototype.addToSchema = function () {
        __block_u_br_1(1);
        {
            __statement_V_i1iR(16);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_V_i1iR(17);
            this.paths = {
                confirm: (__expression_weOqkq(18), this.options.confirmPath || __extro_nBr8G0(19, __intro_ru_qd3(19, this._path).append('_confirm')))
            };
        }
        {
            __statement_V_i1iR(20);
            __extro_nBr8G0(21, __intro_ru_qd3(21, schema).path(this.path, __extro_nBr8G0(22, __intro_ru_qd3(22, _).defaults({
                type: String
            }, this.options))));
        }
        {
            __statement_V_i1iR(23);
            __extro_nBr8G0(24, __intro_ru_qd3(24, schema).pre('save', function (next) {
                __block_u_br_1(2);
                if (__expression_weOqkq(25), !__extro_nBr8G0(26, __intro_ru_qd3(26, this).isModified(field.path))) {
                    __block_u_br_1(3);
                    return __expression_weOqkq(27), (__expression_weOqkq(28), next());
                }
                if (__expression_weOqkq(29), !__extro_nBr8G0(30, __intro_ru_qd3(30, this).get(field.path))) {
                    __block_u_br_1(4);
                    {
                        __statement_V_i1iR(31);
                        __extro_nBr8G0(32, __intro_ru_qd3(32, this).set(field.path, undefined));
                    }
                    return __expression_weOqkq(33), (__expression_weOqkq(34), next());
                }
                {
                    __statement_V_i1iR(35);
                    var item = this;
                }
                {
                    __statement_V_i1iR(36);
                    __extro_nBr8G0(37, __intro_ru_qd3(37, bcrypt).genSalt(field.workFactor, function (err, salt) {
                        __block_u_br_1(5);
                        if (__expression_weOqkq(38), err) {
                            __block_u_br_1(6);
                            return __expression_weOqkq(39), (__expression_weOqkq(40), next(err));
                        }
                        {
                            __statement_V_i1iR(41);
                            __extro_nBr8G0(42, __intro_ru_qd3(42, bcrypt).hash(__extro_nBr8G0(43, __intro_ru_qd3(43, item).get(field.path)), salt, function () {
                                __block_u_br_1(7);
                            }, function (err, hash) {
                                __block_u_br_1(8);
                                if (__expression_weOqkq(44), err) {
                                    __block_u_br_1(9);
                                    return __expression_weOqkq(45), (__expression_weOqkq(46), next(err));
                                }
                                {
                                    __statement_V_i1iR(47);
                                    __extro_nBr8G0(48, __intro_ru_qd3(48, item).set(field.path, hash));
                                }
                                {
                                    __statement_V_i1iR(49);
                                    __expression_weOqkq(50), next();
                                }
                            }));
                        }
                    }));
                }
            }));
        }
        {
            __statement_V_i1iR(51);
            __extro_nBr8G0(52, __intro_ru_qd3(52, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_V_i1iR(53);
    password.prototype.format = function (item) {
        __block_u_br_1(10);
        if (__expression_weOqkq(54), !__extro_nBr8G0(55, __intro_ru_qd3(55, item).get(this.path))) {
            __block_u_br_1(11);
            return __expression_weOqkq(56), '';
        }
        {
            __statement_V_i1iR(57);
            var len = (__expression_weOqkq(58), __extro_nBr8G0(59, __intro_ru_qd3(59, Math).round((__expression_weOqkq(60), __extro_nBr8G0(61, __intro_ru_qd3(61, Math).random()) * 4))) + 6);
        }
        {
            __statement_V_i1iR(62);
            var stars = '';
        }
        for (var i = 0; __expression_weOqkq(63), (__expression_weOqkq(64), i) < (__expression_weOqkq(65), len); __expression_weOqkq(66), i++) {
            __block_u_br_1(12);
            {
                __statement_V_i1iR(67);
                stars += '*';
            }
        }
        return __expression_weOqkq(68), stars;
    };
}
{
    __statement_V_i1iR(69);
    password.prototype.compare = function (item, candidate, callback) {
        __block_u_br_1(13);
        if (__expression_weOqkq(70), 'function' !== (__expression_weOqkq(71), typeof callback)) {
            __block_u_br_1(14);
            {
                __statement_V_i1iR(72);
                throw new Error('Password.compare() requires a callback function.');
            }
        }
        {
            __statement_V_i1iR(73);
            var value = __extro_nBr8G0(74, __intro_ru_qd3(74, item).get(this.path));
        }
        if (__expression_weOqkq(75), !(__expression_weOqkq(76), value)) {
            __block_u_br_1(15);
            return __expression_weOqkq(77), (__expression_weOqkq(78), callback(null, false));
        }
        {
            __statement_V_i1iR(79);
            __extro_nBr8G0(80, __intro_ru_qd3(80, bcrypt).compare(candidate, __extro_nBr8G0(81, __intro_ru_qd3(81, item).get(this.path)), callback));
        }
    };
}
{
    __statement_V_i1iR(82);
    password.prototype.validateInput = function (data, required, item) {
        __block_u_br_1(16);
        if (__expression_weOqkq(83), !(__expression_weOqkq(84), required)) {
            __block_u_br_1(17);
            return __expression_weOqkq(85), true;
        }
        if (__expression_weOqkq(86), item) {
            __block_u_br_1(18);
            return __expression_weOqkq(87), (__expression_weOqkq(90), data[this.path] || __extro_nBr8G0(91, __intro_ru_qd3(91, item).get(this.path))) ? (__expression_weOqkq(88), true) : (__expression_weOqkq(89), false);
        } else {
            __block_u_br_1(19);
            return __expression_weOqkq(92), data[this.path] ? (__expression_weOqkq(93), true) : (__expression_weOqkq(94), false);
        }
    };
}
{
    __statement_V_i1iR(95);
    exports = module.exports = password;
}
