// keystone/.coverdebug/_Users_Jed_Development_Keystone_lib_fieldTypes_localfile.js.js

// Instrumentation Header
{
    var fs = require('fs');
    var __statement_AtaQAU, __expression_VstnoV, __block_wmhSKW;
    var store = require('/Users/Jed/Development/Keystone/node_modules/gulp-coverage/contrib/coverage_store.js');
    
    __statement_AtaQAU = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfile.js');
        fs.writeSync(fd, '{"statement": {"node": ' + i + '}},\n');
    }; 
    
    __expression_VstnoV = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfile.js');
        fs.writeSync(fd, '{"expression": {"node": ' + i + '}},\n');
    }; 
    
    __block_wmhSKW = function(i) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfile.js');
        fs.writeSync(fd, '{"block": ' + i + '},\n');
    }; 
    __intro_dU1xve = function(id, obj) {
        // console.log('__intro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        (typeof obj === 'object' || typeof obj === 'function') &&
            Object.defineProperty && Object.defineProperty(obj, '__instrumented_miss', {enumerable: false, writable: true});
        obj.__instrumented_miss = obj.__instrumented_miss || [];
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (obj.length === 0) {
                // console.log('interim miss: ', id);
                obj.__instrumented_miss[id] = true;
            } else {
                obj.__instrumented_miss[id] = false;
            }
        }
        return obj;
    };
    function isProbablyChainable(obj, id) {
        return obj &&
            obj.__instrumented_miss[id] !== undefined &&
            'number' === typeof obj.length;
    }
    __extro_NfXIUD = function(id, obj) {
        var fd = store.register('/Users/Jed/Development/Keystone/lib/fieldTypes/localfile.js');
        // console.log('__extro: ', id, ', obj.__instrumented_miss: ', obj.__instrumented_miss, ', obj.length: ', obj.length);
        if ('undefined' !== typeof obj && null !== obj && 'undefined' !== typeof obj.__instrumented_miss) {
            if (isProbablyChainable(obj, id) && obj.length === 0 && obj.__instrumented_miss[id]) {
                // if the call was not a "constructor" - i.e. it did not add things to the chainable
                // and it did not return anything from the chainable, it is a miss
                // console.log('miss: ', id);
            } else {
                fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
            }
            obj.__instrumented_miss[id] = undefined;
        } else {
            fs.writeSync(fd, '{"chain": {"node": ' + id + '}},\n');
        }
        return obj;
    };
};
////////////////////////

// Instrumented Code
{
    __statement_AtaQAU(0);
    var fs = (__expression_VstnoV(1), require('fs')), path = (__expression_VstnoV(2), require('path')), _ = (__expression_VstnoV(3), require('underscore')), moment = (__expression_VstnoV(4), require('moment')), async = (__expression_VstnoV(5), require('async')), util = (__expression_VstnoV(6), require('util')), utils = (__expression_VstnoV(7), require('keystone-utils')), super_ = (__expression_VstnoV(8), require('../field'));
}
function localfile(list, path, options) {
    __block_wmhSKW(0);
    {
        __statement_AtaQAU(9);
        this._underscoreMethods = [
            'format',
            'uploadFile'
        ];
    }
    {
        __statement_AtaQAU(10);
        this._pre = {
            move: []
        };
    }
    {
        __statement_AtaQAU(11);
        this._post = {
            move: []
        };
    }
    {
        __statement_AtaQAU(12);
        options.nofilter = true;
    }
    if (options.initial) {
        __block_wmhSKW(1);
        {
            __statement_AtaQAU(13);
            throw new Error((__expression_VstnoV(14), (__expression_VstnoV(15), (__expression_VstnoV(16), (__expression_VstnoV(17), (__expression_VstnoV(18), 'Invalid Configuration\n\n' + 'localfile fields (') + list.key) + '.') + (__expression_VstnoV(19), path)) + ') do not currently support being used as initial fields.\n'));
        }
    }
    {
        __statement_AtaQAU(20);
        __extro_NfXIUD(21, __intro_dU1xve(21, localfile.super_).call(this, list, path, options));
    }
    if (__expression_VstnoV(22), !options.dest) {
        __block_wmhSKW(2);
        {
            __statement_AtaQAU(23);
            throw new Error((__expression_VstnoV(24), (__expression_VstnoV(25), (__expression_VstnoV(26), (__expression_VstnoV(27), (__expression_VstnoV(28), 'Invalid Configuration\n\n' + 'localfile fields (') + list.key) + '.') + (__expression_VstnoV(29), path)) + ') require the "dest" option to be set.'));
        }
    }
    if (__expression_VstnoV(30), options.pre && options.pre.move) {
        __block_wmhSKW(3);
        {
            __statement_AtaQAU(31);
            this._pre.move = __extro_NfXIUD(32, __intro_dU1xve(32, this._pre.move).concat(options.pre.move));
        }
    }
    if (__expression_VstnoV(33), options.post && options.post.move) {
        __block_wmhSKW(4);
        {
            __statement_AtaQAU(34);
            this._post.move = __extro_NfXIUD(35, __intro_dU1xve(35, this._post.move).concat(options.post.move));
        }
    }
}
{
    __statement_AtaQAU(36);
    __extro_NfXIUD(37, __intro_dU1xve(37, util).inherits(localfile, super_));
}
{
    __statement_AtaQAU(38);
    localfile.prototype.pre = function (event, fn) {
        __block_wmhSKW(5);
        if (__expression_VstnoV(39), !this._pre[event]) {
            __block_wmhSKW(6);
            {
                __statement_AtaQAU(40);
                throw new Error((__expression_VstnoV(41), (__expression_VstnoV(42), (__expression_VstnoV(43), (__expression_VstnoV(44), (__expression_VstnoV(45), (__expression_VstnoV(46), (__expression_VstnoV(47), 'localfile (' + this.list.key) + '.') + this.path) + ') error: localfile.pre()\n\n') + 'Event ') + (__expression_VstnoV(48), event)) + ' is not supported.\n'));
            }
        }
        {
            __statement_AtaQAU(49);
            __extro_NfXIUD(50, __intro_dU1xve(50, this._pre[event]).push(fn));
        }
        return __expression_VstnoV(51), this;
    };
}
{
    __statement_AtaQAU(52);
    localfile.prototype.post = function (event, fn) {
        __block_wmhSKW(7);
        if (__expression_VstnoV(53), !this._post[event]) {
            __block_wmhSKW(8);
            {
                __statement_AtaQAU(54);
                throw new Error((__expression_VstnoV(55), (__expression_VstnoV(56), (__expression_VstnoV(57), (__expression_VstnoV(58), (__expression_VstnoV(59), (__expression_VstnoV(60), (__expression_VstnoV(61), 'localfile (' + this.list.key) + '.') + this.path) + ') error: localfile.post()\n\n') + 'Event ') + (__expression_VstnoV(62), event)) + ' is not supported.\n'));
            }
        }
        {
            __statement_AtaQAU(63);
            __extro_NfXIUD(64, __intro_dU1xve(64, this._post[event]).push(fn));
        }
        return __expression_VstnoV(65), this;
    };
}
{
    __statement_AtaQAU(66);
    localfile.prototype.addToSchema = function () {
        __block_wmhSKW(9);
        {
            __statement_AtaQAU(67);
            var field = this, schema = this.list.schema;
        }
        {
            __statement_AtaQAU(68);
            var paths = this.paths = {
                    filename: __extro_NfXIUD(69, __intro_dU1xve(69, this._path).append('.filename')),
                    path: __extro_NfXIUD(70, __intro_dU1xve(70, this._path).append('.path')),
                    size: __extro_NfXIUD(71, __intro_dU1xve(71, this._path).append('.size')),
                    filetype: __extro_NfXIUD(72, __intro_dU1xve(72, this._path).append('.filetype')),
                    exists: __extro_NfXIUD(73, __intro_dU1xve(73, this._path).append('.exists')),
                    upload: __extro_NfXIUD(74, __intro_dU1xve(74, this._path).append('_upload')),
                    action: __extro_NfXIUD(75, __intro_dU1xve(75, this._path).append('_action'))
                };
        }
        {
            __statement_AtaQAU(76);
            var schemaPaths = __extro_NfXIUD(77, __intro_dU1xve(77, this._path).addTo({}, {
                    filename: String,
                    path: String,
                    size: Number,
                    filetype: String
                }));
        }
        {
            __statement_AtaQAU(78);
            __extro_NfXIUD(79, __intro_dU1xve(79, schema).add(schemaPaths));
        }
        {
            __statement_AtaQAU(80);
            var exists = function (item) {
                __block_wmhSKW(10);
                {
                    __statement_AtaQAU(81);
                    var filepath = __extro_NfXIUD(82, __intro_dU1xve(82, item).get(paths.path)), filename = __extro_NfXIUD(83, __intro_dU1xve(83, item).get(paths.filename));
                }
                if (__expression_VstnoV(84), (__expression_VstnoV(85), !(__expression_VstnoV(86), filepath)) || (__expression_VstnoV(87), !(__expression_VstnoV(88), filename))) {
                    __block_wmhSKW(11);
                    return __expression_VstnoV(89), false;
                }
                return __expression_VstnoV(90), __extro_NfXIUD(91, __intro_dU1xve(91, fs).existsSync(__extro_NfXIUD(92, __intro_dU1xve(92, path).join(filepath, filename))));
            };
        }
        {
            __statement_AtaQAU(93);
            __extro_NfXIUD(94, __intro_dU1xve(94, __extro_NfXIUD(95, __intro_dU1xve(95, schema).virtual(paths.exists))).get(function () {
                __block_wmhSKW(12);
                return __expression_VstnoV(96), __extro_NfXIUD(97, __intro_dU1xve(97, schemaMethods.exists).apply(this));
            }));
        }
        {
            __statement_AtaQAU(98);
            var reset = function (item) {
                __block_wmhSKW(13);
                {
                    __statement_AtaQAU(99);
                    __extro_NfXIUD(100, __intro_dU1xve(100, item).set(field.path, {
                        filename: '',
                        path: '',
                        size: 0,
                        filetype: ''
                    }));
                }
            };
        }
        {
            __statement_AtaQAU(101);
            var schemaMethods = {
                    exists: function () {
                        __block_wmhSKW(14);
                        return __expression_VstnoV(102), (__expression_VstnoV(103), exists(this));
                    },
                    reset: function () {
                        __block_wmhSKW(15);
                        {
                            __statement_AtaQAU(104);
                            __expression_VstnoV(105), reset(this);
                        }
                    },
                    delete: function () {
                        __block_wmhSKW(16);
                        if (__expression_VstnoV(106), exists(this)) {
                            __block_wmhSKW(17);
                            {
                                __statement_AtaQAU(107);
                                __extro_NfXIUD(108, __intro_dU1xve(108, fs).unlinkSync(__extro_NfXIUD(109, __intro_dU1xve(109, path).join(__extro_NfXIUD(110, __intro_dU1xve(110, this).get(paths.path)), __extro_NfXIUD(111, __intro_dU1xve(111, this).get(paths.filename))))));
                            }
                        }
                        {
                            __statement_AtaQAU(112);
                            __expression_VstnoV(113), reset(this);
                        }
                    }
                };
        }
        {
            __statement_AtaQAU(114);
            __extro_NfXIUD(115, __intro_dU1xve(115, _).each(schemaMethods, function (fn, key) {
                __block_wmhSKW(18);
                {
                    __statement_AtaQAU(116);
                    __extro_NfXIUD(117, __intro_dU1xve(117, field).underscoreMethod(key, fn));
                }
            }));
        }
        {
            __statement_AtaQAU(118);
            this.apply = function (item, method) {
                __block_wmhSKW(19);
                return __expression_VstnoV(119), __extro_NfXIUD(120, __intro_dU1xve(120, schemaMethods[method]).apply(item, __extro_NfXIUD(121, __intro_dU1xve(121, Array.prototype.slice).call(arguments, 2))));
            };
        }
        {
            __statement_AtaQAU(122);
            __extro_NfXIUD(123, __intro_dU1xve(123, this).bindUnderscoreMethods());
        }
    };
}
{
    __statement_AtaQAU(124);
    localfile.prototype.format = function (item) {
        __block_wmhSKW(20);
        if (__extro_NfXIUD(125, __intro_dU1xve(125, this).hasFormatter())) {
            __block_wmhSKW(21);
            return __expression_VstnoV(126), __extro_NfXIUD(127, __intro_dU1xve(127, this.options).format(item, item[this.path]));
        }
        return __expression_VstnoV(128), __extro_NfXIUD(129, __intro_dU1xve(129, this).href(item));
    };
}
{
    __statement_AtaQAU(130);
    localfile.prototype.hasFormatter = function () {
        __block_wmhSKW(22);
        return __expression_VstnoV(131), (__expression_VstnoV(132), 'function' === (__expression_VstnoV(133), typeof this.options.format));
    };
}
{
    __statement_AtaQAU(134);
    localfile.prototype.href = function (item) {
        __block_wmhSKW(23);
        if (this.options.prefix) {
            __block_wmhSKW(24);
            return __expression_VstnoV(135), (__expression_VstnoV(136), (__expression_VstnoV(137), this.options.prefix + '/') + __extro_NfXIUD(138, __intro_dU1xve(138, item).get(this.paths.filename)));
        }
        return __expression_VstnoV(139), __extro_NfXIUD(140, __intro_dU1xve(140, path).join(__extro_NfXIUD(141, __intro_dU1xve(141, item).get(this.paths.path)), __extro_NfXIUD(142, __intro_dU1xve(142, item).get(this.paths.filename))));
    };
}
{
    __statement_AtaQAU(143);
    localfile.prototype.isModified = function (item) {
        __block_wmhSKW(25);
        return __expression_VstnoV(144), __extro_NfXIUD(145, __intro_dU1xve(145, item).isModified(this.paths.path));
    };
}
{
    __statement_AtaQAU(146);
    localfile.prototype.validateInput = function (data) {
        __block_wmhSKW(26);
        return __expression_VstnoV(147), true;
    };
}
{
    __statement_AtaQAU(148);
    localfile.prototype.updateItem = function (item, data) {
        __block_wmhSKW(27);
    };
}
{
    __statement_AtaQAU(149);
    localfile.prototype.uploadFile = function (item, file, update, callback) {
        __block_wmhSKW(28);
        {
            __statement_AtaQAU(150);
            var field = this, prefix = field.options.datePrefix ? (__expression_VstnoV(151), (__expression_VstnoV(153), __extro_NfXIUD(154, __intro_dU1xve(154, (__expression_VstnoV(155), moment())).format(field.options.datePrefix)) + '-')) : (__expression_VstnoV(152), ''), name = (__expression_VstnoV(156), (__expression_VstnoV(157), prefix) + file.name);
        }
        if (__expression_VstnoV(158), field.options.allowedTypes && (__expression_VstnoV(159), !__extro_NfXIUD(160, __intro_dU1xve(160, _).contains(field.options.allowedTypes, file.type)))) {
            __block_wmhSKW(29);
            return __expression_VstnoV(161), (__expression_VstnoV(162), callback(new Error((__expression_VstnoV(163), 'Unsupported File Type: ' + file.type))));
        }
        if (__expression_VstnoV(164), 'function' === (__expression_VstnoV(165), typeof update)) {
            __block_wmhSKW(30);
            {
                __statement_AtaQAU(166);
                callback = update;
            }
            {
                __statement_AtaQAU(167);
                update = false;
            }
        }
        {
            __statement_AtaQAU(168);
            var doMove = function (callback) {
                __block_wmhSKW(31);
                if (__expression_VstnoV(169), 'function' === (__expression_VstnoV(170), typeof field.options.filename)) {
                    __block_wmhSKW(32);
                    {
                        __statement_AtaQAU(171);
                        name = __extro_NfXIUD(172, __intro_dU1xve(172, field.options).filename(item, name));
                    }
                }
                {
                    __statement_AtaQAU(173);
                    __extro_NfXIUD(174, __intro_dU1xve(174, fs).rename(file.path, __extro_NfXIUD(175, __intro_dU1xve(175, path).join(field.options.dest, name)), function (err) {
                        __block_wmhSKW(33);
                        if (__expression_VstnoV(176), err) {
                            __block_wmhSKW(34);
                            return __expression_VstnoV(177), (__expression_VstnoV(178), callback(err));
                        }
                        {
                            __statement_AtaQAU(179);
                            var fileData = {
                                    filename: name,
                                    path: field.options.dest,
                                    size: file.size,
                                    filetype: file.type
                                };
                        }
                        if (__expression_VstnoV(180), update) {
                            __block_wmhSKW(35);
                            {
                                __statement_AtaQAU(181);
                                __extro_NfXIUD(182, __intro_dU1xve(182, item).set(field.path, fileData));
                            }
                        }
                        {
                            __statement_AtaQAU(183);
                            __expression_VstnoV(184), callback(null, fileData);
                        }
                    }));
                }
            };
        }
        {
            __statement_AtaQAU(185);
            __extro_NfXIUD(186, __intro_dU1xve(186, async).eachSeries(this._pre.move, function (fn, next) {
                __block_wmhSKW(36);
                {
                    __statement_AtaQAU(187);
                    __expression_VstnoV(188), fn(item, file, next);
                }
            }, function (err) {
                __block_wmhSKW(37);
                if (__expression_VstnoV(189), err) {
                    __block_wmhSKW(38);
                    return __expression_VstnoV(190), (__expression_VstnoV(191), callback(err));
                }
                {
                    __statement_AtaQAU(192);
                    __expression_VstnoV(193), doMove(function (err, fileData) {
                        __block_wmhSKW(39);
                        if (__expression_VstnoV(194), err) {
                            __block_wmhSKW(40);
                            return __expression_VstnoV(195), (__expression_VstnoV(196), callback(err));
                        }
                        {
                            __statement_AtaQAU(197);
                            __extro_NfXIUD(198, __intro_dU1xve(198, async).eachSeries(field._post.move, function (fn, next) {
                                __block_wmhSKW(41);
                                {
                                    __statement_AtaQAU(199);
                                    __expression_VstnoV(200), fn(item, file, fileData, next);
                                }
                            }, function (err) {
                                __block_wmhSKW(42);
                                if (__expression_VstnoV(201), err) {
                                    __block_wmhSKW(43);
                                    return __expression_VstnoV(202), (__expression_VstnoV(203), callback(err));
                                }
                                {
                                    __statement_AtaQAU(204);
                                    __expression_VstnoV(205), callback(null, fileData);
                                }
                            }));
                        }
                    });
                }
            }));
        }
    };
}
{
    __statement_AtaQAU(206);
    localfile.prototype.getRequestHandler = function (item, req, paths, callback) {
        __block_wmhSKW(44);
        {
            __statement_AtaQAU(207);
            var field = this;
        }
        if (__extro_NfXIUD(208, __intro_dU1xve(208, utils).isFunction(paths))) {
            __block_wmhSKW(45);
            {
                __statement_AtaQAU(209);
                callback = paths;
            }
            {
                __statement_AtaQAU(210);
                paths = field.paths;
            }
        } else if (__expression_VstnoV(211), !(__expression_VstnoV(212), paths)) {
            __block_wmhSKW(46);
            {
                __statement_AtaQAU(213);
                paths = field.paths;
            }
        }
        {
            __statement_AtaQAU(214);
            callback = (__expression_VstnoV(215), (__expression_VstnoV(216), callback) || function () {
                __block_wmhSKW(47);
            });
        }
        return __expression_VstnoV(217), function () {
            __block_wmhSKW(48);
            if (req.body) {
                __block_wmhSKW(49);
                {
                    __statement_AtaQAU(218);
                    var action = req.body[paths.action];
                }
                if (__extro_NfXIUD(219, __intro_dU1xve(219, /^(delete|reset)$/).test(action))) {
                    __block_wmhSKW(50);
                    {
                        __statement_AtaQAU(220);
                        __extro_NfXIUD(221, __intro_dU1xve(221, field).apply(item, action));
                    }
                }
            }
            if (__expression_VstnoV(222), (__expression_VstnoV(223), req.files && req.files[paths.upload]) && req.files[paths.upload].size) {
                __block_wmhSKW(51);
                return __expression_VstnoV(224), __extro_NfXIUD(225, __intro_dU1xve(225, field).uploadFile(item, req.files[paths.upload], true, callback));
            }
            return __expression_VstnoV(226), (__expression_VstnoV(227), callback());
        };
    };
}
{
    __statement_AtaQAU(228);
    localfile.prototype.handleRequest = function (item, req, paths, callback) {
        __block_wmhSKW(52);
        {
            __statement_AtaQAU(229);
            __expression_VstnoV(230), __extro_NfXIUD(231, __intro_dU1xve(231, this).getRequestHandler(item, req, paths, callback))();
        }
    };
}
{
    __statement_AtaQAU(232);
    exports = module.exports = localfile;
}
